<!DOCTYPE html>

<html lang="en" data-content_root="../../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>dendropy.model.birthdeath &#8212; DendroPy 5.0.2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../../../_static/rtd.css?v=cedc6a2f" />
    
    <script src="../../../_static/documentation_options.js?v=5df4c8d1"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">dendropy.model.birthdeath</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for dendropy.model.birthdeath</h1><div class="highlight"><pre>
<span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="c1">##############################################################################</span>
<span class="c1">##  DendroPy Phylogenetic Computing Library.</span>
<span class="c1">##</span>
<span class="c1">##  Copyright 2010-2015 Jeet Sukumaran and Mark T. Holder.</span>
<span class="c1">##  All rights reserved.</span>
<span class="c1">##</span>
<span class="c1">##  See &quot;LICENSE.rst&quot; for terms and conditions of usage.</span>
<span class="c1">##</span>
<span class="c1">##  If you use this work or any portion thereof in published work,</span>
<span class="c1">##  please cite it as:</span>
<span class="c1">##</span>
<span class="c1">##     Sukumaran, J. and M. T. Holder. 2010. DendroPy: a Python library</span>
<span class="c1">##     for phylogenetic computing. Bioinformatics 26: 1569-1571.</span>
<span class="c1">##</span>
<span class="c1">##############################################################################</span>

<span class="c1">##############################################################################</span>
<span class="c1">## Parts of this code were adapted from:</span>
<span class="c1">##</span>
<span class="c1">##  -   TESS</span>
<span class="c1">##</span>
<span class="c1">##          https://github.com/hoehna/TESS</span>
<span class="c1">##</span>
<span class="c1">##          H{&quot;o}hna S. 2013. Fast simulation of reconstructed phylogenies under</span>
<span class="c1">##          global time-dependent birth--death processes. Bioinformatics, 29(11)</span>
<span class="c1">##          1367-1374.</span>
<span class="c1">##</span>
<span class="c1">##          Copyright (c) 2012- Sebastian Hoehna</span>
<span class="c1">##</span>
<span class="c1">##          TESS is free software; you can redistribute it and/or modify</span>
<span class="c1">##          it under the terms of the GNU Lesser General Public License as</span>
<span class="c1">##          published by the Free Software Foundation; either version 2</span>
<span class="c1">##          of the License, or (at your option) any later version.</span>
<span class="c1">##</span>
<span class="c1">##############################################################################</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Models, modeling and model-fitting of birth-death processes.</span>

<span class="sd">-   Nee, S.  2001.  Inferring speciation rates from phylogenies.</span>
<span class="sd">    Evolution 55:661-668.</span>
<span class="sd">-   Yule, G. U. 1924. A mathematical theory of evolution based on the</span>
<span class="sd">    conclusions of Dr.  J. C. Willis. Phil. Trans. R. Soc. Lond. B</span>
<span class="sd">    213:21-87.</span>
<span class="sd">-   Hoehna, S. (2015). The time-dependent reconstructed evolutionary process</span>
<span class="sd">    with a key-role for mass-extinction events. Journal of theoretical biology,</span>
<span class="sd">    380, 321-331.</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">combinatorics</span>
<span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">probability</span>
<span class="kn">from</span> <span class="nn">dendropy.utility</span> <span class="kn">import</span> <span class="n">GLOBAL_RNG</span>
<span class="kn">from</span> <span class="nn">dendropy.utility.error</span> <span class="kn">import</span> <span class="n">TreeSimTotalExtinctionException</span>
<span class="kn">from</span> <span class="nn">dendropy.utility</span> <span class="kn">import</span> <span class="n">constants</span>
<span class="kn">from</span> <span class="nn">dendropy.utility</span> <span class="kn">import</span> <span class="n">deprecate</span>

<span class="kn">import</span> <span class="nn">dendropy</span>

<div class="viewcode-block" id="birth_death_tree">
<a class="viewcode-back" href="../../../library/treesim.html#dendropy.model.birthdeath.birth_death_tree">[docs]</a>
<span class="k">def</span> <span class="nf">birth_death_tree</span><span class="p">(</span><span class="n">birth_rate</span><span class="p">,</span> <span class="n">death_rate</span><span class="p">,</span> <span class="n">birth_rate_sd</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">death_rate_sd</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns a birth-death tree with birth rate specified by ``birth_rate``, and</span>
<span class="sd">    death rate specified by ``death_rate``, with edge lengths in continuous (real)</span>
<span class="sd">    units.</span>

<span class="sd">    Tree growth is controlled by one or more of the following arguments, of which</span>
<span class="sd">    at least one must be specified:</span>

<span class="sd">        - If ``num_extant_tips`` is given as a keyword argument, tree is grown until the</span>
<span class="sd">          number of EXTANT tips equals this number.</span>
<span class="sd">        - If ``num_extinct_tips`` is given as a keyword argument, tree is grown until the</span>
<span class="sd">          number of EXTINCT tips equals this number.</span>
<span class="sd">        - If ``num_total_tips`` is given as a keyword argument, tree is grown until the</span>
<span class="sd">          number of EXTANT plus EXTINCT tips equals this number.</span>
<span class="sd">        - If &#39;max_time&#39; is given as a keyword argument, tree is grown for</span>
<span class="sd">          a maximum of ``max_time``.</span>
<span class="sd">        - If ``gsa_ntax`` is given then the tree will be simulated up to this number of</span>
<span class="sd">          EXTANT tips (or 0 tips), then a tree will be randomly selected from the</span>
<span class="sd">          intervals which corresond to times at which the tree had exactly ``num_extant_tips``</span>
<span class="sd">          leaves. This allows for simulations according to the &quot;General</span>
<span class="sd">          Sampling Approach&quot; of Hartmann et al. (2010). If this option is</span>
<span class="sd">          specified, then ``num_extant_tips`` MUST be specified and</span>
<span class="sd">          ``num_extinct_tips`` and ``num_total_tips`` CANNOT be specified.</span>

<span class="sd">    If more than one of the above is given, then tree growth will terminate when</span>
<span class="sd">    *any* one of the termination conditions are met.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    birth_rate : float</span>
<span class="sd">        The birth rate.</span>
<span class="sd">    death_rate : float</span>
<span class="sd">        The death rate.</span>
<span class="sd">    birth_rate_sd : float</span>
<span class="sd">        The standard deviation of the normally-distributed mutation added to</span>
<span class="sd">        the birth rate as it is inherited by daughter nodes; if 0, birth rate</span>
<span class="sd">        does not evolve on the tree.</span>
<span class="sd">    death_rate_sd : float</span>
<span class="sd">        The standard deviation of the normally-distributed mutation added to</span>
<span class="sd">        the death rate as it is inherited by daughter nodes; if 0, death rate</span>
<span class="sd">        does not evolve on the tree.</span>

<span class="sd">    Keyword Arguments</span>
<span class="sd">    -----------------</span>

<span class="sd">    num_extant_tips: int</span>
<span class="sd">        If specified, branching process is terminated when number of EXTANT</span>
<span class="sd">        tips equals this number.</span>
<span class="sd">    num_extinct_tips: int</span>
<span class="sd">        If specified, branching process is terminated when number of EXTINCT</span>
<span class="sd">        tips equals this number.</span>
<span class="sd">    num_total_tips: int</span>
<span class="sd">        If specified, branching process is terminated when number of EXTINCT</span>
<span class="sd">        plus EXTANT tips equals this number.</span>
<span class="sd">    max_time: float</span>
<span class="sd">        If specified, branching process is terminated when time reaches or</span>
<span class="sd">        exceeds this value.</span>
<span class="sd">    gsa_ntax: int</span>
<span class="sd">        The General Sampling Approach threshold for number of taxa. See above</span>
<span class="sd">        for details.</span>
<span class="sd">    tree : Tree instance</span>
<span class="sd">        If given, then this tree will be used; otherwise a new one will be created.</span>
<span class="sd">    taxon_namespace : TaxonNamespace instance</span>
<span class="sd">        If given, then this will be assigned to the new tree, and, in addition,</span>
<span class="sd">        taxa assigned to tips will be sourced from or otherwise created with</span>
<span class="sd">        reference to this.</span>
<span class="sd">    is_assign_extant_taxa : bool [default: True]</span>
<span class="sd">        If False, then taxa will not be assigned to extant tips. If True</span>
<span class="sd">        (default), then taxa will be assigned to extant tips. Taxa will be</span>
<span class="sd">        assigned from the specified ``taxon_namespace`` or</span>
<span class="sd">        ``tree.taxon_namespace``. If the number of taxa required exceeds the</span>
<span class="sd">        number of taxa existing in the taxon namespace, new |Taxon| objects</span>
<span class="sd">        will be created as needed and added to the taxon namespace.</span>
<span class="sd">    is_assign_extinct_taxa : bool [default: True]</span>
<span class="sd">        If False, then taxa will not be assigned to extant tips. If True</span>
<span class="sd">        (default), then taxa will be assigned to extant tips. Taxa will be</span>
<span class="sd">        assigned from the specified ``taxon_namespace`` or</span>
<span class="sd">        ``tree.taxon_namespace``. If the number of taxa required exceeds the</span>
<span class="sd">        number of taxa existing in the taxon namespace, new |Taxon| objects</span>
<span class="sd">        will be created as needed and added to the taxon namespace. Note that</span>
<span class="sd">        this option only makes sense if extinct tips are retained (specified via</span>
<span class="sd">        &#39;is_retain_extinct_tips&#39; option), and will otherwise be ignored.</span>
<span class="sd">    is_add_extinct_attr: bool [default: True]</span>
<span class="sd">        If True (default), add an boolean attribute indicating whether or not a</span>
<span class="sd">        node is an extinct tip or not. False will skip this. Name of attribute</span>
<span class="sd">        is set by &#39;extinct_attr_name&#39; argument, defaulting to &#39;is_extinct&#39;.</span>
<span class="sd">        Note that this option only makes sense if extinct tips are retained</span>
<span class="sd">        (specified via &#39;is_retain_extinct_tips&#39; option), and will otherwise be</span>
<span class="sd">        ignored.</span>
<span class="sd">    extinct_attr_name: str [default: &#39;is_extinct&#39;]</span>
<span class="sd">        Name of attribute to add to nodes indicating whether or not tip is extinct.</span>
<span class="sd">        Note that this option only makes sense if extinct tips are retained</span>
<span class="sd">        (specified via &#39;is_retain_extinct_tips&#39; option), and will otherwise be</span>
<span class="sd">        ignored.</span>
<span class="sd">    is_retain_extinct_tips : bool [default: False]</span>
<span class="sd">        If True, extinct tips will be retained on tree. Defaults to False:</span>
<span class="sd">        extinct lineages removed from tree.</span>
<span class="sd">    repeat_until_success: bool [default: True]</span>
<span class="sd">        Under some conditions, it is possible for all lineages on a tree to go</span>
<span class="sd">        extinct. In this case, if this argument is given as |True| (the</span>
<span class="sd">        default), then a new branching process is initiated. If |False|</span>
<span class="sd">        (default), then a TreeSimTotalExtinctionException is raised.</span>
<span class="sd">    rng: random.Random() or equivalent instance</span>
<span class="sd">        A Random() object or equivalent can be passed using the ``rng`` keyword;</span>
<span class="sd">        otherwise GLOBAL_RNG is used.</span>

<span class="sd">    References</span>
<span class="sd">    ----------</span>

<span class="sd">    Hartmann, Wong, and Stadler &quot;Sampling Trees from Evolutionary Models&quot; Systematic Biology. 2010. 59(4). 465-476</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s2">&quot;assign_taxa&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="n">deprecate</span><span class="o">.</span><span class="n">dendropy_deprecation_warning</span><span class="p">(</span>
                <span class="n">message</span><span class="o">=</span><span class="s2">&quot;Deprecated: &#39;assign_taxa&#39; will no longer be supported as an argument to this function. Use &#39;is_assign_extant_taxa&#39; and/or &#39;is_assign_extinct_taxa&#39; instead&quot;</span><span class="p">,</span>
                <span class="n">stacklevel</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="n">a</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;assign_taxa&quot;</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;is_assign_extant_taxa&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">a</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;is_assign_extant_taxa&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">a</span>
    <span class="k">if</span> <span class="s2">&quot;ntax&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="n">deprecate</span><span class="o">.</span><span class="n">dendropy_deprecation_warning</span><span class="p">(</span>
                <span class="n">message</span><span class="o">=</span><span class="s2">&quot;Deprecated: &#39;ntax&#39; is no longer supported as an argument to this function. Use one or more of the following instead: &#39;num_extant_tips&#39;, &#39;num_extinct_tips&#39;, &#39;num_total_tips&#39;, or &#39;max_time&#39;&quot;</span><span class="p">,</span>
                <span class="n">stacklevel</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;num_extant_tips&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;ntax&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="p">((</span><span class="s2">&quot;num_extant_tips&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">)</span>
            <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;num_extinct_tips&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">)</span>
            <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;num_total_tips&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">)</span>
            <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;max_time&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">)</span> <span class="p">):</span>
        <span class="k">if</span> <span class="s2">&quot;taxon_namespace&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="c1">### cannot support legacy approach, b/c ``taxon_namespace`` may grow during function, leading to unpredictable behavior</span>
            <span class="c1"># deprecate.dendropy_deprecation_warning(</span>
            <span class="c1">#         preamble=&quot;Deprecated: The &#39;taxon_namespace&#39; argument can no longer be used to specify a termination condition as a side-effect. Use one or more of the following instead with the length of the taxon namespace instance as a value: &#39;num_extant_tips&#39;, &#39;num_extinct_tips&#39;, or &#39;num_total_tips&#39;&quot;,</span>
            <span class="c1">#         old_construct=&quot;tree = birth_death_tree(\n    ...\n    taxon_namespace=taxon_namespace,\n    ...\n)&quot;,</span>
            <span class="c1">#         new_construct=&quot;tree = birth_death_tree(\n    ...\n    taxon_namespace=taxon_namespace,\n    num_extant_tips=len(taxon_namespace),\n    ...\n)&quot;)</span>
            <span class="c1"># kwargs[&quot;num_extant_tips&quot;] = len(kwargs[&quot;taxon_namespace&quot;])</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The &#39;taxon_namespace&#39; argument can no longer be used to specify a termination condition as a side-effect.&quot;</span>
                             <span class="s2">&quot;Use one or more of the following instead with the length of the taxon namespace instance as a value: &quot;</span>
                             <span class="s2">&quot;&#39;num_extant_tips&#39;, &#39;num_extinct_tips&#39;, or &#39;num_total_tips&#39;.</span><span class="se">\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;That is, instead of:</span><span class="se">\n\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;    tree = birth_death_tree(</span><span class="se">\n</span><span class="s2">        ...</span><span class="se">\n</span><span class="s2">        taxon_namespace=taxon_namespace,</span><span class="se">\n</span><span class="s2">        ...</span><span class="se">\n</span><span class="s2">    )</span><span class="se">\n\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;Use:</span><span class="se">\n\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;    ntax = len(taxon_namespace)</span><span class="se">\n</span><span class="s2">    tree = birth_death_tree(</span><span class="se">\n</span><span class="s2">        ...</span><span class="se">\n</span><span class="s2">        taxon_namespace=taxon_namespace,</span><span class="se">\n</span><span class="s2">        num_extant_tips=ntax,</span><span class="se">\n</span><span class="s2">        ...</span><span class="se">\n</span><span class="s2">    )</span><span class="se">\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Or (recommended):</span><span class="se">\n\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;    tree = birth_death_tree(</span><span class="se">\n</span><span class="s2">        ...</span><span class="se">\n</span><span class="s2">        taxon_namespace=taxon_namespace,</span><span class="se">\n</span><span class="s2">        num_extant_tips=100,</span><span class="se">\n</span><span class="s2">        ...</span><span class="se">\n</span><span class="s2">    )</span><span class="se">\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Note that the taxon namespace instance size may grow during any particular call of the function depending on taxon assignment/creation settings, so&quot;</span>
                             <span class="s2">&quot; for stable and predictable behavor it is important to take a snapshot of the desired taxon namespace size before any call of the function, or, better yet&quot;</span>
                             <span class="s2">&quot; simply pass in a constant value.&quot;</span>
                             <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;One or more of the following must be specified: &#39;num_extant_tips&#39;, &#39;num_extinct_tips&#39;, or &#39;max_time&#39;&quot;</span><span class="p">)</span>
    <span class="n">target_num_extant_tips</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;num_extant_tips&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">target_num_extinct_tips</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;num_extinct_tips&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">target_num_total_tips</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;num_total_tips&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">max_time</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;max_time&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">gsa_ntax</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;gsa_ntax&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">is_add_extinct_attr</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;is_add_extinct_attr&#39;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">extinct_attr_name</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;extinct_attr_name&#39;</span><span class="p">,</span> <span class="s1">&#39;is_extinct&#39;</span><span class="p">)</span>
    <span class="n">is_retain_extinct_tips</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;is_retain_extinct_tips&#39;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
    <span class="n">is_assign_extant_taxa</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;is_assign_extant_taxa&#39;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">is_assign_extinct_taxa</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;is_assign_extinct_taxa&#39;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">repeat_until_success</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;repeat_until_success&#39;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>

    <span class="n">tree</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;tree&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;taxon_namespace&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

    <span class="n">rng</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;rng&#39;</span><span class="p">,</span> <span class="n">GLOBAL_RNG</span><span class="p">)</span>

    <span class="n">ignore_unrecognized_keyword_arguments</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;ignore_unrecognized_keyword_arguments&#39;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">ignore_unrecognized_keyword_arguments</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unsupported keyword arguments: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">kwargs</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>

    <span class="n">terminate_at_full_tree</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">if</span> <span class="n">gsa_ntax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">terminate_at_full_tree</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="c1"># gsa_ntax = 1 + target_num_taxa</span>
    <span class="k">elif</span> <span class="n">target_num_extant_tips</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;If &#39;gsa_ntax&#39; is specified, &#39;num_extant_tips&#39; must be specified&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">target_num_extinct_tips</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;If &#39;gsa_ntax&#39; is specified, &#39;num_extinct_tips&#39; cannot be specified&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">target_num_total_tips</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;If &#39;gsa_ntax&#39; is specified, &#39;num_total_tips&#39; cannot be specified&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">gsa_ntax</span> <span class="o">&lt;</span> <span class="n">target_num_extant_tips</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;gsa_ntax&#39; (</span><span class="si">{}</span><span class="s2">) must be greater than &#39;num_extant_tips&#39; (</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gsa_ntax</span><span class="p">,</span> <span class="n">target_num_extant_tips</span><span class="p">))</span>

    <span class="c1"># initialize tree</span>
    <span class="k">if</span> <span class="n">tree</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">taxon_namespace</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">tree</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="ow">is</span> <span class="n">taxon_namespace</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">taxon_namespace</span>
        <span class="n">extant_tips</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">extinct_tips</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">tree</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">nd</span><span class="o">.</span><span class="n">_child_nodes</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">):</span>
                    <span class="n">extant_tips</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                        <span class="nb">setattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">extinct_tips</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                        <span class="nb">setattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                <span class="nb">setattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">taxon_namespace</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">()</span>
        <span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="o">=</span><span class="n">taxon_namespace</span><span class="p">)</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">is_rooted</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">=</span> <span class="n">birth_rate</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">=</span> <span class="n">death_rate</span>
        <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
            <span class="nb">setattr</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">extant_tips</span> <span class="o">=</span> <span class="p">[</span><span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="p">]</span>
        <span class="n">extinct_tips</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">initial_extant_tip_set</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span>
    <span class="n">initial_extinct_tip_set</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">extinct_tips</span><span class="p">)</span>

    <span class="n">total_time</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1"># for the GSA simulations targetted_time_slices is a list of tuple</span>
    <span class="c1">#   the first element in the tuple is the duration of the amount</span>
    <span class="c1">#   that the simulation spent at the (targetted) number of taxa</span>
    <span class="c1">#   and a list of edge information. The list of edge information includes</span>
    <span class="c1">#   a list of terminal edges in the tree and the length for that edge</span>
    <span class="c1">#   that marks the beginning of the time slice that corresponds to the</span>
    <span class="c1">#   targetted number of taxa.</span>
    <span class="n">targetted_time_slices</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">gsa_ntax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">target_num_extant_tips</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">target_num_extant_tips</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="k">if</span> <span class="n">target_num_extinct_tips</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">extinct_tips</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">target_num_extinct_tips</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="k">if</span> <span class="n">target_num_total_tips</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">extinct_tips</span><span class="p">))</span> <span class="o">&gt;=</span> <span class="n">target_num_total_tips</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="k">if</span> <span class="n">max_time</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">total_time</span> <span class="o">&gt;=</span> <span class="n">max_time</span><span class="p">:</span>
                <span class="k">break</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">gsa_ntax</span><span class="p">:</span>
            <span class="k">break</span>

        <span class="c1"># get vector of birth/death probabilities, and</span>
        <span class="c1"># associate with nodes/events</span>
        <span class="n">event_rates</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">event_nodes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="s1">&#39;birth_rate&#39;</span><span class="p">):</span>
                <span class="n">nd</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">=</span> <span class="n">birth_rate</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="s1">&#39;death_rate&#39;</span><span class="p">):</span>
                <span class="n">nd</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">=</span> <span class="n">death_rate</span>
            <span class="n">event_rates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nd</span><span class="o">.</span><span class="n">birth_rate</span><span class="p">)</span>
            <span class="n">event_nodes</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nd</span><span class="p">,</span> <span class="kc">True</span><span class="p">))</span> <span class="c1"># birth event = True</span>
            <span class="n">event_rates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nd</span><span class="o">.</span><span class="n">death_rate</span><span class="p">)</span>
            <span class="n">event_nodes</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">nd</span><span class="p">,</span> <span class="kc">False</span><span class="p">))</span> <span class="c1"># birth event = False; i.e. death</span>

        <span class="c1"># get total probability of any birth/death</span>
        <span class="n">rate_of_any_event</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">event_rates</span><span class="p">)</span>

        <span class="c1"># waiting time based on above probability</span>
        <span class="n">waiting_time</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">expovariate</span><span class="p">(</span><span class="n">rate_of_any_event</span><span class="p">)</span>

        <span class="k">if</span> <span class="p">(</span> <span class="p">(</span><span class="n">gsa_ntax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span>
                <span class="ow">and</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span> <span class="o">==</span> <span class="n">target_num_extant_tips</span><span class="p">)</span>
                <span class="p">):</span>
            <span class="n">edge_and_start_length</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
                <span class="n">e</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">edge</span>
                <span class="n">edge_and_start_length</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">e</span><span class="p">,</span> <span class="n">e</span><span class="o">.</span><span class="n">length</span><span class="p">))</span>
            <span class="n">targetted_time_slices</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">waiting_time</span><span class="p">,</span> <span class="n">edge_and_start_length</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">terminate_at_full_tree</span><span class="p">:</span>
                <span class="k">break</span>

        <span class="c1"># add waiting time to nodes</span>
        <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">+=</span> <span class="n">waiting_time</span>
            <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
                <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">waiting_time</span>
        <span class="n">total_time</span> <span class="o">+=</span> <span class="n">waiting_time</span>

        <span class="c1"># if event occurs within time constraints</span>
        <span class="k">if</span> <span class="n">max_time</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">total_time</span> <span class="o">&lt;=</span> <span class="n">max_time</span><span class="p">:</span>
            <span class="c1"># normalize probability</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">event_rates</span><span class="p">)):</span>
                <span class="n">event_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">event_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">/</span><span class="n">rate_of_any_event</span>
            <span class="c1"># select node/event and process</span>
            <span class="n">nd</span><span class="p">,</span> <span class="n">birth_event</span> <span class="o">=</span> <span class="n">probability</span><span class="o">.</span><span class="n">weighted_choice</span><span class="p">(</span><span class="n">event_nodes</span><span class="p">,</span> <span class="n">event_rates</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">)</span>
            <span class="n">extant_tips</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">birth_event</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                    <span class="nb">setattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
                <span class="n">c1</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">new_child</span><span class="p">()</span>
                <span class="n">c2</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">new_child</span><span class="p">()</span>
                <span class="n">c1</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">c2</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">c1</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">+</span> <span class="n">rng</span><span class="o">.</span><span class="n">gauss</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">birth_rate_sd</span><span class="p">)</span>
                <span class="n">c1</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">+</span> <span class="n">rng</span><span class="o">.</span><span class="n">gauss</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">death_rate_sd</span><span class="p">)</span>
                <span class="n">c2</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">+</span> <span class="n">rng</span><span class="o">.</span><span class="n">gauss</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">birth_rate_sd</span><span class="p">)</span>
                <span class="n">c2</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">+</span> <span class="n">rng</span><span class="o">.</span><span class="n">gauss</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">death_rate_sd</span><span class="p">)</span>
                <span class="n">extant_tips</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">c1</span><span class="p">)</span>
                <span class="n">extant_tips</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">c2</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">extinct_tips</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                        <span class="nb">setattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># total extinction</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">gsa_ntax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
                        <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">targetted_time_slices</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">):</span>
                            <span class="k">break</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">repeat_until_success</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="n">TreeSimTotalExtinctionException</span><span class="p">()</span>
                    <span class="c1"># We are going to basically restart the simulation because</span>
                    <span class="c1"># the tree has gone extinct (without reaching the specified</span>
                    <span class="c1"># ntax)</span>
                    <span class="n">extant_tips</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">initial_extant_tip_set</span><span class="p">)</span>
                    <span class="n">extinct_tips</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">initial_extinct_tip_set</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                            <span class="nb">setattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
                        <span class="n">nd</span><span class="o">.</span><span class="n">clear_child_nodes</span><span class="p">()</span>
                    <span class="n">total_time</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">if</span> <span class="n">gsa_ntax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">total_duration_at_target_n_tax</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">targetted_time_slices</span><span class="p">:</span>
            <span class="n">total_duration_at_target_n_tax</span> <span class="o">+=</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">random</span><span class="p">()</span><span class="o">*</span><span class="n">total_duration_at_target_n_tax</span>
        <span class="n">selected_slice</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">targetted_time_slices</span><span class="p">):</span>
            <span class="n">r</span> <span class="o">-=</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">r</span> <span class="o">&lt;</span> <span class="mf">0.0</span><span class="p">:</span>
                <span class="n">selected_slice</span> <span class="o">=</span> <span class="n">i</span>
        <span class="k">assert</span><span class="p">(</span><span class="n">selected_slice</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">edges_at_slice</span> <span class="o">=</span> <span class="n">selected_slice</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">last_waiting_time</span> <span class="o">=</span> <span class="n">selected_slice</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">e</span><span class="p">,</span> <span class="n">prev_length</span> <span class="ow">in</span> <span class="n">edges_at_slice</span><span class="p">:</span>
            <span class="n">daughter_nd</span> <span class="o">=</span> <span class="n">e</span><span class="o">.</span><span class="n">head_node</span>
            <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">daughter_nd</span><span class="o">.</span><span class="n">child_nodes</span><span class="p">():</span>
                <span class="n">nd</span><span class="o">.</span><span class="n">_parent_node</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">extinct_tips</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="k">pass</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">extant_tips</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="k">pass</span>
                <span class="k">for</span> <span class="n">desc</span> <span class="ow">in</span> <span class="n">nd</span><span class="o">.</span><span class="n">preorder_iter</span><span class="p">():</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">extant_tips</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">desc</span><span class="p">)</span>
                    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                        <span class="k">pass</span>
            <span class="n">daughter_nd</span><span class="o">.</span><span class="n">clear_child_nodes</span><span class="p">()</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">extinct_tips</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">daughter_nd</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="n">extant_tips</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">daughter_nd</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                <span class="nb">setattr</span><span class="p">(</span><span class="n">daughter_nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
            <span class="n">e</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">prev_length</span> <span class="o">+</span> <span class="n">last_waiting_time</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">is_retain_extinct_tips</span><span class="p">:</span>
        <span class="n">processed_nodes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">extinct_tips</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">processed_nodes</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">processed_nodes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">extinct_tips</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="k">assert</span> <span class="ow">not</span> <span class="n">nd</span><span class="o">.</span><span class="n">_child_nodes</span>
            <span class="k">while</span> <span class="p">(</span><span class="n">nd</span><span class="o">.</span><span class="n">parent_node</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nd</span><span class="o">.</span><span class="n">parent_node</span><span class="o">.</span><span class="n">_child_nodes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">):</span>
                <span class="n">nd</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">parent_node</span>
                <span class="n">processed_nodes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
            <span class="n">tree</span><span class="o">.</span><span class="n">prune_subtree</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">suppress_unifurcations</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">suppress_unifurcations</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">is_assign_extant_taxa</span> <span class="ow">or</span> <span class="n">is_assign_extinct_taxa</span><span class="p">:</span>
        <span class="n">taxon_pool</span> <span class="o">=</span> <span class="p">[</span><span class="n">t</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">taxon_namespace</span><span class="p">]</span>
        <span class="n">rng</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">taxon_pool</span><span class="p">)</span>
        <span class="n">taxon_pool_labels</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">t</span><span class="o">.</span><span class="n">label</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">taxon_pool</span><span class="p">])</span>

        <span class="c1">### ONLY works if in GSA sub-section we remove ALL extant and</span>
        <span class="c1">### extinct nodes beyond time slice: expensive</span>
        <span class="c1">### Furthermore, main reason to use this approach is to have different</span>
        <span class="c1">### label prefixes for extinct vs. extant lineages, but the second time</span>
        <span class="c1">### this function is called with the same taxon namespace or any time</span>
        <span class="c1">### this function is called with a populated taxon namespace, that</span>
        <span class="c1">### aesthetic is lost.</span>
        <span class="c1"># node_pool_labels = (&quot;T&quot;, &quot;X&quot;)</span>
        <span class="c1"># for node_pool_idx, node_pool in enumerate((extant_tips, extinct_tips)):</span>
        <span class="c1">#     for node_idx, nd in enumerate(node_pool):</span>
        <span class="c1">#         if taxon_pool:</span>
        <span class="c1">#             taxon = taxon_pool.pop()</span>
        <span class="c1">#         else:</span>
        <span class="c1">#             taxon = taxon_namespace.require_taxon(&quot;{}{}&quot;.format(node_pool_labels[node_pool_idx], node_idx+1))</span>
        <span class="c1">#         nd.taxon = taxon</span>
        <span class="c1">#         assert not nd._child_nodes</span>

        <span class="n">tlabel_counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">leaf_nodes</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">leaf_nodes</span><span class="p">()</span>
        <span class="n">rng</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">leaf_nodes</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">nd_idx</span><span class="p">,</span> <span class="n">nd</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">leaf_nodes</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">is_assign_extant_taxa</span> <span class="ow">and</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">is_assign_extant_taxa</span> <span class="ow">and</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extinct_tips</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">taxon_pool</span><span class="p">:</span>
                <span class="n">taxon</span> <span class="o">=</span> <span class="n">taxon_pool</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
                    <span class="n">tlabel_counter</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="n">tlabel_counter</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">label</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">taxon_pool_labels</span><span class="p">:</span>
                        <span class="k">break</span>
                <span class="n">taxon</span> <span class="o">=</span> <span class="n">taxon_namespace</span><span class="o">.</span><span class="n">require_taxon</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="n">label</span><span class="p">)</span>
                <span class="n">taxon_pool_labels</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>
            <span class="n">nd</span><span class="o">.</span><span class="n">taxon</span> <span class="o">=</span> <span class="n">taxon</span>
    <span class="k">return</span> <span class="n">tree</span></div>



<span class="c1">#NICOLA: new version of birth-death tree simulation with linear time cost</span>
<span class="c1">#in number of tips instead of quadratic. To obtain this accelleration, however,</span>
<span class="c1">#I have removed variation in birth and death rates.</span>
<span class="k">def</span> <span class="nf">fast_birth_death_tree</span><span class="p">(</span><span class="n">birth_rate</span><span class="p">,</span> <span class="n">death_rate</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="c1">#NICOLA: we are not allowing variation in birth and death rate so to have</span>
    <span class="c1">#increased efficiency.</span>
    <span class="n">birth_rate_sd</span><span class="o">=</span><span class="mf">0.0</span>
    <span class="n">death_rate_sd</span><span class="o">=</span><span class="mf">0.0</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns a birth-death tree with birth rate specified by ``birth_rate``, and</span>
<span class="sd">    death rate specified by ``death_rate``, with edge lengths in continuous (real)</span>
<span class="sd">    units.</span>

<span class="sd">    Tree growth is controlled by one or more of the following arguments, of which</span>
<span class="sd">    at least one must be specified:</span>

<span class="sd">        - If ``num_extant_tips`` is given as a keyword argument, tree is grown until the</span>
<span class="sd">          number of EXTANT tips equals this number.</span>
<span class="sd">        - If ``num_extinct_tips`` is given as a keyword argument, tree is grown until the</span>
<span class="sd">          number of EXTINCT tips equals this number.</span>
<span class="sd">        - If ``num_total_tips`` is given as a keyword argument, tree is grown until the</span>
<span class="sd">          number of EXTANT plus EXTINCT tips equals this number.</span>
<span class="sd">        - If &#39;max_time&#39; is given as a keyword argument, tree is grown for</span>
<span class="sd">          a maximum of ``max_time``.</span>
<span class="sd">        - If ``gsa_ntax`` is given then the tree will be simulated up to this number of</span>
<span class="sd">          EXTANT tips (or 0 tips), then a tree will be randomly selected from the</span>
<span class="sd">          intervals which corresond to times at which the tree had exactly ``num_extant_tips``</span>
<span class="sd">          leaves. This allows for simulations according to the &quot;General</span>
<span class="sd">          Sampling Approach&quot; of Hartmann et al. (2010). If this option is</span>
<span class="sd">          specified, then ``num_extant_tips`` MUST be specified and</span>
<span class="sd">          ``num_extinct_tips`` and ``num_total_tips`` CANNOT be specified.</span>

<span class="sd">    If more than one of the above is given, then tree growth will terminate when</span>
<span class="sd">    *any* one of the termination conditions are met.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    birth_rate : float</span>
<span class="sd">        The birth rate.</span>
<span class="sd">    death_rate : float</span>
<span class="sd">        The death rate.</span>
<span class="sd">    birth_rate_sd : float</span>
<span class="sd">        The standard deviation of the normally-distributed mutation added to</span>
<span class="sd">        the birth rate as it is inherited by daughter nodes; if 0, birth rate</span>
<span class="sd">        does not evolve on the tree.</span>
<span class="sd">    death_rate_sd : float</span>
<span class="sd">        The standard deviation of the normally-distributed mutation added to</span>
<span class="sd">        the death rate as it is inherited by daughter nodes; if 0, death rate</span>
<span class="sd">        does not evolve on the tree.</span>

<span class="sd">    Keyword Arguments</span>
<span class="sd">    -----------------</span>

<span class="sd">    num_extant_tips: int</span>
<span class="sd">        If specified, branching process is terminated when number of EXTANT</span>
<span class="sd">        tips equals this number.</span>
<span class="sd">    num_extinct_tips: int</span>
<span class="sd">        If specified, branching process is terminated when number of EXTINCT</span>
<span class="sd">        tips equals this number.</span>
<span class="sd">    num_total_tips: int</span>
<span class="sd">        If specified, branching process is terminated when number of EXTINCT</span>
<span class="sd">        plus EXTANT tips equals this number.</span>
<span class="sd">    max_time: float</span>
<span class="sd">        If specified, branching process is terminated when time reaches or</span>
<span class="sd">        exceeds this value.</span>
<span class="sd">    gsa_ntax: int</span>
<span class="sd">        The General Sampling Approach threshold for number of taxa. See above</span>
<span class="sd">        for details.</span>
<span class="sd">    tree : Tree instance</span>
<span class="sd">        If given, then this tree will be used; otherwise a new one will be created.</span>
<span class="sd">    taxon_namespace : TaxonNamespace instance</span>
<span class="sd">        If given, then this will be assigned to the new tree, and, in addition,</span>
<span class="sd">        taxa assigned to tips will be sourced from or otherwise created with</span>
<span class="sd">        reference to this.</span>
<span class="sd">    is_assign_extant_taxa : bool [default: True]</span>
<span class="sd">        If False, then taxa will not be assigned to extant tips. If True</span>
<span class="sd">        (default), then taxa will be assigned to extant tips. Taxa will be</span>
<span class="sd">        assigned from the specified ``taxon_namespace`` or</span>
<span class="sd">        ``tree.taxon_namespace``. If the number of taxa required exceeds the</span>
<span class="sd">        number of taxa existing in the taxon namespace, new |Taxon| objects</span>
<span class="sd">        will be created as needed and added to the taxon namespace.</span>
<span class="sd">    is_assign_extinct_taxa : bool [default: True]</span>
<span class="sd">        If False, then taxa will not be assigned to extant tips. If True</span>
<span class="sd">        (default), then taxa will be assigned to extant tips. Taxa will be</span>
<span class="sd">        assigned from the specified ``taxon_namespace`` or</span>
<span class="sd">        ``tree.taxon_namespace``. If the number of taxa required exceeds the</span>
<span class="sd">        number of taxa existing in the taxon namespace, new |Taxon| objects</span>
<span class="sd">        will be created as needed and added to the taxon namespace. Note that</span>
<span class="sd">        this option only makes sense if extinct tips are retained (specified via</span>
<span class="sd">        &#39;is_retain_extinct_tips&#39; option), and will otherwise be ignored.</span>
<span class="sd">    is_add_extinct_attr: bool [default: True]</span>
<span class="sd">        If True (default), add an boolean attribute indicating whether or not a</span>
<span class="sd">        node is an extinct tip or not. False will skip this. Name of attribute</span>
<span class="sd">        is set by &#39;extinct_attr_name&#39; argument, defaulting to &#39;is_extinct&#39;.</span>
<span class="sd">        Note that this option only makes sense if extinct tips are retained</span>
<span class="sd">        (specified via &#39;is_retain_extinct_tips&#39; option), and will otherwise be</span>
<span class="sd">        ignored.</span>
<span class="sd">    extinct_attr_name: str [default: &#39;is_extinct&#39;]</span>
<span class="sd">        Name of attribute to add to nodes indicating whether or not tip is extinct.</span>
<span class="sd">        Note that this option only makes sense if extinct tips are retained</span>
<span class="sd">        (specified via &#39;is_retain_extinct_tips&#39; option), and will otherwise be</span>
<span class="sd">        ignored.</span>
<span class="sd">    is_retain_extinct_tips : bool [default: False]</span>
<span class="sd">        If True, extinct tips will be retained on tree. Defaults to False:</span>
<span class="sd">        extinct lineages removed from tree.</span>
<span class="sd">    repeat_until_success: bool [default: True]</span>
<span class="sd">        Under some conditions, it is possible for all lineages on a tree to go</span>
<span class="sd">        extinct. In this case, if this argument is given as |True| (the</span>
<span class="sd">        default), then a new branching process is initiated. If |False|</span>
<span class="sd">        (default), then a TreeSimTotalExtinctionException is raised.</span>
<span class="sd">    rng: random.Random() or equivalent instance</span>
<span class="sd">        A Random() object or equivalent can be passed using the ``rng`` keyword;</span>
<span class="sd">        otherwise GLOBAL_RNG is used.</span>

<span class="sd">    References</span>
<span class="sd">    ----------</span>

<span class="sd">    Hartmann, Wong, and Stadler &quot;Sampling Trees from Evolutionary Models&quot; Systematic Biology. 2010. 59(4). 465-476</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s2">&quot;assign_taxa&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="n">deprecate</span><span class="o">.</span><span class="n">dendropy_deprecation_warning</span><span class="p">(</span>
                <span class="n">message</span><span class="o">=</span><span class="s2">&quot;Deprecated: &#39;assign_taxa&#39; will no longer be supported as an argument to this function. Use &#39;is_assign_extant_taxa&#39; and/or &#39;is_assign_extinct_taxa&#39; instead&quot;</span><span class="p">,</span>
                <span class="n">stacklevel</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="n">a</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;assign_taxa&quot;</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;is_assign_extant_taxa&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">a</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;is_assign_extant_taxa&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">a</span>
    <span class="k">if</span> <span class="s2">&quot;ntax&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="n">deprecate</span><span class="o">.</span><span class="n">dendropy_deprecation_warning</span><span class="p">(</span>
                <span class="n">message</span><span class="o">=</span><span class="s2">&quot;Deprecated: &#39;ntax&#39; is no longer supported as an argument to this function. Use one or more of the following instead: &#39;num_extant_tips&#39;, &#39;num_extinct_tips&#39;, &#39;num_total_tips&#39;, or &#39;max_time&#39;&quot;</span><span class="p">,</span>
                <span class="n">stacklevel</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;num_extant_tips&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;ntax&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="p">((</span><span class="s2">&quot;num_extant_tips&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">)</span>
            <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;num_extinct_tips&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">)</span>
            <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;num_total_tips&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">)</span>
            <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;max_time&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">)</span> <span class="p">):</span>
        <span class="k">if</span> <span class="s2">&quot;taxon_namespace&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="c1">### cannot support legacy approach, b/c ``taxon_namespace`` may grow during function, leading to unpredictable behavior</span>
            <span class="c1"># deprecate.dendropy_deprecation_warning(</span>
            <span class="c1">#         preamble=&quot;Deprecated: The &#39;taxon_namespace&#39; argument can no longer be used to specify a termination condition as a side-effect. Use one or more of the following instead with the length of the taxon namespace instance as a value: &#39;num_extant_tips&#39;, &#39;num_extinct_tips&#39;, or &#39;num_total_tips&#39;&quot;,</span>
            <span class="c1">#         old_construct=&quot;tree = birth_death_tree(\n    ...\n    taxon_namespace=taxon_namespace,\n    ...\n)&quot;,</span>
            <span class="c1">#         new_construct=&quot;tree = birth_death_tree(\n    ...\n    taxon_namespace=taxon_namespace,\n    num_extant_tips=len(taxon_namespace),\n    ...\n)&quot;)</span>
            <span class="c1"># kwargs[&quot;num_extant_tips&quot;] = len(kwargs[&quot;taxon_namespace&quot;])</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The &#39;taxon_namespace&#39; argument can no longer be used to specify a termination condition as a side-effect.&quot;</span>
                             <span class="s2">&quot;Use one or more of the following instead with the length of the taxon namespace instance as a value: &quot;</span>
                             <span class="s2">&quot;&#39;num_extant_tips&#39;, &#39;num_extinct_tips&#39;, or &#39;num_total_tips&#39;.</span><span class="se">\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;That is, instead of:</span><span class="se">\n\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;    tree = birth_death_tree(</span><span class="se">\n</span><span class="s2">        ...</span><span class="se">\n</span><span class="s2">        taxon_namespace=taxon_namespace,</span><span class="se">\n</span><span class="s2">        ...</span><span class="se">\n</span><span class="s2">    )</span><span class="se">\n\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;Use:</span><span class="se">\n\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;    ntax = len(taxon_namespace)</span><span class="se">\n</span><span class="s2">    tree = birth_death_tree(</span><span class="se">\n</span><span class="s2">        ...</span><span class="se">\n</span><span class="s2">        taxon_namespace=taxon_namespace,</span><span class="se">\n</span><span class="s2">        num_extant_tips=ntax,</span><span class="se">\n</span><span class="s2">        ...</span><span class="se">\n</span><span class="s2">    )</span><span class="se">\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Or (recommended):</span><span class="se">\n\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;    tree = birth_death_tree(</span><span class="se">\n</span><span class="s2">        ...</span><span class="se">\n</span><span class="s2">        taxon_namespace=taxon_namespace,</span><span class="se">\n</span><span class="s2">        num_extant_tips=100,</span><span class="se">\n</span><span class="s2">        ...</span><span class="se">\n</span><span class="s2">    )</span><span class="se">\n</span><span class="s2">&quot;</span>
                             <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Note that the taxon namespace instance size may grow during any particular call of the function depending on taxon assignment/creation settings, so&quot;</span>
                             <span class="s2">&quot; for stable and predictable behavor it is important to take a snapshot of the desired taxon namespace size before any call of the function, or, better yet&quot;</span>
                             <span class="s2">&quot; simply pass in a constant value.&quot;</span>
                             <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;One or more of the following must be specified: &#39;num_extant_tips&#39;, &#39;num_extinct_tips&#39;, or &#39;max_time&#39;&quot;</span><span class="p">)</span>
    <span class="n">target_num_extant_tips</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;num_extant_tips&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">target_num_extinct_tips</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;num_extinct_tips&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">target_num_total_tips</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;num_total_tips&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">max_time</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;max_time&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">gsa_ntax</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;gsa_ntax&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">is_add_extinct_attr</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;is_add_extinct_attr&#39;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">extinct_attr_name</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;extinct_attr_name&#39;</span><span class="p">,</span> <span class="s1">&#39;is_extinct&#39;</span><span class="p">)</span>
    <span class="n">is_retain_extinct_tips</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;is_retain_extinct_tips&#39;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
    <span class="n">is_assign_extant_taxa</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;is_assign_extant_taxa&#39;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">is_assign_extinct_taxa</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;is_assign_extinct_taxa&#39;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">repeat_until_success</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;repeat_until_success&#39;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>

    <span class="n">tree</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;tree&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;taxon_namespace&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

    <span class="n">rng</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;rng&#39;</span><span class="p">,</span> <span class="n">GLOBAL_RNG</span><span class="p">)</span>

    <span class="n">ignore_unrecognized_keyword_arguments</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;ignore_unrecognized_keyword_arguments&#39;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">ignore_unrecognized_keyword_arguments</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unsupported keyword arguments: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">kwargs</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>

    <span class="n">terminate_at_full_tree</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">if</span> <span class="n">gsa_ntax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">terminate_at_full_tree</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="c1"># gsa_ntax = 1 + target_num_taxa</span>
    <span class="k">elif</span> <span class="n">target_num_extant_tips</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;If &#39;gsa_ntax&#39; is specified, &#39;num_extant_tips&#39; must be specified&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">target_num_extinct_tips</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;If &#39;gsa_ntax&#39; is specified, &#39;num_extinct_tups&#39; cannot be specified&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">target_num_total_tips</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;If &#39;gsa_ntax&#39; is specified, &#39;num_total_tips&#39; cannot be specified&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">gsa_ntax</span> <span class="o">&lt;</span> <span class="n">target_num_extant_tips</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;gsa_ntax&#39; (</span><span class="si">{}</span><span class="s2">) must be greater than &#39;num_extant_tips&#39; (</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gsa_ntax</span><span class="p">,</span> <span class="n">target_num_extant_tips</span><span class="p">))</span>

    <span class="c1"># initialize tree</span>
    <span class="k">if</span> <span class="n">tree</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">taxon_namespace</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">tree</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="ow">is</span> <span class="n">taxon_namespace</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">taxon_namespace</span>
        <span class="n">extant_tips</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">extinct_tips</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">tree</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">;</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">nd</span><span class="o">.</span><span class="n">_child_nodes</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">):</span>
                    <span class="n">extant_tips</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
                    <span class="c1">#NICOLA: the reason for this will become clear later</span>
                    <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span><span class="o">=-</span><span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span>
                    <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                        <span class="nb">setattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">extinct_tips</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                        <span class="nb">setattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                <span class="nb">setattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">taxon_namespace</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">()</span>
        <span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="o">=</span><span class="n">taxon_namespace</span><span class="p">)</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">is_rooted</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">=</span> <span class="n">birth_rate</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">=</span> <span class="n">death_rate</span>
        <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
            <span class="nb">setattr</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">extant_tips</span> <span class="o">=</span> <span class="p">[</span><span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="p">]</span>
        <span class="n">extinct_tips</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">initial_extant_tip_set</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span>
    <span class="n">initial_extinct_tip_set</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">extinct_tips</span><span class="p">)</span>
    <span class="c1">#NICOLA: in case we need a restart, let&#39;s record the initial branch lengths.</span>
    <span class="n">initial_lengths</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">initial_extant_tip_set</span><span class="p">:</span>
        <span class="n">initial_lengths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span><span class="p">)</span>

    <span class="n">total_time</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1"># for the GSA simulations targetted_time_slices is a list of tuple</span>
    <span class="c1">#   the first element in the tuple is the duration of the amount</span>
    <span class="c1">#   that the simulation spent at the (targetted) number of taxa</span>
    <span class="c1">#   and a list of edge information. The list of edge information includes</span>
    <span class="c1">#   a list of terminal edges in the tree and the length for that edge</span>
    <span class="c1">#   that marks the beginning of the time slice that corresponds to the</span>
    <span class="c1">#   targetted number of taxa.</span>
    <span class="n">targetted_time_slices</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">gsa_ntax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">target_num_extant_tips</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">target_num_extant_tips</span><span class="p">:</span>
                <span class="c1">#NICOLA: here and in the other cases, update the branch length so to</span>
                <span class="c1">#represent the true branch length (explained better below).</span>
                <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
                        <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span><span class="o">=</span><span class="n">total_time</span><span class="o">-</span><span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span>
                <span class="k">break</span>
            <span class="k">if</span> <span class="n">target_num_extinct_tips</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">extinct_tips</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">target_num_extinct_tips</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
                        <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span><span class="o">=</span><span class="n">total_time</span><span class="o">-</span><span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span>
                <span class="k">break</span>
            <span class="k">if</span> <span class="n">target_num_total_tips</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">extinct_tips</span><span class="p">))</span> <span class="o">&gt;=</span> <span class="n">target_num_total_tips</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
                        <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span><span class="o">=</span><span class="n">total_time</span><span class="o">-</span><span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span>
                <span class="k">break</span>
            <span class="k">if</span> <span class="n">max_time</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">total_time</span> <span class="o">&gt;=</span> <span class="n">max_time</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
                        <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span><span class="o">=</span><span class="n">total_time</span><span class="o">-</span><span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span>
                <span class="k">break</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">gsa_ntax</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
                        <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span><span class="o">=</span><span class="n">total_time</span><span class="o">-</span><span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span>
                <span class="k">break</span>


        <span class="c1"># get vector of birth/death probabilities, and</span>
        <span class="c1"># associate with nodes/events</span>
        <span class="c1">#NICOLA: replaced this</span>
   <span class="c1">#      event_rates = []</span>
<span class="c1">#         event_nodes = []</span>
<span class="c1">#         for nd in extant_tips:</span>
<span class="c1">#             if not hasattr(nd, &#39;birth_rate&#39;):</span>
<span class="c1">#                 nd.birth_rate = birth_rate</span>
<span class="c1">#             if not hasattr(nd, &#39;death_rate&#39;):</span>
<span class="c1">#                 nd.death_rate = death_rate</span>
<span class="c1">#             event_rates.append(nd.birth_rate)</span>
<span class="c1">#             event_nodes.append((nd, True)) # birth event = True</span>
<span class="c1">#             event_rates.append(nd.death_rate)</span>
<span class="c1">#             event_nodes.append((nd, False)) # birth event = False; i.e. death</span>


        <span class="c1"># get total probability of any birth/death</span>
        <span class="c1">#rate_of_any_event = sum(event_rates)</span>
        <span class="c1">#NICOLA: instead, assume that each node has the same birth and death rate, and sample</span>
        <span class="c1">#from the total rate and then sample one node at random from extant ones.</span>
        <span class="c1">#this has constant cost instead of linear in the number of extant taxa.</span>
        <span class="n">rate_of_any_event</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="n">birth_rate</span><span class="o">+</span><span class="n">death_rate</span><span class="p">)</span>

        <span class="c1"># waiting time based on above probability</span>
        <span class="n">waiting_time</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">expovariate</span><span class="p">(</span><span class="n">rate_of_any_event</span><span class="p">)</span>

        <span class="k">if</span> <span class="p">(</span> <span class="p">(</span><span class="n">gsa_ntax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span>
                <span class="ow">and</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span> <span class="o">==</span> <span class="n">target_num_extant_tips</span><span class="p">)</span>
                <span class="p">):</span>
            <span class="n">edge_and_start_length</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
                <span class="n">e</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">edge</span>
                <span class="c1">#NICOLA: modified accordingly</span>
                <span class="n">edge_and_start_length</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">e</span><span class="p">,</span> <span class="n">total_time</span><span class="o">-</span><span class="n">e</span><span class="o">.</span><span class="n">length</span><span class="p">))</span>
            <span class="n">targetted_time_slices</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">waiting_time</span><span class="p">,</span> <span class="n">edge_and_start_length</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">terminate_at_full_tree</span><span class="p">:</span>
                <span class="c1">#NICOLA: terminal update as before</span>
                <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
                        <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span><span class="o">=</span><span class="n">total_time</span><span class="o">-</span><span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span>
                <span class="k">break</span>

        <span class="c1"># add waiting time to nodes</span>
        <span class="c1">#Nicola: this is the part that makes it quadratic in ntax.</span>
        <span class="c1">#instead, we use a trick of initializing nd.edge.length to the time of node creation,</span>
        <span class="c1"># and then we update it to the correct branch length when the node is closed or simulations are terminated.</span>
        <span class="c1">#for nd in extant_tips:</span>
        <span class="c1">#    try:</span>
        <span class="c1">#        nd.edge.length += waiting_time</span>
        <span class="c1">#    except TypeError:</span>
        <span class="c1">#        nd.edge.length = waiting_time</span>
        <span class="n">total_time</span> <span class="o">+=</span> <span class="n">waiting_time</span>

        <span class="c1"># if event occurs within time constraints</span>
        <span class="k">if</span> <span class="n">max_time</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">total_time</span> <span class="o">&lt;=</span> <span class="n">max_time</span><span class="p">:</span>
            <span class="c1"># normalize probability</span>
            <span class="c1">#for i in range(len(event_rates)):</span>
            <span class="c1">#    event_rates[i] = event_rates[i]/rate_of_any_event</span>
            <span class="c1"># select node/event and process</span>
            <span class="c1">#nd, birth_event = probability.weighted_choice(event_nodes, event_rates, rng=rng)</span>
            <span class="c1">#NICOLA: instead of sampling from a categorical distribution (requiring linear time in ntax)</span>
            <span class="c1">#sample a random integer to represent the chosen taxon, and sample if birth or death event.</span>
            <span class="n">taxI</span><span class="o">=</span><span class="n">rng</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">rng</span><span class="o">.</span><span class="n">random</span><span class="p">()</span><span class="o">&lt;</span> <span class="n">birth_rate</span><span class="o">/</span><span class="p">(</span><span class="n">birth_rate</span><span class="o">+</span><span class="n">death_rate</span><span class="p">):</span>
                <span class="n">birth_event</span><span class="o">=</span><span class="kc">True</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">birth_event</span><span class="o">=</span><span class="kc">False</span>
            <span class="n">nd</span><span class="o">=</span><span class="n">extant_tips</span><span class="p">[</span><span class="n">taxI</span><span class="p">]</span>
            <span class="c1">#extant_tips.remove(nd)</span>
            <span class="k">if</span> <span class="n">birth_event</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                    <span class="nb">setattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
                <span class="n">c1</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">new_child</span><span class="p">()</span>
                <span class="n">c2</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">new_child</span><span class="p">()</span>
                <span class="n">extant_tips</span><span class="p">[</span><span class="n">taxI</span><span class="p">]</span><span class="o">=</span><span class="n">c1</span>
                <span class="c1">#NICOLA: we initialize branch lengths to total time;</span>
                <span class="c1">#this may not make sense at first, but when we are done with a node, we</span>
                <span class="c1">#update the branch length to the new current time minus the time of node creation</span>
                <span class="c1">#which gives us the branch length. The advantage is that this takes linear time</span>
                <span class="c1">#instead of quadratic.</span>
                <span class="n">c1</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">total_time</span>
                <span class="n">c2</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">total_time</span>
                <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">total_time</span><span class="o">-</span><span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span>
                <span class="c1">#c1.edge.length = 0</span>
                <span class="c1">#c2.edge.length = 0</span>
                <span class="c1">#NICOLA: here, for speed, we don&#39;t allow variation in birth rates.</span>
                <span class="c1">#It could be possible to allow it, but it would take ntax*log(ntax) time,</span>
                <span class="c1">#which wouldn&#39;t be bad, but it would also require a more complicated approach.</span>
                <span class="c1">#c1.birth_rate = nd.birth_rate + rng.gauss(0, birth_rate_sd)</span>
                <span class="c1">#c1.death_rate = nd.death_rate + rng.gauss(0, death_rate_sd)</span>
                <span class="c1">#c2.birth_rate = nd.birth_rate + rng.gauss(0, birth_rate_sd)</span>
                <span class="c1">#c2.death_rate = nd.death_rate + rng.gauss(0, death_rate_sd)</span>
                <span class="n">c1</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">=</span> <span class="n">birth_rate</span>
                <span class="n">c1</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">=</span> <span class="n">death_rate</span>
                <span class="n">c2</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">=</span> <span class="n">birth_rate</span>
                <span class="n">c2</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">=</span> <span class="n">death_rate</span>
                <span class="c1">#extant_tips.append(c1)</span>
                <span class="n">extant_tips</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">c2</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">del</span> <span class="n">extant_tips</span><span class="p">[</span><span class="n">taxI</span><span class="p">]</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">extant_tips</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">extinct_tips</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                        <span class="nb">setattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># total extinction</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">gsa_ntax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
                        <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">targetted_time_slices</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">):</span>
                            <span class="k">break</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">repeat_until_success</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="n">TreeSimTotalExtinctionException</span><span class="p">()</span>
                    <span class="c1"># We are going to basically restart the simulation because</span>
                    <span class="c1"># the tree has gone extinct (without reaching the specified</span>
                    <span class="c1"># ntax)</span>
                    <span class="n">extant_tips</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">initial_extant_tip_set</span><span class="p">)</span>
                    <span class="n">extinct_tips</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">initial_extinct_tip_set</span><span class="p">)</span>
                    <span class="n">ndIndex</span><span class="o">=</span><span class="mi">0</span>
                    <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                            <span class="nb">setattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
                        <span class="n">nd</span><span class="o">.</span><span class="n">clear_child_nodes</span><span class="p">()</span>
                        <span class="c1">#NICOLA: restore branch length to original value</span>
                        <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span><span class="o">=</span><span class="n">initial_lengths</span><span class="p">[</span><span class="n">ndIndex</span><span class="p">]</span>
                        <span class="n">ndIndex</span><span class="o">+=</span><span class="mi">1</span>
                    <span class="n">total_time</span> <span class="o">=</span> <span class="mi">0</span>


    <span class="k">if</span> <span class="n">gsa_ntax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">total_duration_at_target_n_tax</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">targetted_time_slices</span><span class="p">:</span>
            <span class="n">total_duration_at_target_n_tax</span> <span class="o">+=</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">random</span><span class="p">()</span><span class="o">*</span><span class="n">total_duration_at_target_n_tax</span>
        <span class="n">selected_slice</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">targetted_time_slices</span><span class="p">):</span>
            <span class="n">r</span> <span class="o">-=</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">r</span> <span class="o">&lt;</span> <span class="mf">0.0</span><span class="p">:</span>
                <span class="n">selected_slice</span> <span class="o">=</span> <span class="n">i</span>
        <span class="k">assert</span><span class="p">(</span><span class="n">selected_slice</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">edges_at_slice</span> <span class="o">=</span> <span class="n">selected_slice</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">last_waiting_time</span> <span class="o">=</span> <span class="n">selected_slice</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">e</span><span class="p">,</span> <span class="n">prev_length</span> <span class="ow">in</span> <span class="n">edges_at_slice</span><span class="p">:</span>
            <span class="n">daughter_nd</span> <span class="o">=</span> <span class="n">e</span><span class="o">.</span><span class="n">head_node</span>
            <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">daughter_nd</span><span class="o">.</span><span class="n">child_nodes</span><span class="p">():</span>
                <span class="n">nd</span><span class="o">.</span><span class="n">_parent_node</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">extinct_tips</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="k">pass</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">extant_tips</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="k">pass</span>
                <span class="k">for</span> <span class="n">desc</span> <span class="ow">in</span> <span class="n">nd</span><span class="o">.</span><span class="n">preorder_iter</span><span class="p">():</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">extant_tips</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">desc</span><span class="p">)</span>
                    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                        <span class="k">pass</span>
            <span class="n">daughter_nd</span><span class="o">.</span><span class="n">clear_child_nodes</span><span class="p">()</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">extinct_tips</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">daughter_nd</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="n">extant_tips</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">daughter_nd</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">is_add_extinct_attr</span><span class="p">:</span>
                <span class="nb">setattr</span><span class="p">(</span><span class="n">daughter_nd</span><span class="p">,</span> <span class="n">extinct_attr_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
            <span class="n">e</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">prev_length</span> <span class="o">+</span> <span class="n">last_waiting_time</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">is_retain_extinct_tips</span><span class="p">:</span>
        <span class="n">processed_nodes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">extinct_tips</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">processed_nodes</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">processed_nodes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">extinct_tips</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="k">assert</span> <span class="ow">not</span> <span class="n">nd</span><span class="o">.</span><span class="n">_child_nodes</span>
            <span class="k">while</span> <span class="p">(</span><span class="n">nd</span><span class="o">.</span><span class="n">parent_node</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nd</span><span class="o">.</span><span class="n">parent_node</span><span class="o">.</span><span class="n">_child_nodes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">):</span>
                <span class="n">nd</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">parent_node</span>
                <span class="n">processed_nodes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
            <span class="n">tree</span><span class="o">.</span><span class="n">prune_subtree</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="n">suppress_unifurcations</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">suppress_unifurcations</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">is_assign_extant_taxa</span> <span class="ow">or</span> <span class="n">is_assign_extinct_taxa</span><span class="p">:</span>
        <span class="n">taxon_pool</span> <span class="o">=</span> <span class="p">[</span><span class="n">t</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">taxon_namespace</span><span class="p">]</span>
        <span class="n">rng</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">taxon_pool</span><span class="p">)</span>
        <span class="n">taxon_pool_labels</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">t</span><span class="o">.</span><span class="n">label</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">taxon_pool</span><span class="p">])</span>

        <span class="c1">### ONLY works if in GSA sub-section we remove ALL extant and</span>
        <span class="c1">### extinct nodes beyond time slice: expensive</span>
        <span class="c1">### Furthermore, main reason to use this approach is to have different</span>
        <span class="c1">### label prefixes for extinct vs. extant lineages, but the second time</span>
        <span class="c1">### this function is called with the same taxon namespace or any time</span>
        <span class="c1">### this function is called with a populated taxon namespace, that</span>
        <span class="c1">### aesthetic is lost.</span>
        <span class="c1"># node_pool_labels = (&quot;T&quot;, &quot;X&quot;)</span>
        <span class="c1"># for node_pool_idx, node_pool in enumerate((extant_tips, extinct_tips)):</span>
        <span class="c1">#     for node_idx, nd in enumerate(node_pool):</span>
        <span class="c1">#         if taxon_pool:</span>
        <span class="c1">#             taxon = taxon_pool.pop()</span>
        <span class="c1">#         else:</span>
        <span class="c1">#             taxon = taxon_namespace.require_taxon(&quot;{}{}&quot;.format(node_pool_labels[node_pool_idx], node_idx+1))</span>
        <span class="c1">#         nd.taxon = taxon</span>
        <span class="c1">#         assert not nd._child_nodes</span>

        <span class="n">tlabel_counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">leaf_nodes</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">leaf_nodes</span><span class="p">()</span>
        <span class="n">rng</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">leaf_nodes</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">nd_idx</span><span class="p">,</span> <span class="n">nd</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">leaf_nodes</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">is_assign_extant_taxa</span> <span class="ow">and</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extant_tips</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">is_assign_extant_taxa</span> <span class="ow">and</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">extinct_tips</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">taxon_pool</span><span class="p">:</span>
                <span class="n">taxon</span> <span class="o">=</span> <span class="n">taxon_pool</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
                    <span class="n">tlabel_counter</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="n">tlabel_counter</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">label</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">taxon_pool_labels</span><span class="p">:</span>
                        <span class="k">break</span>
                <span class="n">taxon</span> <span class="o">=</span> <span class="n">taxon_namespace</span><span class="o">.</span><span class="n">require_taxon</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="n">label</span><span class="p">)</span>
                <span class="n">taxon_pool_labels</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>
            <span class="n">nd</span><span class="o">.</span><span class="n">taxon</span> <span class="o">=</span> <span class="n">taxon</span>
    <span class="k">return</span> <span class="n">tree</span>



<div class="viewcode-block" id="discrete_birth_death_tree">
<a class="viewcode-back" href="../../../library/treesim.html#dendropy.model.birthdeath.discrete_birth_death_tree">[docs]</a>
<span class="k">def</span> <span class="nf">discrete_birth_death_tree</span><span class="p">(</span><span class="n">birth_rate</span><span class="p">,</span> <span class="n">death_rate</span><span class="p">,</span> <span class="n">birth_rate_sd</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">death_rate_sd</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns a birth-death tree with birth rate specified by ``birth_rate``, and</span>
<span class="sd">    death rate specified by ``death_rate``, with edge lengths in discrete (integer)</span>
<span class="sd">    units.</span>

<span class="sd">    ``birth_rate_sd`` is the standard deviation of the normally-distributed mutation</span>
<span class="sd">    added to the birth rate as it is inherited by daughter nodes; if 0, birth</span>
<span class="sd">    rate does not evolve on the tree.</span>

<span class="sd">    ``death_rate_sd`` is the standard deviation of the normally-distributed mutation</span>
<span class="sd">    added to the death rate as it is inherited by daughter nodes; if 0, death</span>
<span class="sd">    rate does not evolve on the tree.</span>

<span class="sd">    Tree growth is controlled by one or more of the following arguments, of which</span>
<span class="sd">    at least one must be specified:</span>

<span class="sd">        - If ``ntax`` is given as a keyword argument, tree is grown until the number of</span>
<span class="sd">          tips == ntax.</span>
<span class="sd">        - If ``taxon_namespace`` is given as a keyword argument, tree is grown until the</span>
<span class="sd">          number of tips == len(taxon_namespace), and the taxa are assigned randomly to the</span>
<span class="sd">          tips.</span>
<span class="sd">        - If &#39;max_time&#39; is given as a keyword argument, tree is grown for ``max_time``</span>
<span class="sd">          number of generations.</span>

<span class="sd">    If more than one of the above is given, then tree growth will terminate when</span>
<span class="sd">    *any* of the termination conditions (i.e., number of tips == ``ntax``, or number</span>
<span class="sd">    of tips == len(taxon_namespace) or number of generations = ``max_time``) are met.</span>

<span class="sd">    Also accepts a Tree object (with valid branch lengths) as an argument passed</span>
<span class="sd">    using the keyword ``tree``: if given, then this tree will be used; otherwise</span>
<span class="sd">    a new one will be created.</span>

<span class="sd">    If ``assign_taxa`` is False, then taxa will *not* be assigned to the tips;</span>
<span class="sd">    otherwise (default), taxa will be assigned. If ``taxon_namespace`` is given</span>
<span class="sd">    (``tree.taxon_namespace``, if ``tree`` is given), and the final number of tips on the</span>
<span class="sd">    tree after the termination condition is reached is less then the number of</span>
<span class="sd">    taxa in ``taxon_namespace`` (as will be the case, for example, when</span>
<span class="sd">    ``ntax`` &lt; len(``taxon_namespace``)), then a random subset of taxa in ``taxon_namespace`` will</span>
<span class="sd">    be assigned to the tips of tree. If the number of tips is more than the number</span>
<span class="sd">    of taxa in the ``taxon_namespace``, new Taxon objects will be created and added</span>
<span class="sd">    to the ``taxon_namespace`` if the keyword argument ``create_required_taxa`` is not given as</span>
<span class="sd">    False.</span>

<span class="sd">    Under some conditions, it is possible for all lineages on a tree to go extinct.</span>
<span class="sd">    In this case, if the keyword argument ``repeat_until_success`` is |True|, then a new</span>
<span class="sd">    branching process is initiated.</span>
<span class="sd">    If |False| (default), then a TreeSimTotalExtinctionException is raised.</span>

<span class="sd">    A Random() object or equivalent can be passed using the ``rng`` keyword;</span>
<span class="sd">    otherwise GLOBAL_RNG is used.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s1">&#39;ntax&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span> \
        <span class="ow">and</span> <span class="s1">&#39;taxon_namespace&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span> \
        <span class="ow">and</span> <span class="s1">&#39;max_time&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;At least one of the following must be specified: &#39;ntax&#39;, &#39;taxon_namespace&#39;, or &#39;max_time&#39;&quot;</span><span class="p">)</span>
    <span class="n">target_num_taxa</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">taxon_namespace</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">target_num_gens</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;max_time&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">if</span> <span class="s1">&#39;taxon_namespace&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;taxon_namespace&#39;</span><span class="p">)</span>
        <span class="n">target_num_taxa</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;ntax&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="p">))</span>
    <span class="k">elif</span> <span class="s1">&#39;ntax&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="n">target_num_taxa</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s1">&#39;ntax&#39;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">taxon_namespace</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">()</span>
    <span class="n">repeat_until_success</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;repeat_until_success&#39;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
    <span class="n">rng</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;rng&#39;</span><span class="p">,</span> <span class="n">GLOBAL_RNG</span><span class="p">)</span>

    <span class="c1"># grow tree</span>
    <span class="k">if</span> <span class="s2">&quot;tree&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="n">tree</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s1">&#39;tree&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="s2">&quot;taxon_namespace&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="n">kwargs</span><span class="p">[</span><span class="s1">&#39;taxon_namespace&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">tree</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot specify both ``tree`` and ``taxon_namespace``&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="o">=</span><span class="n">taxon_namespace</span><span class="p">)</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">is_rooted</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">=</span> <span class="n">birth_rate</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">=</span> <span class="n">death_rate</span>
    <span class="n">leaf_nodes</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">leaf_nodes</span><span class="p">()</span>
    <span class="n">num_gens</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">while</span> <span class="p">(</span><span class="n">target_num_taxa</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">leaf_nodes</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">target_num_taxa</span><span class="p">)</span> \
            <span class="ow">and</span> <span class="p">(</span><span class="n">target_num_gens</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">num_gens</span> <span class="o">&lt;</span> <span class="n">target_num_gens</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">leaf_nodes</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="s1">&#39;birth_rate&#39;</span><span class="p">):</span>
                <span class="n">nd</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">=</span> <span class="n">birth_rate</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="s1">&#39;death_rate&#39;</span><span class="p">):</span>
                <span class="n">nd</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">=</span> <span class="n">death_rate</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
                <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">u</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">u</span> <span class="o">&lt;</span> <span class="n">nd</span><span class="o">.</span><span class="n">birth_rate</span><span class="p">:</span>
                <span class="n">c1</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">new_child</span><span class="p">()</span>
                <span class="n">c2</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">new_child</span><span class="p">()</span>
                <span class="n">c1</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">c2</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">c1</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">+</span> <span class="n">rng</span><span class="o">.</span><span class="n">gauss</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">birth_rate_sd</span><span class="p">)</span>
                <span class="n">c1</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">+</span> <span class="n">rng</span><span class="o">.</span><span class="n">gauss</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">death_rate_sd</span><span class="p">)</span>
                <span class="n">c2</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">+</span> <span class="n">rng</span><span class="o">.</span><span class="n">gauss</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">birth_rate_sd</span><span class="p">)</span>
                <span class="n">c2</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">=</span> <span class="n">nd</span><span class="o">.</span><span class="n">death_rate</span> <span class="o">+</span> <span class="n">rng</span><span class="o">.</span><span class="n">gauss</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">death_rate_sd</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">u</span> <span class="o">&gt;</span> <span class="n">nd</span><span class="o">.</span><span class="n">birth_rate</span> <span class="ow">and</span> <span class="n">u</span> <span class="o">&lt;</span> <span class="p">(</span><span class="n">nd</span><span class="o">.</span><span class="n">birth_rate</span> <span class="o">+</span> <span class="n">nd</span><span class="o">.</span><span class="n">death_rate</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">nd</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="p">:</span>
                    <span class="n">tree</span><span class="o">.</span><span class="n">prune_subtree</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
                <span class="k">elif</span> <span class="ow">not</span> <span class="n">repeat_until_success</span><span class="p">:</span>
                    <span class="c1"># all lineages are extinct: raise exception</span>
                    <span class="k">raise</span> <span class="n">TreeSimTotalExtinctionException</span><span class="p">()</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># all lineages are extinct: repeat</span>
                    <span class="n">num_gens</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="n">num_gens</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">leaf_nodes</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">leaf_nodes</span><span class="p">()</span>

    <span class="c1"># If termination condition specified by ntax or taxon_namespace, then the last</span>
    <span class="c1"># split will have a daughter edges of length == 0;</span>
    <span class="c1"># so we continue growing the edges until the next birth/death event *or*</span>
    <span class="c1"># the max number of generations condition is given and met</span>
    <span class="n">gens_to_add</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">while</span> <span class="p">(</span><span class="n">target_num_gens</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">num_gens</span> <span class="o">&lt;</span> <span class="n">target_num_gens</span><span class="p">):</span>
        <span class="n">u</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">u</span> <span class="o">&lt;</span> <span class="p">(</span><span class="n">birth_rate</span> <span class="o">+</span> <span class="n">death_rate</span><span class="p">):</span>
            <span class="k">break</span>
        <span class="n">gens_to_add</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">leaf_nodes</span><span class="p">():</span>
        <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">+=</span> <span class="n">gens_to_add</span>

    <span class="k">if</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;assign_taxa&quot;</span><span class="p">,</span> <span class="kc">True</span><span class="p">):</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">randomly_assign_taxa</span><span class="p">(</span><span class="n">create_required_taxa</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">)</span>

    <span class="c1"># return</span>
    <span class="k">return</span> <span class="n">tree</span></div>


<div class="viewcode-block" id="uniform_pure_birth_tree">
<a class="viewcode-back" href="../../../library/birthdeath.html#dendropy.model.birthdeath.uniform_pure_birth_tree">[docs]</a>
<span class="k">def</span> <span class="nf">uniform_pure_birth_tree</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="p">,</span> <span class="n">birth_rate</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="s2">&quot;Generates a uniform-rate pure-birth process tree. &quot;</span>
    <span class="k">if</span> <span class="n">rng</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">rng</span> <span class="o">=</span> <span class="n">GLOBAL_RNG</span> <span class="c1"># use the global rng by default</span>
    <span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="o">=</span><span class="n">taxon_namespace</span><span class="p">)</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="n">leaf_nodes</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">leaf_nodes</span><span class="p">()</span>
    <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">leaf_nodes</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="p">):</span>
        <span class="n">waiting_time</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">expovariate</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">leaf_nodes</span><span class="p">)</span><span class="o">/</span><span class="n">birth_rate</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">leaf_nodes</span><span class="p">:</span>
            <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">+=</span> <span class="n">waiting_time</span>
        <span class="n">parent_node</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">leaf_nodes</span><span class="p">)</span>
        <span class="n">c1</span> <span class="o">=</span> <span class="n">parent_node</span><span class="o">.</span><span class="n">new_child</span><span class="p">()</span>
        <span class="n">c2</span> <span class="o">=</span> <span class="n">parent_node</span><span class="o">.</span><span class="n">new_child</span><span class="p">()</span>
        <span class="n">c1</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="n">c2</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="n">leaf_nodes</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">leaf_nodes</span><span class="p">()</span>
    <span class="n">leaf_nodes</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">leaf_nodes</span><span class="p">()</span>
    <span class="n">waiting_time</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">expovariate</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">leaf_nodes</span><span class="p">)</span><span class="o">/</span><span class="n">birth_rate</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">leaf_nodes</span><span class="p">:</span>
        <span class="n">nd</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">+=</span> <span class="n">waiting_time</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">leaf_nodes</span><span class="p">):</span>
        <span class="n">leaf</span><span class="o">.</span><span class="n">taxon</span> <span class="o">=</span> <span class="n">taxon_namespace</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">is_rooted</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">return</span> <span class="n">tree</span></div>


<div class="viewcode-block" id="fit_pure_birth_model">
<a class="viewcode-back" href="../../../library/birthdeath.html#dendropy.model.birthdeath.fit_pure_birth_model">[docs]</a>
<span class="k">def</span> <span class="nf">fit_pure_birth_model</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates the maximum-likelihood estimate of the birth rate of a set of *internal* node ages under a Yule (pure-birth) model.</span>

<span class="sd">    Requires either a |Tree| object or an interable of *internal* node ages to be passed in via keyword arguments ``tree`` or ``internal_node_ages``, respectively. The former is more convenient when doing one-off calculations, while the latter is more efficient if the list of internal node ages needs to be used in other places and you already have it calculated and want to avoid re-calculating it here.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    \*\*kwargs : keyword arguments, mandatory</span>

<span class="sd">        Exactly *one* of the following *must* be specified:</span>

<span class="sd">            tree : a |Tree| object.</span>
<span class="sd">                A |Tree| object. The tree needs to be ultrametric for the internal node ages (time from each internal node to the tips) to make sense. The precision by which the ultrametricity is checked can be specified using the ``ultrametricity_precision`` keyword argument (see below). If ``tree`` is given, then ``internal_node_ages`` cannot be given, and vice versa. If ``tree`` is not given, then ``internal_node_ages`` must be given.</span>
<span class="sd">            internal_node_ages : iterable (of numerical values)</span>
<span class="sd">                Iterable of node ages of the internal nodes of a tree, i.e., the list of sum of the edge lengths between each internal node and the tips of the tree. If ``internal_node_ages`` is given, then ``tree`` cannot be given, and vice versa. If ``internal_node_ages`` is not given, then ``tree`` must be given.</span>

<span class="sd">        The following are optional, and are only used if internal node ages are</span>
<span class="sd">        specified (i.e., &#39;internal_node_ages&#39; are passed in):</span>

<span class="sd">            is_node_ages_presorted : bool</span>
<span class="sd">                By default, the vector of node ages are sorted. If this</span>
<span class="sd">                argument is specified as ``True``, then this sorting will be</span>
<span class="sd">                skipped, in which case it is the client code&#39;s responsibility</span>
<span class="sd">                to make sure that the node ages are given in REVERSE order</span>
<span class="sd">                (i.e., oldest nodes -- nodes closer to the root -- given</span>
<span class="sd">                first).</span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    m : dictionary</span>

<span class="sd">    A dictionary with keys being parameter names and values being</span>
<span class="sd">    estimates:</span>

<span class="sd">        &quot;birth_rate&quot;</span>
<span class="sd">            The birth rate.</span>
<span class="sd">        &quot;log_likelihood&quot;</span>
<span class="sd">            The log-likelihood of the model and given birth rate.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>

<span class="sd">    Birth rates can be estimated by passing in trees directly::</span>

<span class="sd">        for idx, tree in enumerate(trees):</span>
<span class="sd">            m = birthdeath.fit_pure_birth_model(tree=tree)</span>
<span class="sd">            print(&quot;Tree {}: birth rate = {} (logL = {})&quot;.format(</span>
<span class="sd">                idx+1, m[&quot;birth_rate&quot;], m[&quot;log_likelihood&quot;]))</span>

<span class="sd">    Or by pre-calculating and passing in a list of node ages::</span>

<span class="sd">        for idx, tree in enumerate(trees):</span>
<span class="sd">            m = birthdeath.fit_pure_birth_model(</span>
<span class="sd">                    internal_node_ages=tree.internal_node_ages())</span>
<span class="sd">            print(&quot;Tree {}: birth rate = {} (logL = {})&quot;.format(</span>
<span class="sd">                idx+1, m[&quot;birth_rate&quot;], m[&quot;log_likelihood&quot;]))</span>


<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    Adapted from the laser package for R:</span>

<span class="sd">        -   Dan Rabosky and Klaus Schliep (2013). laser: Likelihood Analysis of</span>
<span class="sd">            Speciation/Extinction Rates from Phylogenies. R package version</span>
<span class="sd">            2.4-1. http://CRAN.R-project.org/package=laser</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tree</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;tree&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">tree</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">internal_node_ages</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">internal_node_ages</span><span class="p">(</span><span class="n">ultrametricity_precision</span><span class="o">=</span><span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ultrametricity_precision&quot;</span><span class="p">,</span> <span class="mf">1e-6</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">internal_node_ages</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;internal_node_ages&quot;</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;Need to specify &#39;tree&#39; or &#39;internal_node_ages&#39;&quot;</span><span class="p">)</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">internal_node_ages</span>
    <span class="k">if</span> <span class="n">tree</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;is_node_ages_presorted&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">):</span>
        <span class="n">x</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">internal_node_ages</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">st1</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">st2</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">nvec</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">+</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">nv</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">x</span> <span class="k">if</span> <span class="p">(</span><span class="n">i</span> <span class="o">&lt;</span> <span class="n">st1</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">i</span> <span class="o">&gt;=</span> <span class="n">st2</span><span class="p">)]</span>
    <span class="n">lo</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">nvec</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;=</span> <span class="n">st1</span><span class="p">)</span>
    <span class="n">up</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">nvec</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;=</span> <span class="n">st2</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">st1</span> <span class="o">&lt;=</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
        <span class="n">nv</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">st1</span><span class="p">)</span>
        <span class="n">nv</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="n">st2</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">nv</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">nv</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="n">st2</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">nv</span><span class="p">]</span>
    <span class="n">t1</span> <span class="o">=</span> <span class="p">(</span><span class="n">up</span><span class="o">-</span><span class="n">lo</span><span class="p">)</span>
    <span class="n">t2</span> <span class="o">=</span> <span class="p">(</span><span class="n">lo</span><span class="o">*</span><span class="n">nv</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">t3</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span> <span class="n">nv</span><span class="p">[</span><span class="mi">1</span><span class="p">:(</span><span class="n">up</span><span class="o">-</span><span class="n">lo</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span> <span class="p">)</span>
    <span class="n">smax</span> <span class="o">=</span> <span class="n">t1</span><span class="o">/</span><span class="p">(</span><span class="n">t2</span> <span class="o">+</span> <span class="n">t3</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">s1</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="n">lo</span><span class="p">,</span><span class="n">up</span><span class="p">)))</span>
        <span class="n">s2</span> <span class="o">=</span> <span class="p">(</span><span class="n">up</span><span class="o">-</span><span class="n">lo</span><span class="p">)</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">smax</span><span class="p">)</span>
        <span class="n">s3</span> <span class="o">=</span> <span class="n">lo</span> <span class="o">-</span> <span class="n">up</span>
        <span class="n">lh</span> <span class="o">=</span> <span class="n">s1</span> <span class="o">+</span> <span class="n">s2</span> <span class="o">+</span> <span class="n">s3</span>
    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ignore_likelihood_calculation_failure&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Likelihood estimation failure&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">lh</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;-inf&quot;</span><span class="p">)</span>

    <span class="n">result</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;birth_rate&quot;</span> <span class="p">:</span> <span class="n">smax</span><span class="p">,</span>
        <span class="s2">&quot;log_likelihood&quot;</span> <span class="p">:</span> <span class="n">lh</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="fit_pure_birth_model_to_tree">
<a class="viewcode-back" href="../../../library/birthdeath.html#dendropy.model.birthdeath.fit_pure_birth_model_to_tree">[docs]</a>
<span class="k">def</span> <span class="nf">fit_pure_birth_model_to_tree</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">ultrametricity_precision</span><span class="o">=</span><span class="n">constants</span><span class="o">.</span><span class="n">DEFAULT_ULTRAMETRICITY_PRECISION</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates the maximum-likelihood estimate of the birth rate a tree under a</span>
<span class="sd">    Yule (pure-birth) model.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    tree : |Tree| object</span>
<span class="sd">        A tree to be fitted.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    m : dictionary</span>

<span class="sd">    A dictionary with keys being parameter names and values being</span>
<span class="sd">    estimates:</span>

<span class="sd">        -   &quot;birth_rate&quot;</span>
<span class="sd">            The birth rate.</span>
<span class="sd">        -   &quot;log_likelihood&quot;</span>
<span class="sd">            The log-likelihood of the model and given birth rate.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>

<span class="sd">    ::</span>

<span class="sd">        import dendropy</span>
<span class="sd">        from dendropy.model import birthdeath</span>
<span class="sd">        trees = dendropy.TreeList.get_from_path(</span>
<span class="sd">                &quot;pythonidae.nex&quot;, &quot;nexus&quot;)</span>
<span class="sd">        for idx, tree in enumerate(trees):</span>
<span class="sd">            m = birthdeath.fit_pure_birth_model_to_tree(tree)</span>
<span class="sd">            print(&quot;Tree {}: birth rate = {} (logL = {})&quot;.format(</span>
<span class="sd">                idx+1, m[&quot;birth_rate&quot;], m[&quot;log_likelihood&quot;]))</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">fit_pure_birth_model</span><span class="p">(</span><span class="n">tree</span><span class="o">=</span><span class="n">tree</span><span class="p">,</span> <span class="n">ultrametricity_precision</span><span class="o">=</span><span class="n">ultrametricity_precision</span><span class="p">)</span></div>



<div class="viewcode-block" id="birth_death_likelihood">
<a class="viewcode-back" href="../../../library/birthdeath.html#dendropy.model.birthdeath.birth_death_likelihood">[docs]</a>
<span class="k">def</span> <span class="nf">birth_death_likelihood</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates the log-likelihood of a tree (or a set of internal nodes) under</span>
<span class="sd">    a birth death model.</span>

<span class="sd">    Requires either a |Tree| object or an interable of *internal* node</span>
<span class="sd">    ages to be passed in via keyword arguments ``tree`` or ``internal_node_ages``,</span>
<span class="sd">    respectively. The former is more convenient when doing one-off</span>
<span class="sd">    calculations, while the latter is more efficient if the list of internal</span>
<span class="sd">    node ages needs to be used in other places and you already have it</span>
<span class="sd">    calculated and want to avoid re-calculating it here.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    \*\*kwargs : keyword arguments, mandatory</span>

<span class="sd">        Exactly *one* of the following *must* be specified:</span>

<span class="sd">            tree : a |Tree| object.</span>
<span class="sd">                A |Tree| object. The tree needs to be ultrametric for the</span>
<span class="sd">                internal node ages (time from each internal node to the tips)</span>
<span class="sd">                to make sense. The precision by which the ultrametricity is</span>
<span class="sd">                checked can be specified using the ``ultrametricity_precision`` keyword</span>
<span class="sd">                argument (see below). If ``tree`` is given, then</span>
<span class="sd">                ``internal_node_ages`` cannot be given, and vice versa. If ``tree``</span>
<span class="sd">                is not given, then ``internal_node_ages`` must be given.</span>
<span class="sd">            internal_node_ages : iterable (of numerical values)</span>
<span class="sd">                Iterable of node ages of the internal nodes of a tree, i.e., the</span>
<span class="sd">                list of sum of the edge lengths between each internal node and</span>
<span class="sd">                the tips of the tree. If ``internal_node_ages`` is given, then</span>
<span class="sd">                ``tree`` cannot be given, and vice versa. If ``internal_node_ages``</span>
<span class="sd">                is not given, then ``tree`` must be given.</span>

<span class="sd">        The following keyword parameters are mandatory:</span>

<span class="sd">            birth_rate : float</span>
<span class="sd">                The birth rate.</span>
<span class="sd">            death_rate : float</span>
<span class="sd">                The death rate.</span>

<span class="sd">        The following keyword parameters are optional:</span>

<span class="sd">            sampling_probability</span>
<span class="sd">                The probability for a species to be included in the sample. Defaults to 1.0 (all species sampled).</span>
<span class="sd">            sampling_strategy</span>
<span class="sd">                The strategy how samples were obtained. Options are: uniform|diversified|age.</span>
<span class="sd">            is_mrca_included</span>
<span class="sd">                Does the process start with the most recent common ancestor?</span>
<span class="sd">            condition_on : string</span>
<span class="sd">                Do we condition the process on: &quot;time&quot;, &quot;survival&quot;, or &quot;taxa&quot;?</span>

<span class="sd">        The following are optional, and are only used if internal node ages</span>
<span class="sd">        need to be calculated (i.e., &#39;tree&#39; is passed in).</span>

<span class="sd">            ultrametricity_precision : float</span>
<span class="sd">                When calculating the node ages, an error will be raised if the tree in</span>
<span class="sd">                o ultrametric. This error may be due to floating-point or numerical</span>
<span class="sd">                imprecision. You can set the precision of the ultrametricity validation</span>
<span class="sd">                by setting the ``ultrametricity_precision`` parameter. E.g., use</span>
<span class="sd">                ``ultrametricity_precision=0.01`` for a more relaxed precision,</span>
<span class="sd">                down to 2 decimal places. Use ``ultrametricity_precision=False``</span>
<span class="sd">                to disable checking of ultrametricity precision.</span>

<span class="sd">            ignore_likelihood_calculation_failure: bool (default: False)</span>
<span class="sd">                In some cases (typically, abnormal trees, e.g., 1-tip), the</span>
<span class="sd">                likelihood estimation will fail. In this case a ValueError will</span>
<span class="sd">                be raised. If ``ignore_likelihood_calculation_failure`` is</span>
<span class="sd">                |True|, then the function call will still succeed, with the</span>
<span class="sd">                likelihood set to -``inf``.</span>

<span class="sd">        The following are optional, and are only used if internal node ages are</span>
<span class="sd">        specified (i.e., &#39;internal_node_ages&#39; are passed in):</span>

<span class="sd">            is_node_ages_presorted : bool</span>
<span class="sd">                By default, the vector of node ages are sorted. If this</span>
<span class="sd">                argument is specified as ``True``, then this sorting will be</span>
<span class="sd">                skipped, in which case it is the client code&#39;s responsibility</span>
<span class="sd">                to make sure that the node ages are given in REVERSE order</span>
<span class="sd">                (i.e., oldest nodes -- nodes closer to the root -- given</span>
<span class="sd">                first).</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    Lifted directly from the (fantastic!) TESS package for R:</span>

<span class="sd">        H{&quot;o}hna S. 2013. Fast simulation of reconstructed phylogenies under</span>
<span class="sd">        global time-dependent birth--death processes. Bioinformatics, 29(11)</span>
<span class="sd">        1367-1374.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    lnl : float</span>

<span class="sd">    The log-likehood of the tree under the birth-death model.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tree</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;tree&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">tree</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">internal_node_ages</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">internal_node_ages</span><span class="p">(</span><span class="n">ultrametricity_precision</span><span class="o">=</span><span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ultrametricity_precision&quot;</span><span class="p">,</span> <span class="mf">1e-6</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">internal_node_ages</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;internal_node_ages&quot;</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;Need to specify &#39;tree&#39; or &#39;internal_node_ages&#39;&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">tree</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;is_node_ages_presorted&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">):</span>
        <span class="n">internal_node_ages</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">internal_node_ages</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># check for sensible parameter values</span>
    <span class="c1"># if ( lambda &lt;= 0 || mu &lt; 0 || sampling_probability &lt;= 0 || sampling_probability &gt; 1.0) {</span>
    <span class="c1">#     stop(&quot;Invalid parameter values for lambda and mu!&quot;)</span>
    <span class="c1"># }</span>

    <span class="c1"># massExtinctionTimes=c(),</span>
    <span class="c1"># massExtinctionSurvivalProbabilities=c(),</span>
    <span class="c1"># missingSpecies = c(),</span>
    <span class="c1"># timesMissingSpecies = c(),</span>
    <span class="c1"># tess.likelihood 11</span>
    <span class="c1"># samplingProbability=1.0,</span>
    <span class="c1"># samplingStrategy=&quot;uniform&quot;,</span>
    <span class="c1"># MRCA=TRUE,</span>
    <span class="c1"># CONDITION=&quot;survival&quot;,</span>
    <span class="n">birth_rate</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;birth_rate&quot;</span><span class="p">)</span>
    <span class="n">death_rate</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;death_rate&quot;</span><span class="p">)</span>
    <span class="n">massExtinctionTimes</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">massExtinctionSurvivalProbabilities</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">sampling_probability</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;sampling_probability&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
    <span class="n">sampling_strategy</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;sampling_strategy&quot;</span><span class="p">,</span> <span class="s2">&quot;uniform&quot;</span><span class="p">)</span>
    <span class="n">is_mrca_included</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;is_mrca_included&quot;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">condition_on</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;condition_on&quot;</span><span class="p">,</span> <span class="s2">&quot;survival&quot;</span><span class="p">)</span>

    <span class="n">ntax</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">internal_node_ages</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">PRESENT</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">internal_node_ages</span><span class="p">)</span>
    <span class="n">times</span> <span class="o">=</span> <span class="p">[(</span><span class="n">PRESENT</span> <span class="o">-</span> <span class="n">t</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">internal_node_ages</span><span class="p">]</span>

    <span class="c1"># if we condition on the MRCA, then we need to remove the root speciation event</span>
    <span class="k">if</span> <span class="n">is_mrca_included</span><span class="p">:</span>
        <span class="n">times</span> <span class="o">=</span> <span class="n">times</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>

    <span class="c1"># set the uniform taxon sampling probability</span>
    <span class="k">if</span> <span class="n">sampling_strategy</span> <span class="o">==</span> <span class="s2">&quot;uniform&quot;</span><span class="p">:</span>
        <span class="n">rho</span> <span class="o">=</span> <span class="n">sampling_probability</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">rho</span> <span class="o">=</span> <span class="mf">1.0</span>

    <span class="c1"># initialize the log likelihood</span>
    <span class="n">lnl</span> <span class="o">=</span> <span class="mf">0.0</span>

    <span class="c1"># what do we condition on?</span>
    <span class="c1"># did we condition on survival?</span>
    <span class="k">if</span> <span class="n">condition_on</span> <span class="o">==</span> <span class="s2">&quot;survival&quot;</span><span class="p">:</span>
        <span class="n">lnl</span> <span class="o">=</span> <span class="o">-</span> <span class="n">_p_survival_constant</span><span class="p">(</span><span class="n">birth_rate</span><span class="p">,</span><span class="n">death_rate</span><span class="p">,</span><span class="n">massExtinctionTimes</span><span class="p">,</span><span class="n">massExtinctionSurvivalProbabilities</span><span class="p">,</span><span class="n">rho</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="n">PRESENT</span><span class="p">,</span><span class="n">PRESENT</span><span class="p">)</span>

    <span class="c1"># death_rateltiply the probability of a descendant of the initial species</span>
    <span class="n">lnl</span> <span class="o">=</span> <span class="n">lnl</span> <span class="o">+</span> <span class="n">_p1_constant</span><span class="p">(</span>
            <span class="n">birth_rate</span><span class="o">=</span><span class="n">birth_rate</span><span class="p">,</span>
            <span class="n">death_rate</span><span class="o">=</span><span class="n">death_rate</span><span class="p">,</span>
            <span class="n">massExtinctionTimes</span><span class="o">=</span><span class="n">massExtinctionTimes</span><span class="p">,</span>
            <span class="n">massExtinctionSurvivalProbabilities</span><span class="o">=</span><span class="n">massExtinctionSurvivalProbabilities</span><span class="p">,</span>
            <span class="n">sampling_probability</span><span class="o">=</span><span class="n">rho</span><span class="p">,</span>
            <span class="n">t</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">T</span><span class="o">=</span><span class="n">PRESENT</span><span class="p">)</span>

    <span class="c1"># add the survival of a second species if we condition on the MRCA</span>
    <span class="k">if</span> <span class="n">is_mrca_included</span><span class="p">:</span>
        <span class="n">lnl</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">lnl</span>

    <span class="c1"># did we condition on observing n species today</span>
    <span class="k">if</span> <span class="n">condition_on</span> <span class="o">==</span> <span class="s2">&quot;taxa&quot;</span><span class="p">:</span>
        <span class="n">lnl</span> <span class="o">=</span> <span class="n">lnl</span> <span class="o">-</span> <span class="n">_p_N_constant</span><span class="p">(</span>
                <span class="n">birth_rate</span><span class="o">=</span><span class="n">birth_rate</span><span class="p">,</span>
                <span class="n">death_rate</span><span class="o">=</span><span class="n">death_rate</span><span class="p">,</span>
                <span class="n">massExtinctionTimes</span><span class="o">=</span><span class="n">massExtinctionTimes</span><span class="p">,</span>
                <span class="n">massExtinctionSurvivalProbabilities</span><span class="o">=</span><span class="n">massExtinctionSurvivalProbabilities</span><span class="p">,</span>
                <span class="n">sampling_probability</span><span class="o">=</span><span class="n">sampling_probability</span><span class="p">,</span>
                <span class="n">i</span><span class="o">=</span><span class="n">ntax</span><span class="p">,</span>
                <span class="n">s</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                <span class="n">t</span><span class="o">=</span><span class="n">PRESENT</span><span class="p">,</span>
                <span class="n">SURVIVAL</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">MRCA</span><span class="o">=</span><span class="n">is_mrca_included</span><span class="p">)</span>

    <span class="c1"># if we assume diversified sampling, we need to death_rateltiply with the</span>
    <span class="c1"># probability that all missing species happened after the last speciation</span>
    <span class="c1"># event</span>
    <span class="k">if</span> <span class="n">sampling_strategy</span> <span class="o">==</span> <span class="s2">&quot;diversified&quot;</span><span class="p">:</span>
        <span class="c1"># We use equation (5) of Hoehna et al. &quot;Inferring Speciation and Extinction Rates under Different Sampling Schemes&quot;</span>
        <span class="n">lastEvent</span> <span class="o">=</span> <span class="n">times</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">p_0_T</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">_p_survival_constant</span><span class="p">(</span><span class="n">birth_rate</span><span class="p">,</span><span class="n">death_rate</span><span class="p">,</span><span class="n">massExtinctionTimes</span><span class="p">,</span><span class="n">massExtinctionSurvivalProbabilities</span><span class="p">,</span><span class="mf">1.0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="n">PRESENT</span><span class="p">,</span><span class="n">PRESENT</span><span class="p">))</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">((</span><span class="n">death_rate</span><span class="o">-</span><span class="n">birth_rate</span><span class="p">)</span><span class="o">*</span><span class="n">PRESENT</span><span class="p">)</span>
        <span class="n">p_0_t</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">_p_survival_constant</span><span class="p">(</span><span class="n">birth_rate</span><span class="p">,</span><span class="n">death_rate</span><span class="p">,</span><span class="n">massExtinctionTimes</span><span class="p">,</span><span class="n">massExtinctionSurvivalProbabilities</span><span class="p">,</span><span class="mf">1.0</span><span class="p">,</span><span class="n">lastEvent</span><span class="p">,</span><span class="n">PRESENT</span><span class="p">,</span><span class="n">PRESENT</span><span class="p">))</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">((</span><span class="n">death_rate</span><span class="o">-</span><span class="n">birth_rate</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="n">PRESENT</span><span class="o">-</span><span class="n">lastEvent</span><span class="p">))</span>
        <span class="n">F_t</span> <span class="o">=</span> <span class="n">p_0_t</span> <span class="o">/</span> <span class="n">p_0_T</span>
        <span class="c1"># get an estimate of the actual number of taxa</span>
        <span class="n">m</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">ntax</span><span class="p">)</span> <span class="o">/</span> <span class="n">sampling_probability</span><span class="p">)</span>
        <span class="c1"># remove the number of species that we started with</span>
        <span class="k">if</span> <span class="n">is_mrca_included</span><span class="p">:</span>
            <span class="n">k</span> <span class="o">=</span> <span class="mi">2</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">k</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">lnl</span> <span class="o">=</span> <span class="n">lnl</span> <span class="o">+</span> <span class="p">(</span><span class="n">m</span><span class="o">-</span><span class="n">ntax</span><span class="p">)</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">F_t</span><span class="p">)</span> <span class="o">+</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">combinatorics</span><span class="o">.</span><span class="n">choose</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">-</span><span class="n">k</span><span class="p">),</span><span class="nb">int</span><span class="p">(</span><span class="n">ntax</span><span class="o">-</span><span class="n">k</span><span class="p">)))</span>

    <span class="c1"># multiply the probability for the missing species</span>
    <span class="c1"># if len(missing_species) &gt; 0:</span>
    <span class="c1">#     # We use equation (5) of Hoehna et al. &quot;Inferring Speciation and Extinction Rates under Different Sampling Schemes&quot;</span>
    <span class="c1">#     # now iterate over the vector of missing species per interval</span>
    <span class="c1">#     lastEvent = timesMissingSpecies</span>
    <span class="c1">#     p_0_T = 1.0 - math.exp( _p_survival_constant(birth_rate,death_rate,massExtinctionTimes,massExtinctionSurvivalProbabilities,1.0,0,PRESENT,PRESENT,log=TRUE) + ((death_rate-birth_rate)*PRESENT) )</span>
    <span class="c1">#     p_0_t = 1.0 - math.exp( _p_survival_constant(birth_rate,death_rate,massExtinctionTimes,massExtinctionSurvivalProbabilities,1.0,lastEvent,PRESENT,PRESENT,log=TRUE) + ((death_rate-birth_rate)*(PRESENT-lastEvent)) )</span>
    <span class="c1">#     log_F_t = math.log(p_0_t) - math.log(p_0_T)</span>
    <span class="c1">#     # get an estimate of the actual number of taxa</span>
    <span class="c1">#     m = missingSpecies</span>
    <span class="c1">#     # remove the number of species that we started with</span>
    <span class="c1">#     lnl = lnl + sum( m * log_F_t ) #+ lchoose(m-k,ntax-k)</span>


    <span class="c1"># multiply the probability for each speciation time</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">times</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># lnl = lnl + len(times)*math.log(birth_rate) + sum(tess.equations.p1.constant(birth_rate,death_rate,massExtinctionTimes,massExtinctionSurvivalProbabilities,rho,times,PRESENT,log=TRUE))</span>
        <span class="n">lnl</span> <span class="o">+=</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">times</span><span class="p">)</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">birth_rate</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">time</span> <span class="ow">in</span> <span class="n">times</span><span class="p">:</span>
            <span class="n">lnl</span> <span class="o">+=</span> <span class="n">_p1_constant</span><span class="p">(</span>
                    <span class="n">birth_rate</span><span class="o">=</span><span class="n">birth_rate</span><span class="p">,</span>
                    <span class="n">death_rate</span><span class="o">=</span><span class="n">death_rate</span><span class="p">,</span>
                    <span class="n">massExtinctionTimes</span><span class="o">=</span><span class="n">massExtinctionTimes</span><span class="p">,</span>
                    <span class="n">massExtinctionSurvivalProbabilities</span><span class="o">=</span><span class="n">massExtinctionSurvivalProbabilities</span><span class="p">,</span>
                    <span class="n">sampling_probability</span><span class="o">=</span><span class="n">rho</span><span class="p">,</span>
                    <span class="n">t</span><span class="o">=</span><span class="n">time</span><span class="p">,</span>
                    <span class="n">T</span><span class="o">=</span><span class="n">PRESENT</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">lnl</span></div>




<span class="c1">################################################################################</span>
<span class="c1"># From: TESS</span>
<span class="c1">#</span>
<span class="c1">#   https://github.com/hoehna/TESS</span>
<span class="c1">#</span>
<span class="c1">#   H{&quot;o}hna S. 2013. Fast simulation of reconstructed phylogenies under</span>
<span class="c1">#   global time-dependent birth--death processes. Bioinformatics, 29(11)</span>
<span class="c1">#   1367-1374.</span>
<span class="c1">#</span>
<span class="c1">#   Copyright (c) 2012- Sebastian Hoehna</span>
<span class="c1">#</span>
<span class="c1">#   TESS is free software; you can redistribute it and/or modify</span>
<span class="c1">#   it under the terms of the GNU Lesser General Public License as</span>
<span class="c1">#   published by the Free Software Foundation; either version 2</span>
<span class="c1">#   of the License, or (at your option) any later version.</span>
<span class="c1">#</span>
<span class="c1"># @brief  Calculate the probability of n lineage existing at time t</span>
<span class="c1">#         if we started with 1 (or 2) lineage at time s.</span>
<span class="c1">#</span>
<span class="c1">#</span>
<span class="c1"># @see    Equation (3), (5) and (12) in Hoehna, S.: Fast simulation of reconstructed phylogenies</span>
<span class="c1">#         under global, time-dependent birth-death processes. 2013, Bioinformatics</span>
<span class="c1">#</span>
<span class="c1"># @date Last modified: 2013-01-30</span>
<span class="c1"># @author Sebastian Hoehna</span>
<span class="c1"># @version 1.1</span>
<span class="c1"># @since 2012-09-11, version 1.0</span>
<span class="c1">#</span>
<span class="c1"># @param    lambda                                        scalar        speciation rate</span>
<span class="c1"># @param    mu                                            scalar        extinction rate</span>
<span class="c1"># @param    massExtinctionTimes                           vector        timse at which mass-extinctions happen</span>
<span class="c1"># @param    massExtinctionSurvivalProbabilities           vector        survival probability of a mass extinction event</span>
<span class="c1"># @param    samplingProbability                           scalar        probability of uniform sampling at present</span>
<span class="c1"># @param    i                                             scalar        number of lineages</span>
<span class="c1"># @param    s                                             scalar        start time</span>
<span class="c1"># @param    t                                             scalar        stop time</span>
<span class="c1"># @param    MRCA                                          boolean       does the tree start at the mrca?</span>
<span class="c1"># @param    log                                           bool          if in log-scale</span>
<span class="c1"># @return                                                 scalar        log-probability</span>
<span class="c1">#</span>
<span class="c1">################################################################################</span>
<span class="c1"># tess.equations.pN.constant = function(lambda,mu,massExtinctionTimes,massExtinctionSurvivalProbabilities,samplingProbability,i,s,t,SURVIVAL=FALSE,MRCA=FALSE,log=FALSE) {</span>
<span class="k">def</span> <span class="nf">_p_N_constant</span><span class="p">(</span>
        <span class="n">birth_rate</span><span class="p">,</span>
        <span class="n">death_rate</span><span class="p">,</span>
        <span class="n">massExtinctionTimes</span><span class="p">,</span>
        <span class="n">massExtinctionSurvivalProbabilities</span><span class="p">,</span>
        <span class="n">sampling_probability</span><span class="p">,</span>
        <span class="n">i</span><span class="p">,</span>
        <span class="n">s</span><span class="p">,</span>
        <span class="n">t</span><span class="p">,</span>
        <span class="n">SURVIVAL</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">MRCA</span><span class="o">=</span><span class="kc">False</span><span class="p">,):</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span> <span class="c1"># we assume conditioning on survival</span>
        <span class="n">p</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">elif</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">MRCA</span><span class="p">:</span> <span class="c1"># we assume conditioning on survival of the two species</span>
            <span class="n">p</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">SURVIVAL</span><span class="p">:</span>
                <span class="n">p</span> <span class="o">=</span>  <span class="n">_p_survival_constant</span><span class="p">(</span><span class="n">birth_rate</span><span class="p">,</span><span class="n">death_rate</span><span class="p">,</span><span class="n">massExtinctionTimes</span><span class="p">,</span><span class="n">massExtinctionSurvivalProbabilities</span><span class="p">,</span><span class="n">sampling_probability</span><span class="p">,</span><span class="n">s</span><span class="p">,</span><span class="n">t</span><span class="p">,</span><span class="n">t</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">death_rate</span><span class="o">-</span><span class="n">birth_rate</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="n">t</span><span class="o">-</span><span class="n">s</span><span class="p">)</span> <span class="o">-</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">sampling_probability</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">p</span> <span class="o">=</span>  <span class="mi">2</span><span class="o">*</span><span class="n">_p_survival_constant</span><span class="p">(</span><span class="n">birth_rate</span><span class="p">,</span><span class="n">death_rate</span><span class="p">,</span><span class="n">massExtinctionTimes</span><span class="p">,</span><span class="n">massExtinctionSurvivalProbabilities</span><span class="p">,</span><span class="n">sampling_probability</span><span class="p">,</span><span class="n">s</span><span class="p">,</span><span class="n">t</span><span class="p">,</span><span class="n">t</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">death_rate</span><span class="o">-</span><span class="n">birth_rate</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="n">t</span><span class="o">-</span><span class="n">s</span><span class="p">)</span> <span class="o">-</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">sampling_probability</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">p_s</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">_p_survival_constant</span><span class="p">(</span><span class="n">birth_rate</span><span class="p">,</span><span class="n">death_rate</span><span class="p">,</span><span class="n">massExtinctionTimes</span><span class="p">,</span><span class="n">massExtinctionSurvivalProbabilities</span><span class="p">,</span><span class="n">sampling_probability</span><span class="p">,</span><span class="n">s</span><span class="p">,</span><span class="n">t</span><span class="p">,</span><span class="n">t</span><span class="p">))</span>
        <span class="n">r</span>   <span class="o">=</span> <span class="p">(</span><span class="n">death_rate</span><span class="o">-</span><span class="n">birth_rate</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="n">t</span><span class="o">-</span><span class="n">s</span><span class="p">)</span> <span class="o">-</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">sampling_probability</span><span class="p">)</span>
        <span class="c1"># for (j in seq_len(length(massExtinctionTimes)) ) {</span>
        <span class="c1">#     cond =  (s &lt; massExtinctionTimes[j]) &amp; (t &gt;= massExtinctionTimes[j])</span>
        <span class="c1">#     r  = r - ifelse(cond, log(massExtinctionSurvivalProbabilities[j]), 0.0)</span>
        <span class="n">e</span> <span class="o">=</span> <span class="n">p_s</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">e</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">e</span> <span class="o">=</span> <span class="mf">1.0</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">MRCA</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">SURVIVAL</span><span class="p">:</span>
                <span class="n">p</span>   <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">p_s</span><span class="p">)</span> <span class="o">+</span> <span class="n">r</span> <span class="o">+</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">e</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">p</span>   <span class="o">=</span> <span class="mi">2</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">p_s</span><span class="p">)</span> <span class="o">+</span> <span class="n">r</span> <span class="o">+</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">e</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">SURVIVAL</span><span class="p">:</span>
                <span class="n">p</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">p_s</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="o">*</span><span class="n">r</span> <span class="o">+</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">e</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">i</span><span class="o">-</span><span class="mi">2</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">p</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="mi">4</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">p_s</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="o">*</span><span class="n">r</span> <span class="o">+</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">e</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">i</span><span class="o">-</span><span class="mi">2</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">p</span>

<span class="c1">################################################################################</span>
<span class="c1"># From: TESS</span>
<span class="c1">#</span>
<span class="c1">#   https://github.com/hoehna/TESS</span>
<span class="c1">#</span>
<span class="c1">#   H{&quot;o}hna S. 2013. Fast simulation of reconstructed phylogenies under</span>
<span class="c1">#   global time-dependent birth--death processes. Bioinformatics, 29(11)</span>
<span class="c1">#   1367-1374.</span>
<span class="c1">#</span>
<span class="c1">#   Copyright (c) 2012- Sebastian Hoehna</span>
<span class="c1">#</span>
<span class="c1">#   TESS is free software; you can redistribute it and/or modify</span>
<span class="c1">#   it under the terms of the GNU Lesser General Public License as</span>
<span class="c1">#   published by the Free Software Foundation; either version 2</span>
<span class="c1">#   of the License, or (at your option) any later version.</span>
<span class="c1">#</span>
<span class="c1"># @brief  Calculate the probability of survival in the interval [t_low,t_high].</span>
<span class="c1">#</span>
<span class="c1"># See Equation (2) from Hoehna, S., 2013, Fast simulation of reconstructed phylogenies under global, time-dependent birth-death processes</span>
<span class="c1">#</span>
<span class="c1">#</span>
<span class="c1"># @date Last modified: 2013-01-30</span>
<span class="c1"># @author Sebastian Hoehna</span>
<span class="c1"># @version 1.1</span>
<span class="c1"># @since 2012-09-11, version 1.0</span>
<span class="c1">#</span>
<span class="c1"># @param    lambda                                        scalar        speciation rate</span>
<span class="c1"># @param    mu                                            scalar        extinction rate</span>
<span class="c1"># @param    massExtinctionTimes                           vector        timse at which mass-extinctions happen</span>
<span class="c1"># @param    massExtinctionSurvivalProbabilities           vector        survival probability of a mass extinction event</span>
<span class="c1"># @param    samplingProbability                           scalar        probability of uniform sampling at present</span>
<span class="c1"># @param    t_low                                         scalar        starting time</span>
<span class="c1"># @param    t_high                                        scalar        end time</span>
<span class="c1"># @param    T                                             scalar        present time (time goes forward and the origin/MRCA might be 0)</span>
<span class="c1"># @param    log                                           bool          if in log-scale</span>
<span class="c1"># @return                                                 scalar        probability of survival in [t,tau]</span>
<span class="c1">#</span>
<span class="c1">################################################################################</span>
<span class="c1"># tess.equations.pSurvival.constant = function(lambda,mu,massExtinctionTimes,massExtinctionSurvivalProbabilities,samplingProbability,t_low,t_high,T,log=FALSE) {</span>
<span class="k">def</span> <span class="nf">_p_survival_constant</span><span class="p">(</span>
        <span class="n">birth_rate</span><span class="p">,</span>
        <span class="n">death_rate</span><span class="p">,</span>
        <span class="n">massExtinctionTimes</span><span class="p">,</span>
        <span class="n">massExtinctionSurvivalProbabilities</span><span class="p">,</span>
        <span class="n">sampling_probability</span><span class="p">,</span>
        <span class="n">t_low</span><span class="p">,</span>
        <span class="n">t_high</span><span class="p">,</span>
        <span class="n">T</span><span class="p">):</span>

    <span class="c1"># compute the rate</span>
    <span class="n">rate</span> <span class="o">=</span> <span class="n">death_rate</span> <span class="o">-</span> <span class="n">birth_rate</span>

    <span class="c1"># do the integration of int_{t_low}^{t_high} ( death_rate(s) exp(rate(t,s)) ds )</span>
    <span class="c1"># where rate(t,s) = int_{t}^{s} ( death_rate(x)-birth_rate(x) dx ) - sum_{for all t &lt; m_i &lt; s in massExtinctionTimes }( log(massExtinctionSurvivalProbability[i]) )</span>

    <span class="c1"># we compute the integral stepwise for each epoch between mass-extinction events</span>
    <span class="c1"># add mass-extinction</span>
    <span class="n">accudeath_ratelatedMassExtinction</span> <span class="o">=</span> <span class="mf">1.0</span>
    <span class="n">prev_time</span> <span class="o">=</span> <span class="n">t_low</span>
    <span class="n">den</span> <span class="o">=</span> <span class="mf">1.0</span>
    <span class="c1"># if ( length(massExtinctionTimes) &gt; 0 ) {</span>
    <span class="c1">#     for (j in 1:length(massExtinctionTimes) ) {</span>
    <span class="c1">#         cond =  (t_low &lt; massExtinctionTimes[j]) &amp; (t_high &gt;= massExtinctionTimes[j])</span>
    <span class="c1">#         # compute the integral for this time episode until the mass-extinction event</span>
    <span class="c1">#     #       den = den + ifelse(cond, exp(-rate*t_low) * death_rate / (rate * accudeath_ratelatedMassExtinction ) * ( exp(rate* massExtinctionTimes[j]) - exp(rate*prev_time)) , 0 )</span>
    <span class="c1">#         den = den + cond * exp(-rate*t_low) * death_rate / (rate * accudeath_ratelatedMassExtinction ) * ( exp(rate* massExtinctionTimes[j]) - exp(rate*prev_time))</span>
    <span class="c1">#         # store the current time so that we remember from which episode we need to integrate next</span>
    <span class="c1">#     #       prev_time = ifelse(cond, massExtinctionTimes[j], prev_time)</span>
    <span class="c1">#         prev_time[cond] = massExtinctionTimes[j]</span>
    <span class="c1">#         accudeath_ratelatedMassExtinction = accudeath_ratelatedMassExtinction * ifelse(cond, massExtinctionSurvivalProbabilities[j], 1.0)</span>
    <span class="c1">#         # integrate over the tiny time interval of the mass-extinction event itself and add it to the integral</span>
    <span class="c1">#     #       den = den - ifelse(cond, (massExtinctionSurvivalProbabilities[j]-1) / accudeath_ratelatedMassExtinction * exp( rate*(massExtinctionTimes[j] - t_low) ), 0.0 )</span>
    <span class="c1">#         den = den - cond * (massExtinctionSurvivalProbabilities[j]-1) / accudeath_ratelatedMassExtinction * exp( rate*(massExtinctionTimes[j] - t_low) )</span>
    <span class="c1">#     }</span>
    <span class="c1"># }</span>

    <span class="c1"># add the integral of the final epoch until the present time</span>
    <span class="n">den</span> <span class="o">=</span> <span class="n">den</span> <span class="o">+</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">rate</span><span class="o">*</span><span class="n">t_low</span><span class="p">)</span> <span class="o">*</span> <span class="n">death_rate</span> <span class="o">/</span> <span class="p">(</span><span class="n">rate</span> <span class="o">*</span> <span class="n">accudeath_ratelatedMassExtinction</span> <span class="p">)</span> <span class="o">*</span> <span class="p">(</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">rate</span><span class="o">*</span><span class="n">t_high</span><span class="p">)</span> <span class="o">-</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">rate</span><span class="o">*</span><span class="n">prev_time</span><span class="p">))</span>

    <span class="c1"># add sampling</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">sampling_probability</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">t_low</span> <span class="o">&lt;</span> <span class="n">T</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">t_high</span> <span class="o">&gt;=</span> <span class="n">T</span><span class="p">):</span>
        <span class="n">accudeath_ratelatedMassExtinction</span> <span class="o">=</span> <span class="n">accudeath_ratelatedMassExtinction</span> <span class="o">*</span>  <span class="n">sampling_probability</span>
        <span class="n">cond</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">cond</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1">#   den = den - ifelse(cond, (sampling_probability-1)*math.exp( rate*(T-t_low) ) / accudeath_ratelatedMassExtinction, 0.0)</span>
    <span class="n">den</span> <span class="o">=</span> <span class="n">den</span> <span class="o">-</span> <span class="n">cond</span> <span class="o">*</span> <span class="p">(</span><span class="n">sampling_probability</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span> <span class="n">rate</span><span class="o">*</span><span class="p">(</span><span class="n">T</span><span class="o">-</span><span class="n">t_low</span><span class="p">)</span> <span class="p">)</span> <span class="o">/</span> <span class="n">accudeath_ratelatedMassExtinction</span>

    <span class="n">res</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="n">den</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">res</span><span class="p">)</span>


<span class="c1">################################################################################</span>
<span class="c1"># From: TESS</span>
<span class="c1">#</span>
<span class="c1">#   https://github.com/hoehna/TESS</span>
<span class="c1">#</span>
<span class="c1">#   H{&quot;o}hna S. 2013. Fast simulation of reconstructed phylogenies under</span>
<span class="c1">#   global time-dependent birth--death processes. Bioinformatics, 29(11)</span>
<span class="c1">#   1367-1374.</span>
<span class="c1">#</span>
<span class="c1">#   Copyright (c) 2012- Sebastian Hoehna</span>
<span class="c1">#</span>
<span class="c1">#   TESS is free software; you can redistribute it and/or modify</span>
<span class="c1">#   it under the terms of the GNU Lesser General Public License as</span>
<span class="c1">#   published by the Free Software Foundation; either version 2</span>
<span class="c1">#   of the License, or (at your option) any later version.</span>
<span class="c1">#</span>
<span class="c1"># @brief  Calculate the probability of exactly 1 lineage surviving until time T</span>
<span class="c1">#         if we started with 1 lineage at time t.</span>
<span class="c1">#</span>
<span class="c1"># @see    Equation (4) in Hoehna, S.: Fast simulation of reconstructed phylogenies</span>
<span class="c1">#         under global, time-dependent birth-death processes. 2013, Bioinformatics</span>
<span class="c1">#</span>
<span class="c1"># @date Last modified: 2013-01-30</span>
<span class="c1"># @author Sebastian Hoehna</span>
<span class="c1"># @version 1.1</span>
<span class="c1"># @since 2012-09-11, version 1.0</span>
<span class="c1">#</span>
<span class="c1"># @param    lambda                                        scalar        speciation rate</span>
<span class="c1"># @param    mu                                            scalar        extinction rate</span>
<span class="c1"># @param    massExtinctionTimes                           vector        timse at which mass-extinctions happen</span>
<span class="c1"># @param    massExtinctionSurvivalProbabilities           vector        survival probability of a mass extinction event</span>
<span class="c1"># @param    samplingProbability                           scalar        probability of uniform sampling at present</span>
<span class="c1"># @param    t                                             scalar        time</span>
<span class="c1"># @param    T                                             scalar        present time</span>
<span class="c1"># @param    log                                           bool          if in log-scale</span>
<span class="c1"># @return                                                 scalar        ln-probability</span>
<span class="c1">#</span>
<span class="c1">################################################################################</span>
<span class="c1"># tess.equations.p1.constant = function(lambda,mu,massExtinctionTimes,massExtinctionSurvivalProbabilities,samplingProbability,t,T,log=FALSE) {</span>
<span class="k">def</span> <span class="nf">_p1_constant</span><span class="p">(</span>
        <span class="n">birth_rate</span><span class="p">,</span>
        <span class="n">death_rate</span><span class="p">,</span>
        <span class="n">massExtinctionTimes</span><span class="p">,</span>
        <span class="n">massExtinctionSurvivalProbabilities</span><span class="p">,</span>
        <span class="n">sampling_probability</span><span class="p">,</span>
        <span class="n">t</span><span class="p">,</span>
        <span class="n">T</span><span class="p">):</span>
    <span class="c1"># compute the survival probability</span>
    <span class="n">a</span> <span class="o">=</span> <span class="n">_p_survival_constant</span><span class="p">(</span>
            <span class="n">birth_rate</span><span class="o">=</span><span class="n">birth_rate</span><span class="p">,</span>
            <span class="n">death_rate</span><span class="o">=</span><span class="n">death_rate</span><span class="p">,</span>
            <span class="n">massExtinctionTimes</span><span class="o">=</span><span class="n">massExtinctionTimes</span><span class="p">,</span>
            <span class="n">massExtinctionSurvivalProbabilities</span><span class="o">=</span><span class="n">massExtinctionSurvivalProbabilities</span><span class="p">,</span>
            <span class="n">sampling_probability</span><span class="o">=</span><span class="n">sampling_probability</span><span class="p">,</span>
            <span class="n">t_low</span><span class="o">=</span><span class="n">t</span><span class="p">,</span>
            <span class="n">t_high</span><span class="o">=</span><span class="n">T</span><span class="p">,</span>
            <span class="n">T</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
    <span class="c1"># compute the rate</span>
    <span class="n">rate</span> <span class="o">=</span> <span class="p">(</span><span class="n">death_rate</span> <span class="o">-</span> <span class="n">birth_rate</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="n">T</span><span class="o">-</span><span class="n">t</span><span class="p">)</span>
    <span class="c1"># add mass-extinction</span>
    <span class="c1"># for (j in range(len(massExtinctionTimes)) ) :</span>
    <span class="c1">#     rate = rate - ifelse( t &lt; massExtinctionTimes[j] &amp; T &gt;= massExtinctionTimes[j], log(massExtinctionSurvivalProbabilities[j]), 0 )</span>
    <span class="c1"># add sampling</span>
    <span class="n">rate</span> <span class="o">=</span> <span class="n">rate</span> <span class="o">-</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">sampling_probability</span><span class="p">)</span>
    <span class="n">p</span> <span class="o">=</span> <span class="mi">2</span><span class="o">*</span><span class="n">a</span> <span class="o">+</span> <span class="n">rate</span>
    <span class="k">return</span> <span class="n">p</span>
</pre></div>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div style="text-align: center; padding-top: 20px; padding-bottom: 5px; width: 100%;">
    <a href="../../../index.html"><img src="../../../_static/dendropy_logo.png" /></a>
</div><div style="clear:both; width: 100%; height:1px;"></div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script><div style="clear:both; width: 100%; height:1px;"></div><div style="clear:both; width: 100%; height:1px;"></div><div style="clear:both; width: 100%; height:1px;"></div>
    <!-- Documentation -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Documentation</h3>
        <ul>
            <li><a href="../../../primer/downloading.html">Downloading and Installing DendroPy</a></li>
            <li><a href="../../../primer/index.html">The DendroPy Primer</a></li>
            <li><a href="../../../library/index.html">Library API Reference</a></li>
            <li>
                <a href="../../../schemas/index.html">Schemas</a>
                <ul>
                    <li><a href="../../../schemas/fasta.html">FASTA</a></li>
                    <li><a href="../../../schemas/newick.html">Newick</a></li>
                    <li><a href="../../../schemas/nexml.html">NeXML</a></li>
                    <li><a href="../../../schemas/nexus.html">Nexus</a></li>
                    <li><a href="../../../schemas/phylip.html">PHYLIP</a></li>
                </ul>
            </li>
            <li>
                <a href="../../../programs/index.html">Programs</a>
                <ul>
                    <li><a href="../../../programs/sumtrees.html">SumTrees</a></li>
                </ul>
            </li>
            <li><a href="../../../glossary.html">Glossary and Terminological Reference</a></li>
            <li><a href="../../../developer.html">Developer Guide</a></li>
            <li><a href="../../../planning.html">Ongoing Development</a></li>
            <li><a href="../../../changes.html">Change History</a></li>
        </ul>
    </div>

    <!-- Downloads -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Obtaining</h3>
        <ul>
            <li><a target="_blank" href="http://pypi.python.org/pypi/DendroPy">Install from the Python Package Index</a></li>
            <li><a target="_blank" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.2.tar.gz">Download the Source Code Archive</a></li>
            <li><a target="_blank" href="http://github.com/jeetsukumaran/DendroPy">Clone the Source Code Repository</a></li>
        </ul>
    </div>

    <!-- Discussions -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Discussion</span><span style="position: absolute; right: 0; top: 10px "><img src="../../../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot; DendroPy Users&quot; </a>group to follow and participate in discussion, troubleshooting, help, information, suggestions, etc. on the usage and development of the DendroPy phylogenetic computing library.</p>
            <form action="http://groups.google.com/group/dendropy-users/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot;dendropy-users&quot;</a> group, or click <a href="http://groups.google.com/group/dendropy-users?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Announcements -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Announcements</span><span style="position: absolute; right: 0; top: 10px "><img src="../../../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; DendroPy Announcements&quot; </a>group to receive announcements of new releases, updates, changes and other news of interest to DendroPy users and developers.</p>
            <form action="http://groups.google.com/group/dendropy-announce/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; dendropy-announce&quot; </a>group, or click <a href="http://groups.google.com/group/dendropy-announce?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Development -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative; padding-bottom: 15px; margin-bottom:5px;">
        <h3><span style="text-align: left"><a href="https://github.com/jeetsukumaran/DendroPy/">Development</a></span><a href="https://github.com/jeetsukumaran/DendroPy/"><span style="position: absolute; right: 0; top: 10px "><img src="../../../_static/Octocat.png" height="30px" alt="GitHub" /></span></a></h3>
        <div style="margin-top: 15px;">
            <!-- <a href="https://github.com/jeetsukumaran/DendroPy/issues">Issues</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/subscription">Watch</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/fork">Fork</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/stargazers">Star</a> &bull; <a href="https://github.com/jeetsukumaran/">Follow</a> -->
            <ul>
                <li>                <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/issues"><code>Issues</code></a></span> <span style="font-style: italic; font-size:80%;"> - Report bugs or request features</span></li>
                <li>     <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/subscription"><code>&nbsp;Watch</code></a></span> <span style="font-style: italic; font-size:80%;"> - Follow development activity</span></li>
                <li>        <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/fork"><code>&nbsp;&nbsp;Fork</code></a></span> <span style="font-style: italic; font-size:80%;"> - Contribute and collaborate</span></li>
                <li>  <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/stargazers"><code>&nbsp;&nbsp;Star</code></a></span> <span style="font-style: italic; font-size:80%;"> - Throw some glitter, add some glamour</span></li>
        </div>
    </div>


        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">dendropy.model.birthdeath</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2009-2024, Jeet Sukumaran and Mark T. Holder.
    </div>
  </body>
</html>