<!DOCTYPE html>

<html lang="en" data-content_root="../../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>dendropy.model.discrete &#8212; DendroPy 5.0.2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../../../_static/rtd.css?v=cedc6a2f" />
    
    <script src="../../../_static/documentation_options.js?v=5df4c8d1"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">dendropy.model.discrete</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for dendropy.model.discrete</h1><div class="highlight"><pre>
<span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="c1">##############################################################################</span>
<span class="c1">##  DendroPy Phylogenetic Computing Library.</span>
<span class="c1">##</span>
<span class="c1">##  Copyright 2010-2015 Jeet Sukumaran and Mark T. Holder.</span>
<span class="c1">##  All rights reserved.</span>
<span class="c1">##</span>
<span class="c1">##  See &quot;LICENSE.rst&quot; for terms and conditions of usage.</span>
<span class="c1">##</span>
<span class="c1">##  If you use this work or any portion thereof in published work,</span>
<span class="c1">##  please cite it as:</span>
<span class="c1">##</span>
<span class="c1">##     Sukumaran, J. and M. T. Holder. 2010. DendroPy: a Python library</span>
<span class="c1">##     for phylogenetic computing. Bioinformatics 26: 1569-1571.</span>
<span class="c1">##</span>
<span class="c1">##############################################################################</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Models and modeling of discrete character evolution.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">from</span> <span class="nn">dendropy.utility</span> <span class="kn">import</span> <span class="n">GLOBAL_RNG</span>
<span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">probability</span>
<span class="kn">import</span> <span class="nn">dendropy</span>

<span class="c1">############################################################################</span>
<span class="c1">## Character Evolution Modeling</span>

<div class="viewcode-block" id="DiscreteCharacterEvolutionModel">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.DiscreteCharacterEvolutionModel">[docs]</a>
<span class="k">class</span> <span class="nc">DiscreteCharacterEvolutionModel</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="s2">&quot;Base class for discrete character substitution models.&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">state_alphabet</span><span class="p">,</span> <span class="n">stationary_freqs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        __init__ initializes the state_alphabet to define the character type on which</span>
<span class="sd">        this model acts.  The objects random number generator will be ``rng`` or &#39;GLOBAL_RNG&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">state_alphabet</span> <span class="o">=</span> <span class="n">state_alphabet</span>
        <span class="k">if</span> <span class="n">rng</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">rng</span> <span class="o">=</span> <span class="n">GLOBAL_RNG</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">rng</span> <span class="o">=</span> <span class="n">rng</span>

<div class="viewcode-block" id="DiscreteCharacterEvolutionModel.pmatrix">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.DiscreteCharacterEvolutionModel.pmatrix">[docs]</a>
    <span class="k">def</span> <span class="nf">pmatrix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tlen</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1.0</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a matrix of nucleotide substitution</span>
<span class="sd">        probabilities.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span></div>


<div class="viewcode-block" id="DiscreteCharacterEvolutionModel.simulate_descendant_states">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.DiscreteCharacterEvolutionModel.simulate_descendant_states">[docs]</a>
    <span class="k">def</span> <span class="nf">simulate_descendant_states</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
        <span class="n">ancestral_states</span><span class="p">,</span>
        <span class="n">edge_length</span><span class="p">,</span>
        <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
        <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns descendent sequence given ancestral sequence.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">rng</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">rng</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rng</span>
        <span class="n">pmat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pmatrix</span><span class="p">(</span><span class="n">edge_length</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="p">)</span>
        <span class="n">multi</span> <span class="o">=</span> <span class="n">probability</span><span class="o">.</span><span class="n">sample_multinomial</span>
        <span class="n">desc_states</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">state</span> <span class="ow">in</span> <span class="n">ancestral_states</span><span class="p">:</span>
            <span class="n">anc_state_idx</span> <span class="o">=</span> <span class="n">state</span><span class="o">.</span><span class="n">index</span>
            <span class="n">desc_state_idx</span> <span class="o">=</span> <span class="n">multi</span><span class="p">(</span><span class="n">pmat</span><span class="p">[</span><span class="n">anc_state_idx</span><span class="p">],</span> <span class="n">rng</span><span class="p">)</span>
            <span class="n">desc_states</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">state_alphabet</span><span class="p">[</span><span class="n">desc_state_idx</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">desc_states</span></div>
</div>


<div class="viewcode-block" id="DiscreteCharacterEvolver">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.DiscreteCharacterEvolver">[docs]</a>
<span class="k">class</span> <span class="nc">DiscreteCharacterEvolver</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="s2">&quot;Evolves sequences on a tree.&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
     <span class="n">seq_model</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
     <span class="n">mutation_rate</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
     <span class="n">seq_attr</span><span class="o">=</span><span class="s1">&#39;sequences&#39;</span><span class="p">,</span>
     <span class="n">seq_model_attr</span><span class="o">=</span><span class="s2">&quot;seq_model&quot;</span><span class="p">,</span>
     <span class="n">edge_length_attr</span><span class="o">=</span><span class="s2">&quot;length&quot;</span><span class="p">,</span>
     <span class="n">edge_rate_attr</span><span class="o">=</span><span class="s2">&quot;mutation_rate&quot;</span><span class="p">,</span>
     <span class="n">seq_label_attr</span><span class="o">=</span><span class="s1">&#39;taxon&#39;</span><span class="p">):</span>
        <span class="s2">&quot;__init__ sets up meta-data dealing with object nomenclature and semantics.&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_model</span> <span class="o">=</span> <span class="n">seq_model</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_rate</span> <span class="o">=</span> <span class="n">mutation_rate</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_attr</span> <span class="o">=</span> <span class="n">seq_attr</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_model_attr</span> <span class="o">=</span> <span class="n">seq_model_attr</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">edge_length_attr</span> <span class="o">=</span> <span class="n">edge_length_attr</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">edge_rate_attr</span> <span class="o">=</span> <span class="n">edge_rate_attr</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_label_attr</span> <span class="o">=</span> <span class="n">seq_label_attr</span>

<div class="viewcode-block" id="DiscreteCharacterEvolver.evolve_states">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.DiscreteCharacterEvolver.evolve_states">[docs]</a>
    <span class="k">def</span> <span class="nf">evolve_states</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">tree</span><span class="p">,</span>
            <span class="n">seq_len</span><span class="p">,</span>
            <span class="n">root_states</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">simulate_root_states</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">in_place</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Appends a new sequence of length ``seq_len`` to a list at each node</span>
<span class="sd">        in ``tree``.  The attribute name of this list in each node is given</span>
<span class="sd">        by ``seq_attr``. If ``seq_model`` is None, ``tree.seq_model`` or</span>
<span class="sd">        ``seq_model`` at each node must be specified. If ``in_place`` is</span>
<span class="sd">        False, the tree is copied first, otherwise original tree is modified.</span>
<span class="sd">        If ``root_states`` is given, this will be used as the sequence for the root.</span>
<span class="sd">        If not, and if ``simulate_root_states`` is True, then the sequence for the</span>
<span class="sd">        root will be drawn from the stationary distribution of the character model.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">rng</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">rng</span> <span class="o">=</span> <span class="n">GLOBAL_RNG</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">in_place</span><span class="p">:</span>
            <span class="n">tree</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">clone</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="c1"># ==&gt; taxon_namespace_scoped_copy()</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_model</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">seq_model</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_model_attr</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="c1"># loop through edges in preorder (root-&gt;tips)</span>
        <span class="n">n_prev_seq</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># to mollify linter undefined variable warning</span>
        <span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">preorder_edge_iter</span><span class="p">():</span>
            <span class="n">node</span> <span class="o">=</span> <span class="n">edge</span><span class="o">.</span><span class="n">head_node</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">node</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_attr</span><span class="p">):</span>
                <span class="nb">setattr</span><span class="p">(</span><span class="n">node</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_attr</span><span class="p">,</span> <span class="p">[])</span>
            <span class="n">seq_list</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">node</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_attr</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">edge</span><span class="o">.</span><span class="n">tail_node</span><span class="p">:</span>
                <span class="n">par</span> <span class="o">=</span> <span class="n">edge</span><span class="o">.</span><span class="n">tail_node</span>
                <span class="k">assert</span> <span class="n">n_prev_seq</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_list</span><span class="p">)</span> <span class="o">!=</span> <span class="n">n_prev_seq</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;</span><span class="si">%s</span><span class="s2">&#39; length varies among nodes&quot;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_attr</span><span class="p">)</span>
                <span class="n">par_seq</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">par</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_attr</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">seq_model</span>  <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">edge</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_model_attr</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_model</span>
                <span class="n">length</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">edge</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">edge_length_attr</span><span class="p">)</span>
                <span class="n">mutation_rate</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">edge</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">edge_rate_attr</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_rate</span>
                <span class="n">seq_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_model</span><span class="o">.</span><span class="n">simulate_descendant_states</span><span class="p">(</span><span class="n">par_seq</span><span class="p">,</span>
                                                                     <span class="n">length</span><span class="p">,</span>
                                                                     <span class="n">mutation_rate</span><span class="p">,</span>
                                                                     <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># no tail node: root</span>
                <span class="n">n_prev_seq</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_list</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">root_states</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">seq_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">root_states</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">simulate_root_states</span><span class="p">:</span>
                    <span class="n">seq_model</span>  <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">edge</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_model_attr</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_model</span>
                    <span class="n">seq_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_model</span><span class="o">.</span><span class="n">stationary_sample</span><span class="p">(</span><span class="n">seq_len</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">assert</span> <span class="n">n_prev_seq</span> <span class="o">&gt;</span> <span class="mi">0</span>
                    <span class="n">n_prev_seq</span> <span class="o">-=</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">tree</span></div>


<div class="viewcode-block" id="DiscreteCharacterEvolver.extend_char_matrix_with_characters_on_tree">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.DiscreteCharacterEvolver.extend_char_matrix_with_characters_on_tree">[docs]</a>
    <span class="k">def</span> <span class="nf">extend_char_matrix_with_characters_on_tree</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">char_matrix</span><span class="p">,</span>
            <span class="n">tree</span><span class="p">,</span>
            <span class="n">include</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">exclude</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a character matrix with new sequences (or extends sequences of</span>
<span class="sd">        an existing character matrix if provided via ``char_matrix``),</span>
<span class="sd">        where the the sequence for each taxon corresponds to the concatenation</span>
<span class="sd">        of all sequences in the list of sequences associated with tip that</span>
<span class="sd">        references the given taxon.</span>
<span class="sd">        Specific sequences to be included/excluded can be fine-tuned using the</span>
<span class="sd">        ``include`` and ``exclude`` args, where ``include=None`` means to include all</span>
<span class="sd">        by default, and ``exclude=None`` means to exclude all by default.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">leaf_nodes</span><span class="p">():</span>
            <span class="n">cvec</span> <span class="o">=</span> <span class="n">char_matrix</span><span class="p">[</span><span class="n">leaf</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span>
            <span class="n">seq_list</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">leaf</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_attr</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">seq_idx</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">seq_list</span><span class="p">):</span>
                <span class="k">if</span> <span class="p">((</span><span class="n">include</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">seq_idx</span> <span class="ow">in</span> <span class="n">include</span><span class="p">))</span>  \
                    <span class="ow">and</span> <span class="p">((</span><span class="n">exclude</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">seq_idx</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">exclude</span><span class="p">)):</span>
                    <span class="k">for</span> <span class="n">state</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">:</span>
                        <span class="n">cvec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">state</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">char_matrix</span></div>


    <span class="k">def</span> <span class="nf">clean_tree</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tree</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">tree</span><span class="p">:</span>
            <span class="c1"># setattr(nd, self.seq_attr, [])</span>
            <span class="nb">delattr</span><span class="p">(</span><span class="n">nd</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_attr</span><span class="p">)</span></div>


<span class="c1">############################################################################</span>
<span class="c1">## Specialized Models: nucldeotides</span>

<div class="viewcode-block" id="NucleotideCharacterEvolutionModel">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.NucleotideCharacterEvolutionModel">[docs]</a>
<span class="k">class</span> <span class="nc">NucleotideCharacterEvolutionModel</span><span class="p">(</span><span class="n">DiscreteCharacterEvolutionModel</span><span class="p">):</span>
    <span class="s2">&quot;General nucleotide substitution model.&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">base_freqs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">state_alphabet</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="s2">&quot;__init__ calls SeqModel.__init__ and sets the base_freqs field&quot;</span>
        <span class="k">if</span> <span class="n">state_alphabet</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">state_alphabet</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DNA_STATE_ALPHABET</span>
        <span class="n">DiscreteCharacterEvolutionModel</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
                <span class="bp">self</span><span class="p">,</span>
                <span class="n">state_alphabet</span><span class="o">=</span><span class="n">state_alphabet</span><span class="p">,</span>
                <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">base_freqs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span> <span class="o">=</span> <span class="n">base_freqs</span>

<div class="viewcode-block" id="NucleotideCharacterEvolutionModel.stationary_sample">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.NucleotideCharacterEvolutionModel.stationary_sample">[docs]</a>
    <span class="k">def</span> <span class="nf">stationary_sample</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq_len</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a NucleotideSequence() object with length ``length``</span>
<span class="sd">        representing a sample of characters drawn from this model&#39;s</span>
<span class="sd">        stationary distribution.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">probs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span>
        <span class="n">char_state_indices</span> <span class="o">=</span> <span class="p">[</span><span class="n">probability</span><span class="o">.</span><span class="n">sample_multinomial</span><span class="p">(</span><span class="n">probs</span><span class="p">,</span> <span class="n">rng</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">seq_len</span><span class="p">)]</span>
        <span class="k">return</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">state_alphabet</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">char_state_indices</span><span class="p">]</span></div>


<div class="viewcode-block" id="NucleotideCharacterEvolutionModel.is_purine">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.NucleotideCharacterEvolutionModel.is_purine">[docs]</a>
    <span class="k">def</span> <span class="nf">is_purine</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">state_index</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns True if state_index represents a purine (A or G) row or column</span>
<span class="sd">        index: 0, 2</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">state_index</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">0</span></div>


<div class="viewcode-block" id="NucleotideCharacterEvolutionModel.is_pyrimidine">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.NucleotideCharacterEvolutionModel.is_pyrimidine">[docs]</a>
    <span class="k">def</span> <span class="nf">is_pyrimidine</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">state_index</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns True if state_index represents a pyrimidine (C or T) row or column</span>
<span class="sd">        index: 1, 3</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">state_index</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">1</span></div>


<div class="viewcode-block" id="NucleotideCharacterEvolutionModel.is_transversion">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.NucleotideCharacterEvolutionModel.is_transversion">[docs]</a>
    <span class="k">def</span> <span class="nf">is_transversion</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">state1_idx</span><span class="p">,</span> <span class="n">state2_idx</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns True if the change from state1 to state2, as</span>
<span class="sd">        represented by the row or column indices, is a transversional</span>
<span class="sd">        change.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_purine</span><span class="p">(</span><span class="n">state1_idx</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_pyrimidine</span><span class="p">(</span><span class="n">state2_idx</span><span class="p">))</span> \
               <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_pyrimidine</span><span class="p">(</span><span class="n">state1_idx</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_purine</span><span class="p">(</span><span class="n">state2_idx</span><span class="p">))</span></div>


<div class="viewcode-block" id="NucleotideCharacterEvolutionModel.is_purine_transition">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.NucleotideCharacterEvolutionModel.is_purine_transition">[docs]</a>
    <span class="k">def</span> <span class="nf">is_purine_transition</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">state1_idx</span><span class="p">,</span> <span class="n">state2_idx</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns True if the change from state1 to state2, as</span>
<span class="sd">        represented by the row or column indices, is a purine</span>
<span class="sd">        transitional change.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_purine</span><span class="p">(</span><span class="n">state1_idx</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_purine</span><span class="p">(</span><span class="n">state2_idx</span><span class="p">)</span></div>


<div class="viewcode-block" id="NucleotideCharacterEvolutionModel.is_pyrimidine_transition">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.NucleotideCharacterEvolutionModel.is_pyrimidine_transition">[docs]</a>
    <span class="k">def</span> <span class="nf">is_pyrimidine_transition</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">state1_idx</span><span class="p">,</span> <span class="n">state2_idx</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns True if the change from state1 to state2, as</span>
<span class="sd">        represented by the row or column indices, is a pyrimidine</span>
<span class="sd">        transitional change.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_pyrimidine</span><span class="p">(</span><span class="n">state1_idx</span><span class="p">)</span> \
               <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_pyrimidine</span><span class="p">(</span><span class="n">state2_idx</span><span class="p">)</span></div>


<div class="viewcode-block" id="NucleotideCharacterEvolutionModel.is_transition">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.NucleotideCharacterEvolutionModel.is_transition">[docs]</a>
    <span class="k">def</span> <span class="nf">is_transition</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">state1_idx</span><span class="p">,</span> <span class="n">state2_idx</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns True if the change from state1 to state2, as</span>
<span class="sd">        represented by the row or column indices, is a</span>
<span class="sd">        transitional change.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_purine</span><span class="p">(</span><span class="n">state1_idx</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_purine</span><span class="p">(</span><span class="n">state2_idx</span><span class="p">))</span> \
               <span class="ow">or</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">is_pyrimidine</span><span class="p">(</span><span class="n">state1_idx</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_pyrimidine</span><span class="p">(</span><span class="n">state2_idx</span><span class="p">))</span></div>
</div>


<div class="viewcode-block" id="Hky85">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.Hky85">[docs]</a>
<span class="k">class</span> <span class="nc">Hky85</span><span class="p">(</span><span class="n">NucleotideCharacterEvolutionModel</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Hasegawa et al. 1985 model. Implementation following Swofford et</span>
<span class="sd">    al., 1996.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">kappa</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">base_freqs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">state_alphabet</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="s2">&quot;__init__: if no arguments given, defaults to JC69.&quot;</span>
        <span class="k">if</span> <span class="n">state_alphabet</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">state_alphabet</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DNA_STATE_ALPHABET</span>
        <span class="n">NucleotideCharacterEvolutionModel</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
                <span class="bp">self</span><span class="p">,</span>
                <span class="n">base_freqs</span><span class="o">=</span><span class="n">base_freqs</span><span class="p">,</span>
                <span class="n">state_alphabet</span><span class="o">=</span><span class="n">state_alphabet</span><span class="p">,</span>
                <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">correct_rate</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">kappa</span> <span class="o">=</span> <span class="n">kappa</span>

    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">rep</span> <span class="o">=</span> <span class="s2">&quot;kappa=</span><span class="si">%f</span><span class="s2"> bases=</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">kappa</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">rep</span>

<div class="viewcode-block" id="Hky85.corrected_substitution_rate">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.Hky85.corrected_substitution_rate">[docs]</a>
    <span class="k">def</span> <span class="nf">corrected_substitution_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rate</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the factor that we have to multiply to the branch length</span>
<span class="sd">        to make branch lengths proportional to # of substitutions per site.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">correct_rate</span><span class="p">:</span>
            <span class="n">pia</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">pic</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">pig</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
            <span class="n">pit</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">f</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">kappa</span><span class="o">*</span><span class="p">(</span><span class="n">pia</span><span class="o">*</span><span class="n">pig</span> <span class="o">+</span> <span class="n">pic</span><span class="o">*</span><span class="n">pit</span><span class="p">)</span>
            <span class="n">f</span> <span class="o">+=</span> <span class="p">(</span><span class="n">pia</span> <span class="o">+</span> <span class="n">pig</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="n">pic</span> <span class="o">+</span> <span class="n">pit</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">(</span><span class="n">rate</span> <span class="o">*</span> <span class="mf">0.5</span><span class="o">/</span><span class="n">f</span><span class="p">)</span>  <span class="c1"># (rate * 0.5/f)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">rate</span></div>


<div class="viewcode-block" id="Hky85.pij">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.Hky85.pij">[docs]</a>
    <span class="k">def</span> <span class="nf">pij</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">state_i</span><span class="p">,</span> <span class="n">state_j</span><span class="p">,</span> <span class="n">tlen</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1.0</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns probability, p_ij, of going from state i to state j</span>
<span class="sd">        over time tlen at given rate. (tlen * rate = nu, expected</span>
<span class="sd">        number of substitutions)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">nu</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">corrected_substitution_rate</span><span class="p">(</span><span class="n">rate</span><span class="p">)</span> <span class="o">*</span> <span class="n">tlen</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_purine</span><span class="p">(</span><span class="n">state_j</span><span class="p">):</span>
            <span class="n">sumfreqs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sumfreqs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
        <span class="n">factorA</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="p">(</span><span class="n">sumfreqs</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">kappa</span> <span class="o">-</span> <span class="mf">1.0</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">state_i</span> <span class="o">==</span> <span class="n">state_j</span><span class="p">:</span>
            <span class="n">pij</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="n">state_j</span><span class="p">]</span> \
                  <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="n">state_j</span><span class="p">]</span> \
                      <span class="o">*</span> <span class="p">(</span><span class="mf">1.0</span><span class="o">/</span><span class="n">sumfreqs</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="mf">1.0</span> <span class="o">*</span> <span class="n">nu</span><span class="p">)</span> \
                  <span class="o">+</span> <span class="p">((</span><span class="n">sumfreqs</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="n">state_j</span><span class="p">])</span><span class="o">/</span><span class="n">sumfreqs</span><span class="p">)</span> \
                      <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="mf">1.0</span> <span class="o">*</span> <span class="n">nu</span> <span class="o">*</span> <span class="n">factorA</span><span class="p">)</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_transition</span><span class="p">(</span><span class="n">state_i</span><span class="p">,</span> <span class="n">state_j</span><span class="p">):</span>
            <span class="n">pij</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="n">state_j</span><span class="p">]</span> \
                  <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="n">state_j</span><span class="p">]</span> \
                      <span class="o">*</span> <span class="p">(</span><span class="mf">1.0</span><span class="o">/</span><span class="n">sumfreqs</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="mf">1.0</span> <span class="o">*</span> <span class="n">nu</span><span class="p">)</span> \
                  <span class="o">-</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="n">state_j</span><span class="p">]</span> <span class="o">/</span> <span class="n">sumfreqs</span><span class="p">)</span> \
                      <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="mf">1.0</span> <span class="o">*</span> <span class="n">nu</span> <span class="o">*</span> <span class="n">factorA</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">pij</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="n">state_j</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">-</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="mf">1.0</span> <span class="o">*</span> <span class="n">nu</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">pij</span></div>


<div class="viewcode-block" id="Hky85.qmatrix">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.Hky85.qmatrix">[docs]</a>
    <span class="k">def</span> <span class="nf">qmatrix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1.0</span><span class="p">):</span>
        <span class="s2">&quot;Returns the instantaneous rate of change matrix.&quot;</span>
        <span class="n">rate</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">corrected_substitution_rate</span><span class="p">(</span><span class="n">rate</span><span class="p">)</span>
        <span class="n">qmatrix</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">state_i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">4</span><span class="p">):</span>
            <span class="n">qmatrix</span><span class="o">.</span><span class="n">append</span><span class="p">([])</span>
            <span class="k">for</span> <span class="n">state_j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">4</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">state_i</span> <span class="o">==</span> <span class="n">state_j</span><span class="p">:</span>
                    <span class="c1"># we cheat here and insert a placeholder till the</span>
                    <span class="c1"># other cells are calculated</span>
                    <span class="n">qij</span> <span class="o">=</span> <span class="mf">0.0</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_transition</span><span class="p">(</span><span class="n">state_i</span><span class="p">,</span> <span class="n">state_j</span><span class="p">):</span>
                        <span class="n">qij</span> <span class="o">=</span> <span class="n">rate</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">kappa</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="n">state_j</span><span class="p">]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">qij</span> <span class="o">=</span> <span class="n">rate</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">base_freqs</span><span class="p">[</span><span class="n">state_j</span><span class="p">]</span>
                <span class="n">qmatrix</span><span class="p">[</span><span class="n">state_i</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">qij</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">state</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">4</span><span class="p">):</span>
            <span class="n">qmatrix</span><span class="p">[</span><span class="n">state</span><span class="p">][</span><span class="n">state</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mf">1.0</span> <span class="o">*</span> <span class="nb">sum</span><span class="p">(</span><span class="n">qmatrix</span><span class="p">[</span><span class="n">state</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">qmatrix</span></div>


<div class="viewcode-block" id="Hky85.pvector">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.Hky85.pvector">[docs]</a>
    <span class="k">def</span> <span class="nf">pvector</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">tlen</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1.0</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a vector of transition probabilities for a given state</span>
<span class="sd">        over time ``tlen`` at rate ``rate`` for ``state``. (tlen * rate =</span>
<span class="sd">        nu, expected number of substitutions)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">pvec</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># in case later we want to allow characters passed in here</span>
        <span class="n">state_i</span> <span class="o">=</span> <span class="n">state</span>
        <span class="k">for</span> <span class="n">state_j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">4</span><span class="p">):</span>
            <span class="n">pvec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pij</span><span class="p">(</span><span class="n">state_i</span><span class="p">,</span> <span class="n">state_j</span><span class="p">,</span> <span class="n">tlen</span><span class="o">=</span><span class="n">tlen</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="n">rate</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">pvec</span></div>


<div class="viewcode-block" id="Hky85.pmatrix">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.Hky85.pmatrix">[docs]</a>
    <span class="k">def</span> <span class="nf">pmatrix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tlen</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1.0</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a matrix of nucleotide substitution</span>
<span class="sd">        probabilities. Based on analytical solution by Swofford et</span>
<span class="sd">        al., 1996. (tlen * rate = nu, expected number of</span>
<span class="sd">        substitutions)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">pmatrix</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">state_i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">4</span><span class="p">):</span>
            <span class="n">pmatrix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pvector</span><span class="p">(</span><span class="n">state_i</span><span class="p">,</span> <span class="n">tlen</span><span class="o">=</span><span class="n">tlen</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="n">rate</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">pmatrix</span></div>
</div>


<div class="viewcode-block" id="Jc69">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.Jc69">[docs]</a>
<span class="k">class</span> <span class="nc">Jc69</span><span class="p">(</span><span class="n">Hky85</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Jukes-Cantor 1969 model. Specializes HKY85 such that</span>
<span class="sd">    kappa = 1.0, and base frequencies = [0.25, 0.25, 0.25, 0.25].</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">state_alphabet</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="s2">&quot;__init__: uses Hky85.__init__&quot;</span>
        <span class="n">Hky85</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                <span class="n">kappa</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
                <span class="n">base_freqs</span><span class="o">=</span><span class="p">[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">],</span>
                <span class="n">state_alphabet</span><span class="o">=</span><span class="n">state_alphabet</span><span class="p">,</span>
                <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">,</span>
                <span class="p">)</span></div>




<span class="c1">##############################################################################</span>
<span class="c1">## Wrappers for Convenience</span>

<div class="viewcode-block" id="simulate_discrete_char_dataset">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.simulate_discrete_char_dataset">[docs]</a>
<span class="k">def</span> <span class="nf">simulate_discrete_char_dataset</span><span class="p">(</span><span class="n">seq_len</span><span class="p">,</span>
        <span class="n">tree_model</span><span class="p">,</span>
        <span class="n">seq_model</span><span class="p">,</span>
        <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
        <span class="n">root_states</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">dataset</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Wrapper to conveniently generate a DataSet simulated under</span>
<span class="sd">    the given tree and character model.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    seq_len       : int</span>
<span class="sd">        Length of sequence (number of characters).</span>
<span class="sd">    tree_model    : |Tree|</span>
<span class="sd">        Tree on which to simulate.</span>
<span class="sd">    seq_model     : dendropy.model.discrete.NucleotideCharacterEvolutionModel</span>
<span class="sd">        The character substitution model under which to to evolve the</span>
<span class="sd">        characters.</span>
<span class="sd">    mutation_rate : float</span>
<span class="sd">        Mutation *modifier* rate (should be 1.0 if branch lengths on tree</span>
<span class="sd">        reflect true expected number of changes).</span>
<span class="sd">    root_states``   : list</span>
<span class="sd">        Vector of root states (length must equal ``seq_len``).</span>
<span class="sd">    dataset       : |DataSet|</span>
<span class="sd">        If given, the new dendropy.CharacterMatrix object will be</span>
<span class="sd">        added to this (along with a new taxon_namespace if</span>
<span class="sd">        required). Otherwise, a new dendropy.DataSet</span>
<span class="sd">        object will be created.</span>
<span class="sd">    rng           : random number generator</span>
<span class="sd">        If not given, &#39;GLOBAL_RNG&#39; will be used.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>

<span class="sd">    d : |DataSet|</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">dataset</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">dataset</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">tree_model</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">dataset</span><span class="o">.</span><span class="n">taxon_namespaces</span><span class="p">:</span>
        <span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">add_taxon_namespace</span><span class="p">(</span><span class="n">tree_model</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">tree_model</span><span class="o">.</span><span class="n">taxon_namespace</span>
    <span class="n">char_matrix</span> <span class="o">=</span> <span class="n">simulate_discrete_chars</span><span class="p">(</span>
        <span class="n">seq_len</span><span class="o">=</span><span class="n">seq_len</span><span class="p">,</span>
        <span class="n">tree_model</span><span class="o">=</span><span class="n">tree_model</span><span class="p">,</span>
        <span class="n">seq_model</span><span class="o">=</span><span class="n">seq_model</span><span class="p">,</span>
        <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span>
        <span class="n">root_states</span><span class="o">=</span><span class="n">root_states</span><span class="p">,</span>
        <span class="n">char_matrix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">dataset</span><span class="o">.</span><span class="n">add_char_matrix</span><span class="p">(</span><span class="n">char_matrix</span><span class="o">=</span><span class="n">char_matrix</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">dataset</span></div>


<div class="viewcode-block" id="simulate_discrete_chars">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.simulate_discrete_chars">[docs]</a>
<span class="k">def</span> <span class="nf">simulate_discrete_chars</span><span class="p">(</span>
        <span class="n">seq_len</span><span class="p">,</span>
        <span class="n">tree_model</span><span class="p">,</span>
        <span class="n">seq_model</span><span class="p">,</span>
        <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
        <span class="n">root_states</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">char_matrix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">retain_sequences_on_tree</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Wrapper to conveniently generate a characters simulated under</span>
<span class="sd">    the given tree and character model.</span>

<span class="sd">    Since characters will be appended to existing sequences, you can simulate a</span>
<span class="sd">    sequences under a mixed model by calling this method multiple times with</span>
<span class="sd">    different character models and/or different mutation rates, passing</span>
<span class="sd">    in the same ``char_matrix`` object each time.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    seq_len       : int</span>
<span class="sd">        Length of sequence (number of characters).</span>
<span class="sd">    tree_model    : |Tree|</span>
<span class="sd">        Tree on which to simulate.</span>
<span class="sd">    seq_model     : dendropy.model.discrete.NucleotideCharacterEvolutionModel</span>
<span class="sd">        The character substitution model under which to to evolve the</span>
<span class="sd">        characters.</span>
<span class="sd">    mutation_rate : float</span>
<span class="sd">        Mutation *modifier* rate (should be 1.0 if branch lengths on tree</span>
<span class="sd">        reflect true expected number of changes).</span>
<span class="sd">    root_states``   : list</span>
<span class="sd">        Vector of root states (length must equal ``seq_len``).</span>
<span class="sd">    char_matrix   : |DnaCharacterMatrix|</span>
<span class="sd">        If given, new sequences for taxa on ``tree_model`` leaf_nodes will be</span>
<span class="sd">        appended to existing sequences of corresponding taxa in char_matrix; if</span>
<span class="sd">        not, a new |DnaCharacterMatrix| object will be created.</span>
<span class="sd">    retain_sequences_on_tree : bool</span>
<span class="sd">        If |False|, sequence annotations will be cleared from tree after</span>
<span class="sd">        simulation. Set to |True| if you want to, e.g., evolve and accumulate</span>
<span class="sd">        different sequences on tree, or retain information for other purposes.</span>
<span class="sd">    rng           : random number generator</span>
<span class="sd">        If not given, &#39;GLOBAL_RNG&#39; will be used.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    d : a dendropy.datamodel.CharacterMatrix object.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_evolver</span> <span class="o">=</span> <span class="n">DiscreteCharacterEvolver</span><span class="p">(</span><span class="n">seq_model</span><span class="o">=</span><span class="n">seq_model</span><span class="p">,</span>
                               <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">)</span>
    <span class="n">tree</span> <span class="o">=</span> <span class="n">seq_evolver</span><span class="o">.</span><span class="n">evolve_states</span><span class="p">(</span>
        <span class="n">tree</span><span class="o">=</span><span class="n">tree_model</span><span class="p">,</span>
        <span class="n">seq_len</span><span class="o">=</span><span class="n">seq_len</span><span class="p">,</span>
        <span class="n">root_states</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">)</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">migrate_taxon_namespace</span><span class="p">(</span><span class="n">tree_model</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">char_matrix</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">char_matrix</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DnaCharacterMatrix</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="o">=</span><span class="n">tree_model</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">)</span>
        <span class="n">char_matrix</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">tree_model</span><span class="o">.</span><span class="n">taxon_namespace</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">char_matrix</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="ow">is</span> <span class="n">tree_model</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">,</span> <span class="s2">&quot;conflicting taxon sets&quot;</span>
    <span class="n">seq_evolver</span><span class="o">.</span><span class="n">extend_char_matrix_with_characters_on_tree</span><span class="p">(</span>
            <span class="n">char_matrix</span><span class="o">=</span><span class="n">char_matrix</span><span class="p">,</span>
            <span class="n">tree</span><span class="o">=</span><span class="n">tree</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">retain_sequences_on_tree</span><span class="p">:</span>
        <span class="n">seq_evolver</span><span class="o">.</span><span class="n">clean_tree</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">char_matrix</span></div>


<div class="viewcode-block" id="hky85_chars">
<a class="viewcode-back" href="../../../library/discrete.html#dendropy.model.discrete.hky85_chars">[docs]</a>
<span class="k">def</span> <span class="nf">hky85_chars</span><span class="p">(</span>
        <span class="n">seq_len</span><span class="p">,</span>
        <span class="n">tree_model</span><span class="p">,</span>
        <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
        <span class="n">kappa</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
        <span class="n">base_freqs</span><span class="o">=</span><span class="p">[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">],</span>
        <span class="n">root_states</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">char_matrix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">retain_sequences_on_tree</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Convenience class to wrap generation of characters (as a CharacterBlock</span>
<span class="sd">    object) based on the HKY model.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    seq_len       : int</span>
<span class="sd">        Length of sequence (number of characters).</span>
<span class="sd">    tree_model    : |Tree|</span>
<span class="sd">        Tree on which to simulate.</span>
<span class="sd">    mutation_rate : float</span>
<span class="sd">        Mutation *modifier* rate (should be 1.0 if branch lengths on tree</span>
<span class="sd">        reflect true expected number of changes).</span>
<span class="sd">    root_states``   : list</span>
<span class="sd">        Vector of root states (length must equal ``seq_len``).</span>
<span class="sd">    char_matrix   : |DnaCharacterMatrix|</span>
<span class="sd">        If given, new sequences for taxa on ``tree_model`` leaf_nodes will be</span>
<span class="sd">        appended to existing sequences of corresponding taxa in char_matrix; if</span>
<span class="sd">        not, a new |DnaCharacterMatrix| object will be created.</span>
<span class="sd">    retain_sequences_on_tree : bool</span>
<span class="sd">        If |False|, sequence annotations will be cleared from tree after</span>
<span class="sd">        simulation. Set to |True| if you want to, e.g., evolve and accumulate</span>
<span class="sd">        different sequences on tree, or retain information for other purposes.</span>
<span class="sd">    rng           : random number generator</span>
<span class="sd">        If not given, &#39;GLOBAL_RNG&#39; will be used.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    d : |DnaCharacterMatrix|</span>
<span class="sd">        The simulated alignment.</span>

<span class="sd">    Since characters will be appended to existing sequences, you can simulate a</span>
<span class="sd">    sequences under a mixed model by calling this method multiple times with</span>
<span class="sd">    different character model parameter values and/or different mutation</span>
<span class="sd">    rates, passing in the same ``char_matrix`` object each time.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">char_matrix</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">char_matrix</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DnaCharacterMatrix</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="o">=</span><span class="n">tree_model</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">char_matrix</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="ow">is</span> <span class="n">tree_model</span><span class="o">.</span><span class="n">taxon_namespace</span>
    <span class="n">state_alphabet</span> <span class="o">=</span> <span class="n">char_matrix</span><span class="o">.</span><span class="n">default_state_alphabet</span>
    <span class="n">seq_model</span> <span class="o">=</span> <span class="n">Hky85</span><span class="p">(</span>
            <span class="n">kappa</span><span class="o">=</span><span class="n">kappa</span><span class="p">,</span>
            <span class="n">base_freqs</span><span class="o">=</span><span class="n">base_freqs</span><span class="p">,</span>
            <span class="n">state_alphabet</span><span class="o">=</span><span class="n">state_alphabet</span><span class="p">,</span>
            <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">simulate_discrete_chars</span><span class="p">(</span><span class="n">seq_len</span><span class="o">=</span><span class="n">seq_len</span><span class="p">,</span>
                               <span class="n">tree_model</span><span class="o">=</span><span class="n">tree_model</span><span class="p">,</span>
                               <span class="n">seq_model</span><span class="o">=</span><span class="n">seq_model</span><span class="p">,</span>
                               <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span>
                               <span class="n">root_states</span><span class="o">=</span><span class="n">root_states</span><span class="p">,</span>
                               <span class="n">char_matrix</span><span class="o">=</span><span class="n">char_matrix</span><span class="p">,</span>
                               <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">)</span></div>


</pre></div>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div style="text-align: center; padding-top: 20px; padding-bottom: 5px; width: 100%;">
    <a href="../../../index.html"><img src="../../../_static/dendropy_logo.png" /></a>
</div><div style="clear:both; width: 100%; height:1px;"></div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script><div style="clear:both; width: 100%; height:1px;"></div><div style="clear:both; width: 100%; height:1px;"></div><div style="clear:both; width: 100%; height:1px;"></div>
    <!-- Documentation -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Documentation</h3>
        <ul>
            <li><a href="../../../primer/downloading.html">Downloading and Installing DendroPy</a></li>
            <li><a href="../../../primer/index.html">The DendroPy Primer</a></li>
            <li><a href="../../../library/index.html">Library API Reference</a></li>
            <li>
                <a href="../../../schemas/index.html">Schemas</a>
                <ul>
                    <li><a href="../../../schemas/fasta.html">FASTA</a></li>
                    <li><a href="../../../schemas/newick.html">Newick</a></li>
                    <li><a href="../../../schemas/nexml.html">NeXML</a></li>
                    <li><a href="../../../schemas/nexus.html">Nexus</a></li>
                    <li><a href="../../../schemas/phylip.html">PHYLIP</a></li>
                </ul>
            </li>
            <li>
                <a href="../../../programs/index.html">Programs</a>
                <ul>
                    <li><a href="../../../programs/sumtrees.html">SumTrees</a></li>
                </ul>
            </li>
            <li><a href="../../../glossary.html">Glossary and Terminological Reference</a></li>
            <li><a href="../../../developer.html">Developer Guide</a></li>
            <li><a href="../../../planning.html">Ongoing Development</a></li>
            <li><a href="../../../changes.html">Change History</a></li>
        </ul>
    </div>

    <!-- Downloads -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Obtaining</h3>
        <ul>
            <li><a target="_blank" href="http://pypi.python.org/pypi/DendroPy">Install from the Python Package Index</a></li>
            <li><a target="_blank" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.2.tar.gz">Download the Source Code Archive</a></li>
            <li><a target="_blank" href="http://github.com/jeetsukumaran/DendroPy">Clone the Source Code Repository</a></li>
        </ul>
    </div>

    <!-- Discussions -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Discussion</span><span style="position: absolute; right: 0; top: 10px "><img src="../../../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot; DendroPy Users&quot; </a>group to follow and participate in discussion, troubleshooting, help, information, suggestions, etc. on the usage and development of the DendroPy phylogenetic computing library.</p>
            <form action="http://groups.google.com/group/dendropy-users/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot;dendropy-users&quot;</a> group, or click <a href="http://groups.google.com/group/dendropy-users?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Announcements -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Announcements</span><span style="position: absolute; right: 0; top: 10px "><img src="../../../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; DendroPy Announcements&quot; </a>group to receive announcements of new releases, updates, changes and other news of interest to DendroPy users and developers.</p>
            <form action="http://groups.google.com/group/dendropy-announce/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; dendropy-announce&quot; </a>group, or click <a href="http://groups.google.com/group/dendropy-announce?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Development -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative; padding-bottom: 15px; margin-bottom:5px;">
        <h3><span style="text-align: left"><a href="https://github.com/jeetsukumaran/DendroPy/">Development</a></span><a href="https://github.com/jeetsukumaran/DendroPy/"><span style="position: absolute; right: 0; top: 10px "><img src="../../../_static/Octocat.png" height="30px" alt="GitHub" /></span></a></h3>
        <div style="margin-top: 15px;">
            <!-- <a href="https://github.com/jeetsukumaran/DendroPy/issues">Issues</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/subscription">Watch</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/fork">Fork</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/stargazers">Star</a> &bull; <a href="https://github.com/jeetsukumaran/">Follow</a> -->
            <ul>
                <li>                <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/issues"><code>Issues</code></a></span> <span style="font-style: italic; font-size:80%;"> - Report bugs or request features</span></li>
                <li>     <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/subscription"><code>&nbsp;Watch</code></a></span> <span style="font-style: italic; font-size:80%;"> - Follow development activity</span></li>
                <li>        <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/fork"><code>&nbsp;&nbsp;Fork</code></a></span> <span style="font-style: italic; font-size:80%;"> - Contribute and collaborate</span></li>
                <li>  <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/stargazers"><code>&nbsp;&nbsp;Star</code></a></span> <span style="font-style: italic; font-size:80%;"> - Throw some glitter, add some glamour</span></li>
        </div>
    </div>


        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">dendropy.model.discrete</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2009-2024, Jeet Sukumaran and Mark T. Holder.
    </div>
  </body>
</html>