<!DOCTYPE html>

<html lang="en" data-content_root="../../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>dendropy.calculate.phylogeneticdistance &#8212; DendroPy 5.0.2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../../../_static/rtd.css?v=cedc6a2f" />
    
    <script src="../../../_static/documentation_options.js?v=5df4c8d1"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">dendropy.calculate.phylogeneticdistance</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for dendropy.calculate.phylogeneticdistance</h1><div class="highlight"><pre>
<span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="c1">##############################################################################</span>
<span class="c1">##  DendroPy Phylogenetic Computing Library.</span>
<span class="c1">##</span>
<span class="c1">##  Copyright 2010-2015 Jeet Sukumaran and Mark T. Holder.</span>
<span class="c1">##  All rights reserved.</span>
<span class="c1">##</span>
<span class="c1">##  See &quot;LICENSE.rst&quot; for terms and conditions of usage.</span>
<span class="c1">##</span>
<span class="c1">##  If you use this work or any portion thereof in published work,</span>
<span class="c1">##  please cite it as:</span>
<span class="c1">##</span>
<span class="c1">##     Sukumaran, J. and M. T. Holder. 2010. DendroPy: a Python library</span>
<span class="c1">##     for phylogenetic computing. Bioinformatics 26: 1569-1571.</span>
<span class="c1">##</span>
<span class="c1">##############################################################################</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Taxon-to-taxon phylogenetic distances.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">csv</span>
<span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">statistics</span>
<span class="kn">from</span> <span class="nn">dendropy.utility</span> <span class="kn">import</span> <span class="n">GLOBAL_RNG</span>
<span class="kn">from</span> <span class="nn">dendropy.utility</span> <span class="kn">import</span> <span class="n">container</span>
<span class="kn">from</span> <span class="nn">dendropy.utility</span> <span class="kn">import</span> <span class="n">error</span>
<span class="kn">import</span> <span class="nn">dendropy</span>

<div class="viewcode-block" id="PhylogeneticDistanceMatrix">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix">[docs]</a>
<span class="k">class</span> <span class="nc">PhylogeneticDistanceMatrix</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates and maintains patristic distance information of taxa on a tree.</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="PhylogeneticDistanceMatrix.from_tree">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.from_tree">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_tree</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">tree</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates and returns a |PhylogeneticDistanceMatrix| based</span>
<span class="sd">        on the given tree.</span>

<span class="sd">        Note that this creates a &quot;snapshot&quot; of the current state of the tree.</span>
<span class="sd">        Subsequent changes to the tree will not be reflected in</span>
<span class="sd">        |PhylogeneticDistanceMatrix| instances previously created.</span>

<span class="sd">        Also note that syntactically you may prefer to use::</span>

<span class="sd">            pdm = tree.phylogenetic_distance_matrix()</span>

<span class="sd">        instead of::</span>

<span class="sd">            pdm = PhylogeneticDistanceMatrix.from_tree(tree)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        tree : a |Tree| instance</span>
<span class="sd">            The |Tree| from which to get the phylogenetic distances.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pdm : A |PhylogeneticDistanceMatrix| instance</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        ::</span>

<span class="sd">            import dendropy</span>
<span class="sd">            tree = dendropy.Tree.get(path=&quot;tree.nex&quot;,</span>
<span class="sd">                    schema=&quot;nexus&quot;)</span>
<span class="sd">            pdm1 = dendropy.PhylogeneticDistanceMatrix.from_tree(tree)</span>

<span class="sd">            # following is equivalent to above and probably preferred:</span>
<span class="sd">            pdm2 = tree.phylogenetic_distance_matrix()</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">pdm</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">pdm</span><span class="o">.</span><span class="n">compile_from_tree</span><span class="p">(</span><span class="n">tree</span><span class="o">=</span><span class="n">tree</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">pdm</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.from_csv">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.from_csv">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_csv</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span>
            <span class="n">src</span><span class="p">,</span>
            <span class="n">taxon_namespace</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">is_allow_new_taxa</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">is_first_row_column_names</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_first_column_row_names</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">default_data_type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
            <span class="n">label_transform_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="o">**</span><span class="n">csv_reader_kwargs</span>
            <span class="p">):</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Instantiates a new PhylogeneticDistanceMatrix instance with data</span>
<span class="sd">        from an external source.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        src : file or file-like</span>
<span class="sd">            Source of data. This is a token delimited-file (e.g., a</span>
<span class="sd">            comma-delimited or tab-delimited file) providing a table which</span>
<span class="sd">            lists taxon labels in both rows and columns. The cells of the table</span>
<span class="sd">            are numeric (typically real) values that indicate the distance</span>
<span class="sd">            between the taxa of the current row and column. Note that *only*</span>
<span class="sd">            the upper right section of the table is considered. The diagonals</span>
<span class="sd">            values are typically zeroes and, in either case, ignored along with</span>
<span class="sd">            the lower diagonal. Despite being ignored by the</span>
<span class="sd">            PhylogeneticDistanceMatrix object, the values are parsed by the</span>
<span class="sd">            underlying reader and thus have to be valid numerical values.</span>
<span class="sd">        taxon_namespace : |TaxonNamespace| instance</span>
<span class="sd">            The taxon namespace with which to manage taxa. If this has</span>
<span class="sd">            not already been pre-populated with the taxon names, then</span>
<span class="sd">            ``is_allow_new_taxa`` should be set to |True|.</span>
<span class="sd">        is_allow_new_taxa : bool</span>
<span class="sd">            If |False|: we do *not* expect to encounter any new taxa in the</span>
<span class="sd">            data file, and it is an error if we do. If |True|: we do expect to</span>
<span class="sd">            encounter new taxa in the data file. The default value of this</span>
<span class="sd">            depends on the value passed to ``taxon_namespace``. If</span>
<span class="sd">            ``taxon_namespace`` is ``None`` or an empty |TaxonNamespace|</span>
<span class="sd">            instance, then unless explicitly set to |False|,</span>
<span class="sd">            ``is_allow_new_taxa`` will default to |True|: allowing of creation</span>
<span class="sd">            of new taxa corresponding to labels found in the data source. On</span>
<span class="sd">            the other hand, if ``taxon_namespace`` is not None and its value is</span>
<span class="sd">            a |TaxonNamespace| instance with at least one taxon, unless</span>
<span class="sd">            explicitly set to |True|, ``is_allow_new_taxa`` will default to</span>
<span class="sd">            |False|, and it will be an error if taxon labels are found in the</span>
<span class="sd">            data source that do not correspond (exactly) to |Taxon| objects</span>
<span class="sd">            defined in the taxon namespace.  This is to err on the side of</span>
<span class="sd">            caution, to avoid (or rather, highlight) problems due to incorrect</span>
<span class="sd">            or mismatching labels between the data source and the current taxon</span>
<span class="sd">            namespace.</span>
<span class="sd">        is_first_row_column_names : bool</span>
<span class="sd">            By default |True|: assumes that first row lists the taxon names.</span>
<span class="sd">            Set to |False| if there is no header row.</span>
<span class="sd">        is_first_column_row_names : bool</span>
<span class="sd">            By default |True|: assumes that first column lists the taxon names.</span>
<span class="sd">            Set to |False| if there is now row name column.</span>
<span class="sd">        label_transform_fn : function object</span>
<span class="sd">            If not None, this should be a function object that takes a string</span>
<span class="sd">            as an argument and returns another string. This function will be</span>
<span class="sd">            applied to row and column labels before they are matched to taxon</span>
<span class="sd">            labels in the |TaxonNamespace| instance given by</span>
<span class="sd">            ``taxon_namespace``.</span>
<span class="sd">        \*\*csv_reader_kwargs : keyword arguments</span>
<span class="sd">            This arguments will be passed to the underlying CSV reader.</span>
<span class="sd">            The most important one is probably &#39;delimiter&#39;.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pdm : A |PhylogeneticDistanceMatrix| instance</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        ::</span>

<span class="sd">            import dendropy</span>
<span class="sd">            pdm1 = dendropy.PhylogeneticDistanceMatrix.from_csv(</span>
<span class="sd">                    src=open(&quot;data.csv&quot;),</span>
<span class="sd">                    delimiter=&quot;,&quot;)</span>
<span class="sd">            pdm2 = dendropy.PhylogeneticDistanceMatrix.from_csv(</span>
<span class="sd">                    src=open(&quot;data.tsv&quot;),</span>
<span class="sd">                    delimiter=&quot;\t&quot;)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">taxon_namespace</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">is_allow_new_taxa</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">is_allow_new_taxa</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">old_taxon_namespace_mutability</span> <span class="o">=</span> <span class="n">taxon_namespace</span><span class="o">.</span><span class="n">is_mutable</span>
        <span class="n">taxon_namespace</span><span class="o">.</span><span class="n">is_mutable</span> <span class="o">=</span> <span class="n">is_allow_new_taxa</span>
        <span class="n">data_table</span> <span class="o">=</span> <span class="n">container</span><span class="o">.</span><span class="n">DataTable</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span>
                <span class="n">src</span><span class="p">,</span>
                <span class="n">is_first_row_column_names</span><span class="o">=</span><span class="n">is_first_row_column_names</span><span class="p">,</span>
                <span class="n">is_first_column_row_names</span><span class="o">=</span><span class="n">is_first_column_row_names</span><span class="p">,</span>
                <span class="n">default_data_type</span><span class="o">=</span><span class="n">default_data_type</span><span class="p">,</span>
                <span class="n">label_transform_fn</span><span class="o">=</span><span class="n">label_transform_fn</span><span class="p">,</span>
                <span class="o">**</span><span class="n">csv_reader_kwargs</span>
                <span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">is_first_row_column_names</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">is_first_column_row_names</span><span class="p">:</span>
            <span class="n">distances</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">seen_row_labels</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="n">seen_row_taxa_labels</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">i1</span><span class="p">,</span> <span class="n">t1_label</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data_table</span><span class="o">.</span><span class="n">row_name_iter</span><span class="p">()):</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">i1</span><span class="p">:</span>
                    <span class="n">t1</span> <span class="o">=</span> <span class="n">taxon_namespace</span><span class="o">.</span><span class="n">require_taxon</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="n">t1_label</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">t1</span> <span class="o">=</span> <span class="n">taxon_namespace</span><span class="p">[</span><span class="n">i1</span><span class="p">]</span>
                <span class="n">seen_row_labels</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">t1_label</span><span class="p">)</span>
                <span class="k">assert</span> <span class="n">t1</span><span class="o">.</span><span class="n">label</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seen_row_taxa_labels</span>
                <span class="n">seen_row_taxa_labels</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">t1</span><span class="o">.</span><span class="n">label</span><span class="p">)</span>
                <span class="n">distances</span><span class="p">[</span><span class="n">t1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">for</span> <span class="n">i2</span><span class="p">,</span> <span class="n">t2_label</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data_table</span><span class="o">.</span><span class="n">column_name_iter</span><span class="p">()):</span>
                    <span class="k">if</span> <span class="n">t2_label</span> <span class="ow">in</span> <span class="n">seen_row_labels</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">i2</span><span class="p">:</span>
                        <span class="n">t2</span> <span class="o">=</span> <span class="n">taxon_namespace</span><span class="o">.</span><span class="n">require_taxon</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="n">t2_label</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">t2</span> <span class="o">=</span> <span class="n">taxon_namespace</span><span class="p">[</span><span class="n">i2</span><span class="p">]</span>
                    <span class="n">distances</span><span class="p">[</span><span class="n">t1</span><span class="p">][</span><span class="n">t2</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_table</span><span class="p">[</span><span class="n">t1_label</span><span class="p">,</span> <span class="n">t2_label</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">distances</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">seen_taxa</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">if</span> <span class="n">is_first_column_row_names</span><span class="p">:</span>
                <span class="n">name_iter</span> <span class="o">=</span> <span class="n">data_table</span><span class="o">.</span><span class="n">row_name_iter</span><span class="p">()</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">name_iter</span> <span class="o">=</span> <span class="n">data_table</span><span class="o">.</span><span class="n">column_name_iter</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">name_iter</span><span class="p">:</span>
                <span class="n">t1</span> <span class="o">=</span> <span class="n">taxon_namespace</span><span class="o">.</span><span class="n">require_taxon</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="n">label</span><span class="p">)</span>
                <span class="k">assert</span> <span class="n">t1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seen_taxa</span>
                <span class="n">seen_taxa</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">t1</span><span class="p">)</span>
                <span class="n">taxa</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">t1</span><span class="p">)</span>
            <span class="n">seen_row_taxa</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">i1</span><span class="p">,</span> <span class="n">t1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">taxa</span><span class="p">):</span>
                <span class="k">assert</span> <span class="n">t1</span> <span class="ow">in</span> <span class="n">seen_taxa</span>
                <span class="n">seen_row_taxa</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">t1</span><span class="p">)</span>
                <span class="n">distances</span><span class="p">[</span><span class="n">t1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">for</span> <span class="n">i2</span><span class="p">,</span> <span class="n">t2</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">taxa</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">t2</span> <span class="ow">in</span> <span class="n">seen_row_taxa</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="n">distances</span><span class="p">[</span><span class="n">t1</span><span class="p">][</span><span class="n">t2</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_table</span><span class="p">[</span><span class="n">i1</span><span class="p">,</span> <span class="n">i2</span><span class="p">]</span>
        <span class="c1"># else:</span>
            <span class="c1"># raise NotImplementedError()</span>

        <span class="n">taxon_namespace</span><span class="o">.</span><span class="n">is_mutable</span> <span class="o">=</span> <span class="n">old_taxon_namespace_mutability</span>
        <span class="n">pdm</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">()</span>
        <span class="n">pdm</span><span class="o">.</span><span class="n">compile_from_dict</span><span class="p">(</span>
                <span class="n">distances</span><span class="o">=</span><span class="n">distances</span><span class="p">,</span>
                <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">taxon_namespace</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">pdm</span></div>


    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">is_store_path_edges</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_store_path_edges</span> <span class="o">=</span> <span class="n">is_store_path_edges</span>

    <span class="k">def</span> <span class="nf">clear</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_all_distinct_mapped_taxa_pairs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span> <span class="o">=</span> <span class="p">{}</span>

<div class="viewcode-block" id="PhylogeneticDistanceMatrix.compile_from_tree">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.compile_from_tree">[docs]</a>
    <span class="k">def</span> <span class="nf">compile_from_tree</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tree</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates the distances. Note that the path length (in number of</span>
<span class="sd">        steps) between taxa that span the root will be off by one if</span>
<span class="sd">        the tree is unrooted.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">taxon_namespace</span>
        <span class="c1"># for i1, t1 in enumerate(self.taxon_namespace):</span>
        <span class="c1">#     self._taxon_phylogenetic_distances[t1] = {}</span>
        <span class="c1">#     self._taxon_phylogenetic_path_steps[t1] = {}</span>
        <span class="c1">#     self._mrca[t1] = {}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_store_path_edges</span><span class="p">:</span>
            <span class="n">default_pedges</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">default_pedges</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">postorder_node_iter</span><span class="p">():</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span> <span class="o">+=</span> <span class="n">node</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span>
            <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span> <span class="c1"># None for edge length</span>
                <span class="k">pass</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">children</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">child_nodes</span><span class="p">()</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">children</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">node</span><span class="o">.</span><span class="n">desc_paths</span> <span class="o">=</span> <span class="p">{</span><span class="n">node</span> <span class="p">:</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span> <span class="n">default_pedges</span><span class="p">)}</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">node</span><span class="o">.</span><span class="n">desc_paths</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">for</span> <span class="n">cidx1</span><span class="p">,</span> <span class="n">c1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">children</span><span class="p">):</span>
                    <span class="k">for</span> <span class="n">desc1</span><span class="p">,</span> <span class="p">(</span><span class="n">desc1_plen</span><span class="p">,</span> <span class="n">desc1_psteps</span><span class="p">,</span> <span class="n">desc1_pedges</span><span class="p">)</span> <span class="ow">in</span> <span class="n">c1</span><span class="o">.</span><span class="n">desc_paths</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                        <span class="k">if</span> <span class="n">c1</span><span class="o">.</span><span class="n">edge_length</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                            <span class="n">c1_edge_length</span> <span class="o">=</span> <span class="mf">0.0</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">c1_edge_length</span> <span class="o">=</span> <span class="n">c1</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span>
                        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_store_path_edges</span><span class="p">:</span>
                            <span class="n">pedges</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">desc1_pedges</span> <span class="o">+</span> <span class="p">[</span><span class="n">c1</span><span class="o">.</span><span class="n">edge</span><span class="p">])</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">pedges</span> <span class="o">=</span> <span class="n">default_pedges</span>
                        <span class="n">node</span><span class="o">.</span><span class="n">desc_paths</span><span class="p">[</span><span class="n">desc1</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">desc1_plen</span> <span class="o">+</span> <span class="n">c1_edge_length</span><span class="p">,</span> <span class="n">desc1_psteps</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">pedges</span><span class="p">)</span>
                        <span class="k">assert</span> <span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                        <span class="k">if</span> <span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span><span class="p">:</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">)</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span><span class="p">[</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span><span class="p">[</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">][</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span><span class="p">[</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span><span class="p">[</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">][</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
                            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_store_path_edges</span><span class="p">:</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">[</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">[</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">][</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">[</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">:</span> <span class="n">desc1</span><span class="p">}</span>
                        <span class="k">for</span> <span class="n">c2</span> <span class="ow">in</span> <span class="n">children</span><span class="p">[</span><span class="n">cidx1</span><span class="o">+</span><span class="mi">1</span><span class="p">:]:</span>
                            <span class="k">for</span> <span class="n">desc2</span><span class="p">,</span> <span class="p">(</span><span class="n">desc2_plen</span><span class="p">,</span> <span class="n">desc2_psteps</span><span class="p">,</span> <span class="n">desc2_pedges</span><span class="p">)</span> <span class="ow">in</span> <span class="n">c2</span><span class="o">.</span><span class="n">desc_paths</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">desc2</span><span class="o">.</span><span class="n">taxon</span><span class="p">)</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">[</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">][</span><span class="n">desc2</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="n">c1</span><span class="o">.</span><span class="n">parent_node</span>
                                <span class="c1"># self._all_distinct_mapped_taxa_pairs.add( tuple([desc1.taxon, desc2.taxon]) )</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">_all_distinct_mapped_taxa_pairs</span><span class="o">.</span><span class="n">add</span><span class="p">(</span> <span class="nb">frozenset</span><span class="p">([</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">,</span> <span class="n">desc2</span><span class="o">.</span><span class="n">taxon</span><span class="p">])</span> <span class="p">)</span>
                                <span class="k">if</span> <span class="n">c2</span><span class="o">.</span><span class="n">edge_length</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                                    <span class="n">c2_edge_length</span> <span class="o">=</span> <span class="mf">0.0</span>
                                <span class="k">else</span><span class="p">:</span>
                                    <span class="n">c2_edge_length</span> <span class="o">=</span> <span class="n">c2</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span>
                                <span class="n">pat_dist</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">desc_paths</span><span class="p">[</span><span class="n">desc1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">desc2_plen</span> <span class="o">+</span> <span class="n">c2_edge_length</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span><span class="p">[</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">][</span><span class="n">desc2</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="n">pat_dist</span>
                                <span class="n">path_steps</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">desc_paths</span><span class="p">[</span><span class="n">desc1</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">desc2_psteps</span> <span class="o">+</span> <span class="mi">1</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span><span class="p">[</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">][</span><span class="n">desc2</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="n">path_steps</span>
                                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_store_path_edges</span><span class="p">:</span>
                                    <span class="n">pedges</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">desc_paths</span><span class="p">[</span><span class="n">desc1</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">c2</span><span class="o">.</span><span class="n">edge</span><span class="p">]</span> <span class="o">+</span> <span class="n">desc2_pedges</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                                    <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">[</span><span class="n">desc1</span><span class="o">.</span><span class="n">taxon</span><span class="p">][</span><span class="n">desc2</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="n">pedges</span>
                    <span class="k">del</span><span class="p">(</span><span class="n">c1</span><span class="o">.</span><span class="n">desc_paths</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_mirror_lookups</span><span class="p">()</span></div>

        <span class="c1"># assert self._tree_length == tree.length()</span>

    <span class="k">def</span> <span class="nf">compile_from_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">distances</span><span class="p">,</span> <span class="n">taxon_namespace</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">taxon_namespace</span>
        <span class="k">for</span> <span class="n">t1</span> <span class="ow">in</span> <span class="n">distances</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">t1</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span><span class="p">[</span><span class="n">t1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">t2</span> <span class="ow">in</span> <span class="n">distances</span><span class="p">[</span><span class="n">t1</span><span class="p">]:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span><span class="p">[</span><span class="n">t1</span><span class="p">][</span><span class="n">t2</span><span class="p">]</span> <span class="o">=</span> <span class="n">distances</span><span class="p">[</span><span class="n">t1</span><span class="p">][</span><span class="n">t2</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_mirror_lookups</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">_mirror_lookups</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">ddata</span> <span class="ow">in</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">,</span>
                <span class="p">):</span>
            <span class="k">for</span> <span class="n">taxon1</span> <span class="ow">in</span> <span class="n">ddata</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">taxon2</span> <span class="ow">in</span> <span class="n">ddata</span><span class="p">[</span><span class="n">taxon1</span><span class="p">]:</span>
                    <span class="c1"># assert taxon1 is not taxon2</span>
                    <span class="k">if</span> <span class="n">taxon2</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ddata</span><span class="p">:</span>
                        <span class="n">ddata</span><span class="p">[</span><span class="n">taxon2</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                    <span class="n">ddata</span><span class="p">[</span><span class="n">taxon2</span><span class="p">][</span><span class="n">taxon1</span><span class="p">]</span> <span class="o">=</span> <span class="n">ddata</span><span class="p">[</span><span class="n">taxon1</span><span class="p">][</span><span class="n">taxon2</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">taxon1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">taxon2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">taxon2</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">[</span><span class="n">taxon2</span><span class="p">][</span><span class="n">taxon1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">if</span> <span class="n">taxon1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">[</span><span class="n">taxon2</span><span class="p">]:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">[</span><span class="n">taxon2</span><span class="p">][</span><span class="n">taxon1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="nb">reversed</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">[</span><span class="n">taxon1</span><span class="p">][</span><span class="n">taxon2</span><span class="p">]))</span>

    <span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">o</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">o</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">return</span> <span class="p">(</span><span class="kc">True</span>
                <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span> <span class="o">==</span> <span class="n">o</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">)</span>
                <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_all_distinct_mapped_taxa_pairs</span> <span class="o">==</span> <span class="n">o</span><span class="o">.</span><span class="n">_all_distinct_mapped_taxa_pairs</span><span class="p">)</span>
                <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span> <span class="o">==</span> <span class="n">o</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span><span class="p">)</span>
                <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span> <span class="o">==</span> <span class="n">o</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span><span class="p">)</span>
                <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span> <span class="o">==</span> <span class="n">o</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">)</span>
                <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span> <span class="o">==</span> <span class="n">o</span><span class="o">.</span><span class="n">_mrca</span><span class="p">)</span>
                <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span> <span class="o">==</span> <span class="n">o</span><span class="o">.</span><span class="n">_tree_length</span><span class="p">)</span>
                <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span> <span class="o">==</span> <span class="n">o</span><span class="o">.</span><span class="n">_num_edges</span><span class="p">)</span>
                <span class="p">)</span>

    <span class="k">def</span> <span class="fm">__hash__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">id</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxon1</span><span class="p">,</span> <span class="n">taxon2</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">patristic_distance</span><span class="p">(</span><span class="n">taxon1</span><span class="p">,</span> <span class="n">taxon2</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__copy__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">clone</span><span class="p">()</span>

    <span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">taxon</span>

    <span class="k">def</span> <span class="nf">clone</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">o</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">()</span>
        <span class="n">o</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxon_namespace</span>
        <span class="n">o</span><span class="o">.</span><span class="n">_mapped_taxa</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">)</span>
        <span class="n">o</span><span class="o">.</span><span class="n">_all_distinct_mapped_taxa_pairs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_all_distinct_mapped_taxa_pairs</span><span class="p">)</span>
        <span class="n">o</span><span class="o">.</span><span class="n">_tree_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span>
        <span class="n">o</span><span class="o">.</span><span class="n">_num_edges</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span>
        <span class="k">for</span> <span class="n">src</span><span class="p">,</span> <span class="n">dest</span> <span class="ow">in</span> <span class="p">(</span>
                <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span><span class="p">,</span> <span class="n">o</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span><span class="p">,),</span>
                <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span><span class="p">,</span> <span class="n">o</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span><span class="p">,),</span>
                <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">,</span> <span class="n">o</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">,),</span>
                <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">,</span> <span class="n">o</span><span class="o">.</span><span class="n">_mrca</span><span class="p">,),</span>
                <span class="p">):</span>
            <span class="k">for</span> <span class="n">t1</span> <span class="ow">in</span> <span class="n">src</span><span class="p">:</span>
                <span class="n">dest</span><span class="p">[</span><span class="n">t1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">for</span> <span class="n">t2</span> <span class="ow">in</span> <span class="n">src</span><span class="p">[</span><span class="n">t1</span><span class="p">]:</span>
                    <span class="n">dest</span><span class="p">[</span><span class="n">t1</span><span class="p">][</span><span class="n">t2</span><span class="p">]</span> <span class="o">=</span> <span class="n">src</span><span class="p">[</span><span class="n">t1</span><span class="p">][</span><span class="n">t2</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">o</span>

<div class="viewcode-block" id="PhylogeneticDistanceMatrix.mrca">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.mrca">[docs]</a>
    <span class="k">def</span> <span class="nf">mrca</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxon1</span><span class="p">,</span> <span class="n">taxon2</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns MRCA of two taxon objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">[</span><span class="n">taxon1</span><span class="p">][</span><span class="n">taxon2</span><span class="p">]</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.distance">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.distance">[docs]</a>
    <span class="k">def</span> <span class="nf">distance</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">taxon1</span><span class="p">,</span>
            <span class="n">taxon2</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns distance between taxon1 and taxon2.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">is_weighted_edge_distances</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">patristic_distance</span><span class="p">(</span><span class="n">taxon1</span><span class="p">,</span> <span class="n">taxon2</span><span class="p">,</span> <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">path_edge_count</span><span class="p">(</span><span class="n">taxon1</span><span class="p">,</span> <span class="n">taxon2</span><span class="p">,</span> <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">)</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.patristic_distance">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.patristic_distance">[docs]</a>
    <span class="k">def</span> <span class="nf">patristic_distance</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxon1</span><span class="p">,</span> <span class="n">taxon2</span><span class="p">,</span> <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns patristic distance between two taxon objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">taxon1</span> <span class="ow">is</span> <span class="n">taxon2</span><span class="p">:</span>
            <span class="k">return</span> <span class="mf">0.0</span>
        <span class="n">d</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span><span class="p">[</span><span class="n">taxon1</span><span class="p">][</span><span class="n">taxon2</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">is_normalize_by_tree_size</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">d</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">d</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.path_edge_count">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.path_edge_count">[docs]</a>
    <span class="k">def</span> <span class="nf">path_edge_count</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxon1</span><span class="p">,</span> <span class="n">taxon2</span><span class="p">,</span> <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the number of edges between two taxon objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">taxon1</span> <span class="ow">is</span> <span class="n">taxon2</span><span class="p">:</span>
            <span class="k">return</span> <span class="mi">0</span>
        <span class="n">d</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span><span class="p">[</span><span class="n">taxon1</span><span class="p">][</span><span class="n">taxon2</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">is_normalize_by_tree_size</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">d</span><span class="p">)</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">d</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.path_edges">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.path_edges">[docs]</a>
    <span class="k">def</span> <span class="nf">path_edges</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxon1</span><span class="p">,</span> <span class="n">taxon2</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the edges between two taxon objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_edges</span><span class="p">[</span><span class="n">taxon1</span><span class="p">][</span><span class="n">taxon2</span><span class="p">]</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.distances">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.distances">[docs]</a>
    <span class="k">def</span> <span class="nf">distances</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns list of patristic distances.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">dmatrix</span><span class="p">,</span> <span class="n">normalization_factor</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_distance_matrix_and_normalization_factor</span><span class="p">(</span>
                <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="n">is_weighted_edge_distances</span><span class="p">,</span>
                <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">,</span>
                <span class="p">)</span>
        <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">t1</span><span class="p">,</span> <span class="n">t2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_all_distinct_mapped_taxa_pairs</span><span class="p">:</span>
            <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dmatrix</span><span class="p">[</span><span class="n">t1</span><span class="p">][</span><span class="n">t2</span><span class="p">]</span><span class="o">/</span><span class="n">normalization_factor</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">results</span></div>


    <span class="k">def</span> <span class="nf">max_pairwise_distance_taxa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">is_weighted_edge_distances</span><span class="p">:</span>
            <span class="n">dists</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">dists</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span>
        <span class="n">max_dist</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">max_dist_taxa</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">t1</span><span class="p">,</span> <span class="n">t2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_all_distinct_mapped_taxa_pairs</span><span class="p">:</span>
            <span class="n">pat_dist</span> <span class="o">=</span> <span class="n">dists</span><span class="p">[</span><span class="n">t1</span><span class="p">][</span><span class="n">t2</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">max_dist</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">pat_dist</span> <span class="o">&gt;</span> <span class="n">max_dist</span><span class="p">:</span>
                <span class="n">max_dist</span> <span class="o">=</span> <span class="n">pat_dist</span>
                <span class="n">max_dist_taxa</span> <span class="o">=</span> <span class="p">(</span><span class="n">t1</span><span class="p">,</span> <span class="n">t2</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">max_dist_taxa</span>

<div class="viewcode-block" id="PhylogeneticDistanceMatrix.sum_of_distances">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.sum_of_distances">[docs]</a>
    <span class="k">def</span> <span class="nf">sum_of_distances</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns sum of patristic distances on tree.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">distances</span><span class="p">(</span><span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="n">is_weighted_edge_distances</span><span class="p">,</span><span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">))</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.taxon_iter">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.taxon_iter">[docs]</a>
    <span class="k">def</span> <span class="nf">taxon_iter</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filter_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Iterates over taxa in matrix. Note that this could be a subset of the taxa in</span>
<span class="sd">        the associated taxon namespace.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">t1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">filter_fn</span> <span class="ow">or</span> <span class="n">filter_fn</span><span class="p">(</span><span class="n">t1</span><span class="p">):</span>
                <span class="k">yield</span> <span class="n">t1</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.distinct_taxon_pair_iter">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.distinct_taxon_pair_iter">[docs]</a>
    <span class="k">def</span> <span class="nf">distinct_taxon_pair_iter</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filter_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Iterates over all distinct pairs of taxa in matrix.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">t1</span><span class="p">,</span> <span class="n">t2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_all_distinct_mapped_taxa_pairs</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">filter_fn</span> <span class="ow">or</span> <span class="p">(</span><span class="n">filter_fn</span><span class="p">(</span><span class="n">t1</span><span class="p">)</span> <span class="ow">and</span> <span class="n">filter_fn</span><span class="p">(</span><span class="n">t2</span><span class="p">)):</span>
                <span class="k">yield</span> <span class="n">t1</span><span class="p">,</span> <span class="n">t2</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.mean_pairwise_distance">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.mean_pairwise_distance">[docs]</a>
    <span class="k">def</span> <span class="nf">mean_pairwise_distance</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">filter_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates the phylogenetic ecology statistic &quot;MPD&quot;[1,2] for the tree</span>
<span class="sd">        (only considering taxa for which ``filter_fn`` returns True when</span>
<span class="sd">        applied if ``filter_fn`` is specified).</span>

<span class="sd">        The mean pairwise distance (mpd) is given by:</span>

<span class="sd">            .. math::</span>
<span class="sd">                mpd = \frac{ \sum_{i}^{n} \sum_{j}^{n} \delta_{i,j} }{n \choose 2},</span>

<span class="sd">        where :math:`i \neq j`, :math:`\delta_{i,j}` is the phylogenetic</span>
<span class="sd">        distance between species :math:`i` and :math:`j`, and :math:`n` is the number</span>
<span class="sd">        of species in the sample.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filter_fn : function object or None</span>
<span class="sd">            If |None|, then all leaves will be considered. Otherwise should</span>
<span class="sd">            be a function object that takes a Taxon instance as an argument and</span>
<span class="sd">            returns |True| if it is to be included in the calculation or</span>
<span class="sd">            |False| otherwise.</span>
<span class="sd">            In trees sampled from multiple communites, ``filter_fn`` can be</span>
<span class="sd">            used to restrict the calculation to only one community based on</span>
<span class="sd">            some criteria.</span>
<span class="sd">        is_weighted_edge_distances : bool</span>
<span class="sd">            If |True| then the edge-weighted distance, i.e., considering edge</span>
<span class="sd">            lengths, is returned. Otherwise the the path steps or the number of</span>
<span class="sd">            edges rather then the sum of is_weighted_edge_distances edges, connecting two</span>
<span class="sd">            taxa is considered.</span>
<span class="sd">        is_normalize_by_tree_size : bool</span>
<span class="sd">            If |True| then the results are normalized by the total tree length</span>
<span class="sd">            or steps/edges (depending on whether edge-weighted or unweighted</span>
<span class="sd">            distances are used, respectively). Otherwise, raw distances are</span>
<span class="sd">            used.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        mpd : float</span>
<span class="sd">            The Mean Pairwise Distance (MPD) statistic for the daata.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        ::</span>

<span class="sd">            import dendropy</span>
<span class="sd">            tree = dendropy.Tree.get(path=&quot;data.nex&quot;,</span>
<span class="sd">                    schema=&quot;nexus&quot;)</span>
<span class="sd">            pdm = dendropy.PhylogeneticDistanceMatrix(tree)</span>

<span class="sd">            # consider all tips</span>
<span class="sd">            mpd1 = pdm.mean_pairwise_distance()</span>

<span class="sd">            # only tips within the same community, based on the node annotation</span>
<span class="sd">            # &quot;community&quot;</span>
<span class="sd">            mpds_by_community = {}</span>
<span class="sd">            for community_label in (&quot;1&quot;, &quot;2&quot;, &quot;3&quot;,):</span>
<span class="sd">                filter_fn = lambda x: x.annotations[&quot;community&quot;] == community_label</span>
<span class="sd">                mpd = pdm.mean_pairwise_distance(filter_fn=filter_fn)</span>
<span class="sd">                mpds_by_community[community_label] = mpd</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>

<span class="sd">        [1] Webb, C.O. 2000. Exploring the phylogenetic structure of</span>
<span class="sd">        ecological communities: An example for rainforest trees. The</span>
<span class="sd">        American Naturalist 156: 145-155.</span>

<span class="sd">        [2] Swenson, N.G. Functional and Phylogenetic Ecology in R.</span>


<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">comparison_regime</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">distinct_taxon_pair_iter</span><span class="p">(</span><span class="n">filter_fn</span><span class="o">=</span><span class="n">filter_fn</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_mean_pairwise_distance</span><span class="p">(</span>
                <span class="n">comparison_regime</span><span class="o">=</span><span class="n">comparison_regime</span><span class="p">,</span>
                <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="n">is_weighted_edge_distances</span><span class="p">,</span>
                <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">,</span>
                <span class="p">)</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.mean_nearest_taxon_distance">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.mean_nearest_taxon_distance">[docs]</a>
    <span class="k">def</span> <span class="nf">mean_nearest_taxon_distance</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">filter_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates the phylogenetic ecology statistic &quot;MNTD&quot;[1,2] for the tree</span>
<span class="sd">        (only considering taxa for which ``filter_fn`` returns True when</span>
<span class="sd">        applied if ``filter_fn`` is specified).</span>

<span class="sd">        The mean nearest taxon distance (mntd) is given by:</span>

<span class="sd">            .. math::</span>
<span class="sd">                mntd = \frac{ \sum_{i}^{n} min(\delta_{i,j}) }{n},</span>

<span class="sd">        where :math:`i \neq j`, :math:`\delta_{i,j}` is the phylogenetic</span>
<span class="sd">        distance between species :math:`i` and :math:`j`, and :math:`n` is the number</span>
<span class="sd">        of species in the sample.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filter_fn : function object or None</span>
<span class="sd">            If |None|, then all leaves will be considered. Otherwise should</span>
<span class="sd">            be a function object that takes a Taxon instance as an argument and</span>
<span class="sd">            returns |True| if it is to be included in the calculation or</span>
<span class="sd">            |False| otherwise.</span>
<span class="sd">            In trees sampled from multiple communites, ``filter_fn`` can be</span>
<span class="sd">            used to restrict the calculation to only one community based on</span>
<span class="sd">            some criteria.</span>
<span class="sd">        is_weighted_edge_distances : bool</span>
<span class="sd">            If |True| then the edge-weighted distance, i.e., considering edge</span>
<span class="sd">            lengths, is returned. Otherwise the the path steps or the number of</span>
<span class="sd">            edges rather then the sum of is_weighted_edge_distances edges, connecting two</span>
<span class="sd">            taxa is considered.</span>
<span class="sd">        is_normalize_by_tree_size : bool</span>
<span class="sd">            If |True| then the results are normalized by the total tree length</span>
<span class="sd">            or steps/edges (depending on whether edge-weighted or unweighted</span>
<span class="sd">            distances are used, respectively). Otherwise, raw distances are</span>
<span class="sd">            used.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        mntd : float</span>
<span class="sd">            The Mean Nearest Taxon Distance (MNTD) statistic for the daata.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        ::</span>

<span class="sd">            import dendropy</span>
<span class="sd">            tree = dendropy.Tree.get(path=&quot;data.nex&quot;,</span>
<span class="sd">                    schema=&quot;nexus&quot;)</span>
<span class="sd">            pdm = dendropy.PhylogeneticDistanceMatrix(tree)</span>

<span class="sd">            # consider all tips</span>
<span class="sd">            mntd = pdm.mean_nearest_taxon_distance()</span>

<span class="sd">            # only tips within the same community, based on the node annotation</span>
<span class="sd">            # &quot;community&quot;</span>
<span class="sd">            mntds_by_community = {}</span>
<span class="sd">            for community_label in (&quot;1&quot;, &quot;2&quot;, &quot;3&quot;,):</span>
<span class="sd">                filter_fn = lambda x: x.annotations[&quot;community&quot;] == community_label</span>
<span class="sd">                mntd = pdm.mean_pairwise_distance(filter_fn=filter_fn)</span>
<span class="sd">                mntds_by_community[community_label] = mntd</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>

<span class="sd">        [1] Webb, C.O. 2000. Exploring the phylogenetic structure of</span>
<span class="sd">        ecological communities: An example for rainforest trees. The</span>
<span class="sd">        American Naturalist 156: 145-155.</span>

<span class="sd">        [2] Swenson, N.G. Functional and Phylogenetic Ecology in R.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">comparison_regime</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_taxon_to_all_other_taxa_comparisons</span><span class="p">(</span><span class="n">filter_fn</span><span class="o">=</span><span class="n">filter_fn</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_mean_nearest_taxon_distance</span><span class="p">(</span>
                <span class="n">comparison_regime</span><span class="o">=</span><span class="n">comparison_regime</span><span class="p">,</span>
                <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="n">is_weighted_edge_distances</span><span class="p">,</span>
                <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">,)</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.standardized_effect_size_mean_pairwise_distance">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.standardized_effect_size_mean_pairwise_distance">[docs]</a>
    <span class="k">def</span> <span class="nf">standardized_effect_size_mean_pairwise_distance</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">assemblage_memberships</span><span class="p">,</span>
            <span class="n">num_randomization_replicates</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">is_skip_single_taxon_assemblages</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">null_model_type</span><span class="o">=</span><span class="s2">&quot;taxa.label&quot;</span><span class="p">,</span>
            <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the standardized effect size value for the MPD statistic under</span>
<span class="sd">        a null model under various community compositions.</span>

<span class="sd">        The S.E.S. is given by:</span>

<span class="sd">            .. math::</span>
<span class="sd">                SES(statistic) = \frac{observed - mean(model_{null})}{sd(model_{null})}</span>

<span class="sd">        This removes any bias associated with the decrease in variance in the</span>
<span class="sd">        MPD statistic value as species richness increases to the point where</span>
<span class="sd">        communities become saturated. Equivalent to -1 times the Nearest</span>
<span class="sd">        Relative Index (NRI) when using phylogenetic distances.</span>

<span class="sd">        In contrast to the function calculating the non-standardized effect</span>
<span class="sd">        size version of this statistic, which uses filter function to specify</span>
<span class="sd">        the subset of taxa to be considerd, here a collection of (multiple)</span>
<span class="sd">        sets of taxa constituting a community is specified. This to allow</span>
<span class="sd">        calculation of the null model statistic across all community sets for</span>
<span class="sd">        each randomization replicate.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        assemblage_memberships : iterable of iterable of |Taxon| objects</span>
<span class="sd">            A collection of collections, e.g. a list of sets, with the elements</span>
<span class="sd">            of each set being |Taxon| instances. Each set specifies the</span>
<span class="sd">            composition of a community. The standardized effect size of this</span>
<span class="sd">            statistic will be calculated for each community as specified by a</span>
<span class="sd">            set of |Taxon| instances.</span>
<span class="sd">        num_randomization_replicates : int</span>
<span class="sd">            Number of randomization replicates.</span>
<span class="sd">        is_weighted_edge_distances: bool</span>
<span class="sd">            If ``True`` then edge lengths will be considered for distances.</span>
<span class="sd">            Otherwise, just the number of edges.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        r : list of results</span>
<span class="sd">            A list of results, with each result corresponding to a community</span>
<span class="sd">            set given in ``assemblage_memberships``. Each result consists of a named</span>
<span class="sd">            tuple with the following elements:</span>

<span class="sd">                -   obs       : the observed value of the statistic</span>
<span class="sd">                -   null_model_mean : the mean value of the statistic under the null</span>
<span class="sd">                                model</span>
<span class="sd">                -   null_model_sd   : the standard deviation of the statistic under</span>
<span class="sd">                                the null model</span>
<span class="sd">                -   z         : the standardized effect value of the statistic</span>
<span class="sd">                                (= SES as defined in [1] above)</span>
<span class="sd">                -   p         : the p-value of the observed value of the</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        ::</span>

<span class="sd">            import dendropy</span>
<span class="sd">            tree = dendropy.Tree.get_from_path(</span>
<span class="sd">                    src=&quot;data/community.tree.newick&quot;,</span>
<span class="sd">                    schema=&quot;newick&quot;,</span>
<span class="sd">                    rooting=&quot;force-rooted&quot;)</span>
<span class="sd">            pdm = tree.phylogenetic_distance_matrix()</span>
<span class="sd">            assemblage_membership_definitions = pdm.assemblage_membership_definitions_from_csv(&quot;data/comm1.csv&quot;)</span>
<span class="sd">            results = pdm.standardized_effect_size_mean_pairwise_distance(assemblage_memberships=assemblage_membership_definitions.values())</span>
<span class="sd">            print(results)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">assemblage_memberships</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">assemblage_memberships</span> <span class="o">=</span> <span class="p">[</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">)</span> <span class="p">]</span>
        <span class="n">comparison_regimes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">assemblage_membership</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">assemblage_memberships</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">assemblage_membership</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">is_skip_single_taxon_assemblages</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">error</span><span class="o">.</span><span class="n">SingleTaxonAssemblageException</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">assemblage_membership</span><span class="p">))</span>
            <span class="n">filter_fn</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">taxon</span><span class="p">:</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">assemblage_membership</span>
            <span class="n">comparison_regime</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">distinct_taxon_pair_iter</span><span class="p">(</span><span class="n">filter_fn</span><span class="o">=</span><span class="n">filter_fn</span><span class="p">))</span>
            <span class="n">comparison_regimes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">comparison_regime</span><span class="p">)</span>
        <span class="n">results</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_standardized_effect_size</span><span class="p">(</span>
                <span class="n">statisticf_name</span><span class="o">=</span><span class="s2">&quot;_calculate_mean_pairwise_distance&quot;</span><span class="p">,</span>
                <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="n">is_weighted_edge_distances</span><span class="p">,</span>
                <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">,</span>
                <span class="n">comparison_regimes</span><span class="o">=</span><span class="n">comparison_regimes</span><span class="p">,</span>
                <span class="n">null_model_type</span><span class="o">=</span><span class="n">null_model_type</span><span class="p">,</span>
                <span class="n">num_randomization_replicates</span><span class="o">=</span><span class="n">num_randomization_replicates</span><span class="p">,</span>
                <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">results</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.standardized_effect_size_mean_nearest_taxon_distance">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.standardized_effect_size_mean_nearest_taxon_distance">[docs]</a>
    <span class="k">def</span> <span class="nf">standardized_effect_size_mean_nearest_taxon_distance</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">assemblage_memberships</span><span class="p">,</span>
            <span class="n">num_randomization_replicates</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">is_skip_single_taxon_assemblages</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">null_model_type</span><span class="o">=</span><span class="s2">&quot;taxa.label&quot;</span><span class="p">,</span>
            <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the standardized effect size value for the MNTD statistic under</span>
<span class="sd">        a null model under various community compositions.</span>

<span class="sd">        The S.E.S. is given by:</span>

<span class="sd">            .. math::</span>
<span class="sd">                SES(statistic) = \frac{observed - mean(model_{null})}{sd(model_{null})}</span>

<span class="sd">        This removes any bias associated with the decrease in variance in the</span>
<span class="sd">        MPD statistic value as species richness increases to the point where</span>
<span class="sd">        communities become saturated. Equivalent to -1 times the Nearest Taxon</span>
<span class="sd">        Index when using phylogenetic distances.</span>

<span class="sd">        In contrast to the function calculating the non-standardized effect</span>
<span class="sd">        size version of this statistic, which uses filter function to specify</span>
<span class="sd">        the subset of taxa to be considerd, here a collection of (multiple)</span>
<span class="sd">        sets of taxa constituting a community is specified. This to allow</span>
<span class="sd">        calculation of the null model statistic across all community sets for</span>
<span class="sd">        each randomization replicate.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        assemblage_memberships : iterable of iterable of |Taxon| objects</span>
<span class="sd">            A collection of collections, e.g. a list of sets, with the elements</span>
<span class="sd">            of each set being |Taxon| instances. Each set specifies the</span>
<span class="sd">            composition of a community. The standardized effect size of this</span>
<span class="sd">            statistic will be calculated for each community as specified by a</span>
<span class="sd">            set of |Taxon| instances.</span>
<span class="sd">        num_randomization_replicates : int</span>
<span class="sd">            Number of randomization replicates.</span>
<span class="sd">        is_weighted_edge_distances: bool</span>
<span class="sd">            If ``True`` then edge lengths will be considered for distances.</span>
<span class="sd">            Otherwise, just the number of edges.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        r : list of results</span>
<span class="sd">            A list of results, with each result corresponding to a community</span>
<span class="sd">            set given in ``assemblage_memberships``. Each result consists of a named</span>
<span class="sd">            tuple with the following elements:</span>

<span class="sd">                -   obs       : the observed value of the statistic</span>
<span class="sd">                -   null_model_mean : the mean value of the statistic under the null</span>
<span class="sd">                                model</span>
<span class="sd">                -   null_model_sd   : the standard deviation of the statistic under</span>
<span class="sd">                                the null model</span>
<span class="sd">                -   z         : the standardized effect value of the statistic</span>
<span class="sd">                                (= SES as defined in [1] above)</span>
<span class="sd">                -   p         : the p-value of the observed value of the</span>
<span class="sd">                                statistic under the null model.</span>
<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        ::</span>

<span class="sd">            import dendropy</span>
<span class="sd">            tree = dendropy.Tree.get_from_path(</span>
<span class="sd">                    src=&quot;data/community.tree.newick&quot;,</span>
<span class="sd">                    schema=&quot;newick&quot;,</span>
<span class="sd">                    rooting=&quot;force-rooted&quot;)</span>
<span class="sd">            pdm = dendropy.PhylogeneticDistanceMatrix.from_tree(tree)</span>
<span class="sd">            assemblage_memberships = pdm.assemblage_membership_definitions_from_csv(&quot;data/comm1.csv&quot;)</span>
<span class="sd">            results = pdm.standardized_effect_size_mean_nearest_taxon_distance(assemblage_memberships=assemblage_memberships)</span>
<span class="sd">            print(results)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">assemblage_memberships</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">assemblage_memberships</span> <span class="o">=</span> <span class="p">[</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">)</span> <span class="p">]</span>
        <span class="n">comparison_regimes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">assemblage_membership</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">assemblage_memberships</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">assemblage_membership</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">is_skip_single_taxon_assemblages</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">error</span><span class="o">.</span><span class="n">SingleTaxonAssemblageException</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">assemblage_membership</span><span class="p">))</span>
            <span class="n">filter_fn</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">taxon</span><span class="p">:</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">assemblage_membership</span>
            <span class="n">comparison_regime</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_taxon_to_all_other_taxa_comparisons</span><span class="p">(</span><span class="n">filter_fn</span><span class="o">=</span><span class="n">filter_fn</span><span class="p">)</span>
            <span class="n">comparison_regimes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">comparison_regime</span><span class="p">)</span>
        <span class="n">results</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_standardized_effect_size</span><span class="p">(</span>
                <span class="n">statisticf_name</span><span class="o">=</span><span class="s2">&quot;_calculate_mean_nearest_taxon_distance&quot;</span><span class="p">,</span>
                <span class="n">comparison_regimes</span><span class="o">=</span><span class="n">comparison_regimes</span><span class="p">,</span>
                <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="n">is_weighted_edge_distances</span><span class="p">,</span>
                <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">,</span>
                <span class="n">null_model_type</span><span class="o">=</span><span class="n">null_model_type</span><span class="p">,</span>
                <span class="n">num_randomization_replicates</span><span class="o">=</span><span class="n">num_randomization_replicates</span><span class="p">,</span>
                <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">results</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.shuffle_taxa">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.shuffle_taxa">[docs]</a>
    <span class="k">def</span> <span class="nf">shuffle_taxa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">is_shuffle_phylogenetic_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_shuffle_phylogenetic_path_steps</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_shuffle_mrca</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Randomly shuffles taxa in-situ.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">rng</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">rng</span> <span class="o">=</span> <span class="n">GLOBAL_RNG</span>
        <span class="n">reordered_taxa</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">)</span>
        <span class="n">rng</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">reordered_taxa</span><span class="p">)</span>
        <span class="n">current_to_shuffled_taxon_map</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">,</span> <span class="n">reordered_taxa</span><span class="p">))</span>
        <span class="n">to_shuffle</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">is_shuffle_phylogenetic_distances</span><span class="p">:</span>
            <span class="n">to_shuffle</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;_taxon_phylogenetic_distances&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">is_shuffle_phylogenetic_path_steps</span><span class="p">:</span>
            <span class="n">to_shuffle</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;_taxon_phylogenetic_path_steps&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">is_shuffle_mrca</span><span class="p">:</span>
            <span class="n">to_shuffle</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;_mrca&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">attr_name</span> <span class="ow">in</span> <span class="n">to_shuffle</span><span class="p">:</span>
            <span class="n">src</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">attr_name</span><span class="p">)</span>
            <span class="n">dest</span> <span class="o">=</span> <span class="p">{}</span>

            <span class="c1">## 5m8.076s</span>
            <span class="c1"># for t1, t2 in self._all_distinct_mapped_taxa_pairs:</span>
            <span class="c1">#     x1 = current_to_shuffled_taxon_map[t1]</span>
            <span class="c1">#     x2 = current_to_shuffled_taxon_map[t2]</span>
            <span class="c1">#     d = src[t1][t2]</span>
            <span class="c1">#     try:</span>
            <span class="c1">#         dest[x1][x2] = d</span>
            <span class="c1">#     except KeyError:</span>
            <span class="c1">#         dest[x1] = {x2: d}</span>
            <span class="c1">#         if t1 in src[t1]:</span>
            <span class="c1">#             dest[x1][x1] = src[t1][t1]</span>
            <span class="c1">#     try:</span>
            <span class="c1">#         dest[x2][x1] = d</span>
            <span class="c1">#     except KeyError:</span>
            <span class="c1">#         dest[x2] = {x1: d}</span>
            <span class="c1">#         if t2 in src[t2]:</span>
            <span class="c1">#             dest[x2][x2] = src[t2][t2]</span>
            <span class="c1"># setattr(self, attr_name, dest)</span>

            <span class="c1"># 4m48.025s</span>
            <span class="k">for</span> <span class="n">t1</span> <span class="ow">in</span> <span class="n">src</span><span class="p">:</span>
                <span class="n">x1</span> <span class="o">=</span> <span class="n">current_to_shuffled_taxon_map</span><span class="p">[</span><span class="n">t1</span><span class="p">]</span>
                <span class="n">dest</span><span class="p">[</span><span class="n">x1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">for</span> <span class="n">t2</span> <span class="ow">in</span> <span class="n">src</span><span class="p">[</span><span class="n">t1</span><span class="p">]:</span>
                    <span class="n">x2</span> <span class="o">=</span> <span class="n">current_to_shuffled_taxon_map</span><span class="p">[</span><span class="n">t2</span><span class="p">]</span>
                    <span class="n">dest</span><span class="p">[</span><span class="n">x1</span><span class="p">][</span><span class="n">x2</span><span class="p">]</span> <span class="o">=</span> <span class="n">src</span><span class="p">[</span><span class="n">t1</span><span class="p">][</span><span class="n">t2</span><span class="p">]</span>
            <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">attr_name</span><span class="p">,</span> <span class="n">dest</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">current_to_shuffled_taxon_map</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.nj_tree">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.nj_tree">[docs]</a>
    <span class="k">def</span> <span class="nf">nj_tree</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">tree_factory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns an Neighbor-Joining (NJ) tree based on the distances in the matrix.</span>

<span class="sd">        Calculates and returns a tree under the Neighbor-Joining algorithm of</span>
<span class="sd">        Saitou and Nei (1987) for the data in the matrix.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        is_weighted_edge_distances: bool</span>
<span class="sd">            If ``True`` then edge lengths will be considered for distances.</span>
<span class="sd">            Otherwise, just the number of edges.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        t : |Tree|</span>
<span class="sd">            A |Tree| instance corresponding to the Neighbor-Joining (NJ) tree</span>
<span class="sd">            for this data.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        ::</span>

<span class="sd">            import dendropy</span>

<span class="sd">            # Read data from a CSV file into a PhylogeneticDistanceMatrix</span>
<span class="sd">            # object</span>
<span class="sd">            with open(&quot;distance_matrix.csv&quot;) as src:</span>
<span class="sd">                pdm = dendropy.PhylogeneticDistanceMatrix.from_csv(</span>
<span class="sd">                        src,</span>
<span class="sd">                        is_first_row_column_names=True,</span>
<span class="sd">                        is_first_column_row_names=True,</span>
<span class="sd">                        is_allow_new_taxa=True,</span>
<span class="sd">                        delimiter=&quot;,&quot;,</span>
<span class="sd">                        )</span>

<span class="sd">            # Calculate the tree</span>
<span class="sd">            nj_tree = pdm.nj_tree()</span>

<span class="sd">            # Print it</span>
<span class="sd">            print(nj_tree.as_string(&quot;nexus&quot;))</span>


<span class="sd">        References</span>
<span class="sd">        ----------</span>
<span class="sd">        Saitou, N. and Nei, M. (1987) The neighbor-joining method: a new method</span>
<span class="sd">        for reconstructing phylogenetic trees. Molecular Biology and Evolution,</span>
<span class="sd">        4: 406-425.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">is_weighted_edge_distances</span><span class="p">:</span>
            <span class="n">original_dmatrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">original_dmatrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span>
        <span class="k">if</span> <span class="n">tree_factory</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">tree_factory</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span>
        <span class="n">tree</span> <span class="o">=</span> <span class="n">tree_factory</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">)</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">is_rooted</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="c1"># initialize node pool</span>
        <span class="n">node_pool</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">t1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">:</span>
            <span class="n">nd</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">node_factory</span><span class="p">()</span>
            <span class="n">nd</span><span class="o">.</span><span class="n">taxon</span> <span class="o">=</span> <span class="n">t1</span>
            <span class="n">nd</span><span class="o">.</span><span class="n">_nj_distances</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">node_pool</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>

        <span class="c1"># initialize factor</span>
        <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">)</span>

        <span class="c1"># cache calculations</span>
        <span class="k">for</span> <span class="n">nd1</span> <span class="ow">in</span> <span class="n">node_pool</span><span class="p">:</span>
            <span class="n">nd1</span><span class="o">.</span><span class="n">_nj_xsub</span> <span class="o">=</span> <span class="mf">0.0</span>
            <span class="k">for</span> <span class="n">nd2</span> <span class="ow">in</span> <span class="n">node_pool</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">nd1</span> <span class="ow">is</span> <span class="n">nd2</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">d</span> <span class="o">=</span> <span class="n">original_dmatrix</span><span class="p">[</span><span class="n">nd1</span><span class="o">.</span><span class="n">taxon</span><span class="p">][</span><span class="n">nd2</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span>
                <span class="n">nd1</span><span class="o">.</span><span class="n">_nj_distances</span><span class="p">[</span><span class="n">nd2</span><span class="p">]</span> <span class="o">=</span> <span class="n">d</span>
                <span class="n">nd1</span><span class="o">.</span><span class="n">_nj_xsub</span> <span class="o">+=</span> <span class="n">d</span>

        <span class="k">while</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>

            <span class="c1"># calculate the Q-matrix</span>
            <span class="n">min_q</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">nodes_to_join</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">for</span> <span class="n">idx1</span><span class="p">,</span> <span class="n">nd1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">node_pool</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]):</span>
                <span class="k">for</span> <span class="n">idx2</span><span class="p">,</span> <span class="n">nd2</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">node_pool</span><span class="p">[</span><span class="n">idx1</span><span class="o">+</span><span class="mi">1</span><span class="p">:]):</span>
                    <span class="n">v1</span> <span class="o">=</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span> <span class="o">*</span> <span class="n">nd1</span><span class="o">.</span><span class="n">_nj_distances</span><span class="p">[</span><span class="n">nd2</span><span class="p">]</span>
                    <span class="n">qvalue</span> <span class="o">=</span> <span class="n">v1</span> <span class="o">-</span> <span class="n">nd1</span><span class="o">.</span><span class="n">_nj_xsub</span> <span class="o">-</span> <span class="n">nd2</span><span class="o">.</span><span class="n">_nj_xsub</span>
                    <span class="k">if</span> <span class="n">min_q</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">qvalue</span> <span class="o">&lt;</span> <span class="n">min_q</span><span class="p">:</span>
                        <span class="n">min_q</span> <span class="o">=</span> <span class="n">qvalue</span>
                        <span class="n">nodes_to_join</span> <span class="o">=</span> <span class="p">(</span><span class="n">nd1</span><span class="p">,</span> <span class="n">nd2</span><span class="p">)</span>

            <span class="c1"># create the new node</span>
            <span class="n">new_node</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">node_factory</span><span class="p">()</span>

            <span class="c1"># attach it to the tree</span>
            <span class="k">for</span> <span class="n">node_to_join</span> <span class="ow">in</span> <span class="n">nodes_to_join</span><span class="p">:</span>
                <span class="n">new_node</span><span class="o">.</span><span class="n">add_child</span><span class="p">(</span><span class="n">node_to_join</span><span class="p">)</span>
                <span class="n">node_pool</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">node_to_join</span><span class="p">)</span>

            <span class="c1"># calculate the distances for the new node</span>
            <span class="n">new_node</span><span class="o">.</span><span class="n">_nj_distances</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">new_node</span><span class="o">.</span><span class="n">_nj_xsub</span> <span class="o">=</span> <span class="mf">0.0</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">node_pool</span><span class="p">:</span>
                <span class="c1"># actual node-to-node distances</span>
                <span class="n">v1</span> <span class="o">=</span> <span class="mf">0.0</span>
                <span class="k">for</span> <span class="n">node_to_join</span> <span class="ow">in</span> <span class="n">nodes_to_join</span><span class="p">:</span>
                    <span class="n">v1</span> <span class="o">+=</span> <span class="n">node</span><span class="o">.</span><span class="n">_nj_distances</span><span class="p">[</span><span class="n">node_to_join</span><span class="p">]</span>
                <span class="n">v3</span> <span class="o">=</span> <span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">_nj_distances</span><span class="p">[</span><span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
                <span class="n">dist</span> <span class="o">=</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="p">(</span><span class="n">v1</span> <span class="o">-</span> <span class="n">v3</span><span class="p">)</span>
                <span class="n">new_node</span><span class="o">.</span><span class="n">_nj_distances</span><span class="p">[</span><span class="n">node</span><span class="p">]</span> <span class="o">=</span> <span class="n">dist</span>
                <span class="n">node</span><span class="o">.</span><span class="n">_nj_distances</span><span class="p">[</span><span class="n">new_node</span><span class="p">]</span> <span class="o">=</span> <span class="n">dist</span>

                <span class="c1"># Adjust/recalculate the values needed for the Q-matrix</span>
                <span class="c1"># calculations</span>
                <span class="n">new_node</span><span class="o">.</span><span class="n">_nj_xsub</span> <span class="o">+=</span> <span class="n">dist</span>
                <span class="n">node</span><span class="o">.</span><span class="n">_nj_xsub</span> <span class="o">+=</span> <span class="n">dist</span>
                <span class="k">for</span> <span class="n">node_to_join</span> <span class="ow">in</span> <span class="n">nodes_to_join</span><span class="p">:</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">_nj_xsub</span> <span class="o">-=</span> <span class="n">node_to_join</span><span class="o">.</span><span class="n">_nj_distances</span><span class="p">[</span><span class="n">node</span><span class="p">]</span>

            <span class="c1"># calculate the branch lengths</span>
            <span class="k">if</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
                <span class="n">v1</span> <span class="o">=</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">_nj_distances</span><span class="p">[</span><span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
                <span class="n">v4</span>  <span class="o">=</span> <span class="mf">1.0</span><span class="o">/</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="p">(</span><span class="n">n</span><span class="o">-</span><span class="mi">2</span><span class="p">))</span> <span class="o">*</span> <span class="p">(</span><span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">_nj_xsub</span> <span class="o">-</span> <span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">_nj_xsub</span><span class="p">)</span>
                <span class="n">delta_f</span> <span class="o">=</span> <span class="n">v1</span> <span class="o">+</span> <span class="n">v4</span>
                <span class="n">delta_g</span> <span class="o">=</span> <span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">_nj_distances</span><span class="p">[</span><span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">-</span> <span class="n">delta_f</span>
                <span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">delta_f</span>
                <span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">delta_g</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">d</span> <span class="o">=</span> <span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">_nj_distances</span><span class="p">[</span><span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
                <span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">d</span> <span class="o">/</span> <span class="mi">2</span>
                <span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">d</span> <span class="o">/</span> <span class="mi">2</span>

            <span class="c1"># clean up</span>
            <span class="k">for</span> <span class="n">node_to_join</span> <span class="ow">in</span> <span class="n">nodes_to_join</span><span class="p">:</span>
                <span class="k">del</span> <span class="n">node_to_join</span><span class="o">.</span><span class="n">_nj_distances</span>
                <span class="k">del</span> <span class="n">node_to_join</span><span class="o">.</span><span class="n">_nj_xsub</span>

            <span class="c1"># add the new node to the pool of nodes</span>
            <span class="n">node_pool</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_node</span><span class="p">)</span>

            <span class="c1"># adjust count</span>
            <span class="n">n</span> <span class="o">-=</span> <span class="mi">1</span>

        <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span> <span class="o">=</span> <span class="n">node_pool</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">del</span> <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">_nj_distances</span>
        <span class="k">del</span> <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">_nj_xsub</span>
        <span class="k">return</span> <span class="n">tree</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.upgma_tree">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.upgma_tree">[docs]</a>
    <span class="k">def</span> <span class="nf">upgma_tree</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">tree_factory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns an Unweighted Pair Group Method with Arithmetic Mean (UPGMA) tree</span>
<span class="sd">        based on the distances in the matrix.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        is_weighted_edge_distances: bool</span>
<span class="sd">            If ``True`` then edge lengths will be considered for distances.</span>
<span class="sd">            Otherwise, just the number of edges.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        t : |Tree|</span>
<span class="sd">            A |Tree| instance corresponding to the UPGMA tree</span>
<span class="sd">            for this data.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        ::</span>

<span class="sd">            import dendropy</span>

<span class="sd">            # Read data from a CSV file into a PhylogeneticDistanceMatrix</span>
<span class="sd">            # object</span>
<span class="sd">            with open(&quot;distance_matrix.csv&quot;) as src:</span>
<span class="sd">                pdm = dendropy.PhylogeneticDistanceMatrix.from_csv(</span>
<span class="sd">                        src,</span>
<span class="sd">                        is_first_row_column_names=True,</span>
<span class="sd">                        is_first_column_row_names=True,</span>
<span class="sd">                        is_allow_new_taxa=True,</span>
<span class="sd">                        delimiter=&quot;,&quot;,</span>
<span class="sd">                        )</span>

<span class="sd">            # Calculate the tree</span>
<span class="sd">            upgma_tree = pdm.upgma_tree()</span>

<span class="sd">            # Print it</span>
<span class="sd">            print(upgma_tree.as_string(&quot;nexus&quot;))</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">is_weighted_edge_distances</span><span class="p">:</span>
            <span class="n">original_dmatrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">original_dmatrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span>
        <span class="k">if</span> <span class="n">tree_factory</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">tree_factory</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span>
        <span class="n">tree</span> <span class="o">=</span> <span class="n">tree_factory</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">)</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">is_rooted</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">node_pool</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">t1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">:</span>
            <span class="n">nd</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">node_factory</span><span class="p">()</span>
            <span class="n">nd</span><span class="o">.</span><span class="n">taxon</span> <span class="o">=</span> <span class="n">t1</span>
            <span class="n">nd</span><span class="o">.</span><span class="n">_upgma_cluster</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">nd</span><span class="p">])</span>
            <span class="n">nd</span><span class="o">.</span><span class="n">_upgma_distance_from_tip</span> <span class="o">=</span> <span class="mf">0.0</span>
            <span class="n">nd</span><span class="o">.</span><span class="n">_upgma_distances</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">node_pool</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nd</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">idx1</span><span class="p">,</span> <span class="n">nd1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">node_pool</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]):</span>
            <span class="k">for</span> <span class="n">idx2</span><span class="p">,</span> <span class="n">nd2</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">node_pool</span><span class="p">[</span><span class="n">idx1</span><span class="o">+</span><span class="mi">1</span><span class="p">:]):</span>
                <span class="n">d</span> <span class="o">=</span> <span class="n">original_dmatrix</span><span class="p">[</span><span class="n">nd1</span><span class="o">.</span><span class="n">taxon</span><span class="p">][</span><span class="n">nd2</span><span class="o">.</span><span class="n">taxon</span><span class="p">]</span>
                <span class="n">nd1</span><span class="o">.</span><span class="n">_upgma_distances</span><span class="p">[</span><span class="n">nd2</span><span class="p">]</span> <span class="o">=</span> <span class="n">d</span>
                <span class="n">nd2</span><span class="o">.</span><span class="n">_upgma_distances</span><span class="p">[</span><span class="n">nd1</span><span class="p">]</span> <span class="o">=</span> <span class="n">d</span>
        <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">node_pool</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">min_distance</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">nodes_to_join</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">for</span> <span class="n">idx1</span><span class="p">,</span> <span class="n">nd1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">node_pool</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]):</span>
                <span class="k">for</span> <span class="n">idx2</span><span class="p">,</span> <span class="n">nd2</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">node_pool</span><span class="p">[</span><span class="n">idx1</span><span class="o">+</span><span class="mi">1</span><span class="p">:]):</span>
                    <span class="n">d</span> <span class="o">=</span> <span class="n">nd1</span><span class="o">.</span><span class="n">_upgma_distances</span><span class="p">[</span><span class="n">nd2</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">min_distance</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">d</span> <span class="o">&lt;</span> <span class="n">min_distance</span><span class="p">:</span>
                        <span class="n">nodes_to_join</span> <span class="o">=</span> <span class="p">(</span><span class="n">nd1</span><span class="p">,</span> <span class="n">nd2</span><span class="p">)</span>
                        <span class="n">min_distance</span> <span class="o">=</span> <span class="n">d</span>
            <span class="n">new_node</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">node_factory</span><span class="p">()</span>
            <span class="n">new_node</span><span class="o">.</span><span class="n">_upgma_cluster</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="n">new_node</span><span class="o">.</span><span class="n">_upgma_distances</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">elen</span> <span class="o">=</span> <span class="n">min_distance</span> <span class="o">/</span> <span class="mf">2.0</span>
            <span class="k">for</span> <span class="n">node_to_join</span> <span class="ow">in</span> <span class="n">nodes_to_join</span><span class="p">:</span>
                <span class="n">new_node</span><span class="o">.</span><span class="n">add_child</span><span class="p">(</span><span class="n">node_to_join</span><span class="p">)</span>
                <span class="n">new_node</span><span class="o">.</span><span class="n">_upgma_cluster</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">node_to_join</span><span class="o">.</span><span class="n">_upgma_cluster</span><span class="p">)</span>
                <span class="n">node_to_join</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">elen</span> <span class="o">-</span> <span class="n">node_to_join</span><span class="o">.</span><span class="n">_upgma_distance_from_tip</span>
                <span class="n">node_pool</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">node_to_join</span><span class="p">)</span>
            <span class="n">new_node</span><span class="o">.</span><span class="n">_upgma_distance_from_tip</span> <span class="o">=</span> <span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="o">+</span> <span class="n">nodes_to_join</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">_upgma_distance_from_tip</span>
            <span class="k">for</span> <span class="n">idx1</span><span class="p">,</span> <span class="n">nd1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">node_pool</span><span class="p">):</span>
                <span class="n">d1</span> <span class="o">=</span> <span class="mf">0.0</span>
                <span class="n">count</span> <span class="o">=</span> <span class="mf">0.0</span>
                <span class="k">for</span> <span class="n">node_to_join</span> <span class="ow">in</span> <span class="n">nodes_to_join</span><span class="p">:</span>
                    <span class="n">d2</span> <span class="o">=</span> <span class="n">node_to_join</span><span class="o">.</span><span class="n">_upgma_distances</span><span class="p">[</span><span class="n">nd1</span><span class="p">]</span>
                    <span class="n">xc</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">node_to_join</span><span class="o">.</span><span class="n">_upgma_cluster</span><span class="p">)</span>
                    <span class="n">d1</span> <span class="o">+=</span> <span class="p">(</span><span class="n">d2</span> <span class="o">*</span> <span class="n">xc</span><span class="p">)</span>
                    <span class="n">count</span> <span class="o">+=</span> <span class="n">xc</span>
                <span class="n">d</span> <span class="o">=</span> <span class="n">d1</span> <span class="o">/</span> <span class="n">count</span>
                <span class="n">nd1</span><span class="o">.</span><span class="n">_upgma_distances</span><span class="p">[</span><span class="n">new_node</span><span class="p">]</span> <span class="o">=</span> <span class="n">d</span>
                <span class="n">new_node</span><span class="o">.</span><span class="n">_upgma_distances</span><span class="p">[</span><span class="n">nd1</span><span class="p">]</span> <span class="o">=</span> <span class="n">d</span>
            <span class="k">for</span> <span class="n">node_to_join</span> <span class="ow">in</span> <span class="n">nodes_to_join</span><span class="p">:</span>
                <span class="k">del</span> <span class="n">node_to_join</span><span class="o">.</span><span class="n">_upgma_cluster</span>
                <span class="k">del</span> <span class="n">node_to_join</span><span class="o">.</span><span class="n">_upgma_distance_from_tip</span>
                <span class="k">del</span> <span class="n">node_to_join</span><span class="o">.</span><span class="n">_upgma_distances</span>
            <span class="n">node_pool</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_node</span><span class="p">)</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span> <span class="o">=</span> <span class="n">node_pool</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">del</span> <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">_upgma_cluster</span>
        <span class="k">del</span> <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">_upgma_distance_from_tip</span>
        <span class="k">del</span> <span class="n">tree</span><span class="o">.</span><span class="n">seed_node</span><span class="o">.</span><span class="n">_upgma_distances</span>
        <span class="k">return</span> <span class="n">tree</span></div>


<div class="viewcode-block" id="PhylogeneticDistanceMatrix.as_data_table">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.as_data_table">[docs]</a>
    <span class="k">def</span> <span class="nf">as_data_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns this as a table.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">is_weighted_edge_distances</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">patristic_distance</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">path_edge_count</span>
        <span class="n">dt</span> <span class="o">=</span> <span class="n">container</span><span class="o">.</span><span class="n">DataTable</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">t1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">:</span>
            <span class="n">dt</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">row_name</span><span class="o">=</span><span class="n">t1</span><span class="o">.</span><span class="n">label</span><span class="p">)</span>
            <span class="n">dt</span><span class="o">.</span><span class="n">add_column</span><span class="p">(</span><span class="n">column_name</span><span class="o">=</span><span class="n">t1</span><span class="o">.</span><span class="n">label</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">t1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">t2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">:</span>
                <span class="n">dt</span><span class="p">[</span><span class="n">t1</span><span class="o">.</span><span class="n">label</span><span class="p">,</span> <span class="n">t2</span><span class="o">.</span><span class="n">label</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">(</span><span class="n">t1</span><span class="p">,</span> <span class="n">t2</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">dt</span></div>


    <span class="k">def</span> <span class="nf">write_csv</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">out</span><span class="p">,</span>
            <span class="n">is_first_row_column_names</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_first_column_row_names</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">label_transform_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="o">**</span><span class="n">csv_writer_kwargs</span>
            <span class="p">):</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">dest</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">dest</span> <span class="o">=</span> <span class="n">out</span>
        <span class="k">if</span> <span class="n">label_transform_fn</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">label_transform_fn</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span>
        <span class="n">dmatrix</span><span class="p">,</span> <span class="n">normalization_factor</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_distance_matrix_and_normalization_factor</span><span class="p">(</span>
                <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="n">is_weighted_edge_distances</span><span class="p">,</span>
                <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">,</span>
                <span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;delimiter&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">csv_writer_kwargs</span><span class="p">:</span>
            <span class="n">csv_writer_kwargs</span><span class="p">[</span><span class="s2">&quot;delimiter&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="n">csv_writer_kwargs</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">is_first_row_column_names</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">if</span> <span class="n">is_first_column_row_names</span><span class="p">:</span>
                <span class="n">row</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">:</span>
                <span class="n">row</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">label_transform_fn</span><span class="p">(</span><span class="n">taxon</span><span class="o">.</span><span class="n">label</span><span class="p">))</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
            <span class="c1"># dest.write(delimiter.join(row))</span>
            <span class="c1"># dest.write(&quot;\n&quot;)</span>
        <span class="k">for</span> <span class="n">taxon1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">if</span> <span class="n">is_first_column_row_names</span><span class="p">:</span>
                <span class="n">row</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">label_transform_fn</span><span class="p">(</span><span class="n">taxon1</span><span class="o">.</span><span class="n">label</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">taxon2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">:</span>
                <span class="n">d</span> <span class="o">=</span> <span class="n">dmatrix</span><span class="p">[</span><span class="n">taxon1</span><span class="p">][</span><span class="n">taxon2</span><span class="p">]</span> <span class="o">/</span> <span class="n">normalization_factor</span>
                <span class="n">row</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">d</span><span class="p">))</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
            <span class="c1"># dest.write(delimiter.join(row))</span>
            <span class="c1"># dest.write(&quot;\n&quot;)</span>

<div class="viewcode-block" id="PhylogeneticDistanceMatrix.assemblage_membership_definitions_from_csv">
<a class="viewcode-back" href="../../../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.assemblage_membership_definitions_from_csv">[docs]</a>
    <span class="k">def</span> <span class="nf">assemblage_membership_definitions_from_csv</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">src</span><span class="p">,</span>
            <span class="n">default_data_type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
            <span class="o">**</span><span class="n">csv_reader_kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Convenience method to return list of community sets from a delimited</span>
<span class="sd">        file that lists taxon (labels) in columns and community</span>
<span class="sd">        presence/absences or abundances in rows.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">src</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">src</span><span class="p">)</span> <span class="k">as</span> <span class="n">srcf</span><span class="p">:</span>
                <span class="n">data_table</span> <span class="o">=</span> <span class="n">container</span><span class="o">.</span><span class="n">DataTable</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span>
                        <span class="n">src</span><span class="p">,</span>
                        <span class="n">default_data_type</span><span class="o">=</span><span class="n">default_data_type</span><span class="p">,</span>
                        <span class="o">**</span><span class="n">csv_reader_kwargs</span>
                        <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">data_table</span> <span class="o">=</span> <span class="n">container</span><span class="o">.</span><span class="n">DataTable</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span>
                    <span class="n">src</span><span class="p">,</span>
                    <span class="n">default_data_type</span><span class="o">=</span><span class="n">default_data_type</span><span class="p">,</span>
                    <span class="o">**</span><span class="n">csv_reader_kwargs</span>
                    <span class="p">)</span>
        <span class="n">mapped_taxon_labels</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">taxon</span><span class="o">.</span><span class="n">label</span> <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxon_iter</span><span class="p">()])</span>
        <span class="k">for</span> <span class="n">column_name</span> <span class="ow">in</span> <span class="n">data_table</span><span class="o">.</span><span class="n">column_name_iter</span><span class="p">():</span>
            <span class="k">assert</span> <span class="n">column_name</span> <span class="ow">in</span> <span class="n">mapped_taxon_labels</span>
        <span class="n">assemblage_memberships</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">row_name</span> <span class="ow">in</span> <span class="n">data_table</span><span class="o">.</span><span class="n">row_name_iter</span><span class="p">():</span>
            <span class="n">assemblage_membership</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxon_iter</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">data_table</span><span class="p">[</span><span class="n">row_name</span><span class="p">,</span> <span class="n">taxon</span><span class="o">.</span><span class="n">label</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">assemblage_membership</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">taxon</span><span class="p">)</span>
            <span class="n">assemblage_memberships</span><span class="p">[</span><span class="n">row_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">assemblage_membership</span>
        <span class="k">return</span> <span class="n">assemblage_memberships</span></div>


    <span class="k">def</span> <span class="nf">_get_taxon_to_all_other_taxa_comparisons</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filter_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="n">permutations</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">taxon1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">:</span>
            <span class="c1"># permutations[taxon1] = []</span>
            <span class="k">if</span> <span class="n">filter_fn</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">filter_fn</span><span class="p">(</span><span class="n">taxon1</span><span class="p">):</span>
                <span class="k">continue</span>
            <span class="k">for</span> <span class="n">taxon2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped_taxa</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">taxon1</span> <span class="ow">is</span> <span class="n">taxon2</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">if</span> <span class="n">filter_fn</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">filter_fn</span><span class="p">(</span><span class="n">taxon2</span><span class="p">):</span>
                    <span class="k">continue</span>
                <span class="n">permutations</span><span class="p">[</span><span class="n">taxon1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">taxon2</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">permutations</span>

    <span class="k">def</span> <span class="nf">_get_distance_matrix_and_normalization_factor</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">is_weighted_edge_distances</span><span class="p">:</span>
            <span class="n">dmatrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_distances</span>
            <span class="k">if</span> <span class="n">is_normalize_by_tree_size</span><span class="p">:</span>
                <span class="n">normalization_factor</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">normalization_factor</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">dmatrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxon_phylogenetic_path_steps</span>
            <span class="k">if</span> <span class="n">is_normalize_by_tree_size</span><span class="p">:</span>
                <span class="n">normalization_factor</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">normalization_factor</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="k">return</span> <span class="n">dmatrix</span><span class="p">,</span> <span class="n">normalization_factor</span>

    <span class="k">def</span> <span class="nf">_calculate_mean_pairwise_distance</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">comparison_regime</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="p">):</span>
        <span class="n">dmatrix</span><span class="p">,</span> <span class="n">normalization_factor</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_distance_matrix_and_normalization_factor</span><span class="p">(</span>
                <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="n">is_weighted_edge_distances</span><span class="p">,</span>
                <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">,)</span>
        <span class="n">distances</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">taxon1</span><span class="p">,</span> <span class="n">taxon2</span> <span class="ow">in</span> <span class="n">comparison_regime</span><span class="p">:</span>
            <span class="n">distances</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dmatrix</span><span class="p">[</span><span class="n">taxon1</span><span class="p">][</span><span class="n">taxon2</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">distances</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">distances</span><span class="p">)</span> <span class="o">/</span> <span class="n">normalization_factor</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">distances</span><span class="p">)</span> <span class="o">*</span> <span class="mf">1.0</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">error</span><span class="o">.</span><span class="n">NullAssemblageException</span><span class="p">(</span><span class="s2">&quot;No taxa in assemblage&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_calculate_mean_nearest_taxon_distance</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">comparison_regime</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="p">):</span>
        <span class="n">dmatrix</span><span class="p">,</span> <span class="n">normalization_factor</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_distance_matrix_and_normalization_factor</span><span class="p">(</span>
                <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="n">is_weighted_edge_distances</span><span class="p">,</span>
                <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">,)</span>
        <span class="n">distances</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">taxon1</span> <span class="ow">in</span> <span class="n">comparison_regime</span><span class="p">:</span>
            <span class="c1"># subdistances = [dmatrix[taxon1][taxon2] for taxon2 in comparison_regime[taxon1]]</span>
            <span class="c1"># distances.append(min(subdistances))</span>
            <span class="n">min_distance</span> <span class="o">=</span> <span class="n">dmatrix</span><span class="p">[</span><span class="n">taxon1</span><span class="p">][</span><span class="n">comparison_regime</span><span class="p">[</span><span class="n">taxon1</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>
            <span class="k">for</span> <span class="n">taxon2</span> <span class="ow">in</span> <span class="n">comparison_regime</span><span class="p">[</span><span class="n">taxon1</span><span class="p">][</span><span class="mi">1</span><span class="p">:]:</span>
                <span class="n">d</span> <span class="o">=</span> <span class="n">dmatrix</span><span class="p">[</span><span class="n">taxon1</span><span class="p">][</span><span class="n">taxon2</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">d</span> <span class="o">&lt;</span> <span class="n">min_distance</span><span class="p">:</span>
                    <span class="n">min_distance</span> <span class="o">=</span> <span class="n">d</span>
            <span class="n">distances</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">min_distance</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">distances</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">distances</span><span class="p">)</span> <span class="o">/</span> <span class="n">normalization_factor</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">distances</span><span class="p">)</span> <span class="o">*</span> <span class="mf">1.0</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">error</span><span class="o">.</span><span class="n">NullAssemblageException</span><span class="p">(</span><span class="s2">&quot;No taxa in assemblage&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_calculate_standardized_effect_size</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">statisticf_name</span><span class="p">,</span>
            <span class="n">comparison_regimes</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="p">,</span>
            <span class="n">null_model_type</span><span class="o">=</span><span class="s2">&quot;taxa.label&quot;</span><span class="p">,</span>
            <span class="n">num_randomization_replicates</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
            <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="n">result_type</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">namedtuple</span><span class="p">(</span><span class="s2">&quot;PhylogeneticCommunityStandardizedEffectSizeStatisticCalculationResult&quot;</span><span class="p">,</span>
                <span class="p">[</span><span class="s2">&quot;obs&quot;</span><span class="p">,</span> <span class="s2">&quot;null_model_mean&quot;</span><span class="p">,</span> <span class="s2">&quot;null_model_sd&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span> <span class="s2">&quot;rank&quot;</span><span class="p">,</span> <span class="s2">&quot;p&quot;</span><span class="p">,])</span>
        <span class="n">statisticf_kwargs</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;is_weighted_edge_distances&quot;</span><span class="p">:</span> <span class="n">is_weighted_edge_distances</span><span class="p">,</span>
            <span class="s2">&quot;is_normalize_by_tree_size&quot;</span><span class="p">:</span> <span class="n">is_normalize_by_tree_size</span>
        <span class="p">}</span>
        <span class="n">observed_stat_values</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">null_model_stat_values</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">null_model_matrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">clone</span><span class="p">()</span>
        <span class="k">assert</span> <span class="n">null_model_matrix</span> <span class="o">==</span> <span class="bp">self</span>
        <span class="k">if</span> <span class="n">is_weighted_edge_distances</span><span class="p">:</span>
            <span class="n">is_shuffle_phylogenetic_distances</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">is_shuffle_phylogenetic_path_steps</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">is_shuffle_phylogenetic_distances</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="n">is_shuffle_phylogenetic_path_steps</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">for</span> <span class="n">rep_idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_randomization_replicates</span><span class="p">):</span>
            <span class="n">null_model_matrix</span><span class="o">.</span><span class="n">shuffle_taxa</span><span class="p">(</span>
                    <span class="n">is_shuffle_phylogenetic_distances</span><span class="o">=</span><span class="n">is_shuffle_phylogenetic_distances</span><span class="p">,</span>
                    <span class="n">is_shuffle_phylogenetic_path_steps</span><span class="o">=</span><span class="n">is_shuffle_phylogenetic_distances</span><span class="p">,</span>
                    <span class="n">is_shuffle_mrca</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                    <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">comparison_regime_idx</span><span class="p">,</span> <span class="n">comparison_regime</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">comparison_regimes</span><span class="p">):</span>
                <span class="n">statisticf_kwargs</span><span class="p">[</span><span class="s2">&quot;comparison_regime&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">comparison_regime</span>
                <span class="k">if</span> <span class="n">rep_idx</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">observed_stat_values</span><span class="p">[</span><span class="n">comparison_regime_idx</span><span class="p">]</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">statisticf_name</span><span class="p">)(</span><span class="o">**</span><span class="n">statisticf_kwargs</span><span class="p">)</span>
                    <span class="n">null_model_stat_values</span><span class="p">[</span><span class="n">comparison_regime_idx</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">stat_value</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">null_model_matrix</span><span class="p">,</span> <span class="n">statisticf_name</span><span class="p">)(</span><span class="o">**</span><span class="n">statisticf_kwargs</span><span class="p">)</span>
                <span class="n">null_model_stat_values</span><span class="p">[</span><span class="n">comparison_regime_idx</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">stat_value</span><span class="p">)</span>
        <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">comparison_regime_idx</span><span class="p">,</span> <span class="n">comparison_regime</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">comparison_regimes</span><span class="p">):</span>
            <span class="n">obs_value</span> <span class="o">=</span> <span class="n">observed_stat_values</span><span class="p">[</span><span class="n">comparison_regime_idx</span><span class="p">]</span>
            <span class="n">stat_values</span> <span class="o">=</span> <span class="n">null_model_stat_values</span><span class="p">[</span><span class="n">comparison_regime_idx</span><span class="p">]</span>
            <span class="n">null_model_mean</span><span class="p">,</span> <span class="n">null_model_var</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">mean_and_sample_variance</span><span class="p">(</span><span class="n">stat_values</span><span class="p">)</span>
            <span class="n">rank</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">rank</span><span class="p">(</span>
                    <span class="n">value_to_be_ranked</span><span class="o">=</span><span class="n">obs_value</span><span class="p">,</span>
                    <span class="n">value_providing_rank</span><span class="o">=</span><span class="n">stat_values</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">null_model_var</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">null_model_sd</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">null_model_var</span><span class="p">)</span>
                <span class="n">z</span> <span class="o">=</span> <span class="p">(</span><span class="n">obs_value</span> <span class="o">-</span> <span class="n">null_model_mean</span><span class="p">)</span> <span class="o">/</span> <span class="n">null_model_sd</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">null_model_sd</span> <span class="o">=</span> <span class="mf">0.0</span>
                <span class="n">z</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">p</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">rank</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">stat_values</span><span class="p">)</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">result_type</span><span class="p">(</span>
                    <span class="n">obs</span><span class="o">=</span><span class="n">obs_value</span><span class="p">,</span>
                    <span class="n">null_model_mean</span><span class="o">=</span><span class="n">null_model_mean</span><span class="p">,</span>
                    <span class="n">null_model_sd</span><span class="o">=</span><span class="n">null_model_sd</span><span class="p">,</span>
                    <span class="n">z</span><span class="o">=</span><span class="n">z</span><span class="p">,</span>
                    <span class="n">rank</span><span class="o">=</span><span class="n">rank</span><span class="p">,</span>
                    <span class="n">p</span><span class="o">=</span><span class="n">p</span><span class="p">)</span>
            <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">results</span></div>


<span class="k">class</span> <span class="nc">NodeDistanceMatrix</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_tree</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">tree</span><span class="p">):</span>
        <span class="n">ndm</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">()</span>
        <span class="n">ndm</span><span class="o">.</span><span class="n">compile_from_tree</span><span class="p">(</span><span class="n">tree</span><span class="o">=</span><span class="n">tree</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ndm</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">clear</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">def</span> <span class="nf">compile_from_tree</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tree</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">node1</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">postorder_node_iter</span><span class="p">():</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span> <span class="o">+=</span> <span class="n">node1</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span>
            <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span> <span class="c1"># None for edge length</span>
                <span class="k">pass</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">node1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">node1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">node1</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">}</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">node1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">node1</span><span class="p">:</span> <span class="mi">0</span><span class="p">}</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">[</span><span class="n">node1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">node1</span><span class="p">:</span> <span class="n">node1</span><span class="p">}</span>
            <span class="n">children</span> <span class="o">=</span> <span class="n">node1</span><span class="o">.</span><span class="n">child_nodes</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">ch_idx</span><span class="p">,</span> <span class="n">ch1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">children</span><span class="p">):</span>
                <span class="n">ch1_elen</span> <span class="o">=</span> <span class="n">ch1</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="k">if</span> <span class="n">ch1</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="mf">0.0</span>
                <span class="k">for</span> <span class="n">ch1_subtree_node</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">ch1</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
                    <span class="k">if</span> <span class="n">ch1_subtree_node</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">node1</span><span class="p">]:</span>
                        <span class="n">d</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">ch1</span><span class="p">][</span><span class="n">ch1_subtree_node</span><span class="p">]</span> <span class="o">+</span> <span class="n">ch1_elen</span>
                        <span class="n">d2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">ch1</span><span class="p">][</span><span class="n">ch1_subtree_node</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">ch1_subtree_node</span><span class="p">]</span> <span class="o">=</span> <span class="n">d</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">ch1_subtree_node</span><span class="p">][</span><span class="n">node1</span><span class="p">]</span> <span class="o">=</span> <span class="n">d</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">ch1_subtree_node</span><span class="p">]</span> <span class="o">=</span> <span class="n">d2</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">ch1_subtree_node</span><span class="p">][</span><span class="n">node1</span><span class="p">]</span> <span class="o">=</span> <span class="n">d2</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">ch1</span><span class="p">]</span> <span class="o">=</span> <span class="n">ch1_elen</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">ch1</span><span class="p">][</span><span class="n">node1</span><span class="p">]</span> <span class="o">=</span> <span class="n">ch1_elen</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">ch1</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">ch1</span><span class="p">][</span><span class="n">node1</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="k">for</span> <span class="n">ch2</span> <span class="ow">in</span> <span class="n">children</span><span class="p">[</span><span class="n">ch_idx</span><span class="o">+</span><span class="mi">1</span><span class="p">:]:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">[</span><span class="n">ch1</span><span class="p">][</span><span class="n">ch2</span><span class="p">]</span> <span class="o">=</span> <span class="n">node1</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">[</span><span class="n">ch2</span><span class="p">][</span><span class="n">ch1</span><span class="p">]</span> <span class="o">=</span> <span class="n">node1</span>
                    <span class="n">ch2_elen</span> <span class="o">=</span> <span class="n">ch2</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="k">if</span> <span class="n">ch2</span><span class="o">.</span><span class="n">edge</span><span class="o">.</span><span class="n">length</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="mf">0.0</span>
                    <span class="n">d</span> <span class="o">=</span> <span class="n">ch1_elen</span> <span class="o">+</span> <span class="n">ch2_elen</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">ch1</span><span class="p">][</span><span class="n">ch2</span><span class="p">]</span> <span class="o">=</span> <span class="n">d</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">ch2</span><span class="p">][</span><span class="n">ch1</span><span class="p">]</span> <span class="o">=</span> <span class="n">d</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">ch1</span><span class="p">][</span><span class="n">ch2</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">ch2</span><span class="p">][</span><span class="n">ch1</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
            <span class="c1"># Below is ugly, ugly, ugly. Basic idea is to link nodes of each</span>
            <span class="c1"># the subtrees of each of the child nodes of node1. Assumes</span>
            <span class="c1"># that any pairwise comparison of nodes descending from node1</span>
            <span class="c1"># (as given by nodes in a pairwise comparison with node1) not</span>
            <span class="c1"># already made have their MRCA at node1.</span>
            <span class="k">for</span> <span class="n">snd1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">node1</span><span class="p">]:</span>
                <span class="k">for</span> <span class="n">snd2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">node1</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="n">snd1</span> <span class="ow">is</span> <span class="n">snd2</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="k">if</span> <span class="n">snd1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">snd1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">snd1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                    <span class="k">if</span> <span class="n">snd2</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">snd2</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">snd2</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                    <span class="k">if</span> <span class="n">snd2</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">snd1</span><span class="p">]:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">snd1</span><span class="p">][</span><span class="n">snd2</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">snd1</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">snd2</span><span class="p">]</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">snd1</span><span class="p">][</span><span class="n">snd2</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">snd1</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">snd2</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">snd1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">snd2</span><span class="p">]:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">snd2</span><span class="p">][</span><span class="n">snd1</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">snd1</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">snd2</span><span class="p">]</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">snd2</span><span class="p">][</span><span class="n">snd1</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">snd1</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">snd2</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">snd1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">[</span><span class="n">snd1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                    <span class="k">if</span> <span class="n">snd2</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">[</span><span class="n">snd2</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                    <span class="k">if</span> <span class="n">snd2</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">[</span><span class="n">snd1</span><span class="p">]:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">[</span><span class="n">snd1</span><span class="p">][</span><span class="n">snd2</span><span class="p">]</span> <span class="o">=</span> <span class="n">node1</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">[</span><span class="n">snd2</span><span class="p">][</span><span class="n">snd1</span><span class="p">]</span> <span class="o">=</span> <span class="n">node1</span>

    <span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">o</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">node_namespace</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">o</span><span class="o">.</span><span class="n">node_namespace</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">return</span> <span class="p">(</span><span class="kc">True</span>
                <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span> <span class="o">==</span> <span class="n">o</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">)</span>
                <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span> <span class="o">==</span> <span class="n">o</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">)</span>
                <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span> <span class="o">==</span> <span class="n">o</span><span class="o">.</span><span class="n">_mrca</span><span class="p">)</span>
                <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span> <span class="o">==</span> <span class="n">o</span><span class="o">.</span><span class="n">_tree_length</span><span class="p">)</span>
                <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span> <span class="o">==</span> <span class="n">o</span><span class="o">.</span><span class="n">_num_edges</span><span class="p">)</span>
                <span class="p">)</span>

    <span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">node</span>

    <span class="k">def</span> <span class="fm">__hash__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">id</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node1</span><span class="p">,</span> <span class="n">node2</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">patristic_distance</span><span class="p">(</span><span class="n">node1</span><span class="p">,</span> <span class="n">node2</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__copy__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">clone</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">clone</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">o</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">()</span>
        <span class="n">o</span><span class="o">.</span><span class="n">_tree_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span>
        <span class="n">o</span><span class="o">.</span><span class="n">_num_edges</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span>
        <span class="k">for</span> <span class="n">src</span><span class="p">,</span> <span class="n">dest</span> <span class="ow">in</span> <span class="p">(</span>
                <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">,</span> <span class="n">o</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">,),</span>
                <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">,</span> <span class="n">o</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">,),</span>
                <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">,</span> <span class="n">o</span><span class="o">.</span><span class="n">_mrca</span><span class="p">,),</span>
                <span class="p">):</span>
            <span class="k">for</span> <span class="n">t1</span> <span class="ow">in</span> <span class="n">src</span><span class="p">:</span>
                <span class="n">dest</span><span class="p">[</span><span class="n">t1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">for</span> <span class="n">t2</span> <span class="ow">in</span> <span class="n">src</span><span class="p">[</span><span class="n">t1</span><span class="p">]:</span>
                    <span class="n">dest</span><span class="p">[</span><span class="n">t1</span><span class="p">][</span><span class="n">t2</span><span class="p">]</span> <span class="o">=</span> <span class="n">src</span><span class="p">[</span><span class="n">t1</span><span class="p">][</span><span class="n">t2</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">o</span>

    <span class="k">def</span> <span class="nf">mrca</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node1</span><span class="p">,</span> <span class="n">node2</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns MRCA of two node objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mrca</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">node2</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">distance</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">node1</span><span class="p">,</span>
            <span class="n">node2</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns distance between node1 and node2.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">is_weighted_edge_distances</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">patristic_distance</span><span class="p">(</span><span class="n">node1</span><span class="p">,</span> <span class="n">node2</span><span class="p">,</span> <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">path_edge_count</span><span class="p">(</span><span class="n">node1</span><span class="p">,</span> <span class="n">node2</span><span class="p">,</span> <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">patristic_distance</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node1</span><span class="p">,</span> <span class="n">node2</span><span class="p">,</span> <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns patristic distance between two node objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">node1</span> <span class="ow">is</span> <span class="n">node2</span><span class="p">:</span>
            <span class="k">return</span> <span class="mf">0.0</span>
        <span class="n">d</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">node2</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">is_normalize_by_tree_size</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">d</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">d</span>

    <span class="k">def</span> <span class="nf">path_edge_count</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node1</span><span class="p">,</span> <span class="n">node2</span><span class="p">,</span> <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the number of edges between two node objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">node1</span> <span class="ow">is</span> <span class="n">node2</span><span class="p">:</span>
            <span class="k">return</span> <span class="mi">0</span>
        <span class="n">d</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">node2</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">is_normalize_by_tree_size</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">d</span><span class="p">)</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">d</span>

    <span class="k">def</span> <span class="nf">distances</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns list of patristic distances.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">dmatrix</span><span class="p">,</span> <span class="n">normalization_factor</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_distance_matrix_and_normalization_factor</span><span class="p">(</span>
                <span class="n">is_weighted_edge_distances</span><span class="o">=</span><span class="n">is_weighted_edge_distances</span><span class="p">,</span>
                <span class="n">is_normalize_by_tree_size</span><span class="o">=</span><span class="n">is_normalize_by_tree_size</span><span class="p">,</span>
                <span class="p">)</span>
        <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">nodes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">dmatrix</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="k">for</span> <span class="n">node_idx1</span><span class="p">,</span> <span class="n">node1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">nodes</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]):</span>
            <span class="k">for</span> <span class="n">node_idx2</span><span class="p">,</span> <span class="n">node2</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">nodes</span><span class="p">[</span><span class="n">node_idx1</span><span class="o">+</span><span class="mi">1</span><span class="p">:]):</span>
                <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dmatrix</span><span class="p">[</span><span class="n">node1</span><span class="p">][</span><span class="n">node2</span><span class="p">]</span><span class="o">/</span><span class="n">normalization_factor</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">results</span>

    <span class="k">def</span> <span class="nf">_get_distance_matrix_and_normalization_factor</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">is_weighted_edge_distances</span><span class="p">,</span>
            <span class="n">is_normalize_by_tree_size</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">is_weighted_edge_distances</span><span class="p">:</span>
            <span class="n">dmatrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_distances</span>
            <span class="k">if</span> <span class="n">is_normalize_by_tree_size</span><span class="p">:</span>
                <span class="n">normalization_factor</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_tree_length</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">normalization_factor</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">dmatrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_node_phylogenetic_path_steps</span>
            <span class="k">if</span> <span class="n">is_normalize_by_tree_size</span><span class="p">:</span>
                <span class="n">normalization_factor</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_edges</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">normalization_factor</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="k">return</span> <span class="n">dmatrix</span><span class="p">,</span> <span class="n">normalization_factor</span>

</pre></div>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div style="text-align: center; padding-top: 20px; padding-bottom: 5px; width: 100%;">
    <a href="../../../index.html"><img src="../../../_static/dendropy_logo.png" /></a>
</div><div style="clear:both; width: 100%; height:1px;"></div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script><div style="clear:both; width: 100%; height:1px;"></div><div style="clear:both; width: 100%; height:1px;"></div><div style="clear:both; width: 100%; height:1px;"></div>
    <!-- Documentation -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Documentation</h3>
        <ul>
            <li><a href="../../../primer/downloading.html">Downloading and Installing DendroPy</a></li>
            <li><a href="../../../primer/index.html">The DendroPy Primer</a></li>
            <li><a href="../../../library/index.html">Library API Reference</a></li>
            <li>
                <a href="../../../schemas/index.html">Schemas</a>
                <ul>
                    <li><a href="../../../schemas/fasta.html">FASTA</a></li>
                    <li><a href="../../../schemas/newick.html">Newick</a></li>
                    <li><a href="../../../schemas/nexml.html">NeXML</a></li>
                    <li><a href="../../../schemas/nexus.html">Nexus</a></li>
                    <li><a href="../../../schemas/phylip.html">PHYLIP</a></li>
                </ul>
            </li>
            <li>
                <a href="../../../programs/index.html">Programs</a>
                <ul>
                    <li><a href="../../../programs/sumtrees.html">SumTrees</a></li>
                </ul>
            </li>
            <li><a href="../../../glossary.html">Glossary and Terminological Reference</a></li>
            <li><a href="../../../developer.html">Developer Guide</a></li>
            <li><a href="../../../planning.html">Ongoing Development</a></li>
            <li><a href="../../../changes.html">Change History</a></li>
        </ul>
    </div>

    <!-- Downloads -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Obtaining</h3>
        <ul>
            <li><a target="_blank" href="http://pypi.python.org/pypi/DendroPy">Install from the Python Package Index</a></li>
            <li><a target="_blank" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.2.tar.gz">Download the Source Code Archive</a></li>
            <li><a target="_blank" href="http://github.com/jeetsukumaran/DendroPy">Clone the Source Code Repository</a></li>
        </ul>
    </div>

    <!-- Discussions -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Discussion</span><span style="position: absolute; right: 0; top: 10px "><img src="../../../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot; DendroPy Users&quot; </a>group to follow and participate in discussion, troubleshooting, help, information, suggestions, etc. on the usage and development of the DendroPy phylogenetic computing library.</p>
            <form action="http://groups.google.com/group/dendropy-users/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot;dendropy-users&quot;</a> group, or click <a href="http://groups.google.com/group/dendropy-users?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Announcements -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Announcements</span><span style="position: absolute; right: 0; top: 10px "><img src="../../../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; DendroPy Announcements&quot; </a>group to receive announcements of new releases, updates, changes and other news of interest to DendroPy users and developers.</p>
            <form action="http://groups.google.com/group/dendropy-announce/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; dendropy-announce&quot; </a>group, or click <a href="http://groups.google.com/group/dendropy-announce?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Development -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative; padding-bottom: 15px; margin-bottom:5px;">
        <h3><span style="text-align: left"><a href="https://github.com/jeetsukumaran/DendroPy/">Development</a></span><a href="https://github.com/jeetsukumaran/DendroPy/"><span style="position: absolute; right: 0; top: 10px "><img src="../../../_static/Octocat.png" height="30px" alt="GitHub" /></span></a></h3>
        <div style="margin-top: 15px;">
            <!-- <a href="https://github.com/jeetsukumaran/DendroPy/issues">Issues</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/subscription">Watch</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/fork">Fork</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/stargazers">Star</a> &bull; <a href="https://github.com/jeetsukumaran/">Follow</a> -->
            <ul>
                <li>                <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/issues"><code>Issues</code></a></span> <span style="font-style: italic; font-size:80%;"> - Report bugs or request features</span></li>
                <li>     <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/subscription"><code>&nbsp;Watch</code></a></span> <span style="font-style: italic; font-size:80%;"> - Follow development activity</span></li>
                <li>        <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/fork"><code>&nbsp;&nbsp;Fork</code></a></span> <span style="font-style: italic; font-size:80%;"> - Contribute and collaborate</span></li>
                <li>  <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/stargazers"><code>&nbsp;&nbsp;Star</code></a></span> <span style="font-style: italic; font-size:80%;"> - Throw some glitter, add some glamour</span></li>
        </div>
    </div>


        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">dendropy.calculate.phylogeneticdistance</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2009-2024, Jeet Sukumaran and Mark T. Holder.
    </div>
  </body>
</html>