<!DOCTYPE html>

<html lang="en" data-content_root="../../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>dendropy.calculate.popgenstat &#8212; DendroPy 5.0.2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../../../_static/rtd.css?v=cedc6a2f" />
    
    <script src="../../../_static/documentation_options.js?v=5df4c8d1"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">dendropy.calculate.popgenstat</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for dendropy.calculate.popgenstat</h1><div class="highlight"><pre>
<span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="c1">##############################################################################</span>
<span class="c1">##  DendroPy Phylogenetic Computing Library.</span>
<span class="c1">##</span>
<span class="c1">##  Copyright 2010-2015 Jeet Sukumaran and Mark T. Holder.</span>
<span class="c1">##  All rights reserved.</span>
<span class="c1">##</span>
<span class="c1">##  See &quot;LICENSE.rst&quot; for terms and conditions of usage.</span>
<span class="c1">##</span>
<span class="c1">##  If you use this work or any portion thereof in published work,</span>
<span class="c1">##  please cite it as:</span>
<span class="c1">##</span>
<span class="c1">##     Sukumaran, J. and M. T. Holder. 2010. DendroPy: a Python library</span>
<span class="c1">##     for phylogenetic computing. Bioinformatics 26: 1569-1571.</span>
<span class="c1">##</span>
<span class="c1">##############################################################################</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Population genetic statistics.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">dendropy</span>
<span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">combinatorics</span>

<span class="c1">###############################################################################</span>
<span class="c1">## internal functions: generally taking lower-level data, such as sequences etc.</span>
<span class="c1">###############################################################################</span>

<span class="k">def</span> <span class="nf">_count_differences</span><span class="p">(</span><span class="n">char_sequences</span><span class="p">,</span> <span class="n">state_alphabet</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns pair of values: total number of pairwise differences observed between</span>
<span class="sd">    all sequences, and mean number of pairwise differences pair base.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sum_diff</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="n">mean_diff</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="n">sq_diff</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="n">comps</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1">#Check that all sequences are the same length</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">char_sequences</span><span class="p">]))</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;sequences of unequal length&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">ignore_uncertain</span><span class="p">:</span>
        <span class="n">attr</span> <span class="o">=</span> <span class="s2">&quot;fundamental_indexes_with_gaps_as_missing&quot;</span>
        <span class="n">_states_to_ignore</span> <span class="o">=</span> <span class="p">[</span><span class="n">state_alphabet</span><span class="o">.</span><span class="n">gap_state</span><span class="p">,</span> <span class="n">state_alphabet</span><span class="o">.</span><span class="n">no_data_state</span><span class="p">]</span>
        <span class="n">states_to_ignore</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="nb">getattr</span><span class="p">(</span><span class="n">char</span><span class="p">,</span> <span class="n">attr</span><span class="p">)</span> <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">_states_to_ignore</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">attr</span> <span class="o">=</span> <span class="s2">&quot;fundamental_indexes&quot;</span>
        <span class="n">states_to_ignore</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

    <span class="n">reduced_char_sequences</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">char_sequences</span><span class="p">:</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="p">[</span><span class="nb">getattr</span><span class="p">(</span><span class="n">char</span><span class="p">,</span> <span class="n">attr</span><span class="p">)</span> <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">sequence</span><span class="p">]</span>
        <span class="n">reduced_char_sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">vidx</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">reduced_char_sequences</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">reduced_char_sequences</span><span class="p">[</span><span class="n">vidx</span><span class="o">+</span><span class="mi">1</span><span class="p">:]:</span>
            <span class="n">diff</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">counted</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">comps</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">for</span> <span class="n">cidx</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">i</span><span class="p">):</span>
                <span class="n">c1</span> <span class="o">=</span> <span class="n">c</span>
                <span class="n">c2</span> <span class="o">=</span> <span class="n">j</span><span class="p">[</span><span class="n">cidx</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">c1</span> <span class="ow">in</span> <span class="n">states_to_ignore</span> <span class="ow">or</span> <span class="n">c2</span> <span class="ow">in</span> <span class="n">states_to_ignore</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">counted</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">c1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">c2</span><span class="p">:</span>
                    <span class="n">diff</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">sum_diff</span> <span class="o">+=</span> <span class="nb">float</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span>
            <span class="c1"># If counted &lt; 0, this means that there is sites between these sequences</span>
            <span class="c1"># in which both are not ignored: i.e., one or the other has a gap</span>
            <span class="c1"># or an uncertain character. We consider this to mean (maybe</span>
            <span class="c1"># somewhat paradoxically) that there are no sites that are</span>
            <span class="c1"># different between the sequences. Put less paradoxically: there</span>
            <span class="c1"># are no non-ignored sites that are different between the</span>
            <span class="c1"># sequences.</span>
            <span class="n">mean_diff</span> <span class="o">+=</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span> <span class="o">/</span> <span class="n">counted</span><span class="p">)</span> <span class="k">if</span> <span class="n">counted</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="nb">float</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span>
            <span class="n">sq_diff</span> <span class="o">+=</span> <span class="p">(</span><span class="n">diff</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sum_diff</span><span class="p">,</span> <span class="n">mean_diff</span> <span class="o">/</span> <span class="n">comps</span><span class="p">,</span> <span class="n">sq_diff</span>

<span class="k">def</span> <span class="nf">_nucleotide_diversity</span><span class="p">(</span><span class="n">char_sequences</span><span class="p">,</span> <span class="n">state_alphabet</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns $\pi$, the proportional nucleotide diversity, calculated for a</span>
<span class="sd">    list of character sequences.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">_count_differences</span><span class="p">(</span><span class="n">char_sequences</span><span class="p">,</span> <span class="n">state_alphabet</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>

<span class="k">def</span> <span class="nf">_average_number_of_pairwise_differences</span><span class="p">(</span><span class="n">char_sequences</span><span class="p">,</span> <span class="n">state_alphabet</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns $k$ (Tajima 1983; Wakely 1996), calculated for a set of sequences:</span>

<span class="sd">    k = \frac{\right(\sum \sum \k_{ij}\left)}{n \choose 2}</span>

<span class="sd">    where $k_{ij}$ is the number of pairwise differences between the</span>
<span class="sd">    $i$th and $j$th sequence, and $n$ is the number of DNA sequences</span>
<span class="sd">    sampled.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sum_diff</span><span class="p">,</span> <span class="n">mean_diff</span><span class="p">,</span> <span class="n">sq_diff</span> <span class="o">=</span> <span class="n">_count_differences</span><span class="p">(</span><span class="n">char_sequences</span><span class="p">,</span> <span class="n">state_alphabet</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sum_diff</span> <span class="o">/</span> <span class="n">combinatorics</span><span class="o">.</span><span class="n">choose</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">char_sequences</span><span class="p">),</span> <span class="mi">2</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">_num_segregating_sites</span><span class="p">(</span><span class="n">char_sequences</span><span class="p">,</span> <span class="n">state_alphabet</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns the raw number of segregating sites (polymorphic sites).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">s</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="n">ignore_uncertain</span><span class="p">:</span>
        <span class="n">attr</span> <span class="o">=</span> <span class="s2">&quot;fundamental_indexes_with_gaps_as_missing&quot;</span>
        <span class="n">_states_to_ignore</span> <span class="o">=</span> <span class="p">[</span><span class="n">state_alphabet</span><span class="o">.</span><span class="n">gap_state</span><span class="p">,</span> <span class="n">state_alphabet</span><span class="o">.</span><span class="n">no_data_state</span><span class="p">]</span>
        <span class="n">states_to_ignore</span><span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="nb">getattr</span><span class="p">(</span><span class="n">char</span><span class="p">,</span> <span class="n">attr</span><span class="p">)</span> <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">_states_to_ignore</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">attr</span> <span class="o">=</span> <span class="s2">&quot;fundamental_indexes&quot;</span>
        <span class="n">states_to_ignore</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">char_sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">char_sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
            <span class="n">c2</span> <span class="o">=</span> <span class="n">v</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">f1</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">attr</span><span class="p">)</span>
            <span class="n">f2</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">c2</span><span class="p">,</span> <span class="n">attr</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">f1</span> <span class="ow">in</span> <span class="n">states_to_ignore</span> <span class="ow">or</span> <span class="n">f2</span> <span class="ow">in</span> <span class="n">states_to_ignore</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">f1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">f2</span><span class="p">:</span>
                <span class="n">s</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">break</span>
    <span class="k">return</span> <span class="n">s</span>

<span class="k">def</span> <span class="nf">_tajimas_d</span><span class="p">(</span><span class="n">num_sequences</span><span class="p">,</span> <span class="n">avg_num_pairwise_differences</span><span class="p">,</span> <span class="n">num_segregating_sites</span><span class="p">):</span>

    <span class="c1">### VERIFICATION ###</span>
    <span class="c1">###</span>
    <span class="c1">### Given: num_sequences = 10, num_pairwise_differences = 3.888889, num_segregating_sites = 16</span>
    <span class="c1">###  i.e.: tajimas_d(10, 3.888889, 16)  == -1.44617198561</span>
    <span class="c1">###  Then:    a1 == 2.82896825397</span>
    <span class="c1">###           a2 == 1.53976773117</span>
    <span class="c1">###           b1 == 0.407407407407</span>
    <span class="c1">###           b2 == 0.279012345679</span>
    <span class="c1">###           c1 == 0.0539216450284</span>
    <span class="c1">###           c2 == 0.0472267720013</span>
    <span class="c1">###           e1 == 0.0190605338016</span>
    <span class="c1">###           e2 == 0.0049489277699</span>
    <span class="c1">###           D ==  -1.44617198561</span>

    <span class="n">a1</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="mf">1.0</span><span class="o">/</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_sequences</span><span class="p">)])</span>
    <span class="n">a2</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="mf">1.0</span><span class="o">/</span><span class="p">(</span><span class="n">i</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_sequences</span><span class="p">)])</span>
    <span class="n">b1</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">num_sequences</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="mi">3</span><span class="o">*</span><span class="p">(</span><span class="n">num_sequences</span><span class="o">-</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">b2</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="p">(</span> <span class="p">(</span><span class="n">num_sequences</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span> <span class="o">+</span> <span class="n">num_sequences</span> <span class="o">+</span> <span class="mi">3</span> <span class="p">))</span> <span class="o">/</span> <span class="p">(</span><span class="mi">9</span><span class="o">*</span><span class="n">num_sequences</span><span class="o">*</span><span class="p">(</span><span class="n">num_sequences</span><span class="o">-</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">c1</span> <span class="o">=</span> <span class="n">b1</span> <span class="o">-</span> <span class="mf">1.0</span><span class="o">/</span><span class="n">a1</span>
    <span class="n">c2</span> <span class="o">=</span> <span class="n">b2</span> <span class="o">-</span> <span class="nb">float</span><span class="p">(</span><span class="n">num_sequences</span><span class="o">+</span><span class="mi">2</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">a1</span> <span class="o">*</span> <span class="n">num_sequences</span><span class="p">)</span> <span class="o">+</span> <span class="nb">float</span><span class="p">(</span><span class="n">a2</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">a1</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">e1</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">c1</span><span class="p">)</span> <span class="o">/</span> <span class="n">a1</span>
    <span class="n">e2</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">c2</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span> <span class="p">(</span><span class="n">a1</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span> <span class="o">+</span> <span class="n">a2</span> <span class="p">)</span>
    <span class="n">D</span> <span class="o">=</span> <span class="p">(</span>
        <span class="nb">float</span><span class="p">(</span><span class="n">avg_num_pairwise_differences</span> <span class="o">-</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">num_segregating_sites</span><span class="p">)</span><span class="o">/</span><span class="n">a1</span><span class="p">))</span>
        <span class="o">/</span> <span class="n">math</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span>
            <span class="p">(</span><span class="n">e1</span> <span class="o">*</span> <span class="n">num_segregating_sites</span> <span class="p">)</span>
          <span class="o">+</span> <span class="p">((</span><span class="n">e2</span> <span class="o">*</span> <span class="n">num_segregating_sites</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">num_segregating_sites</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="p">))</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">D</span>

<span class="c1">###############################################################################</span>
<span class="c1">## friendlier-functions, generally taking a CharacterMatrix</span>
<span class="c1">###############################################################################</span>

<div class="viewcode-block" id="num_segregating_sites">
<a class="viewcode-back" href="../../../library/popgenstat.html#dendropy.calculate.popgenstat.num_segregating_sites">[docs]</a>
<span class="k">def</span> <span class="nf">num_segregating_sites</span><span class="p">(</span><span class="n">char_matrix</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns the raw number of segregating sites (polymorphic sites).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">_num_segregating_sites</span><span class="p">(</span>
            <span class="n">char_matrix</span><span class="o">.</span><span class="n">sequences</span><span class="p">(),</span>
            <span class="n">char_matrix</span><span class="o">.</span><span class="n">default_state_alphabet</span><span class="p">,</span>
            <span class="n">ignore_uncertain</span><span class="p">)</span></div>


<div class="viewcode-block" id="average_number_of_pairwise_differences">
<a class="viewcode-back" href="../../../library/popgenstat.html#dendropy.calculate.popgenstat.average_number_of_pairwise_differences">[docs]</a>
<span class="k">def</span> <span class="nf">average_number_of_pairwise_differences</span><span class="p">(</span><span class="n">char_matrix</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns $k$, calculated for a character block.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">_average_number_of_pairwise_differences</span><span class="p">(</span><span class="n">char_matrix</span><span class="o">.</span><span class="n">sequences</span><span class="p">(),</span> <span class="n">char_matrix</span><span class="o">.</span><span class="n">default_state_alphabet</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="p">)</span></div>


<div class="viewcode-block" id="nucleotide_diversity">
<a class="viewcode-back" href="../../../library/popgenstat.html#dendropy.calculate.popgenstat.nucleotide_diversity">[docs]</a>
<span class="k">def</span> <span class="nf">nucleotide_diversity</span><span class="p">(</span><span class="n">char_matrix</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns $\pi$, calculated for a character block.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">_nucleotide_diversity</span><span class="p">(</span><span class="n">char_matrix</span><span class="o">.</span><span class="n">sequences</span><span class="p">(),</span> <span class="n">char_matrix</span><span class="o">.</span><span class="n">default_state_alphabet</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="p">)</span></div>


<div class="viewcode-block" id="tajimas_d">
<a class="viewcode-back" href="../../../library/popgenstat.html#dendropy.calculate.popgenstat.tajimas_d">[docs]</a>
<span class="k">def</span> <span class="nf">tajimas_d</span><span class="p">(</span><span class="n">char_matrix</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns Tajima&#39;s D.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sequences</span> <span class="o">=</span> <span class="n">char_matrix</span><span class="o">.</span><span class="n">sequences</span><span class="p">()</span>
    <span class="n">num_sequences</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
    <span class="n">avg_num_pairwise_differences</span> <span class="o">=</span> <span class="n">_average_number_of_pairwise_differences</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">char_matrix</span><span class="o">.</span><span class="n">default_state_alphabet</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="o">=</span><span class="n">ignore_uncertain</span><span class="p">)</span>
    <span class="n">num_segregating_sites</span> <span class="o">=</span> <span class="n">_num_segregating_sites</span><span class="p">(</span>
            <span class="n">sequences</span><span class="p">,</span>
            <span class="n">char_matrix</span><span class="o">.</span><span class="n">default_state_alphabet</span><span class="p">,</span>
            <span class="n">ignore_uncertain</span><span class="o">=</span><span class="n">ignore_uncertain</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">_tajimas_d</span><span class="p">(</span><span class="n">num_sequences</span><span class="p">,</span> <span class="n">avg_num_pairwise_differences</span><span class="p">,</span> <span class="n">num_segregating_sites</span><span class="p">)</span></div>


<div class="viewcode-block" id="wattersons_theta">
<a class="viewcode-back" href="../../../library/popgenstat.html#dendropy.calculate.popgenstat.wattersons_theta">[docs]</a>
<span class="k">def</span> <span class="nf">wattersons_theta</span><span class="p">(</span><span class="n">char_matrix</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns Watterson&#39;s Theta (per sequence)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sequences</span> <span class="o">=</span> <span class="n">char_matrix</span><span class="o">.</span><span class="n">sequences</span><span class="p">()</span>
    <span class="n">num_segregating_sites</span> <span class="o">=</span> <span class="n">_num_segregating_sites</span><span class="p">(</span>
            <span class="n">sequences</span><span class="p">,</span>
            <span class="n">char_matrix</span><span class="o">.</span><span class="n">default_state_alphabet</span><span class="p">,</span>
            <span class="n">ignore_uncertain</span><span class="o">=</span><span class="n">ignore_uncertain</span><span class="p">)</span>
    <span class="n">a1</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="mf">1.0</span><span class="o">/</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequences</span><span class="p">))])</span>
    <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">num_segregating_sites</span><span class="p">)</span> <span class="o">/</span> <span class="n">a1</span></div>


<span class="c1">###############################################################################</span>
<span class="c1">## Classes</span>
<span class="c1">###############################################################################</span>

<span class="k">class</span> <span class="nc">PopulationPairSummaryStatistics</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pop1_seqs</span><span class="p">,</span> <span class="n">pop2_seqs</span><span class="p">,</span> <span class="n">ignore_uncertain</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pop1_seqs</span> <span class="o">=</span> <span class="n">pop1_seqs</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pop2_seqs</span> <span class="o">=</span> <span class="n">pop2_seqs</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">combined_seqs</span> <span class="o">=</span> <span class="n">pop1_seqs</span> <span class="o">+</span> <span class="n">pop2_seqs</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ignore_uncertain</span> <span class="o">=</span> <span class="n">ignore_uncertain</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">state_alphabet</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DNA_STATE_ALPHABET</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">average_number_of_pairwise_differences</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">average_number_of_pairwise_differences_between</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">average_number_of_pairwise_differences_within</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">average_number_of_pairwise_differences_net</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_segregating_sites</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">wattersons_theta</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">wakeleys_psi</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tajimas_d</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ignore_uncertain</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">state_attr</span> <span class="o">=</span> <span class="s2">&quot;fundamental_indexes_with_gaps_as_missing&quot;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">states_to_ignore</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">state_alphabet</span><span class="o">.</span><span class="n">gap_state</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">state_alphabet</span><span class="o">.</span><span class="n">no_data_state</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">state_attr</span> <span class="o">=</span> <span class="s2">&quot;fundamental_indexes&quot;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">states_to_ignore</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">calc</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">calc</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a summary of a set of sequences that can be partitioned into</span>
<span class="sd">        the list of lists of taxa given by ``taxon_groups``.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">diffs_x</span><span class="p">,</span> <span class="n">mean_diffs_x</span><span class="p">,</span> <span class="n">sq_diff_x</span> <span class="o">=</span> <span class="n">_count_differences</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pop1_seqs</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">state_alphabet</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ignore_uncertain</span><span class="p">)</span>
        <span class="n">diffs_y</span><span class="p">,</span> <span class="n">mean_diffs_y</span><span class="p">,</span> <span class="n">sq_diff_y</span> <span class="o">=</span> <span class="n">_count_differences</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pop2_seqs</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">state_alphabet</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ignore_uncertain</span><span class="p">)</span>
        <span class="n">d_x</span> <span class="o">=</span> <span class="n">diffs_x</span> <span class="o">/</span> <span class="n">combinatorics</span><span class="o">.</span><span class="n">choose</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pop1_seqs</span><span class="p">),</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">d_y</span> <span class="o">=</span> <span class="n">diffs_y</span> <span class="o">/</span> <span class="n">combinatorics</span><span class="o">.</span><span class="n">choose</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pop2_seqs</span><span class="p">),</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">d_xy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_average_number_of_pairwise_differences_between_populations</span><span class="p">()</span>
        <span class="n">s2_x</span> <span class="o">=</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">sq_diff_x</span><span class="p">)</span> <span class="o">/</span> <span class="n">combinatorics</span><span class="o">.</span><span class="n">choose</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pop1_seqs</span><span class="p">),</span> <span class="mi">2</span><span class="p">)</span> <span class="p">)</span> <span class="o">-</span> <span class="p">(</span><span class="n">d_x</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">s2_y</span> <span class="o">=</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">sq_diff_y</span><span class="p">)</span> <span class="o">/</span> <span class="n">combinatorics</span><span class="o">.</span><span class="n">choose</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pop2_seqs</span><span class="p">),</span> <span class="mi">2</span><span class="p">)</span> <span class="p">)</span> <span class="o">-</span> <span class="p">(</span><span class="n">d_y</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">s2_xy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_variance_of_pairwise_differences_between_populations</span><span class="p">(</span><span class="n">d_xy</span><span class="p">)</span>
        <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_seqs</span><span class="p">)</span>
        <span class="n">n_x</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pop1_seqs</span><span class="p">))</span>
        <span class="n">n_y</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pop2_seqs</span><span class="p">))</span>
        <span class="n">a</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">n</span> <span class="o">*</span> <span class="p">(</span><span class="n">n</span><span class="o">-</span><span class="mi">1</span><span class="p">))</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">n_x</span> <span class="o">*</span> <span class="p">(</span><span class="n">n_x</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>
        <span class="n">ay</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">n_y</span> <span class="o">*</span> <span class="p">(</span><span class="n">n_y</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>
        <span class="n">k</span> <span class="o">=</span> <span class="n">_average_number_of_pairwise_differences</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_seqs</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">state_alphabet</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ignore_uncertain</span><span class="p">)</span>
        <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_seqs</span><span class="p">)</span>

        <span class="c1"># Hickerson 2006: pi #</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">average_number_of_pairwise_differences</span> <span class="o">=</span> <span class="n">k</span>

        <span class="c1"># Hickerson 2006: pi_b #</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">average_number_of_pairwise_differences_between</span> <span class="o">=</span> <span class="n">d_xy</span>

        <span class="c1"># Hickerson 2006: pi_w #</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">average_number_of_pairwise_differences_within</span> <span class="o">=</span> <span class="n">d_x</span> <span class="o">+</span> <span class="n">d_y</span>

        <span class="c1"># Hickerson 2006: pi_net #</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">average_number_of_pairwise_differences_net</span> <span class="o">=</span> <span class="n">d_xy</span> <span class="o">-</span> <span class="p">(</span><span class="n">d_x</span> <span class="o">+</span> <span class="n">d_y</span><span class="p">)</span>

        <span class="c1"># Hickerson 2006: S #</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_segregating_sites</span> <span class="o">=</span> <span class="n">_num_segregating_sites</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">combined_seqs</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">state_alphabet</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ignore_uncertain</span><span class="p">)</span>

        <span class="c1"># Hickerson 2006: theta #</span>
        <span class="n">a1</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="mf">1.0</span><span class="o">/</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">n</span><span class="p">)])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">wattersons_theta</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">num_segregating_sites</span><span class="p">)</span> <span class="o">/</span> <span class="n">a1</span>

        <span class="c1"># Wakeley 1996 #</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">wakeleys_psi</span> <span class="o">=</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">a</span><span class="p">))</span> <span class="o">*</span> <span class="p">(</span> <span class="n">ax</span> <span class="o">*</span> <span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">s2_x</span><span class="p">)</span><span class="o">/</span><span class="n">d_x</span><span class="p">)</span> <span class="o">+</span> <span class="n">ay</span> <span class="o">*</span> <span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">s2_y</span><span class="p">)</span><span class="o">/</span><span class="n">d_y</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">n_x</span> <span class="o">*</span> <span class="n">n_y</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">s2_xy</span><span class="p">)</span><span class="o">/</span><span class="n">k</span><span class="p">))</span>

        <span class="c1"># Tajima&#39;s D #</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tajimas_d</span> <span class="o">=</span> <span class="n">_tajimas_d</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">average_number_of_pairwise_differences</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_segregating_sites</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_average_number_of_pairwise_differences_between_populations</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Implements Eq (3) of:</span>

<span class="sd">        Wakeley, J. 1996. Distinguishing migration from isolation using the</span>
<span class="sd">        variance of pairwise differences. Theoretical Population Biology 49:</span>
<span class="sd">        369-386.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">diffs</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">sx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pop1_seqs</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">sy</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pop2_seqs</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">cidx</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sx</span><span class="p">):</span>
                    <span class="n">c1</span> <span class="o">=</span> <span class="n">c</span>
                    <span class="n">c2</span> <span class="o">=</span> <span class="n">sy</span><span class="p">[</span><span class="n">cidx</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">c1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">states_to_ignore</span> <span class="ow">or</span> <span class="n">c2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">states_to_ignore</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="n">f1</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">state_attr</span><span class="p">)</span>
                    <span class="n">f2</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">c2</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">state_attr</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">f1</span> <span class="o">!=</span> <span class="n">f2</span><span class="p">:</span>
                        <span class="n">diffs</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">dxy</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pop1_seqs</span><span class="p">)</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pop2_seqs</span><span class="p">))</span> <span class="o">*</span> <span class="nb">float</span><span class="p">(</span><span class="n">diffs</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">dxy</span>

    <span class="k">def</span> <span class="nf">_variance_of_pairwise_differences_between_populations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mean_diff</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Implements Eq (10) of:</span>

<span class="sd">        Wakeley, J. 1996. Distinguishing migration from isolation using the</span>
<span class="sd">        variance of pairwise differences. Theoretical Population Biology 49:</span>
<span class="sd">        369-386.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ss_diffs</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">sx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pop1_seqs</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">sy</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pop2_seqs</span><span class="p">:</span>
                <span class="n">diffs</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="k">for</span> <span class="n">cidx</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sx</span><span class="p">):</span>
                    <span class="n">c1</span> <span class="o">=</span> <span class="n">c</span>
                    <span class="n">c2</span> <span class="o">=</span> <span class="n">sy</span><span class="p">[</span><span class="n">cidx</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">c1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">states_to_ignore</span> <span class="ow">or</span> <span class="n">c2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">states_to_ignore</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="n">f1</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">state_attr</span><span class="p">)</span>
                    <span class="n">f2</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">c2</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">state_attr</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">f1</span> <span class="o">!=</span> <span class="n">f2</span><span class="p">:</span>
                        <span class="n">diffs</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">ss_diffs</span> <span class="o">+=</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">diffs</span> <span class="o">-</span> <span class="n">mean_diff</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">ss_diffs</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pop1_seqs</span><span class="p">)</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pop2_seqs</span><span class="p">))</span>

<div class="viewcode-block" id="derived_state_matrix">
<a class="viewcode-back" href="../../../library/popgenstat.html#dendropy.calculate.popgenstat.derived_state_matrix">[docs]</a>
<span class="k">def</span> <span class="nf">derived_state_matrix</span><span class="p">(</span>
        <span class="n">char_matrix</span><span class="p">,</span>
        <span class="n">ancestral_sequence</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">derived_state_alphabet</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">ignore_uncertain</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a list of CharDataSequence objects, and a reference ancestral sequence,</span>
<span class="sd">    this returns a list of strings corresponding to the list of CharDataSequence</span>
<span class="sd">    objects, where a &#39;0&#39; indicates the ancestral state and &#39;1&#39; a derived state.</span>

<span class="sd">    e.g.</span>

<span class="sd">        Given:</span>
<span class="sd">                GGCTAATCTGA</span>
<span class="sd">                GCTTTTTCTGA</span>
<span class="sd">                GCTCTCTCTTC</span>

<span class="sd">        with ancestral sequence:</span>
<span class="sd">                GGTTAATCTGA</span>

<span class="sd">        this returns:</span>
<span class="sd">                0010000000</span>
<span class="sd">                0000110000</span>
<span class="sd">                0001110011</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">derived_state_alphabet</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">derived_state_alphabet</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">StateAlphabet</span><span class="p">(</span>
                <span class="n">fundamental_states</span><span class="o">=</span><span class="s2">&quot;01&quot;</span><span class="p">,</span>
                <span class="n">polymorphic_states</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">ambiguous_states</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">no_data_symbol</span><span class="o">=</span><span class="s2">&quot;?&quot;</span><span class="p">,</span>
                <span class="n">gap_symbol</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
    <span class="n">derived_matrix</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">StandardCharacterMatrix</span><span class="p">(</span>
            <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">char_matrix</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">,</span>
            <span class="n">default_state_alphabet</span><span class="o">=</span><span class="n">derived_state_alphabet</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">ignore_uncertain</span><span class="p">:</span>
        <span class="n">attr</span> <span class="o">=</span> <span class="s2">&quot;fundamental_indexes_with_gaps_as_missing&quot;</span>
        <span class="n">states_to_ignore</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">char_matrix</span><span class="o">.</span><span class="n">default_state_alphabet</span><span class="o">.</span><span class="n">gap_state</span><span class="p">,</span> <span class="n">char_matrix</span><span class="o">.</span><span class="n">default_state_alphabet</span><span class="o">.</span><span class="n">no_data_state</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">attr</span> <span class="o">=</span> <span class="s2">&quot;fundamental_indexes&quot;</span>
        <span class="n">states_to_ignore</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">ancestral_sequence</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">ancestral_sequence</span> <span class="o">=</span> <span class="n">char_matrix</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">ancestral_fundamental_ids</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">c1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ancestral_sequence</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">c1</span> <span class="ow">in</span> <span class="n">states_to_ignore</span><span class="p">:</span>
            <span class="n">ancestral_fundamental_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ancestral_fundamental_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">attr</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">char_matrix</span><span class="p">:</span>
        <span class="n">s1</span> <span class="o">=</span>  <span class="n">char_matrix</span><span class="p">[</span><span class="n">taxon</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">c2</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">s1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">ancestral_fundamental_ids</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">c2</span> <span class="ow">in</span> <span class="n">states_to_ignore</span><span class="p">:</span>
                <span class="n">derived_matrix</span><span class="p">[</span><span class="n">taxon</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">derived_matrix</span><span class="o">.</span><span class="n">default_state_alphabet</span><span class="p">[</span><span class="s2">&quot;?&quot;</span><span class="p">])</span>
                <span class="k">continue</span>
            <span class="n">f2</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">c2</span><span class="p">,</span> <span class="n">attr</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">f2</span> <span class="o">==</span> <span class="n">ancestral_fundamental_ids</span><span class="p">[</span><span class="n">idx</span><span class="p">]:</span>
                <span class="n">derived_matrix</span><span class="p">[</span><span class="n">taxon</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">derived_matrix</span><span class="o">.</span><span class="n">default_state_alphabet</span><span class="p">[</span><span class="s2">&quot;0&quot;</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">derived_matrix</span><span class="p">[</span><span class="n">taxon</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">derived_matrix</span><span class="o">.</span><span class="n">default_state_alphabet</span><span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">derived_matrix</span></div>


<div class="viewcode-block" id="unfolded_site_frequency_spectrum">
<a class="viewcode-back" href="../../../library/popgenstat.html#dendropy.calculate.popgenstat.unfolded_site_frequency_spectrum">[docs]</a>
<span class="k">def</span> <span class="nf">unfolded_site_frequency_spectrum</span><span class="p">(</span>
        <span class="n">char_matrix</span><span class="p">,</span>
        <span class="n">ancestral_sequence</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">ignore_uncertain</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">pad</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns the site frequency spectrum of list of CharDataSequence objects given by char_sequences,</span>
<span class="sd">    with reference to the ancestral sequence given by ancestral_seq. If ancestral_seq</span>
<span class="sd">    is None, then the first sequence in char_sequences is taken to be the ancestral</span>
<span class="sd">    sequence.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">dsm</span> <span class="o">=</span> <span class="n">derived_state_matrix</span><span class="p">(</span>
            <span class="n">char_matrix</span><span class="o">=</span><span class="n">char_matrix</span><span class="p">,</span>
            <span class="n">ancestral_sequence</span><span class="o">=</span><span class="n">ancestral_sequence</span><span class="p">,</span>
            <span class="n">derived_state_alphabet</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">ignore_uncertain</span><span class="o">=</span><span class="n">ignore_uncertain</span><span class="p">,</span>
            <span class="p">)</span>
    <span class="n">sites</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">dsm</span><span class="o">.</span><span class="n">sequences</span><span class="p">())</span> <span class="c1"># transpose</span>
    <span class="n">freqs</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">pad</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">char_matrix</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="n">freqs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">sites</span><span class="p">:</span>
        <span class="n">p</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">s</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">symbol</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">p</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">freqs</span><span class="p">:</span>
            <span class="n">freqs</span><span class="p">[</span><span class="n">p</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">freqs</span><span class="p">[</span><span class="n">p</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">freqs</span></div>

</pre></div>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div style="text-align: center; padding-top: 20px; padding-bottom: 5px; width: 100%;">
    <a href="../../../index.html"><img src="../../../_static/dendropy_logo.png" /></a>
</div><div style="clear:both; width: 100%; height:1px;"></div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script><div style="clear:both; width: 100%; height:1px;"></div><div style="clear:both; width: 100%; height:1px;"></div><div style="clear:both; width: 100%; height:1px;"></div>
    <!-- Documentation -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Documentation</h3>
        <ul>
            <li><a href="../../../primer/downloading.html">Downloading and Installing DendroPy</a></li>
            <li><a href="../../../primer/index.html">The DendroPy Primer</a></li>
            <li><a href="../../../library/index.html">Library API Reference</a></li>
            <li>
                <a href="../../../schemas/index.html">Schemas</a>
                <ul>
                    <li><a href="../../../schemas/fasta.html">FASTA</a></li>
                    <li><a href="../../../schemas/newick.html">Newick</a></li>
                    <li><a href="../../../schemas/nexml.html">NeXML</a></li>
                    <li><a href="../../../schemas/nexus.html">Nexus</a></li>
                    <li><a href="../../../schemas/phylip.html">PHYLIP</a></li>
                </ul>
            </li>
            <li>
                <a href="../../../programs/index.html">Programs</a>
                <ul>
                    <li><a href="../../../programs/sumtrees.html">SumTrees</a></li>
                </ul>
            </li>
            <li><a href="../../../glossary.html">Glossary and Terminological Reference</a></li>
            <li><a href="../../../developer.html">Developer Guide</a></li>
            <li><a href="../../../planning.html">Ongoing Development</a></li>
            <li><a href="../../../changes.html">Change History</a></li>
        </ul>
    </div>

    <!-- Downloads -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Obtaining</h3>
        <ul>
            <li><a target="_blank" href="http://pypi.python.org/pypi/DendroPy">Install from the Python Package Index</a></li>
            <li><a target="_blank" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.2.tar.gz">Download the Source Code Archive</a></li>
            <li><a target="_blank" href="http://github.com/jeetsukumaran/DendroPy">Clone the Source Code Repository</a></li>
        </ul>
    </div>

    <!-- Discussions -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Discussion</span><span style="position: absolute; right: 0; top: 10px "><img src="../../../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot; DendroPy Users&quot; </a>group to follow and participate in discussion, troubleshooting, help, information, suggestions, etc. on the usage and development of the DendroPy phylogenetic computing library.</p>
            <form action="http://groups.google.com/group/dendropy-users/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot;dendropy-users&quot;</a> group, or click <a href="http://groups.google.com/group/dendropy-users?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Announcements -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Announcements</span><span style="position: absolute; right: 0; top: 10px "><img src="../../../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; DendroPy Announcements&quot; </a>group to receive announcements of new releases, updates, changes and other news of interest to DendroPy users and developers.</p>
            <form action="http://groups.google.com/group/dendropy-announce/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; dendropy-announce&quot; </a>group, or click <a href="http://groups.google.com/group/dendropy-announce?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Development -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative; padding-bottom: 15px; margin-bottom:5px;">
        <h3><span style="text-align: left"><a href="https://github.com/jeetsukumaran/DendroPy/">Development</a></span><a href="https://github.com/jeetsukumaran/DendroPy/"><span style="position: absolute; right: 0; top: 10px "><img src="../../../_static/Octocat.png" height="30px" alt="GitHub" /></span></a></h3>
        <div style="margin-top: 15px;">
            <!-- <a href="https://github.com/jeetsukumaran/DendroPy/issues">Issues</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/subscription">Watch</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/fork">Fork</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/stargazers">Star</a> &bull; <a href="https://github.com/jeetsukumaran/">Follow</a> -->
            <ul>
                <li>                <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/issues"><code>Issues</code></a></span> <span style="font-style: italic; font-size:80%;"> - Report bugs or request features</span></li>
                <li>     <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/subscription"><code>&nbsp;Watch</code></a></span> <span style="font-style: italic; font-size:80%;"> - Follow development activity</span></li>
                <li>        <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/fork"><code>&nbsp;&nbsp;Fork</code></a></span> <span style="font-style: italic; font-size:80%;"> - Contribute and collaborate</span></li>
                <li>  <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/stargazers"><code>&nbsp;&nbsp;Star</code></a></span> <span style="font-style: italic; font-size:80%;"> - Throw some glitter, add some glamour</span></li>
        </div>
    </div>


        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">dendropy.calculate.popgenstat</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2009-2024, Jeet Sukumaran and Mark T. Holder.
    </div>
  </body>
</html>