<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>dendropy.datamodel.datasetmodel: Datasets – Aggregate Collections of Taxon, Character, and Tree Data &#8212; DendroPy 5.0.5 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/rtd.css?v=cedc6a2f" />
    
    <script src="../_static/documentation_options.js?v=967d581c"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="dendropy.calculate.phylogeneticdistance: Phylogenetic Distance Calculations and Operations" href="phylogeneticdistance.html" />
    <link rel="prev" title="dendropy.datamodel.charmatrixmodel: Character Sequences and Matrices" href="charmatrixmodel.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="phylogeneticdistance.html" title="dendropy.calculate.phylogeneticdistance: Phylogenetic Distance Calculations and Operations"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="charmatrixmodel.html" title="dendropy.datamodel.charmatrixmodel: Character Sequences and Matrices"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.5 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">DendroPy Library Reference</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href=""><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.datamodel.datasetmodel</span></code>: Datasets – Aggregate Collections of Taxon, Character, and Tree Data</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="dendropy-datamodel-datasetmodel-datasets-aggregate-collections-of-taxon-character-and-tree-data">
<h1><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.datamodel.datasetmodel</span></code>: Datasets – Aggregate Collections of Taxon, Character, and Tree Data<a class="headerlink" href="#dendropy-datamodel-datasetmodel-datasets-aggregate-collections-of-taxon-character-and-tree-data" title="Link to this heading">¶</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dendropy.datamodel.datasetmodel.</span></span><span class="sig-name descname"><span class="pre">DataSet</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet" title="Link to this definition">¶</a></dt>
<dd><p>A phylogenetic data object that coordinates collections of
<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>, <a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>, and (various kinds of)
<a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a> objects.</p>
<p>A <a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> has three attributes:</p>
<blockquote>
<div><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">taxon_namespaces</span></code></dt><dd><p>A list of <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> objects, each representing
a distinct namespace for operational taxononomic unit concept
definitions.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">tree_lists</span></code></dt><dd><p>A list of <a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> objects, each representing a
collection of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">char_matrices</span></code></dt><dd><p>A list of <a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a>-derived objects (e.g.
<a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.DnaCharacterMatrix" title="dendropy.datamodel.charmatrixmodel.DnaCharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">DnaCharacterMatrix</span></code></a>).</p>
</dd>
</dl>
</div></blockquote>
<p>Multiple <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> objects within a <a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> are
allowed so as to support reading/loading of data from external sources that
have multiple independent taxon namespaces defined within the same source
or document (e.g., a Mesquite file with multiple taxa blocks, or a NeXML
file with multiple OTU sections). Ideally, however, this would not
be how data is managed. Recommended idiomatic usage would be to use a
<a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> to manage multiple types of data that all share and
reference the same, single taxon namespace.</p>
<p>This convention can be enforced by setting the DataSet instance to
“attached taxon namespace” mode:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="p">()</span>
<span class="n">tns</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">()</span>
<span class="n">ds</span><span class="o">.</span><span class="n">attach_taxon_namespace</span><span class="p">(</span><span class="n">tns</span><span class="p">)</span>
</pre></div>
</div>
<p>After setting this mode, all subsequent data read or created will be
coerced to use the same, common operational taxonomic unit concept
namespace.</p>
<p>Note that unless there is a need to collect and serialize a collection of
data to the same file or external source, it is probably better
semantically to use more specific data structures (e.g., a
<a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> object for trees or a <a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.DnaCharacterMatrix" title="dendropy.datamodel.charmatrixmodel.DnaCharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">DnaCharacterMatrix</span></code></a>
object for an alignment). Similarly, when deserializing an external
data source, if just a single type or collection of data is needed (e.g.,
the collection of trees from a file that includes both trees and an
alignment), then it is semantically cleaner to deserialize the data
into a more specific structure (e.g., a <a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> to get all the
trees). However, when deserializing a mixed external data source
with, e.g. multiple alignments or trees and one or more alignments, and you
need to access/use more than a single collection, it is more efficient to
read the entire data source at once into a <a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object and then
independently extract the data objects as you need them from the various
collections.</p>
<p>The constructor can take one argument. This can either be another
<a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> instance or an iterable of <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>,
<a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>, or <a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a>-derived instances.</p>
<p>In the former case, the newly-constructed <a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> will be a
shallow-copy clone of the argument.</p>
<p>In the latter case, the newly-constructed <a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> will have
the elements of the iterable added to the respective collections
(<code class="docutils literal notranslate"><span class="pre">taxon_namespaces</span></code>, <code class="docutils literal notranslate"><span class="pre">tree_lists</span></code>, or <code class="docutils literal notranslate"><span class="pre">char_matrices</span></code>, as
appropriate). This is essentially like calling <a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.add" title="dendropy.datamodel.datasetmodel.DataSet.add"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DataSet.add</span></code></a>
on each element separately.</p>
<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.add">
<span class="sig-name descname"><span class="pre">add</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_object</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.add"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.add" title="Link to this definition">¶</a></dt>
<dd><p>Generic add for TaxonNamespace, TreeList or CharacterMatrix objects.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.add_char_matrix">
<span class="sig-name descname"><span class="pre">add_char_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char_matrix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.add_char_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.add_char_matrix" title="Link to this definition">¶</a></dt>
<dd><p>Adds a <a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a> or <a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a>-derived
instance to this dataset if it is not already there.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>char_matrix</strong> (<a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a>) – The <a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a> object to be added.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.add_taxon_namespace">
<span class="sig-name descname"><span class="pre">add_taxon_namespace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon_namespace</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.add_taxon_namespace"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.add_taxon_namespace" title="Link to this definition">¶</a></dt>
<dd><p>Adds a taxonomic unit concept namespace represented by a
<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> instance to this dataset if it is not already
there.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>taxon_namespace</strong> (<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>) – The <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> object to be added.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.add_taxon_set">
<span class="sig-name descname"><span class="pre">add_taxon_set</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon_set</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.add_taxon_set"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.add_taxon_set" title="Link to this definition">¶</a></dt>
<dd><p>DEPRECATED: Use <a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.add_taxon_namespace" title="dendropy.datamodel.datasetmodel.DataSet.add_taxon_namespace"><code class="xref any py py-meth docutils literal notranslate"><span class="pre">add_taxon_namespace()</span></code></a> instead.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.add_tree_list">
<span class="sig-name descname"><span class="pre">add_tree_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.add_tree_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.add_tree_list" title="Link to this definition">¶</a></dt>
<dd><p>Adds a <a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> instance to this dataset if it is not already
there.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>tree_list</strong> (<a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>) – The <a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> object to be added.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.as_string">
<span class="sig-name descname"><span class="pre">as_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.as_string" title="Link to this definition">¶</a></dt>
<dd><p>Composes and returns string representation of the data.</p>
<p><strong>Mandatory Schema-Specification Keyword Argument:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>schema</strong> (<em>str</em>) – Identifier of format of data. See
“<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional Schema-Specific Keyword Arguments:</strong></p>
<blockquote>
<div><p>These provide control over how the data is formatted, and supported
argument names and values depend on the schema as specified by the
value passed as the “<code class="docutils literal notranslate"><span class="pre">schema</span></code>” argument. See “<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more
details.</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.attach_taxon_namespace">
<span class="sig-name descname"><span class="pre">attach_taxon_namespace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon_namespace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.attach_taxon_namespace"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.attach_taxon_namespace" title="Link to this definition">¶</a></dt>
<dd><p>Forces all read() calls of this DataSet to use the same <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>. If
<code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> If <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> is None, then a new <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> will be
created, added to <code class="docutils literal notranslate"><span class="pre">self.taxon_namespaces</span></code>, and that is the <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> that will be
attached.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.attach_taxon_set">
<span class="sig-name descname"><span class="pre">attach_taxon_set</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon_set</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.attach_taxon_set"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.attach_taxon_set" title="Link to this definition">¶</a></dt>
<dd><p>DEPRECATED: Use <a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.attach_taxon_namespace" title="dendropy.datamodel.datasetmodel.DataSet.attach_taxon_namespace"><code class="xref any py py-meth docutils literal notranslate"><span class="pre">attach_taxon_namespace()</span></code></a> instead.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.clone">
<span class="sig-name descname"><span class="pre">clone</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">depth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.clone" title="Link to this definition">¶</a></dt>
<dd><p>Creates and returns a copy of <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>depth</strong> (<em>integer</em>) – <p>The depth of the copy:</p>
<blockquote>
<div><ul class="simple">
<li><p>0: shallow-copy: All member objects are references,
except for :attr:<code class="docutils literal notranslate"><span class="pre">annotation_set</span></code> of top-level object and
member <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects: these are full,
independent instances (though any complex objects in the
<code class="docutils literal notranslate"><span class="pre">value</span></code> field of <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects are also
just references).</p></li>
<li><p>1: taxon-namespace-scoped copy: All member objects are full
independent instances, <em>except</em> for <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>
and <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> instances: these are references.</p></li>
<li><p>2: Exhaustive deep-copy: all objects are cloned.</p></li>
</ul>
</div></blockquote>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.copy_annotations_from">
<span class="sig-name descname"><span class="pre">copy_annotations_from</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attribute_object_mapper</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.copy_annotations_from" title="Link to this definition">¶</a></dt>
<dd><p>Copies annotations from <code class="docutils literal notranslate"><span class="pre">other</span></code>, which must be of <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotable" title="dendropy.datamodel.basemodel.Annotable"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotable</span></code></a>
type.</p>
<p>Copies are deep-copies, in that the <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects added
to the <code class="docutils literal notranslate"><span class="pre">annotation_set</span></code> <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a> collection of <code class="docutils literal notranslate"><span class="pre">self</span></code> are
independent copies of those in the <code class="docutils literal notranslate"><span class="pre">annotate_set</span></code> collection of
<code class="docutils literal notranslate"><span class="pre">other</span></code>. However, dynamic bound-attribute annotations retain references
to the original objects as given in <code class="docutils literal notranslate"><span class="pre">other</span></code>, which may or may not be
desirable. This is handled by updated the objects to which attributes
are bound via mappings found in <code class="docutils literal notranslate"><span class="pre">attribute_object_mapper</span></code>.
In dynamic bound-attribute annotations, the <code class="docutils literal notranslate"><span class="pre">_value</span></code> attribute of the
annotations object (<code class="xref py py-attr docutils literal notranslate"><span class="pre">Annotation._value</span></code>) is a tuple consisting of
“<code class="docutils literal notranslate"><span class="pre">(obj,</span> <span class="pre">attr_name)</span></code>”, which instructs the <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> object to
return “<code class="docutils literal notranslate"><span class="pre">getattr(obj,</span> <span class="pre">attr_name)</span></code>” (via: “<code class="docutils literal notranslate"><span class="pre">getattr(*self._value)</span></code>”)
when returning the value of the Annotation. “<code class="docutils literal notranslate"><span class="pre">obj</span></code>” is typically the object
to which the <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a> belongs (i.e., <code class="docutils literal notranslate"><span class="pre">self</span></code>). When a copy
of <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> is created, the object reference given in the
first element of the <code class="docutils literal notranslate"><span class="pre">_value</span></code> tuple of dynamic bound-attribute
annotations are unchanged, unless the id of the object reference is fo</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotable" title="dendropy.datamodel.basemodel.Annotable"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotable</span></code></a>) – Source of annotations to copy.</p></li>
<li><p><strong>attribute_object_mapper</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a>) – Like the <code class="docutils literal notranslate"><span class="pre">memo</span></code> of <code class="docutils literal notranslate"><span class="pre">__deepcopy__</span></code>, maps object id’s to objects. The
purpose of this is to update the parent or owner objects of dynamic
attribute annotations.
If a dynamic attribute <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a>
gives object <code class="docutils literal notranslate"><span class="pre">x</span></code> as the parent or owner of the attribute (that is,
the first element of the <code class="xref py py-attr docutils literal notranslate"><span class="pre">Annotation._value</span></code> tuple is
<code class="docutils literal notranslate"><span class="pre">other</span></code>) and <code class="docutils literal notranslate"><span class="pre">id(x)</span></code> is found in <code class="docutils literal notranslate"><span class="pre">attribute_object_mapper</span></code>,
then in the copy the owner of the attribute is changed to
<code class="docutils literal notranslate"><span class="pre">attribute_object_mapper[id(x)]</span></code>.
If <code class="docutils literal notranslate"><span class="pre">attribute_object_mapper</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), then the
following mapping is automatically inserted: <code class="docutils literal notranslate"><span class="pre">id(other):</span> <span class="pre">self</span></code>.
That is, any references to <code class="docutils literal notranslate"><span class="pre">other</span></code> in any <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a>
object will be remapped to <code class="docutils literal notranslate"><span class="pre">self</span></code>.  If really no reattribution
mappings are desired, then an empty dictionary should be passed
instead.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.deep_copy_annotations_from">
<span class="sig-name descname"><span class="pre">deep_copy_annotations_from</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.deep_copy_annotations_from" title="Link to this definition">¶</a></dt>
<dd><p>Note that all references to <code class="docutils literal notranslate"><span class="pre">other</span></code> in any annotation value (and
sub-annotation, and sub-sub-sub-annotation, etc.) will be
replaced with references to <code class="docutils literal notranslate"><span class="pre">self</span></code>. This may not always make sense
(i.e., a reference to a particular entity may be absolute regardless of
context).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.detach_taxon_namespace">
<span class="sig-name descname"><span class="pre">detach_taxon_namespace</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.detach_taxon_namespace"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.detach_taxon_namespace" title="Link to this definition">¶</a></dt>
<dd><p>Relaxes constraint forcing all</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.detach_taxon_set">
<span class="sig-name descname"><span class="pre">detach_taxon_set</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.detach_taxon_set"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.detach_taxon_set" title="Link to this definition">¶</a></dt>
<dd><p>DEPRECATED: Use <a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.detach_taxon_namespace" title="dendropy.datamodel.datasetmodel.DataSet.detach_taxon_namespace"><code class="xref any py py-meth docutils literal notranslate"><span class="pre">detach_taxon_namespace()</span></code></a> instead.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.get">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.get"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.get" title="Link to this definition">¶</a></dt>
<dd><p>Instantiate and return a <em>new</em> <a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> object from a data source.</p>
<p><strong>Mandatory Source-Specification Keyword Argument (Exactly One Required):</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>file</strong> (<em>file</em>) – File or file-like object of data opened for reading.</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path to file of data.</p></li>
<li><p><strong>url</strong> (<em>str</em>) – URL of data.</p></li>
<li><p><strong>data</strong> (<em>str</em>) – Data given directly.</p></li>
</ul>
</div></blockquote>
<p><strong>Mandatory Schema-Specification Keyword Argument:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>schema</strong> (<em>str</em>) – Identifier of format of data given by the
“<code class="docutils literal notranslate"><span class="pre">file</span></code>”, “<code class="docutils literal notranslate"><span class="pre">path</span></code>”, “<code class="docutils literal notranslate"><span class="pre">data</span></code>”, or “<code class="docutils literal notranslate"><span class="pre">url</span></code>” argument
specified above: “<a class="reference internal" href="../schemas/fasta.html"><span class="doc">fasta</span></a>”, “<a class="reference internal" href="../schemas/newick.html"><span class="doc">newick</span></a>”, “<a class="reference internal" href="../schemas/nexus.html"><span class="doc">nexus</span></a>”, or
“<a class="reference internal" href="../schemas/nexml.html"><span class="doc">nexml</span></a>”, “<a class="reference internal" href="../schemas/phylip.html"><span class="doc">phylip</span></a>”, etc. See “<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional General Keyword Arguments:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>exclude_trees</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then all tree data in the data
source will be skipped.</p></li>
<li><p><strong>exclude_chars</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then all character
data in the data source will be skipped.</p></li>
<li><p><strong>taxon_namespace</strong> (<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>) – The <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>
instance to use to <a class="reference internal" href="../primer/taxa.html"><span class="doc">manage the taxon names</span></a>.
If not specified, a new one will be created.</p></li>
<li><p><strong>ignore_unrecognized_keyword_arguments</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>,
then unsupported or unrecognized keyword arguments will not
result in an error. Default is <code class="docutils literal notranslate"><span class="pre">False</span></code>: unsupported keyword
arguments will result in an error.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional Schema-Specific Keyword Arguments:</strong></p>
<blockquote>
<div><p>These provide control over how the data is interpreted and
processed, and supported argument names and values depend on
the schema as specified by the value passed as the “<code class="docutils literal notranslate"><span class="pre">schema</span></code>”
argument. See “<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p>
</div></blockquote>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dataset1</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">path</span><span class="o">=</span><span class="s2">&quot;pythonidae.chars_and_trees.nex&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="n">dataset2</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">url</span><span class="o">=</span><span class="s2">&quot;http://purl.org/phylo/treebase/phylows/study/TB2:S1925?format=nexml&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.get_from_path">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_from_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.get_from_path" title="Link to this definition">¶</a></dt>
<dd><p>Factory method to return new object of this class from file
specified by string <code class="docutils literal notranslate"><span class="pre">src</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>src</strong> (<em>string</em>) – Full file path to source of data.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pdo</strong> (<em>phylogenetic data object</em>) – New instance of object, constructed and populated from data given
in source.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.get_from_stream">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_from_stream</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.get_from_stream" title="Link to this definition">¶</a></dt>
<dd><p>Factory method to return new object of this class from file-like object
<code class="docutils literal notranslate"><span class="pre">src</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>src</strong> (<em>file</em><em> or </em><em>file-like</em>) – Source of data.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pdo</strong> (<em>phylogenetic data object</em>) – New instance of object, constructed and populated from data given
in source.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.get_from_string">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.get_from_string" title="Link to this definition">¶</a></dt>
<dd><p>Factory method to return new object of this class from string <code class="docutils literal notranslate"><span class="pre">src</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>src</strong> (<em>string</em>) – Data as a string.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pdo</strong> (<em>phylogenetic data object</em>) – New instance of object, constructed and populated from data given
in source.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.get_from_url">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_from_url</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strip_markup</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.get_from_url" title="Link to this definition">¶</a></dt>
<dd><p>Factory method to return a new object of this class from
URL given by <code class="docutils literal notranslate"><span class="pre">src</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>src</strong> (<em>string</em>) – URL of location providing source of data.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pdo</strong> (<em>phylogenetic data object</em>) – New instance of object, constructed and populated from data given
in source.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.get_tree_list">
<span class="sig-name descname"><span class="pre">get_tree_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.get_tree_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.get_tree_list" title="Link to this definition">¶</a></dt>
<dd><p>Returns a TreeList object specified by label.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.new_char_matrix">
<span class="sig-name descname"><span class="pre">new_char_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char_matrix_type</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.new_char_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.new_char_matrix" title="Link to this definition">¶</a></dt>
<dd><p>Creation and accession of new <a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a> (of class
<code class="docutils literal notranslate"><span class="pre">char_matrix_type</span></code>) into <code class="docutils literal notranslate"><span class="pre">chars</span></code> of self.”</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.new_taxon_namespace">
<span class="sig-name descname"><span class="pre">new_taxon_namespace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.new_taxon_namespace"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.new_taxon_namespace" title="Link to this definition">¶</a></dt>
<dd><p>Creates a new <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> object, according to the arguments given
(passed to <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref any py py-class docutils literal notranslate"><span class="pre">TaxonNamespace()</span></code></a>), and adds it to this <a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.new_taxon_set">
<span class="sig-name descname"><span class="pre">new_taxon_set</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.new_taxon_set"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.new_taxon_set" title="Link to this definition">¶</a></dt>
<dd><p>DEPRECATED: Use <a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.new_taxon_namespace" title="dendropy.datamodel.datasetmodel.DataSet.new_taxon_namespace"><code class="xref any py py-meth docutils literal notranslate"><span class="pre">new_taxon_namespace()</span></code></a> instead.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.new_tree_list">
<span class="sig-name descname"><span class="pre">new_tree_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.new_tree_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.new_tree_list" title="Link to this definition">¶</a></dt>
<dd><p>Creates a new <a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> instance, adds it to this DataSet.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>*args</strong> (<em>positional arguments</em>) – Passed directly to <a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> constructor.</p></li>
<li><p><strong>**kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Passed directly to <a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> constructor.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>t</strong> (<em>|TreeList|</em>) – The new <a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> instance created.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.read">
<span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.read"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.read" title="Link to this definition">¶</a></dt>
<dd><p>Add data to <code class="docutils literal notranslate"><span class="pre">self</span></code> from data source.</p>
<p><strong>Mandatory Source-Specification Keyword Argument (Exactly One Required):</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>file</strong> (<em>file</em>) – File or file-like object of data opened for reading.</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path to file of data.</p></li>
<li><p><strong>url</strong> (<em>str</em>) – URL of data.</p></li>
<li><p><strong>data</strong> (<em>str</em>) – Data given directly.</p></li>
</ul>
</div></blockquote>
<p><strong>Mandatory Schema-Specification Keyword Argument:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>schema</strong> (<em>str</em>) – Identifier of format of data given by the
“<code class="docutils literal notranslate"><span class="pre">file</span></code>”, “<code class="docutils literal notranslate"><span class="pre">path</span></code>”, “<code class="docutils literal notranslate"><span class="pre">data</span></code>”, or “<code class="docutils literal notranslate"><span class="pre">url</span></code>” argument
specified above: “<a class="reference internal" href="../schemas/newick.html"><span class="doc">newick</span></a>”, “<a class="reference internal" href="../schemas/nexus.html"><span class="doc">nexus</span></a>”, or “<a class="reference internal" href="../schemas/nexml.html"><span class="doc">nexml</span></a>”. See
“<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional General Keyword Arguments:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>exclude_trees</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then all tree data in the data
source will be skipped.</p></li>
<li><p><strong>exclude_chars</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then all character
data in the data source will be skipped.</p></li>
<li><p><strong>taxon_namespace</strong> (<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>) – The <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>
instance to use to <a class="reference internal" href="../primer/taxa.html"><span class="doc">manage the taxon names</span></a>.
If not specified, a new one will be created unless the DataSet
object is in attached taxon namespace mode
(<code class="docutils literal notranslate"><span class="pre">self.attached_taxon_namespace</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code> but assigned
to a specific <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> instance).</p></li>
<li><p><strong>ignore_unrecognized_keyword_arguments</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>,
then unsupported or unrecognized keyword arguments will not
result in an error. Default is <code class="docutils literal notranslate"><span class="pre">False</span></code>: unsupported keyword
arguments will result in an error.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional Schema-Specific Keyword Arguments:</strong></p>
<blockquote>
<div><p>These provide control over how the data is interpreted and
processed, and supported argument names and values depend on
the schema as specified by the value passed as the “<code class="docutils literal notranslate"><span class="pre">schema</span></code>”
argument. See “<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p>
</div></blockquote>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="p">()</span>
<span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">path</span><span class="o">=</span><span class="s2">&quot;pythonidae.chars_and_trees.nex&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">url</span><span class="o">=</span><span class="s2">&quot;http://purl.org/phylo/treebase/phylows/study/TB2:S1925?format=nexml&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.read_from_path">
<span class="sig-name descname"><span class="pre">read_from_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.read_from_path" title="Link to this definition">¶</a></dt>
<dd><p>Reads data from file specified by <code class="docutils literal notranslate"><span class="pre">filepath</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filepath</strong> (<em>file</em><em> or </em><em>file-like</em>) – Full file path to source of data.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>n</strong> (<code class="docutils literal notranslate"><span class="pre">tuple</span></code> [integer]) – A value indicating size of data read, where “size” depends on
the object:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>: <strong>undefined</strong></p></li>
<li><p><a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>: number of trees</p></li>
<li><p><a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a>: number of sequences</p></li>
<li><p><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>: <code class="docutils literal notranslate"><span class="pre">tuple</span></code> (number of taxon namespaces, number of tree lists, number of matrices)</p></li>
</ul>
</div></blockquote>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.read_from_stream">
<span class="sig-name descname"><span class="pre">read_from_stream</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.read_from_stream" title="Link to this definition">¶</a></dt>
<dd><p>Reads from file (exactly equivalent to just <code class="docutils literal notranslate"><span class="pre">read()</span></code>, provided
here as a separate method for completeness.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fileobj</strong> (<em>file</em><em> or </em><em>file-like</em>) – Source of data.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>n</strong> (<code class="docutils literal notranslate"><span class="pre">tuple</span></code> [integer]) – A value indicating size of data read, where “size” depends on
the object:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>: <strong>undefined</strong></p></li>
<li><p><a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>: number of trees</p></li>
<li><p><a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a>: number of sequences</p></li>
<li><p><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>: <code class="docutils literal notranslate"><span class="pre">tuple</span></code> (number of taxon namespaces, number of tree lists, number of matrices)</p></li>
</ul>
</div></blockquote>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.read_from_string">
<span class="sig-name descname"><span class="pre">read_from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.read_from_string" title="Link to this definition">¶</a></dt>
<dd><p>Reads a string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>src_str</strong> (<em>string</em>) – Data as a string.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>n</strong> (<code class="docutils literal notranslate"><span class="pre">tuple</span></code> [integer]) – A value indicating size of data read, where “size” depends on
the object:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>: <strong>undefined</strong></p></li>
<li><p><a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>: number of trees</p></li>
<li><p><a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a>: number of sequences</p></li>
<li><p><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>: <code class="docutils literal notranslate"><span class="pre">tuple</span></code> (number of taxon namespaces, number of tree lists, number of matrices)</p></li>
</ul>
</div></blockquote>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.read_from_url">
<span class="sig-name descname"><span class="pre">read_from_url</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.read_from_url" title="Link to this definition">¶</a></dt>
<dd><p>Reads a URL source.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>src</strong> (<em>string</em>) – URL of location providing source of data.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>n</strong> (<code class="docutils literal notranslate"><span class="pre">tuple</span></code> [integer]) – A value indicating size of data read, where “size” depends on
the object:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>: <strong>undefined</strong></p></li>
<li><p><a class="reference internal" href="treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>: number of trees</p></li>
<li><p><a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a>: number of sequences</p></li>
<li><p><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>: <code class="docutils literal notranslate"><span class="pre">tuple</span></code> (number of taxon namespaces, number of tree lists, number of matrices)</p></li>
</ul>
</div></blockquote>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.taxon_namespace_scoped_copy">
<span class="sig-name descname"><span class="pre">taxon_namespace_scoped_copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">memo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.taxon_namespace_scoped_copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.taxon_namespace_scoped_copy" title="Link to this definition">¶</a></dt>
<dd><p>Cloning level: 1.
Taxon-namespace-scoped copy: All member objects are full independent
instances, <em>except</em> for <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> and <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a>
objects: these are preserved as references.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.unify_taxon_namespaces">
<span class="sig-name descname"><span class="pre">unify_taxon_namespaces</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon_namespace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">case_sensitive_label_mapping</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attach_taxon_namespace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/datasetmodel.html#DataSet.unify_taxon_namespaces"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.unify_taxon_namespaces" title="Link to this definition">¶</a></dt>
<dd><p>Reindices taxa across all subcomponents, mapping to single taxon set.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.write">
<span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.write" title="Link to this definition">¶</a></dt>
<dd><p>Writes out <code class="docutils literal notranslate"><span class="pre">self</span></code> in <code class="docutils literal notranslate"><span class="pre">schema</span></code> format.</p>
<p><strong>Mandatory Destination-Specification Keyword Argument (Exactly One of the Following Required):</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>file</strong> (<em>file</em>) – File or file-like object opened for writing.</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path to file to which to write.</p></li>
</ul>
</div></blockquote>
<p><strong>Mandatory Schema-Specification Keyword Argument:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>schema</strong> (<em>str</em>) – Identifier of format of data. See
“<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional Schema-Specific Keyword Arguments:</strong></p>
<blockquote>
<div><p>These provide control over how the data is formatted, and supported
argument names and values depend on the schema as specified by the
value passed as the “<code class="docutils literal notranslate"><span class="pre">schema</span></code>” argument. See “<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more
details.</p>
</div></blockquote>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Using a file path:</span>
<span class="n">d</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;path/to/file.dat&quot;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>

<span class="c1"># Using an open file:</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;path/to/file.dat&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">d</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="n">f</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.write_to_path">
<span class="sig-name descname"><span class="pre">write_to_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.write_to_path" title="Link to this definition">¶</a></dt>
<dd><p>Writes to file specified by <code class="docutils literal notranslate"><span class="pre">dest</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.datasetmodel.DataSet.write_to_stream">
<span class="sig-name descname"><span class="pre">write_to_stream</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.datasetmodel.DataSet.write_to_stream" title="Link to this definition">¶</a></dt>
<dd><p>Writes to file-like object <code class="docutils literal notranslate"><span class="pre">dest</span></code>.</p>
</dd></dl>

</dd></dl>

</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div style="text-align: center; padding-top: 20px; padding-bottom: 5px; width: 100%;">
    <a href="../index.html"><img src="../_static/dendropy_logo.png" /></a>
</div><div style="clear:both; width: 100%; height:1px;"></div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h3><a href="../index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.datamodel.datasetmodel</span></code>: Datasets – Aggregate Collections of Taxon, Character, and Tree Data</a><ul>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet"><code class="docutils literal notranslate"><span class="pre">DataSet</span></code></a><ul>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.add"><code class="docutils literal notranslate"><span class="pre">DataSet.add</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.add_char_matrix"><code class="docutils literal notranslate"><span class="pre">DataSet.add_char_matrix</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.add_taxon_namespace"><code class="docutils literal notranslate"><span class="pre">DataSet.add_taxon_namespace</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.add_taxon_set"><code class="docutils literal notranslate"><span class="pre">DataSet.add_taxon_set</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.add_tree_list"><code class="docutils literal notranslate"><span class="pre">DataSet.add_tree_list</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.as_string"><code class="docutils literal notranslate"><span class="pre">DataSet.as_string</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.attach_taxon_namespace"><code class="docutils literal notranslate"><span class="pre">DataSet.attach_taxon_namespace</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.attach_taxon_set"><code class="docutils literal notranslate"><span class="pre">DataSet.attach_taxon_set</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.clone"><code class="docutils literal notranslate"><span class="pre">DataSet.clone</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.copy_annotations_from"><code class="docutils literal notranslate"><span class="pre">DataSet.copy_annotations_from</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.deep_copy_annotations_from"><code class="docutils literal notranslate"><span class="pre">DataSet.deep_copy_annotations_from</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.detach_taxon_namespace"><code class="docutils literal notranslate"><span class="pre">DataSet.detach_taxon_namespace</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.detach_taxon_set"><code class="docutils literal notranslate"><span class="pre">DataSet.detach_taxon_set</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.get"><code class="docutils literal notranslate"><span class="pre">DataSet.get</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.get_from_path"><code class="docutils literal notranslate"><span class="pre">DataSet.get_from_path</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.get_from_stream"><code class="docutils literal notranslate"><span class="pre">DataSet.get_from_stream</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.get_from_string"><code class="docutils literal notranslate"><span class="pre">DataSet.get_from_string</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.get_from_url"><code class="docutils literal notranslate"><span class="pre">DataSet.get_from_url</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.get_tree_list"><code class="docutils literal notranslate"><span class="pre">DataSet.get_tree_list</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.new_char_matrix"><code class="docutils literal notranslate"><span class="pre">DataSet.new_char_matrix</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.new_taxon_namespace"><code class="docutils literal notranslate"><span class="pre">DataSet.new_taxon_namespace</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.new_taxon_set"><code class="docutils literal notranslate"><span class="pre">DataSet.new_taxon_set</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.new_tree_list"><code class="docutils literal notranslate"><span class="pre">DataSet.new_tree_list</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.read"><code class="docutils literal notranslate"><span class="pre">DataSet.read</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.read_from_path"><code class="docutils literal notranslate"><span class="pre">DataSet.read_from_path</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.read_from_stream"><code class="docutils literal notranslate"><span class="pre">DataSet.read_from_stream</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.read_from_string"><code class="docutils literal notranslate"><span class="pre">DataSet.read_from_string</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.read_from_url"><code class="docutils literal notranslate"><span class="pre">DataSet.read_from_url</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.taxon_namespace_scoped_copy"><code class="docutils literal notranslate"><span class="pre">DataSet.taxon_namespace_scoped_copy</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.unify_taxon_namespaces"><code class="docutils literal notranslate"><span class="pre">DataSet.unify_taxon_namespaces</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.write"><code class="docutils literal notranslate"><span class="pre">DataSet.write</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.write_to_path"><code class="docutils literal notranslate"><span class="pre">DataSet.write_to_path</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.datasetmodel.DataSet.write_to_stream"><code class="docutils literal notranslate"><span class="pre">DataSet.write_to_stream</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

  </div><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h4>Previous topic</h4>
    <p class="topless"><a href="charmatrixmodel.html"
                          title="previous chapter"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.datamodel.charmatrixmodel</span></code>: Character Sequences and Matrices</a></p>
  </div>
  <div>
    <h4>Next topic</h4>
    <p class="topless"><a href="phylogeneticdistance.html"
                          title="next chapter"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.calculate.phylogeneticdistance</span></code>: Phylogenetic Distance Calculations and Operations</a></p>
  </div><div style="clear:both; width: 100%; height:1px;"></div>
    <!-- Documentation -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Documentation</h3>
        <ul>
            <li><a href="../primer/downloading.html">Downloading and Installing DendroPy</a></li>
            <li><a href="../primer/index.html">The DendroPy Primer</a></li>
            <li><a href="index.html">Library API Reference</a></li>
            <li>
                <a href="../schemas/index.html">Schemas</a>
                <ul>
                    <li><a href="../schemas/fasta.html">FASTA</a></li>
                    <li><a href="../schemas/newick.html">Newick</a></li>
                    <li><a href="../schemas/nexml.html">NeXML</a></li>
                    <li><a href="../schemas/nexus.html">Nexus</a></li>
                    <li><a href="../schemas/phylip.html">PHYLIP</a></li>
                </ul>
            </li>
            <li>
                <a href="../programs/index.html">Programs</a>
                <ul>
                    <li><a href="../programs/sumtrees.html">SumTrees</a></li>
                </ul>
            </li>
            <li><a href="../glossary.html">Glossary and Terminological Reference</a></li>
            <li><a href="../developer.html">Developer Guide</a></li>
            <li><a href="../planning.html">Ongoing Development</a></li>
            <li><a href="../changes.html">Change History</a></li>
        </ul>
    </div>

    <!-- Downloads -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Obtaining</h3>
        <ul>
            <li><a target="_blank" href="http://pypi.python.org/pypi/DendroPy">Install from the Python Package Index</a></li>
            <li><a target="_blank" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.5.tar.gz">Download the Source Code Archive</a></li>
            <li><a target="_blank" href="http://github.com/jeetsukumaran/DendroPy">Clone the Source Code Repository</a></li>
        </ul>
    </div>

    <!-- Discussions -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Discussion</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot; DendroPy Users&quot; </a>group to follow and participate in discussion, troubleshooting, help, information, suggestions, etc. on the usage and development of the DendroPy phylogenetic computing library.</p>
            <form action="http://groups.google.com/group/dendropy-users/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot;dendropy-users&quot;</a> group, or click <a href="http://groups.google.com/group/dendropy-users?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Announcements -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Announcements</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; DendroPy Announcements&quot; </a>group to receive announcements of new releases, updates, changes and other news of interest to DendroPy users and developers.</p>
            <form action="http://groups.google.com/group/dendropy-announce/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; dendropy-announce&quot; </a>group, or click <a href="http://groups.google.com/group/dendropy-announce?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Development -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative; padding-bottom: 15px; margin-bottom:5px;">
        <h3><span style="text-align: left"><a href="https://github.com/jeetsukumaran/DendroPy/">Development</a></span><a href="https://github.com/jeetsukumaran/DendroPy/"><span style="position: absolute; right: 0; top: 10px "><img src="../_static/Octocat.png" height="30px" alt="GitHub" /></span></a></h3>
        <div style="margin-top: 15px;">
            <!-- <a href="https://github.com/jeetsukumaran/DendroPy/issues">Issues</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/subscription">Watch</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/fork">Fork</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/stargazers">Star</a> &bull; <a href="https://github.com/jeetsukumaran/">Follow</a> -->
            <ul>
                <li>                <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/issues"><code>Issues</code></a></span> <span style="font-style: italic; font-size:80%;"> - Report bugs or request features</span></li>
                <li>     <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/subscription"><code>&nbsp;Watch</code></a></span> <span style="font-style: italic; font-size:80%;"> - Follow development activity</span></li>
                <li>        <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/fork"><code>&nbsp;&nbsp;Fork</code></a></span> <span style="font-style: italic; font-size:80%;"> - Contribute and collaborate</span></li>
                <li>  <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/stargazers"><code>&nbsp;&nbsp;Star</code></a></span> <span style="font-style: italic; font-size:80%;"> - Throw some glitter, add some glamour</span></li>
        </div>
    </div>


        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="phylogeneticdistance.html" title="dendropy.calculate.phylogeneticdistance: Phylogenetic Distance Calculations and Operations"
             >next</a> |</li>
        <li class="right" >
          <a href="charmatrixmodel.html" title="dendropy.datamodel.charmatrixmodel: Character Sequences and Matrices"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.5 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >DendroPy Library Reference</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href=""><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.datamodel.datasetmodel</span></code>: Datasets – Aggregate Collections of Taxon, Character, and Tree Data</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2009-2025, Jeet Sukumaran and Mark T. Holder.
    </div>
  </body>
</html>