<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>dendropy.calculate.phylogeneticdistance: Phylogenetic Distance Calculations and Operations &#8212; DendroPy 5.0.2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/rtd.css?v=cedc6a2f" />
    
    <script src="../_static/documentation_options.js?v=5df4c8d1"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="dendropy.model.birthdeath: The Birth-Death and Related Processes" href="birthdeath.html" />
    <link rel="prev" title="dendropy.datamodel.datasetmodel: Datasets – Aggregate Collections of Taxon, Character, and Tree Data" href="datasetmodel.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="birthdeath.html" title="dendropy.model.birthdeath: The Birth-Death and Related Processes"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="datasetmodel.html" title="dendropy.datamodel.datasetmodel: Datasets – Aggregate Collections of Taxon, Character, and Tree Data"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">DendroPy Library Reference</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href=""><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.calculate.phylogeneticdistance</span></code>: Phylogenetic Distance Calculations and Operations</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="module-dendropy.calculate.phylogeneticdistance">
<span id="dendropy-calculate-phylogeneticdistance-phylogenetic-distance-calculations-and-operations"></span><h1><a class="reference internal" href="#module-dendropy.calculate.phylogeneticdistance" title="dendropy.calculate.phylogeneticdistance"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.calculate.phylogeneticdistance</span></code></a>: Phylogenetic Distance Calculations and Operations<a class="headerlink" href="#module-dendropy.calculate.phylogeneticdistance" title="Link to this heading">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<section id="the-phylogeneticdistancematrix-class">
<h2>The <a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix" title="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix</span></code></a> Class<a class="headerlink" href="#the-phylogeneticdistancematrix-class" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dendropy.calculate.phylogeneticdistance.</span></span><span class="sig-name descname"><span class="pre">PhylogeneticDistanceMatrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">is_store_path_edges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix" title="Link to this definition">¶</a></dt>
<dd><p>Calculates and maintains patristic distance information of taxa on a tree.</p>
<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.as_data_table">
<span class="sig-name descname"><span class="pre">as_data_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">is_weighted_edge_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.as_data_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.as_data_table" title="Link to this definition">¶</a></dt>
<dd><p>Returns this as a table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.assemblage_membership_definitions_from_csv">
<span class="sig-name descname"><span class="pre">assemblage_membership_definitions_from_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_data_type=&lt;class</span> <span class="pre">'float'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">**csv_reader_kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.assemblage_membership_definitions_from_csv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.assemblage_membership_definitions_from_csv" title="Link to this definition">¶</a></dt>
<dd><p>Convenience method to return list of community sets from a delimited
file that lists taxon (labels) in columns and community
presence/absences or abundances in rows.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.compile_from_tree">
<span class="sig-name descname"><span class="pre">compile_from_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.compile_from_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.compile_from_tree" title="Link to this definition">¶</a></dt>
<dd><p>Calculates the distances. Note that the path length (in number of
steps) between taxa that span the root will be off by one if
the tree is unrooted.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.distance">
<span class="sig-name descname"><span class="pre">distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_weighted_edge_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_normalize_by_tree_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.distance" title="Link to this definition">¶</a></dt>
<dd><p>Returns distance between taxon1 and taxon2.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.distances">
<span class="sig-name descname"><span class="pre">distances</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">is_weighted_edge_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_normalize_by_tree_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.distances"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.distances" title="Link to this definition">¶</a></dt>
<dd><p>Returns list of patristic distances.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.distinct_taxon_pair_iter">
<span class="sig-name descname"><span class="pre">distinct_taxon_pair_iter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filter_fn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.distinct_taxon_pair_iter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.distinct_taxon_pair_iter" title="Link to this definition">¶</a></dt>
<dd><p>Iterates over all distinct pairs of taxa in matrix.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.from_csv">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_namespace=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_allow_new_taxa=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_first_row_column_names=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_first_column_row_names=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_data_type=&lt;class</span> <span class="pre">'float'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_transform_fn=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">**csv_reader_kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.from_csv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.from_csv" title="Link to this definition">¶</a></dt>
<dd><p>Instantiates a new PhylogeneticDistanceMatrix instance with data
from an external source.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>src</strong> (<em>file</em><em> or </em><em>file-like</em>) – Source of data. This is a token delimited-file (e.g., a
comma-delimited or tab-delimited file) providing a table which
lists taxon labels in both rows and columns. The cells of the table
are numeric (typically real) values that indicate the distance
between the taxa of the current row and column. Note that <em>only</em>
the upper right section of the table is considered. The diagonals
values are typically zeroes and, in either case, ignored along with
the lower diagonal. Despite being ignored by the
PhylogeneticDistanceMatrix object, the values are parsed by the
underlying reader and thus have to be valid numerical values.</p></li>
<li><p><strong>taxon_namespace</strong> (<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> instance) – The taxon namespace with which to manage taxa. If this has
not already been pre-populated with the taxon names, then
<code class="docutils literal notranslate"><span class="pre">is_allow_new_taxa</span></code> should be set to <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p></li>
<li><p><strong>is_allow_new_taxa</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code>: we do <em>not</em> expect to encounter any new taxa in the
data file, and it is an error if we do. If <code class="docutils literal notranslate"><span class="pre">True</span></code>: we do expect to
encounter new taxa in the data file. The default value of this
depends on the value passed to <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code>. If
<code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code> or an empty <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>
instance, then unless explicitly set to <code class="docutils literal notranslate"><span class="pre">False</span></code>,
<code class="docutils literal notranslate"><span class="pre">is_allow_new_taxa</span></code> will default to <code class="docutils literal notranslate"><span class="pre">True</span></code>: allowing of creation
of new taxa corresponding to labels found in the data source. On
the other hand, if <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> is not None and its value is
a <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> instance with at least one taxon, unless
explicitly set to <code class="docutils literal notranslate"><span class="pre">True</span></code>, <code class="docutils literal notranslate"><span class="pre">is_allow_new_taxa</span></code> will default to
<code class="docutils literal notranslate"><span class="pre">False</span></code>, and it will be an error if taxon labels are found in the
data source that do not correspond (exactly) to <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects
defined in the taxon namespace.  This is to err on the side of
caution, to avoid (or rather, highlight) problems due to incorrect
or mismatching labels between the data source and the current taxon
namespace.</p></li>
<li><p><strong>is_first_row_column_names</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – By default <code class="docutils literal notranslate"><span class="pre">True</span></code>: assumes that first row lists the taxon names.
Set to <code class="docutils literal notranslate"><span class="pre">False</span></code> if there is no header row.</p></li>
<li><p><strong>is_first_column_row_names</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – By default <code class="docutils literal notranslate"><span class="pre">True</span></code>: assumes that first column lists the taxon names.
Set to <code class="docutils literal notranslate"><span class="pre">False</span></code> if there is now row name column.</p></li>
<li><p><strong>label_transform_fn</strong> (<em>function object</em>) – If not None, this should be a function object that takes a string
as an argument and returns another string. This function will be
applied to row and column labels before they are matched to taxon
labels in the <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> instance given by
<code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code>.</p></li>
<li><p><strong>**csv_reader_kwargs</strong> (<em>keyword arguments</em>) – This arguments will be passed to the underlying CSV reader.
The most important one is probably ‘delimiter’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pdm</strong> (<em>A |PhylogeneticDistanceMatrix| instance</em>)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">pdm1</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">PhylogeneticDistanceMatrix</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span>
        <span class="n">src</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="s2">&quot;data.csv&quot;</span><span class="p">),</span>
        <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">)</span>
<span class="n">pdm2</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">PhylogeneticDistanceMatrix</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span>
        <span class="n">src</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="s2">&quot;data.tsv&quot;</span><span class="p">),</span>
        <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.from_tree">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.from_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.from_tree" title="Link to this definition">¶</a></dt>
<dd><p>Creates and returns a <a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix" title="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix</span></code></a> based
on the given tree.</p>
<p>Note that this creates a “snapshot” of the current state of the tree.
Subsequent changes to the tree will not be reflected in
<a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix" title="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix</span></code></a> instances previously created.</p>
<p>Also note that syntactically you may prefer to use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pdm</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">phylogenetic_distance_matrix</span><span class="p">()</span>
</pre></div>
</div>
<p>instead of:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pdm</span> <span class="o">=</span> <span class="n">PhylogeneticDistanceMatrix</span><span class="o">.</span><span class="n">from_tree</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>tree</strong> (a <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance) – The <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> from which to get the phylogenetic distances.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pdm</strong> (<em>A |PhylogeneticDistanceMatrix| instance</em>)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;tree.nex&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="n">pdm1</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">PhylogeneticDistanceMatrix</span><span class="o">.</span><span class="n">from_tree</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>

<span class="c1"># following is equivalent to above and probably preferred:</span>
<span class="n">pdm2</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">phylogenetic_distance_matrix</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.mean_nearest_taxon_distance">
<span class="sig-name descname"><span class="pre">mean_nearest_taxon_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filter_fn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_weighted_edge_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_normalize_by_tree_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.mean_nearest_taxon_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.mean_nearest_taxon_distance" title="Link to this definition">¶</a></dt>
<dd><p>Calculates the phylogenetic ecology statistic “MNTD”[1,2] for the tree
(only considering taxa for which <code class="docutils literal notranslate"><span class="pre">filter_fn</span></code> returns True when
applied if <code class="docutils literal notranslate"><span class="pre">filter_fn</span></code> is specified).</p>
<p>The mean nearest taxon distance (mntd) is given by:</p>
<blockquote>
<div><div class="math notranslate nohighlight">
\[mntd = \frac{ \sum_{i}^{n} min(\delta_{i,j}) }{n},\]</div>
</div></blockquote>
<p>where <span class="math notranslate nohighlight">\(i \neq j\)</span>, <span class="math notranslate nohighlight">\(\delta_{i,j}\)</span> is the phylogenetic
distance between species <span class="math notranslate nohighlight">\(i\)</span> and <span class="math notranslate nohighlight">\(j\)</span>, and <span class="math notranslate nohighlight">\(n\)</span> is the number
of species in the sample.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filter_fn</strong> (<em>function object</em><em> or </em><em>None</em>) – If <code class="docutils literal notranslate"><span class="pre">None</span></code>, then all leaves will be considered. Otherwise should
be a function object that takes a Taxon instance as an argument and
returns <code class="docutils literal notranslate"><span class="pre">True</span></code> if it is to be included in the calculation or
<code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.
In trees sampled from multiple communites, <code class="docutils literal notranslate"><span class="pre">filter_fn</span></code> can be
used to restrict the calculation to only one community based on
some criteria.</p></li>
<li><p><strong>is_weighted_edge_distances</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> then the edge-weighted distance, i.e., considering edge
lengths, is returned. Otherwise the the path steps or the number of
edges rather then the sum of is_weighted_edge_distances edges, connecting two
taxa is considered.</p></li>
<li><p><strong>is_normalize_by_tree_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> then the results are normalized by the total tree length
or steps/edges (depending on whether edge-weighted or unweighted
distances are used, respectively). Otherwise, raw distances are
used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mntd</strong> (<em>float</em>) – The Mean Nearest Taxon Distance (MNTD) statistic for the daata.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;data.nex&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="n">pdm</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">PhylogeneticDistanceMatrix</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>

<span class="c1"># consider all tips</span>
<span class="n">mntd</span> <span class="o">=</span> <span class="n">pdm</span><span class="o">.</span><span class="n">mean_nearest_taxon_distance</span><span class="p">()</span>

<span class="c1"># only tips within the same community, based on the node annotation</span>
<span class="c1"># &quot;community&quot;</span>
<span class="n">mntds_by_community</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">community_label</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;3&quot;</span><span class="p">,):</span>
    <span class="n">filter_fn</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;community&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">community_label</span>
    <span class="n">mntd</span> <span class="o">=</span> <span class="n">pdm</span><span class="o">.</span><span class="n">mean_pairwise_distance</span><span class="p">(</span><span class="n">filter_fn</span><span class="o">=</span><span class="n">filter_fn</span><span class="p">)</span>
    <span class="n">mntds_by_community</span><span class="p">[</span><span class="n">community_label</span><span class="p">]</span> <span class="o">=</span> <span class="n">mntd</span>
</pre></div>
</div>
<p class="rubric">References</p>
<p>[1] Webb, C.O. 2000. Exploring the phylogenetic structure of
ecological communities: An example for rainforest trees. The
American Naturalist 156: 145-155.</p>
<p>[2] Swenson, N.G. Functional and Phylogenetic Ecology in R.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.mean_pairwise_distance">
<span class="sig-name descname"><span class="pre">mean_pairwise_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filter_fn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_weighted_edge_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_normalize_by_tree_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.mean_pairwise_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.mean_pairwise_distance" title="Link to this definition">¶</a></dt>
<dd><p>Calculates the phylogenetic ecology statistic “MPD”[1,2] for the tree
(only considering taxa for which <code class="docutils literal notranslate"><span class="pre">filter_fn</span></code> returns True when
applied if <code class="docutils literal notranslate"><span class="pre">filter_fn</span></code> is specified).</p>
<p>The mean pairwise distance (mpd) is given by:</p>
<blockquote>
<div><div class="math notranslate nohighlight">
\[mpd = \frac{ \sum_{i}^{n} \sum_{j}^{n} \delta_{i,j} }{n \choose 2},\]</div>
</div></blockquote>
<p>where <span class="math notranslate nohighlight">\(i \neq j\)</span>, <span class="math notranslate nohighlight">\(\delta_{i,j}\)</span> is the phylogenetic
distance between species <span class="math notranslate nohighlight">\(i\)</span> and <span class="math notranslate nohighlight">\(j\)</span>, and <span class="math notranslate nohighlight">\(n\)</span> is the number
of species in the sample.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filter_fn</strong> (<em>function object</em><em> or </em><em>None</em>) – If <code class="docutils literal notranslate"><span class="pre">None</span></code>, then all leaves will be considered. Otherwise should
be a function object that takes a Taxon instance as an argument and
returns <code class="docutils literal notranslate"><span class="pre">True</span></code> if it is to be included in the calculation or
<code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.
In trees sampled from multiple communites, <code class="docutils literal notranslate"><span class="pre">filter_fn</span></code> can be
used to restrict the calculation to only one community based on
some criteria.</p></li>
<li><p><strong>is_weighted_edge_distances</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> then the edge-weighted distance, i.e., considering edge
lengths, is returned. Otherwise the the path steps or the number of
edges rather then the sum of is_weighted_edge_distances edges, connecting two
taxa is considered.</p></li>
<li><p><strong>is_normalize_by_tree_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> then the results are normalized by the total tree length
or steps/edges (depending on whether edge-weighted or unweighted
distances are used, respectively). Otherwise, raw distances are
used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mpd</strong> (<em>float</em>) – The Mean Pairwise Distance (MPD) statistic for the daata.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;data.nex&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="n">pdm</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">PhylogeneticDistanceMatrix</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>

<span class="c1"># consider all tips</span>
<span class="n">mpd1</span> <span class="o">=</span> <span class="n">pdm</span><span class="o">.</span><span class="n">mean_pairwise_distance</span><span class="p">()</span>

<span class="c1"># only tips within the same community, based on the node annotation</span>
<span class="c1"># &quot;community&quot;</span>
<span class="n">mpds_by_community</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">community_label</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;3&quot;</span><span class="p">,):</span>
    <span class="n">filter_fn</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;community&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">community_label</span>
    <span class="n">mpd</span> <span class="o">=</span> <span class="n">pdm</span><span class="o">.</span><span class="n">mean_pairwise_distance</span><span class="p">(</span><span class="n">filter_fn</span><span class="o">=</span><span class="n">filter_fn</span><span class="p">)</span>
    <span class="n">mpds_by_community</span><span class="p">[</span><span class="n">community_label</span><span class="p">]</span> <span class="o">=</span> <span class="n">mpd</span>
</pre></div>
</div>
<p class="rubric">References</p>
<p>[1] Webb, C.O. 2000. Exploring the phylogenetic structure of
ecological communities: An example for rainforest trees. The
American Naturalist 156: 145-155.</p>
<p>[2] Swenson, N.G. Functional and Phylogenetic Ecology in R.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.mrca">
<span class="sig-name descname"><span class="pre">mrca</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.mrca"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.mrca" title="Link to this definition">¶</a></dt>
<dd><p>Returns MRCA of two taxon objects.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.nj_tree">
<span class="sig-name descname"><span class="pre">nj_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">is_weighted_edge_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tree_factory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.nj_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.nj_tree" title="Link to this definition">¶</a></dt>
<dd><p>Returns an Neighbor-Joining (NJ) tree based on the distances in the matrix.</p>
<p>Calculates and returns a tree under the Neighbor-Joining algorithm of
Saitou and Nei (1987) for the data in the matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>is_weighted_edge_distances</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> then edge lengths will be considered for distances.
Otherwise, just the number of edges.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>t</strong> (<em>|Tree|</em>) – A <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance corresponding to the Neighbor-Joining (NJ) tree
for this data.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>

<span class="c1"># Read data from a CSV file into a PhylogeneticDistanceMatrix</span>
<span class="c1"># object</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;distance_matrix.csv&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">src</span><span class="p">:</span>
    <span class="n">pdm</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">PhylogeneticDistanceMatrix</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span>
            <span class="n">src</span><span class="p">,</span>
            <span class="n">is_first_row_column_names</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_first_column_row_names</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_allow_new_taxa</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
            <span class="p">)</span>

<span class="c1"># Calculate the tree</span>
<span class="n">nj_tree</span> <span class="o">=</span> <span class="n">pdm</span><span class="o">.</span><span class="n">nj_tree</span><span class="p">()</span>

<span class="c1"># Print it</span>
<span class="nb">print</span><span class="p">(</span><span class="n">nj_tree</span><span class="o">.</span><span class="n">as_string</span><span class="p">(</span><span class="s2">&quot;nexus&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p class="rubric">References</p>
<p>Saitou, N. and Nei, M. (1987) The neighbor-joining method: a new method
for reconstructing phylogenetic trees. Molecular Biology and Evolution,
4: 406-425.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.path_edge_count">
<span class="sig-name descname"><span class="pre">path_edge_count</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_normalize_by_tree_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.path_edge_count"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.path_edge_count" title="Link to this definition">¶</a></dt>
<dd><p>Returns the number of edges between two taxon objects.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.path_edges">
<span class="sig-name descname"><span class="pre">path_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.path_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.path_edges" title="Link to this definition">¶</a></dt>
<dd><p>Returns the edges between two taxon objects.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.patristic_distance">
<span class="sig-name descname"><span class="pre">patristic_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_normalize_by_tree_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.patristic_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.patristic_distance" title="Link to this definition">¶</a></dt>
<dd><p>Returns patristic distance between two taxon objects.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.shuffle_taxa">
<span class="sig-name descname"><span class="pre">shuffle_taxa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">is_shuffle_phylogenetic_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_shuffle_phylogenetic_path_steps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_shuffle_mrca</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.shuffle_taxa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.shuffle_taxa" title="Link to this definition">¶</a></dt>
<dd><p>Randomly shuffles taxa in-situ.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.standardized_effect_size_mean_nearest_taxon_distance">
<span class="sig-name descname"><span class="pre">standardized_effect_size_mean_nearest_taxon_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assemblage_memberships</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_randomization_replicates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_weighted_edge_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_normalize_by_tree_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_skip_single_taxon_assemblages</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">null_model_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'taxa.label'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.standardized_effect_size_mean_nearest_taxon_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.standardized_effect_size_mean_nearest_taxon_distance" title="Link to this definition">¶</a></dt>
<dd><p>Returns the standardized effect size value for the MNTD statistic under
a null model under various community compositions.</p>
<p>The S.E.S. is given by:</p>
<blockquote>
<div><div class="math notranslate nohighlight">
\[SES(statistic) = \frac{observed - mean(model_{null})}{sd(model_{null})}\]</div>
</div></blockquote>
<p>This removes any bias associated with the decrease in variance in the
MPD statistic value as species richness increases to the point where
communities become saturated. Equivalent to -1 times the Nearest Taxon
Index when using phylogenetic distances.</p>
<p>In contrast to the function calculating the non-standardized effect
size version of this statistic, which uses filter function to specify
the subset of taxa to be considerd, here a collection of (multiple)
sets of taxa constituting a community is specified. This to allow
calculation of the null model statistic across all community sets for
each randomization replicate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>assemblage_memberships</strong> (iterable of iterable of <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects) – A collection of collections, e.g. a list of sets, with the elements
of each set being <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> instances. Each set specifies the
composition of a community. The standardized effect size of this
statistic will be calculated for each community as specified by a
set of <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> instances.</p></li>
<li><p><strong>num_randomization_replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a>) – Number of randomization replicates.</p></li>
<li><p><strong>is_weighted_edge_distances</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> then edge lengths will be considered for distances.
Otherwise, just the number of edges.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>r</strong> (<em>list of results</em>) – A list of results, with each result corresponding to a community
set given in <code class="docutils literal notranslate"><span class="pre">assemblage_memberships</span></code>. Each result consists of a named
tuple with the following elements:</p>
<blockquote>
<div><ul class="simple">
<li><p>obs       : the observed value of the statistic</p></li>
<li><dl class="simple">
<dt>null_model_mean<span class="classifier">the mean value of the statistic under the null</span></dt><dd><p>model</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>null_model_sd<span class="classifier">the standard deviation of the statistic under</span></dt><dd><p>the null model</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>z<span class="classifier">the standardized effect value of the statistic</span></dt><dd><p>(= SES as defined in [1] above)</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>p<span class="classifier">the p-value of the observed value of the</span></dt><dd><p>statistic under the null model.</p>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span>
        <span class="n">src</span><span class="o">=</span><span class="s2">&quot;data/community.tree.newick&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">,</span>
        <span class="n">rooting</span><span class="o">=</span><span class="s2">&quot;force-rooted&quot;</span><span class="p">)</span>
<span class="n">pdm</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">PhylogeneticDistanceMatrix</span><span class="o">.</span><span class="n">from_tree</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
<span class="n">assemblage_memberships</span> <span class="o">=</span> <span class="n">pdm</span><span class="o">.</span><span class="n">assemblage_membership_definitions_from_csv</span><span class="p">(</span><span class="s2">&quot;data/comm1.csv&quot;</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">pdm</span><span class="o">.</span><span class="n">standardized_effect_size_mean_nearest_taxon_distance</span><span class="p">(</span><span class="n">assemblage_memberships</span><span class="o">=</span><span class="n">assemblage_memberships</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.standardized_effect_size_mean_pairwise_distance">
<span class="sig-name descname"><span class="pre">standardized_effect_size_mean_pairwise_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assemblage_memberships</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_randomization_replicates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_weighted_edge_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_normalize_by_tree_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_skip_single_taxon_assemblages</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">null_model_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'taxa.label'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.standardized_effect_size_mean_pairwise_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.standardized_effect_size_mean_pairwise_distance" title="Link to this definition">¶</a></dt>
<dd><p>Returns the standardized effect size value for the MPD statistic under
a null model under various community compositions.</p>
<p>The S.E.S. is given by:</p>
<blockquote>
<div><div class="math notranslate nohighlight">
\[SES(statistic) = \frac{observed - mean(model_{null})}{sd(model_{null})}\]</div>
</div></blockquote>
<p>This removes any bias associated with the decrease in variance in the
MPD statistic value as species richness increases to the point where
communities become saturated. Equivalent to -1 times the Nearest
Relative Index (NRI) when using phylogenetic distances.</p>
<p>In contrast to the function calculating the non-standardized effect
size version of this statistic, which uses filter function to specify
the subset of taxa to be considerd, here a collection of (multiple)
sets of taxa constituting a community is specified. This to allow
calculation of the null model statistic across all community sets for
each randomization replicate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>assemblage_memberships</strong> (iterable of iterable of <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects) – A collection of collections, e.g. a list of sets, with the elements
of each set being <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> instances. Each set specifies the
composition of a community. The standardized effect size of this
statistic will be calculated for each community as specified by a
set of <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> instances.</p></li>
<li><p><strong>num_randomization_replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a>) – Number of randomization replicates.</p></li>
<li><p><strong>is_weighted_edge_distances</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> then edge lengths will be considered for distances.
Otherwise, just the number of edges.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>r</strong> (<em>list of results</em>) – A list of results, with each result corresponding to a community
set given in <code class="docutils literal notranslate"><span class="pre">assemblage_memberships</span></code>. Each result consists of a named
tuple with the following elements:</p>
<blockquote>
<div><ul class="simple">
<li><p>obs       : the observed value of the statistic</p></li>
<li><dl class="simple">
<dt>null_model_mean<span class="classifier">the mean value of the statistic under the null</span></dt><dd><p>model</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>null_model_sd<span class="classifier">the standard deviation of the statistic under</span></dt><dd><p>the null model</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>z<span class="classifier">the standardized effect value of the statistic</span></dt><dd><p>(= SES as defined in [1] above)</p>
</dd>
</dl>
</li>
<li><p>p         : the p-value of the observed value of the</p></li>
</ul>
</div></blockquote>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span>
        <span class="n">src</span><span class="o">=</span><span class="s2">&quot;data/community.tree.newick&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">,</span>
        <span class="n">rooting</span><span class="o">=</span><span class="s2">&quot;force-rooted&quot;</span><span class="p">)</span>
<span class="n">pdm</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">phylogenetic_distance_matrix</span><span class="p">()</span>
<span class="n">assemblage_membership_definitions</span> <span class="o">=</span> <span class="n">pdm</span><span class="o">.</span><span class="n">assemblage_membership_definitions_from_csv</span><span class="p">(</span><span class="s2">&quot;data/comm1.csv&quot;</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">pdm</span><span class="o">.</span><span class="n">standardized_effect_size_mean_pairwise_distance</span><span class="p">(</span><span class="n">assemblage_memberships</span><span class="o">=</span><span class="n">assemblage_membership_definitions</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.sum_of_distances">
<span class="sig-name descname"><span class="pre">sum_of_distances</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">is_weighted_edge_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_normalize_by_tree_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.sum_of_distances"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.sum_of_distances" title="Link to this definition">¶</a></dt>
<dd><p>Returns sum of patristic distances on tree.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.taxon_iter">
<span class="sig-name descname"><span class="pre">taxon_iter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filter_fn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.taxon_iter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.taxon_iter" title="Link to this definition">¶</a></dt>
<dd><p>Iterates over taxa in matrix. Note that this could be a subset of the taxa in
the associated taxon namespace.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.upgma_tree">
<span class="sig-name descname"><span class="pre">upgma_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">is_weighted_edge_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tree_factory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/calculate/phylogeneticdistance.html#PhylogeneticDistanceMatrix.upgma_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.upgma_tree" title="Link to this definition">¶</a></dt>
<dd><p>Returns an Unweighted Pair Group Method with Arithmetic Mean (UPGMA) tree
based on the distances in the matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>is_weighted_edge_distances</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> then edge lengths will be considered for distances.
Otherwise, just the number of edges.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>t</strong> (<em>|Tree|</em>) – A <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance corresponding to the UPGMA tree
for this data.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>

<span class="c1"># Read data from a CSV file into a PhylogeneticDistanceMatrix</span>
<span class="c1"># object</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;distance_matrix.csv&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">src</span><span class="p">:</span>
    <span class="n">pdm</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">PhylogeneticDistanceMatrix</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span>
            <span class="n">src</span><span class="p">,</span>
            <span class="n">is_first_row_column_names</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_first_column_row_names</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">is_allow_new_taxa</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
            <span class="p">)</span>

<span class="c1"># Calculate the tree</span>
<span class="n">upgma_tree</span> <span class="o">=</span> <span class="n">pdm</span><span class="o">.</span><span class="n">upgma_tree</span><span class="p">()</span>

<span class="c1"># Print it</span>
<span class="nb">print</span><span class="p">(</span><span class="n">upgma_tree</span><span class="o">.</span><span class="n">as_string</span><span class="p">(</span><span class="s2">&quot;nexus&quot;</span><span class="p">))</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div style="text-align: center; padding-top: 20px; padding-bottom: 5px; width: 100%;">
    <a href="../index.html"><img src="../_static/dendropy_logo.png" /></a>
</div><div style="clear:both; width: 100%; height:1px;"></div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h3><a href="../index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.calculate.phylogeneticdistance</span></code>: Phylogenetic Distance Calculations and Operations</a><ul>
<li><a class="reference internal" href="#the-phylogeneticdistancematrix-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix</span></code> Class</a><ul>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix</span></code></a><ul>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.as_data_table"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.as_data_table</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.assemblage_membership_definitions_from_csv"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.assemblage_membership_definitions_from_csv</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.compile_from_tree"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.compile_from_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.distance"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.distance</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.distances"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.distances</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.distinct_taxon_pair_iter"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.distinct_taxon_pair_iter</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.from_csv"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.from_csv</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.from_tree"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.from_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.mean_nearest_taxon_distance"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.mean_nearest_taxon_distance</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.mean_pairwise_distance"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.mean_pairwise_distance</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.mrca"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.mrca</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.nj_tree"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.nj_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.path_edge_count"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.path_edge_count</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.path_edges"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.path_edges</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.patristic_distance"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.patristic_distance</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.shuffle_taxa"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.shuffle_taxa</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.standardized_effect_size_mean_nearest_taxon_distance"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.standardized_effect_size_mean_nearest_taxon_distance</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.standardized_effect_size_mean_pairwise_distance"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.standardized_effect_size_mean_pairwise_distance</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.sum_of_distances"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.sum_of_distances</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.taxon_iter"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.taxon_iter</span></code></a></li>
<li><a class="reference internal" href="#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix.upgma_tree"><code class="docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix.upgma_tree</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

  </div><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h4>Previous topic</h4>
    <p class="topless"><a href="datasetmodel.html"
                          title="previous chapter"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.datamodel.datasetmodel</span></code>: Datasets – Aggregate Collections of Taxon, Character, and Tree Data</a></p>
  </div>
  <div>
    <h4>Next topic</h4>
    <p class="topless"><a href="birthdeath.html"
                          title="next chapter"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.model.birthdeath</span></code>: The Birth-Death and Related Processes</a></p>
  </div><div style="clear:both; width: 100%; height:1px;"></div>
    <!-- Documentation -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Documentation</h3>
        <ul>
            <li><a href="../primer/downloading.html">Downloading and Installing DendroPy</a></li>
            <li><a href="../primer/index.html">The DendroPy Primer</a></li>
            <li><a href="index.html">Library API Reference</a></li>
            <li>
                <a href="../schemas/index.html">Schemas</a>
                <ul>
                    <li><a href="../schemas/fasta.html">FASTA</a></li>
                    <li><a href="../schemas/newick.html">Newick</a></li>
                    <li><a href="../schemas/nexml.html">NeXML</a></li>
                    <li><a href="../schemas/nexus.html">Nexus</a></li>
                    <li><a href="../schemas/phylip.html">PHYLIP</a></li>
                </ul>
            </li>
            <li>
                <a href="../programs/index.html">Programs</a>
                <ul>
                    <li><a href="../programs/sumtrees.html">SumTrees</a></li>
                </ul>
            </li>
            <li><a href="../glossary.html">Glossary and Terminological Reference</a></li>
            <li><a href="../developer.html">Developer Guide</a></li>
            <li><a href="../planning.html">Ongoing Development</a></li>
            <li><a href="../changes.html">Change History</a></li>
        </ul>
    </div>

    <!-- Downloads -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Obtaining</h3>
        <ul>
            <li><a target="_blank" href="http://pypi.python.org/pypi/DendroPy">Install from the Python Package Index</a></li>
            <li><a target="_blank" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.2.tar.gz">Download the Source Code Archive</a></li>
            <li><a target="_blank" href="http://github.com/jeetsukumaran/DendroPy">Clone the Source Code Repository</a></li>
        </ul>
    </div>

    <!-- Discussions -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Discussion</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot; DendroPy Users&quot; </a>group to follow and participate in discussion, troubleshooting, help, information, suggestions, etc. on the usage and development of the DendroPy phylogenetic computing library.</p>
            <form action="http://groups.google.com/group/dendropy-users/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot;dendropy-users&quot;</a> group, or click <a href="http://groups.google.com/group/dendropy-users?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Announcements -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Announcements</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; DendroPy Announcements&quot; </a>group to receive announcements of new releases, updates, changes and other news of interest to DendroPy users and developers.</p>
            <form action="http://groups.google.com/group/dendropy-announce/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; dendropy-announce&quot; </a>group, or click <a href="http://groups.google.com/group/dendropy-announce?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Development -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative; padding-bottom: 15px; margin-bottom:5px;">
        <h3><span style="text-align: left"><a href="https://github.com/jeetsukumaran/DendroPy/">Development</a></span><a href="https://github.com/jeetsukumaran/DendroPy/"><span style="position: absolute; right: 0; top: 10px "><img src="../_static/Octocat.png" height="30px" alt="GitHub" /></span></a></h3>
        <div style="margin-top: 15px;">
            <!-- <a href="https://github.com/jeetsukumaran/DendroPy/issues">Issues</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/subscription">Watch</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/fork">Fork</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/stargazers">Star</a> &bull; <a href="https://github.com/jeetsukumaran/">Follow</a> -->
            <ul>
                <li>                <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/issues"><code>Issues</code></a></span> <span style="font-style: italic; font-size:80%;"> - Report bugs or request features</span></li>
                <li>     <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/subscription"><code>&nbsp;Watch</code></a></span> <span style="font-style: italic; font-size:80%;"> - Follow development activity</span></li>
                <li>        <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/fork"><code>&nbsp;&nbsp;Fork</code></a></span> <span style="font-style: italic; font-size:80%;"> - Contribute and collaborate</span></li>
                <li>  <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/stargazers"><code>&nbsp;&nbsp;Star</code></a></span> <span style="font-style: italic; font-size:80%;"> - Throw some glitter, add some glamour</span></li>
        </div>
    </div>


        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="birthdeath.html" title="dendropy.model.birthdeath: The Birth-Death and Related Processes"
             >next</a> |</li>
        <li class="right" >
          <a href="datasetmodel.html" title="dendropy.datamodel.datasetmodel: Datasets – Aggregate Collections of Taxon, Character, and Tree Data"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >DendroPy Library Reference</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href=""><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.calculate.phylogeneticdistance</span></code>: Phylogenetic Distance Calculations and Operations</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2009-2024, Jeet Sukumaran and Mark T. Holder.
    </div>
  </body>
</html>