<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>dendropy.datamodel.treecollectionmodel: Collections of Trees &#8212; DendroPy 5.0.5 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/rtd.css?v=cedc6a2f" />
    
    <script src="../_static/documentation_options.js?v=967d581c"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="dendropy.datamodel.charstatemodel: Character State Identities and Alphabets" href="charstatemodel.html" />
    <link rel="prev" title="dendropy.datamodel.treemodel: Trees" href="treemodel.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="charstatemodel.html" title="dendropy.datamodel.charstatemodel: Character State Identities and Alphabets"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="treemodel.html" title="dendropy.datamodel.treemodel: Trees"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.5 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">DendroPy Library Reference</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href=""><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.datamodel.treecollectionmodel</span></code>: Collections of Trees</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="module-dendropy.datamodel.treecollectionmodel">
<span id="dendropy-datamodel-treecollectionmodel-collections-of-trees"></span><h1><a class="reference internal" href="#module-dendropy.datamodel.treecollectionmodel" title="dendropy.datamodel.treecollectionmodel"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.datamodel.treecollectionmodel</span></code></a>: Collections of Trees<a class="headerlink" href="#module-dendropy.datamodel.treecollectionmodel" title="Link to this heading">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<section id="the-treelist-class">
<h2>The <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> Class<a class="headerlink" href="#the-treelist-class" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dendropy.datamodel.treecollectionmodel.</span></span><span class="sig-name descname"><span class="pre">TreeList</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="Link to this definition">¶</a></dt>
<dd><p>A collection of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects, all referencing the same “universe” of
opeational taxonomic unit concepts through the same <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>
object reference.</p>
<p>Constructs a new <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> object, populating it with any iterable
container with Tree object members passed as unnamed argument, or from
a data source if <code class="docutils literal notranslate"><span class="pre">stream</span></code> and <code class="docutils literal notranslate"><span class="pre">schema</span></code> are passed.</p>
<p>If passed an iterable container, the objects in that container must be
of type <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> (or derived). If the container is of type <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>,
then, because each <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> object must have the same <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>
reference as the containing <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>, the trees in the container
passed as an initialization argument will be <strong>deep</strong>-copied (except
for associated <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> and <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects, which will
be shallow-copied). If the container is any other type of
iterable, then the <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects will be <strong>shallow</strong>-copied.</p>
<p><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> objects can directly thus be instantiated in the
following ways:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># /usr/bin/env python</span>

<span class="kn">from</span> <span class="nn">dendropy</span> <span class="kn">import</span> <span class="n">TaxonNamespace</span><span class="p">,</span> <span class="n">Tree</span><span class="p">,</span> <span class="n">TreeList</span>

<span class="c1"># instantiate an empty tree</span>
<span class="n">tlst1</span> <span class="o">=</span> <span class="n">TreeList</span><span class="p">()</span>

<span class="c1"># TreeList objects can be instantiated from an external data source</span>
<span class="c1"># using the &#39;get()&#39; factory class method</span>

<span class="n">tlst2</span> <span class="o">=</span> <span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;treefile.tre&#39;</span><span class="p">,</span> <span class="s1">&#39;rU&#39;</span><span class="p">),</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>
<span class="n">tlst3</span> <span class="o">=</span> <span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s1">&#39;sometrees.nexus&#39;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="n">tlst4</span> <span class="o">=</span> <span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s2">&quot;((A,B),(C,D));((A,C),(B,D));&quot;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>

<span class="c1"># can also call `read()` on a TreeList object; each read adds</span>
<span class="c1"># (appends) the tree(s) found to the TreeList</span>
<span class="n">tlst5</span> <span class="o">=</span> <span class="n">TreeList</span><span class="p">()</span>
<span class="n">tlst5</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;boot1.tre&#39;</span><span class="p">,</span> <span class="s1">&#39;rU&#39;</span><span class="p">),</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>
<span class="n">tlst5</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;boot3.tre&quot;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>
<span class="n">tlst5</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s2">&quot;((A,B),(C,D));((A,C),(B,D));&quot;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>

<span class="c1"># populated from list of Tree objects</span>
<span class="n">tlist6_1</span> <span class="o">=</span> <span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="s2">&quot;((A,B),(C,D))&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>
<span class="n">tlist6_2</span> <span class="o">=</span> <span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="s2">&quot;((A,C),(B,D))&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>
<span class="n">tlist6</span> <span class="o">=</span> <span class="n">TreeList</span><span class="p">([</span><span class="n">tlist5_1</span><span class="p">,</span> <span class="n">tlist5_2</span><span class="p">])</span>

<span class="c1"># passing keywords to underlying tree parser</span>
<span class="n">tlst8</span> <span class="o">=</span> <span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
                 <span class="n">data</span><span class="o">=</span><span class="s2">&quot;((A,B),(C,D));((A,C),(B,D));&quot;</span><span class="p">,</span>
                 <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">,</span>
                 <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">tlst3</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">,</span>
                 <span class="n">rooting</span><span class="o">=</span><span class="s2">&quot;force-rooted&quot;</span><span class="p">,</span>
                 <span class="n">extract_comment_metadata</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                 <span class="n">store_tree_weights</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                 <span class="n">preserve_underscores</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Subsets of trees can be read. Note that in most cases, the entire</span>
<span class="c1"># data source is parsed, so this is not more efficient than reading</span>
<span class="c1"># all the trees and then manually-extracting them later; just more</span>
<span class="c1"># convenient</span>

<span class="c1"># skip the *first* 100 trees in the *first* (offset=0) collection of trees</span>
<span class="n">trees</span> <span class="o">=</span> <span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
            <span class="n">path</span><span class="o">=</span><span class="s2">&quot;mcmc.tre&quot;</span><span class="p">,</span>
            <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">,</span>
            <span class="n">collection_offset</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">tree_offset</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="c1"># get the *last* 10 trees in the *second* (offset=1) collection of trees</span>
<span class="n">trees</span> <span class="o">=</span> <span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
            <span class="n">path</span><span class="o">=</span><span class="s2">&quot;mcmc.tre&quot;</span><span class="p">,</span>
            <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">,</span>
            <span class="n">collection_offset</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">tree_offset</span><span class="o">=-</span><span class="mi">10</span><span class="p">)</span>

<span class="c1"># get the last 10 trees in the second-to-last collection of trees</span>
<span class="n">trees</span> <span class="o">=</span> <span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
            <span class="n">path</span><span class="o">=</span><span class="s2">&quot;mcmc.tre&quot;</span><span class="p">,</span>
            <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">,</span>
            <span class="n">collection_offset</span><span class="o">=-</span><span class="mi">2</span><span class="p">,</span>
            <span class="n">tree_offset</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="c1"># Slices give shallow-copy: trees are references</span>
<span class="n">tlst4copy0a</span> <span class="o">=</span> <span class="n">t4</span><span class="p">[:]</span>
<span class="k">assert</span> <span class="n">tlst4copy0a</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">is</span> <span class="n">t4</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">tlst4copy0b</span> <span class="o">=</span> <span class="n">t4</span><span class="p">[:</span><span class="mi">4</span><span class="p">]</span>
<span class="k">assert</span> <span class="n">tlst4copy0b</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">is</span> <span class="n">t4</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

<span class="c1"># &#39;Taxon-namespace-scoped&#39; copy:</span>
<span class="c1"># I.e., Deep-copied objects but taxa and taxon namespace</span>
<span class="c1"># are copied as references</span>
<span class="n">tlst4copy1a</span> <span class="o">=</span> <span class="n">TreeList</span><span class="p">(</span><span class="n">t4</span><span class="p">)</span>
<span class="n">tlst4copy1b</span> <span class="o">=</span> <span class="n">TreeList</span><span class="p">([</span><span class="n">Tree</span><span class="p">(</span><span class="n">t</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">tlst5</span><span class="p">])</span>
<span class="k">assert</span> <span class="n">tlst4copy1a</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">tlst4</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># True</span>
<span class="k">assert</span> <span class="n">tlst4copy1a</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="ow">is</span> <span class="n">tlst4</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="c1"># True</span>
<span class="k">assert</span> <span class="n">tlst4copy1b</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">tlst4</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># True</span>
<span class="k">assert</span> <span class="n">tlst4copy1b</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="ow">is</span> <span class="n">tlst4</span><span class="o">.</span><span class="n">taxon_namespace</span> <span class="c1"># True</span>
</pre></div>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.DEFAULT_TREE_TYPE">
<span class="sig-name descname"><span class="pre">DEFAULT_TREE_TYPE</span></span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.DEFAULT_TREE_TYPE" title="Link to this definition">¶</a></dt>
<dd><p>alias of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel._tree.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.__add__">
<span class="sig-name descname"><span class="pre">__add__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.__add__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.__add__" title="Link to this definition">¶</a></dt>
<dd><p>Creates and returns new <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> with clones of all trees in <code class="docutils literal notranslate"><span class="pre">self</span></code>
as well as all <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects in <code class="docutils literal notranslate"><span class="pre">other</span></code>.  If <code class="docutils literal notranslate"><span class="pre">other</span></code> is a
<a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>, then the trees are <em>cloned</em> and migrated into
<code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code>; otherwise, the original objects are migrated into
<code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> and added directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> (iterable of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects) – </p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>tlist</strong> (<em>|TreeList| object</em>) – <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> object containing clones of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects
in <code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.__getitem__" title="Link to this definition">¶</a></dt>
<dd><p>If <code class="docutils literal notranslate"><span class="pre">index</span></code> is an integer, then <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> object at position <code class="docutils literal notranslate"><span class="pre">index</span></code>
is returned. If <code class="docutils literal notranslate"><span class="pre">index</span></code> is a slice, then a <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> is returned
with references (i.e., not copies or clones, but the actual original
instances themselves) to <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects in the positions given
by the slice. The <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> is the same as <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<em>integer</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#slice" title="(in Python v3.13)"><em>slice</em></a>) – Index or slice.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>t</strong> (<em>|Tree| object or |TreeList| object</em>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.__iadd__">
<span class="sig-name descname"><span class="pre">__iadd__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.__iadd__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.__iadd__" title="Link to this definition">¶</a></dt>
<dd><p>In-place addition of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects in <code class="docutils literal notranslate"><span class="pre">other</span></code> to <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">other</span></code> is a <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>, then the trees are <em>copied</em>
and migrated into <code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code>; otherwise, the original
objects are migrated into <code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> and added directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> (iterable of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects) – </p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>``self``</strong> (<em>|TreeList|</em>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_import_strategy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'migrate'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.append"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.append" title="Link to this definition">¶</a></dt>
<dd><p>Adds a <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> object, <code class="docutils literal notranslate"><span class="pre">tree</span></code>, to the collection.</p>
<p>The <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> reference of <code class="docutils literal notranslate"><span class="pre">tree</span></code> will be set to that of
<code class="docutils literal notranslate"><span class="pre">self</span></code>.  Any <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects associated with nodes in <code class="docutils literal notranslate"><span class="pre">tree</span></code>
that are not already in <code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> will be handled
according to <code class="docutils literal notranslate"><span class="pre">taxon_import_strategy</span></code>:</p>
<blockquote>
<div><ul class="simple">
<li><dl class="simple">
<dt>‘migrate’</dt><dd><p><a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects associated with <code class="docutils literal notranslate"><span class="pre">tree</span></code> that are not already
in <code class="docutils literal notranslate"><span class="pre">self.taxon_nameaspace</span></code> will be remapped based on their
labels, with new :class|Taxon| objects being reconstructed if
none with matching labels are found. Specifically,
<a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree.migrate_taxon_namespace" title="dendropy.datamodel.treemodel.Tree.migrate_taxon_namespace"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dendropy.datamodel.treemodel.Tree.migrate_taxon_namespace</span></code></a>
will be called on <code class="docutils literal notranslate"><span class="pre">tree</span></code>, where <code class="docutils literal notranslate"><span class="pre">kwargs</span></code> is as passed to this
function.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>‘add’</dt><dd><p><a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects associated with <code class="docutils literal notranslate"><span class="pre">tree</span></code> that are not already
in <code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> will be added. Note that this might
result in <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects with duplicate labels as no
attempt at mapping to existing <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects based on
label-matching is done.</p>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (A <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance) – The <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> object to be added.</p></li>
<li><p><strong>taxon_import_strategy</strong> (<em>string</em>) – If <code class="docutils literal notranslate"><span class="pre">tree</span></code> is associated with a different <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>,
this argument determines how new <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects in <code class="docutils literal notranslate"><span class="pre">tree</span></code>
are handled: ‘migrate’ or ‘add’. See above for details.</p></li>
<li><p><strong>**kwargs</strong> (<em>keyword arguments</em>) – These arguments will be passed directly to
‘migrate_taxon_namespace()’ method call on <code class="docutils literal notranslate"><span class="pre">tree</span></code>.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.migrate_taxon_namespace</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.as_string">
<span class="sig-name descname"><span class="pre">as_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.as_string" title="Link to this definition">¶</a></dt>
<dd><p>Composes and returns string representation of the data.</p>
<p><strong>Mandatory Schema-Specification Keyword Argument:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>schema</strong> (<em>str</em>) – Identifier of format of data. See
“<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional Schema-Specific Keyword Arguments:</strong></p>
<blockquote>
<div><p>These provide control over how the data is formatted, and supported
argument names and values depend on the schema as specified by the
value passed as the “<code class="docutils literal notranslate"><span class="pre">schema</span></code>” argument. See “<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more
details.</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.as_tree_array">
<span class="sig-name descname"><span class="pre">as_tree_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.as_tree_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.as_tree_array" title="Link to this definition">¶</a></dt>
<dd><p>Return <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray" title="dendropy.datamodel.treecollectionmodel.TreeArray"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeArray</span></code></a> collecting information on splits in contained
trees. Keyword arguments get passed directly to <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray" title="dendropy.datamodel.treecollectionmodel.TreeArray"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeArray</span></code></a>
constructor.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.clone">
<span class="sig-name descname"><span class="pre">clone</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">depth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.clone" title="Link to this definition">¶</a></dt>
<dd><p>Creates and returns a copy of <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>depth</strong> (<em>integer</em>) – <p>The depth of the copy:</p>
<blockquote>
<div><ul class="simple">
<li><p>0: shallow-copy: All member objects are references,
except for :attr:<code class="docutils literal notranslate"><span class="pre">annotation_set</span></code> of top-level object and
member <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects: these are full,
independent instances (though any complex objects in the
<code class="docutils literal notranslate"><span class="pre">value</span></code> field of <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects are also
just references).</p></li>
<li><p>1: taxon-namespace-scoped copy: All member objects are full
independent instances, <em>except</em> for <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>
and <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> instances: these are references.</p></li>
<li><p>2: Exhaustive deep-copy: all objects are cloned.</p></li>
</ul>
</div></blockquote>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.consensus">
<span class="sig-name descname"><span class="pre">consensus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">min_freq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_bipartitions_updated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summarize_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.consensus"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.consensus" title="Link to this definition">¶</a></dt>
<dd><p>Returns a consensus tree of all trees in self, with minumum frequency
of bipartition to be added to the consensus tree given by <code class="docutils literal notranslate"><span class="pre">min_freq</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.copy_annotations_from">
<span class="sig-name descname"><span class="pre">copy_annotations_from</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attribute_object_mapper</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.copy_annotations_from" title="Link to this definition">¶</a></dt>
<dd><p>Copies annotations from <code class="docutils literal notranslate"><span class="pre">other</span></code>, which must be of <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotable" title="dendropy.datamodel.basemodel.Annotable"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotable</span></code></a>
type.</p>
<p>Copies are deep-copies, in that the <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects added
to the <code class="docutils literal notranslate"><span class="pre">annotation_set</span></code> <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a> collection of <code class="docutils literal notranslate"><span class="pre">self</span></code> are
independent copies of those in the <code class="docutils literal notranslate"><span class="pre">annotate_set</span></code> collection of
<code class="docutils literal notranslate"><span class="pre">other</span></code>. However, dynamic bound-attribute annotations retain references
to the original objects as given in <code class="docutils literal notranslate"><span class="pre">other</span></code>, which may or may not be
desirable. This is handled by updated the objects to which attributes
are bound via mappings found in <code class="docutils literal notranslate"><span class="pre">attribute_object_mapper</span></code>.
In dynamic bound-attribute annotations, the <code class="docutils literal notranslate"><span class="pre">_value</span></code> attribute of the
annotations object (<code class="xref py py-attr docutils literal notranslate"><span class="pre">Annotation._value</span></code>) is a tuple consisting of
“<code class="docutils literal notranslate"><span class="pre">(obj,</span> <span class="pre">attr_name)</span></code>”, which instructs the <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> object to
return “<code class="docutils literal notranslate"><span class="pre">getattr(obj,</span> <span class="pre">attr_name)</span></code>” (via: “<code class="docutils literal notranslate"><span class="pre">getattr(*self._value)</span></code>”)
when returning the value of the Annotation. “<code class="docutils literal notranslate"><span class="pre">obj</span></code>” is typically the object
to which the <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a> belongs (i.e., <code class="docutils literal notranslate"><span class="pre">self</span></code>). When a copy
of <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> is created, the object reference given in the
first element of the <code class="docutils literal notranslate"><span class="pre">_value</span></code> tuple of dynamic bound-attribute
annotations are unchanged, unless the id of the object reference is fo</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotable" title="dendropy.datamodel.basemodel.Annotable"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotable</span></code></a>) – Source of annotations to copy.</p></li>
<li><p><strong>attribute_object_mapper</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><em>dict</em></a>) – Like the <code class="docutils literal notranslate"><span class="pre">memo</span></code> of <code class="docutils literal notranslate"><span class="pre">__deepcopy__</span></code>, maps object id’s to objects. The
purpose of this is to update the parent or owner objects of dynamic
attribute annotations.
If a dynamic attribute <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a>
gives object <code class="docutils literal notranslate"><span class="pre">x</span></code> as the parent or owner of the attribute (that is,
the first element of the <code class="xref py py-attr docutils literal notranslate"><span class="pre">Annotation._value</span></code> tuple is
<code class="docutils literal notranslate"><span class="pre">other</span></code>) and <code class="docutils literal notranslate"><span class="pre">id(x)</span></code> is found in <code class="docutils literal notranslate"><span class="pre">attribute_object_mapper</span></code>,
then in the copy the owner of the attribute is changed to
<code class="docutils literal notranslate"><span class="pre">attribute_object_mapper[id(x)]</span></code>.
If <code class="docutils literal notranslate"><span class="pre">attribute_object_mapper</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), then the
following mapping is automatically inserted: <code class="docutils literal notranslate"><span class="pre">id(other):</span> <span class="pre">self</span></code>.
That is, any references to <code class="docutils literal notranslate"><span class="pre">other</span></code> in any <a class="reference internal" href="basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a>
object will be remapped to <code class="docutils literal notranslate"><span class="pre">self</span></code>.  If really no reattribution
mappings are desired, then an empty dictionary should be passed
instead.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.deep_copy_annotations_from">
<span class="sig-name descname"><span class="pre">deep_copy_annotations_from</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.deep_copy_annotations_from" title="Link to this definition">¶</a></dt>
<dd><p>Note that all references to <code class="docutils literal notranslate"><span class="pre">other</span></code> in any annotation value (and
sub-annotation, and sub-sub-sub-annotation, etc.) will be
replaced with references to <code class="docutils literal notranslate"><span class="pre">self</span></code>. This may not always make sense
(i.e., a reference to a particular entity may be absolute regardless of
context).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.extend">
<span class="sig-name descname"><span class="pre">extend</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.extend"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.extend" title="Link to this definition">¶</a></dt>
<dd><p>In-place addition of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects in <code class="docutils literal notranslate"><span class="pre">other</span></code> to <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">other</span></code> is a <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>, then the trees are <em>copied</em>
and migrated into <code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code>; otherwise, the original
objects are migrated into <code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> and added directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> (iterable of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects) – </p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>``self``</strong> (<em>|TreeList|</em>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.frequency_of_bipartition">
<span class="sig-name descname"><span class="pre">frequency_of_bipartition</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.frequency_of_bipartition"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.frequency_of_bipartition" title="Link to this definition">¶</a></dt>
<dd><p>Given a bipartition specified as:</p>
<blockquote>
<div><ul class="simple">
<li><p>a <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Bipartition" title="dendropy.datamodel.treemodel.Bipartition"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bipartition</span></code></a> instance given the keyword ‘bipartition’</p></li>
<li><p>a split bitmask given the keyword ‘split_bitmask’</p></li>
<li><p>a list of <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects given with the keyword <code class="docutils literal notranslate"><span class="pre">taxa</span></code></p></li>
<li><p>a list of taxon labels given with the keyword <code class="docutils literal notranslate"><span class="pre">labels</span></code></p></li>
</ul>
</div></blockquote>
<p>this function returns the proportion of trees in self
in which the split is found.</p>
<p>If the tree(s) in the collection are unrooted, then the bipartition
will be normalized for the comparison.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.frequency_of_split">
<span class="sig-name descname"><span class="pre">frequency_of_split</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.frequency_of_split"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.frequency_of_split" title="Link to this definition">¶</a></dt>
<dd><p>DEPRECATED: use ‘frequency_of_bipartition()’ instead.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.get">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.get"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.get" title="Link to this definition">¶</a></dt>
<dd><p>Instantiate and return a <em>new</em> <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> object from a data source.</p>
<p><strong>Mandatory Source-Specification Keyword Argument (Exactly One Required):</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>file</strong> (<em>file</em>) – File or file-like object of data opened for reading.</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path to file of data.</p></li>
<li><p><strong>url</strong> (<em>str</em>) – URL of data.</p></li>
<li><p><strong>data</strong> (<em>str</em>) – Data given directly.</p></li>
</ul>
</div></blockquote>
<p><strong>Mandatory Schema-Specification Keyword Argument:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>schema</strong> (<em>str</em>) – Identifier of format of data given by the
“<code class="docutils literal notranslate"><span class="pre">file</span></code>”, “<code class="docutils literal notranslate"><span class="pre">path</span></code>”, “<code class="docutils literal notranslate"><span class="pre">data</span></code>”, or “<code class="docutils literal notranslate"><span class="pre">url</span></code>” argument
specified above: “<a class="reference internal" href="../schemas/newick.html"><span class="doc">newick</span></a>”, “<a class="reference internal" href="../schemas/nexus.html"><span class="doc">nexus</span></a>”, or “<a class="reference internal" href="../schemas/nexml.html"><span class="doc">nexml</span></a>”. See
“<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional General Keyword Arguments:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>label</strong> (<em>str</em>) – Name or identifier to be assigned to the new
object; if not given, will be assigned the one specified in the
data source, or <code class="docutils literal notranslate"><span class="pre">None</span></code> otherwise.</p></li>
<li><p><strong>taxon_namespace</strong> (<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>) – The <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>
instance to use to <a class="reference internal" href="../primer/taxa.html"><span class="doc">manage the taxon names</span></a>.
If not specified, a new one will be created.</p></li>
<li><p><strong>collection_offset</strong> (<em>int</em>) – 0-based index of tree block or
collection in source to be parsed. If not specified then the
first collection (offset = 0) is assumed.</p></li>
<li><p><strong>tree_offset</strong> (<em>int</em>) – 0-based index of first tree within the
collection specified by <code class="docutils literal notranslate"><span class="pre">collection_offset</span></code> to be parsed (i.e.,
skipping the first <code class="docutils literal notranslate"><span class="pre">tree_offset</span></code> trees). If not
specified, then the first tree (offset = 0) is assumed (i.e., no
trees within the specified collection will be skipped). Use this
to specify, e.g. a burn-in.</p></li>
<li><p><strong>ignore_unrecognized_keyword_arguments</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>,
then unsupported or unrecognized keyword arguments will not
result in an error. Default is <code class="docutils literal notranslate"><span class="pre">False</span></code>: unsupported keyword
arguments will result in an error.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional Schema-Specific Keyword Arguments:</strong></p>
<blockquote>
<div><p>These provide control over how the data is interpreted and
processed, and supported argument names and values depend on
the schema as specified by the value passed as the “<code class="docutils literal notranslate"><span class="pre">schema</span></code>”
argument. See “<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p>
</div></blockquote>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tlst1</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">file</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;treefile.tre&#39;</span><span class="p">,</span> <span class="s1">&#39;rU&#39;</span><span class="p">),</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>
<span class="n">tlst2</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">path</span><span class="o">=</span><span class="s1">&#39;sometrees.nexus&#39;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">,</span>
        <span class="n">collection_offset</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">tree_offset</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">tlst3</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="s2">&quot;((A,B),(C,D));((A,C),(B,D));&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>
<span class="n">tree4</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">dendropy</span><span class="o">.</span><span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">url</span><span class="o">=</span><span class="s2">&quot;http://api.opentreeoflife.org/v2/study/pg_1144/tree/tree2324.nex&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.get_from_path">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_from_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.get_from_path" title="Link to this definition">¶</a></dt>
<dd><p>Factory method to return new object of this class from file
specified by string <code class="docutils literal notranslate"><span class="pre">src</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>src</strong> (<em>string</em>) – Full file path to source of data.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pdo</strong> (<em>phylogenetic data object</em>) – New instance of object, constructed and populated from data given
in source.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.get_from_stream">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_from_stream</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.get_from_stream" title="Link to this definition">¶</a></dt>
<dd><p>Factory method to return new object of this class from file-like object
<code class="docutils literal notranslate"><span class="pre">src</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>src</strong> (<em>file</em><em> or </em><em>file-like</em>) – Source of data.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pdo</strong> (<em>phylogenetic data object</em>) – New instance of object, constructed and populated from data given
in source.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.get_from_string">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.get_from_string" title="Link to this definition">¶</a></dt>
<dd><p>Factory method to return new object of this class from string <code class="docutils literal notranslate"><span class="pre">src</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>src</strong> (<em>string</em>) – Data as a string.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pdo</strong> (<em>phylogenetic data object</em>) – New instance of object, constructed and populated from data given
in source.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.get_from_url">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_from_url</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strip_markup</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.get_from_url" title="Link to this definition">¶</a></dt>
<dd><p>Factory method to return a new object of this class from
URL given by <code class="docutils literal notranslate"><span class="pre">src</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>src</strong> (<em>string</em>) – URL of location providing source of data.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pdo</strong> (<em>phylogenetic data object</em>) – New instance of object, constructed and populated from data given
in source.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_import_strategy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'migrate'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.insert"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.insert" title="Link to this definition">¶</a></dt>
<dd><p>Inserts a <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> object, <code class="docutils literal notranslate"><span class="pre">tree</span></code>, into the collection before
<code class="docutils literal notranslate"><span class="pre">index</span></code>.</p>
<p>The <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> reference of <code class="docutils literal notranslate"><span class="pre">tree</span></code> will be set to that of
<code class="docutils literal notranslate"><span class="pre">self</span></code>.  Any <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects associated with nodes in <code class="docutils literal notranslate"><span class="pre">tree</span></code>
that are not already in <code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> will be handled
according to <code class="docutils literal notranslate"><span class="pre">taxon_import_strategy</span></code>:</p>
<blockquote>
<div><ul class="simple">
<li><dl class="simple">
<dt>‘migrate’</dt><dd><p><a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects associated with <code class="docutils literal notranslate"><span class="pre">tree</span></code> that are not already
in <code class="docutils literal notranslate"><span class="pre">self.taxon_nameaspace</span></code> will be remapped based on their
labels, with new :class|Taxon| objects being reconstructed if
none with matching labels are found. Specifically,
<a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree.migrate_taxon_namespace" title="dendropy.datamodel.treemodel.Tree.migrate_taxon_namespace"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dendropy.datamodel.treemodel.Tree.migrate_taxon_namespace</span></code></a>
will be called on <code class="docutils literal notranslate"><span class="pre">tree</span></code>, where <code class="docutils literal notranslate"><span class="pre">kwargs</span></code> is as passed to
this function.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>‘add’</dt><dd><p><a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects associated with <code class="docutils literal notranslate"><span class="pre">tree</span></code> that are not already
in <code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> will be added. Note that this might
result in <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects with duplicate labels as no
attempt at mapping to existing <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects based on
label-matching is done.</p>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>index</strong> (<em>integer</em>) – Position before which to insert <code class="docutils literal notranslate"><span class="pre">tree</span></code>.</p></li>
<li><p><strong>tree</strong> (A <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance) – The <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> object to be added.</p></li>
<li><p><strong>taxon_import_strategy</strong> (<em>string</em>) – If <code class="docutils literal notranslate"><span class="pre">tree</span></code> is associated with a different <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>,
this argument determines how new <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects in <code class="docutils literal notranslate"><span class="pre">tree</span></code>
are handled: ‘migrate’ or ‘add’. See above for details.</p></li>
<li><p><strong>**kwargs</strong> (<em>keyword arguments</em>) – These arguments will be passed directly to
‘migrate_taxon_namespace()’ method call on <code class="docutils literal notranslate"><span class="pre">tree</span></code>.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.migrate_taxon_namespace</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.maximum_product_of_split_support_tree">
<span class="sig-name descname"><span class="pre">maximum_product_of_split_support_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_external_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'log_product_of_split_support'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.maximum_product_of_split_support_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.maximum_product_of_split_support_tree" title="Link to this definition">¶</a></dt>
<dd><p>Return the tree with that maximizes the product of split supports, also
known as the “Maximum Clade Credibility Tree” or MCCT.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>include_external_splits</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then non-internal split posteriors will be included in
the score. Defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>: these are skipped. This should only
make a difference when dealing with splits collected from trees of
different leaf sets.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mcct_tree</strong> (<em>Tree</em>) – Tree that maximizes the product of split supports.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.maximum_sum_of_split_support_tree">
<span class="sig-name descname"><span class="pre">maximum_sum_of_split_support_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_external_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sum_of_split_support'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.maximum_sum_of_split_support_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.maximum_sum_of_split_support_tree" title="Link to this definition">¶</a></dt>
<dd><p>Return the tree with that maximizes the <em>sum</em> of split supports.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>include_external_splits</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then non-internal split posteriors will be included in
the score. Defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>: these are skipped. This should only
make a difference when dealing with splits collected from trees of
different leaf sets.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mcct_tree</strong> (<em>Tree</em>) – Tree that maximizes the sum of split supports.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.migrate_taxon_namespace">
<span class="sig-name descname"><span class="pre">migrate_taxon_namespace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon_namespace</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unify_taxa_by_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_mapping_memo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.migrate_taxon_namespace" title="Link to this definition">¶</a></dt>
<dd><p>Move this object and all members to a new operational taxonomic unit
concept namespace scope.</p>
<p>Current <code class="xref py py-attr docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> value will be replaced with value
given in <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> if this is not <code class="docutils literal notranslate"><span class="pre">None</span></code>, or a new
<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> object. Following this,
<code class="docutils literal notranslate"><span class="pre">reconstruct_taxon_namespace()</span></code> will be called: each distinct
<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> object associated with <code class="docutils literal notranslate"><span class="pre">self</span></code> or members of <code class="docutils literal notranslate"><span class="pre">self</span></code> that
is not alread in <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> will be replaced with a new
<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> object that will be created with the same label and
added to <code class="xref py py-attr docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code>.  Calling this method results in
the object (and all its member objects) being associated with a new,
independent taxon namespace.</p>
<p>Label mapping case sensitivity follows the
<code class="docutils literal notranslate"><span class="pre">self.taxon_namespace.is_case_sensitive</span></code> setting. If
<code class="docutils literal notranslate"><span class="pre">False</span></code> and <code class="docutils literal notranslate"><span class="pre">unify_taxa_by_label</span></code> is also <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the
establishment of correspondence between <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects in the
old and new namespaces with be based on case-insensitive matching of
labels. E.g., if there are four <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects with labels
‘Foo’, ‘Foo’, ‘FOO’, and ‘FoO’ in the old namespace, then all objects
that reference these will reference a single new <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> object
in the new namespace (with a label some existing casing variant of
‘foo’). If <code class="docutils literal notranslate"><span class="pre">True</span></code>: if <code class="docutils literal notranslate"><span class="pre">unify_taxa_by_label</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>,
<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects with labels identical except in case will be
considered distinct.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>taxon_namespace</strong> (<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>) – The <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> into the scope of which this object
will be moved.</p></li>
<li><p><strong>unify_taxa_by_label</strong> (<em>boolean</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then references to distinct <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects with
identical labels in the current namespace will be replaced with a
reference to a single <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> object in the new namespace.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>: references to distinct <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects will
remain distinct, even if the labels are the same.</p></li>
<li><p><strong>taxon_mapping_memo</strong> (<em>dictionary</em>) – Similar to <code class="docutils literal notranslate"><span class="pre">memo</span></code> of deepcopy, this is a dictionary that maps
<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects in the old namespace to corresponding
<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects in the new namespace. Mostly for interal
use when migrating complex data to a new namespace. Note that
any mappings here take precedence over all other options: if a
<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> object in the old namespace is found in this
dictionary, the counterpart in the new namespace will be whatever
value is mapped, regardless of, e.g. label values.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Use this method to move an object from one taxon namespace to
another.</p>
<p>For example, to get a copy of an object associated with another taxon
namespace and associate it with a different namespace:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get handle to the new TaxonNamespace</span>
<span class="n">other_taxon_namespace</span> <span class="o">=</span> <span class="n">some_other_data</span><span class="o">.</span><span class="n">taxon_namespace</span>

<span class="c1"># Get a taxon-namespace scoped copy of a tree</span>
<span class="c1"># in another namespace</span>
<span class="n">t2</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">(</span><span class="n">t1</span><span class="p">)</span>

<span class="c1"># Replace taxon namespace of copy</span>
<span class="n">t2</span><span class="o">.</span><span class="n">migrate_taxon_namespace</span><span class="p">(</span><span class="n">other_taxon_namespace</span><span class="p">)</span>
</pre></div>
</div>
<p>You can also use this method to get a copy of a structure and then
move it to a new namespace:</p>
<blockquote>
<div><p>t2 = Tree(t1)
t2.migrate_taxon_namespace(TaxonNamespace())</p>
<p># Note: the same effect can be achived by:
t3 = copy.deepcopy(t1)</p>
</div></blockquote>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.reconstruct_taxon_namespace" title="dendropy.datamodel.treecollectionmodel.TreeList.reconstruct_taxon_namespace"><code class="xref py py-meth docutils literal notranslate"><span class="pre">reconstruct_taxon_namespace</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.poll_taxa">
<span class="sig-name descname"><span class="pre">poll_taxa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.poll_taxa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.poll_taxa" title="Link to this definition">¶</a></dt>
<dd><p>Returns a set populated with all of <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> instances associated
with <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>taxa</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#set" title="(in Python v3.13)"><em>set</em></a><em>(</em><em>)</em>) – Set to populate. If not specified, a new one will be created.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>taxa</strong> (<em>set[|Taxon|]</em>) – Set of taxa associated with <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.purge_taxon_namespace">
<span class="sig-name descname"><span class="pre">purge_taxon_namespace</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.purge_taxon_namespace" title="Link to this definition">¶</a></dt>
<dd><p>Remove all <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> instances in <code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> that are
not associated with <code class="docutils literal notranslate"><span class="pre">self</span></code> or any item in <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.read">
<span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.read"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.read" title="Link to this definition">¶</a></dt>
<dd><p>Add <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects to existing <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> from data source providing
one or more collections of trees.</p>
<p><strong>Mandatory Source-Specification Keyword Argument (Exactly One Required):</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>file</strong> (<em>file</em>) – File or file-like object of data opened for reading.</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path to file of data.</p></li>
<li><p><strong>url</strong> (<em>str</em>) – URL of data.</p></li>
<li><p><strong>data</strong> (<em>str</em>) – Data given directly.</p></li>
</ul>
</div></blockquote>
<p><strong>Mandatory Schema-Specification Keyword Argument:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>schema</strong> (<em>str</em>) – Identifier of format of data given by the
“<code class="docutils literal notranslate"><span class="pre">file</span></code>”, “<code class="docutils literal notranslate"><span class="pre">path</span></code>”, “<code class="docutils literal notranslate"><span class="pre">data</span></code>”, or “<code class="docutils literal notranslate"><span class="pre">url</span></code>” argument
specified above: “<a class="reference internal" href="../schemas/newick.html"><span class="doc">newick</span></a>”, “<a class="reference internal" href="../schemas/nexus.html"><span class="doc">nexus</span></a>”, or “<a class="reference internal" href="../schemas/nexml.html"><span class="doc">nexml</span></a>”. See
“<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional General Keyword Arguments:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>collection_offset</strong> (<em>int</em>) – 0-based index of tree block or
collection in source to be parsed. If not specified then the
first collection (offset = 0) is assumed.</p></li>
<li><p><strong>tree_offset</strong> (<em>int</em>) – 0-based index of first tree within the
collection specified by <code class="docutils literal notranslate"><span class="pre">collection_offset</span></code> to be parsed (i.e.,
skipping the first <code class="docutils literal notranslate"><span class="pre">tree_offset</span></code> trees). If not
specified, then the first tree (offset = 0) is assumed (i.e., no
trees within the specified collection will be skipped). Use this
to specify, e.g. a burn-in.</p></li>
<li><p><strong>ignore_unrecognized_keyword_arguments</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>,
then unsupported or unrecognized keyword arguments will not
result in an error. Default is <code class="docutils literal notranslate"><span class="pre">False</span></code>: unsupported keyword
arguments will result in an error.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional Schema-Specific Keyword Arguments:</strong></p>
<blockquote>
<div><p>These provide control over how the data is interpreted and
processed, and supported argument names and values depend on
the schema as specified by the value passed as the “<code class="docutils literal notranslate"><span class="pre">schema</span></code>”
argument. See “<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p>
</div></blockquote>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tlist</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TreeList</span><span class="p">()</span>
<span class="n">tlist</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">file</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;treefile.tre&#39;</span><span class="p">,</span> <span class="s1">&#39;rU&#39;</span><span class="p">),</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">,</span>
        <span class="n">tree_offset</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">tlist</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">path</span><span class="o">=</span><span class="s1">&#39;sometrees.nexus&#39;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">,</span>
        <span class="n">collection_offset</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">tree_offset</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">tlist</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="s2">&quot;((A,B),(C,D));((A,C),(B,D));&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>
<span class="n">tlist</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">url</span><span class="o">=</span><span class="s2">&quot;http://api.opentreeoflife.org/v2/study/pg_1144/tree/tree2324.nex&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.read_from_path">
<span class="sig-name descname"><span class="pre">read_from_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.read_from_path" title="Link to this definition">¶</a></dt>
<dd><p>Reads data from file specified by <code class="docutils literal notranslate"><span class="pre">filepath</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filepath</strong> (<em>file</em><em> or </em><em>file-like</em>) – Full file path to source of data.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>n</strong> (<code class="docutils literal notranslate"><span class="pre">tuple</span></code> [integer]) – A value indicating size of data read, where “size” depends on
the object:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>: <strong>undefined</strong></p></li>
<li><p><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>: number of trees</p></li>
<li><p><a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a>: number of sequences</p></li>
<li><p><a class="reference internal" href="datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>: <code class="docutils literal notranslate"><span class="pre">tuple</span></code> (number of taxon namespaces, number of tree lists, number of matrices)</p></li>
</ul>
</div></blockquote>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.read_from_stream">
<span class="sig-name descname"><span class="pre">read_from_stream</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.read_from_stream" title="Link to this definition">¶</a></dt>
<dd><p>Reads from file (exactly equivalent to just <code class="docutils literal notranslate"><span class="pre">read()</span></code>, provided
here as a separate method for completeness.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fileobj</strong> (<em>file</em><em> or </em><em>file-like</em>) – Source of data.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>n</strong> (<code class="docutils literal notranslate"><span class="pre">tuple</span></code> [integer]) – A value indicating size of data read, where “size” depends on
the object:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>: <strong>undefined</strong></p></li>
<li><p><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>: number of trees</p></li>
<li><p><a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a>: number of sequences</p></li>
<li><p><a class="reference internal" href="datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>: <code class="docutils literal notranslate"><span class="pre">tuple</span></code> (number of taxon namespaces, number of tree lists, number of matrices)</p></li>
</ul>
</div></blockquote>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.read_from_string">
<span class="sig-name descname"><span class="pre">read_from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.read_from_string" title="Link to this definition">¶</a></dt>
<dd><p>Reads a string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>src_str</strong> (<em>string</em>) – Data as a string.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>n</strong> (<code class="docutils literal notranslate"><span class="pre">tuple</span></code> [integer]) – A value indicating size of data read, where “size” depends on
the object:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>: <strong>undefined</strong></p></li>
<li><p><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>: number of trees</p></li>
<li><p><a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a>: number of sequences</p></li>
<li><p><a class="reference internal" href="datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>: <code class="docutils literal notranslate"><span class="pre">tuple</span></code> (number of taxon namespaces, number of tree lists, number of matrices)</p></li>
</ul>
</div></blockquote>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.read_from_url">
<span class="sig-name descname"><span class="pre">read_from_url</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">src</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.read_from_url" title="Link to this definition">¶</a></dt>
<dd><p>Reads a URL source.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>src</strong> (<em>string</em>) – URL of location providing source of data.</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – Specification of data format (e.g., “nexus”).</p></li>
<li><p><strong>kwargs</strong> (<em>keyword arguments</em><em>, </em><em>optional</em>) – Arguments to customize parsing, instantiation, processing, and
accession of objects read from the data source, including schema-
or format-specific handling. These will be passed to the underlying
schema-specific reader for handling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>n</strong> (<code class="docutils literal notranslate"><span class="pre">tuple</span></code> [integer]) – A value indicating size of data read, where “size” depends on
the object:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>: <strong>undefined</strong></p></li>
<li><p><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>: number of trees</p></li>
<li><p><a class="reference internal" href="charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a>: number of sequences</p></li>
<li><p><a class="reference internal" href="datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>: <code class="docutils literal notranslate"><span class="pre">tuple</span></code> (number of taxon namespaces, number of tree lists, number of matrices)</p></li>
</ul>
</div></blockquote>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.reconstruct_taxon_namespace">
<span class="sig-name descname"><span class="pre">reconstruct_taxon_namespace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">unify_taxa_by_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_mapping_memo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.reconstruct_taxon_namespace"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.reconstruct_taxon_namespace" title="Link to this definition">¶</a></dt>
<dd><p>Repopulates the current taxon namespace with new taxon objects,
preserving labels. Each distinct <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> object associated with
<code class="docutils literal notranslate"><span class="pre">self</span></code> or members of <code class="docutils literal notranslate"><span class="pre">self</span></code> that is not already in
<code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> will be replaced with a new <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a>
object that will be created with the same label and added to
<code class="xref py py-attr docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code>.</p>
<p>Label mapping case sensitivity follows the
<code class="docutils literal notranslate"><span class="pre">self.taxon_namespace.is_case_sensitive</span></code> setting. If
<code class="docutils literal notranslate"><span class="pre">False</span></code> and <code class="docutils literal notranslate"><span class="pre">unify_taxa_by_label</span></code> is also <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the
establishment of correspondence between <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects in the
old and new namespaces with be based on case-insensitive matching of
labels. E.g., if there are four <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects with labels
‘Foo’, ‘Foo’, ‘FOO’, and ‘FoO’ in the old namespace, then all objects
that reference these will reference a single new <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> object
in the new namespace (with a label some existing casing variant of
‘foo’). If <code class="docutils literal notranslate"><span class="pre">True</span></code>: if <code class="docutils literal notranslate"><span class="pre">unify_taxa_by_label</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>,
<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects with labels identical except in case will be
considered distinct.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Existing <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects in <code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> are <em>not</em>
removed. This method should thus only be called <em>only</em> when
<code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> has been changed. In fact, typical usage would
not involve calling this method directly, but rather through</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>unify_taxa_by_label</strong> (<em>boolean</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then references to distinct <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects with
identical labels in the current namespace will be replaced with a
reference to a single <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> object in the new namespace.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>: references to distinct <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects will
remain distinct, even if the labels are the same.</p></li>
<li><p><strong>taxon_mapping_memo</strong> (<em>dictionary</em>) – Similar to <code class="docutils literal notranslate"><span class="pre">memo</span></code> of deepcopy, this is a dictionary that maps
<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects in the old namespace to corresponding
<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects in the new namespace. Mostly for interal
use when migrating complex data to a new namespace.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.reindex_subcomponent_taxa">
<span class="sig-name descname"><span class="pre">reindex_subcomponent_taxa</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.reindex_subcomponent_taxa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.reindex_subcomponent_taxa" title="Link to this definition">¶</a></dt>
<dd><p>DEPRECATED: Use <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.reconstruct_taxon_namespace" title="dendropy.datamodel.treecollectionmodel.TreeList.reconstruct_taxon_namespace"><code class="xref py py-meth docutils literal notranslate"><span class="pre">reconstruct_taxon_namespace</span></code></a> instead.
Derived classes should override this to ensure that their various
components, attributes and members all refer to the same <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>
object as <code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code>, and that <code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> has all
the <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects in the various members.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.reindex_taxa">
<span class="sig-name descname"><span class="pre">reindex_taxa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon_namespace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.reindex_taxa" title="Link to this definition">¶</a></dt>
<dd><p>DEPRECATED: Use <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.migrate_taxon_namespace" title="dendropy.datamodel.treecollectionmodel.TreeList.migrate_taxon_namespace"><code class="xref any py py-meth docutils literal notranslate"><span class="pre">migrate_taxon_namespace()</span></code></a> instead.
Rebuilds <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> from scratch, or assigns <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects from
given <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> object <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> based on label values.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.split_distribution">
<span class="sig-name descname"><span class="pre">split_distribution</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">is_bipartitions_updated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_edge_length_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.split_distribution"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.split_distribution" title="Link to this definition">¶</a></dt>
<dd><p>Return <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution" title="dendropy.datamodel.treecollectionmodel.SplitDistribution"><code class="xref any py py-class docutils literal notranslate"><span class="pre">SplitDistribution</span></code></a> collecting information on splits in
contained trees. Keyword arguments get passed directly to
<a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution" title="dendropy.datamodel.treecollectionmodel.SplitDistribution"><code class="xref any py py-class docutils literal notranslate"><span class="pre">SplitDistribution</span></code></a> constructor.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.taxon_namespace_scoped_copy">
<span class="sig-name descname"><span class="pre">taxon_namespace_scoped_copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">memo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.taxon_namespace_scoped_copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.taxon_namespace_scoped_copy" title="Link to this definition">¶</a></dt>
<dd><p>Cloning level: 1.
Taxon-namespace-scoped copy: All member objects are full independent
instances, <em>except</em> for <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> and <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a>
objects: these are preserved as references.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.tree_factory">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">tree_factory</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.tree_factory"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.tree_factory" title="Link to this definition">¶</a></dt>
<dd><p>Creates and returns a <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> of a type that this list understands how to
manage.</p>
<p>Deriving classes can override this to provide for custom Tree-type
object lists. You can simple override the class-level variable
<a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.DEFAULT_TREE_TYPE" title="dendropy.datamodel.treecollectionmodel.TreeList.DEFAULT_TREE_TYPE"><code class="xref any py py-attr docutils literal notranslate"><span class="pre">DEFAULT_TREE_TYPE</span></code></a> in your derived class if the constructor signature
of the alternate tree type is the same as <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>.
If you want to have a TreeList <em>instance</em> that generates
custom trees (i.e., as opposed to a TreeList-ish <em>class</em> of instances),
set the <code class="docutils literal notranslate"><span class="pre">tree_type</span></code> attribute of the TreeList instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>*args</strong> (<em>positional arguments</em>) – Passed directly to constructor of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>.</p></li>
<li><p><strong>**kwargs</strong> (<em>keyword arguments</em>) – Passed directly to constructor of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><em>A |Tree| object.</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.update_taxon_namespace">
<span class="sig-name descname"><span class="pre">update_taxon_namespace</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.update_taxon_namespace"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.update_taxon_namespace" title="Link to this definition">¶</a></dt>
<dd><p>All <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects associated with <code class="docutils literal notranslate"><span class="pre">self</span></code> or members of <code class="docutils literal notranslate"><span class="pre">self</span></code>
that are not in <code class="docutils literal notranslate"><span class="pre">self.taxon_namespace</span></code> will be added. Note that, unlike
<code class="docutils literal notranslate"><span class="pre">reconstruct_taxon_namespace</span></code>, no new <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects
will be created.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.write">
<span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.write" title="Link to this definition">¶</a></dt>
<dd><p>Writes out <code class="docutils literal notranslate"><span class="pre">self</span></code> in <code class="docutils literal notranslate"><span class="pre">schema</span></code> format.</p>
<p><strong>Mandatory Destination-Specification Keyword Argument (Exactly One of the Following Required):</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>file</strong> (<em>file</em>) – File or file-like object opened for writing.</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path to file to which to write.</p></li>
</ul>
</div></blockquote>
<p><strong>Mandatory Schema-Specification Keyword Argument:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>schema</strong> (<em>str</em>) – Identifier of format of data. See
“<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional Schema-Specific Keyword Arguments:</strong></p>
<blockquote>
<div><p>These provide control over how the data is formatted, and supported
argument names and values depend on the schema as specified by the
value passed as the “<code class="docutils literal notranslate"><span class="pre">schema</span></code>” argument. See “<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more
details.</p>
</div></blockquote>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Using a file path:</span>
<span class="n">d</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;path/to/file.dat&quot;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>

<span class="c1"># Using an open file:</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;path/to/file.dat&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">d</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="n">f</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.write_to_path">
<span class="sig-name descname"><span class="pre">write_to_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.write_to_path" title="Link to this definition">¶</a></dt>
<dd><p>Writes to file specified by <code class="docutils literal notranslate"><span class="pre">dest</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.write_to_stream">
<span class="sig-name descname"><span class="pre">write_to_stream</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.write_to_stream" title="Link to this definition">¶</a></dt>
<dd><p>Writes to file-like object <code class="docutils literal notranslate"><span class="pre">dest</span></code>.</p>
</dd></dl>

</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id0">
<span class="sig-prename descclassname"><span class="pre">TreeList.</span></span><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">\*\*kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeList.read"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id0" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeList.put">
<span class="sig-prename descclassname"><span class="pre">TreeList.</span></span><span class="sig-name descname"><span class="pre">put</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">\*\*kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeList.put" title="Link to this definition">¶</a></dt>
<dd><p>Write out collection of trees to file.</p>
<dl class="field-list simple">
<dt class="field-odd">Mandatory Destimation-Specification Keyword Arguments (one and exactly one of the following required)<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>file</em>) – File or file-like object opened for writing.</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path to file to which to write.</p></li>
</ul>
</dd>
<dt class="field-even">Mandatory Schema-Specification Keyword Argument<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>schema</strong> (<em>str</em>) – Identifier of format of data given by the “<code class="docutils literal notranslate"><span class="pre">file</span></code>”, “<code class="docutils literal notranslate"><span class="pre">path</span></code>”, “<code class="docutils literal notranslate"><span class="pre">data</span></code>”, or “<code class="docutils literal notranslate"><span class="pre">url</span></code>” argument specified above: “<a class="reference internal" href="../schemas/newick.html"><span class="doc">newick</span></a>”, “<a class="reference internal" href="../schemas/nexus.html"><span class="doc">nexus</span></a>”, or “<a class="reference internal" href="../schemas/nexml.html"><span class="doc">nexml</span></a>”.</p></li>
</ul>
</dd>
<dt class="field-odd">Optional Schema-Specific Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><dl class="simple">
<dt>These provide control over how the data is formatted, and supported argument names and values depend on the schema as specified by the value passed as the “<code class="docutils literal notranslate"><span class="pre">schema</span></code>” argument:</dt><dd><ul>
<li><p><a class="reference internal" href="../schemas/newick.html#schema-specific-keyword-arguments-writing-newick"><span class="std std-ref">“newick” (Newick)</span></a></p></li>
<li><p><a class="reference internal" href="../schemas/nexus.html#schema-specific-keyword-arguments-writing-nexus"><span class="std std-ref">“nexus” (Nexus)</span></a></p></li>
<li><p><a class="reference internal" href="../schemas/nexml.html#schema-specific-keyword-arguments-writing-nexml"><span class="std std-ref">“nexml” (NeXML)</span></a></p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="the-treearray-class">
<h2>The <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray" title="dendropy.datamodel.treecollectionmodel.TreeArray"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeArray</span></code></a> Class<a class="headerlink" href="#the-treearray-class" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dendropy.datamodel.treecollectionmodel.</span></span><span class="sig-name descname"><span class="pre">TreeArray</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon_namespace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_rooted_trees</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_edge_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_node_ages</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_tree_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ultrametricity_precision</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_force_max_age</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_label_age_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray" title="Link to this definition">¶</a></dt>
<dd><p>High-performance collection of tree structures.</p>
<p>Storage of minimal tree structural information as represented by toplogy
and edge lengths, minimizing memory and processing time.
This class stores trees as collections of splits and edge lengths. All
other information, such as labels, metadata annotations, etc. will be
discarded. A full <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance can be reconstructed as needed
from the structural information stored by this class, at the cost of
computation time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>taxon_namespace</strong> (<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>) – The operational taxonomic unit concept namespace to manage taxon
references.</p></li>
<li><p><strong>is_rooted_trees</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If not set, then it will be set based on the rooting state of the
first tree added. If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then trying to add an unrooted tree
will result in an error. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, then trying to add a rooted
tree will result in an error.</p></li>
<li><p><strong>ignore_edge_lengths</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then edge lengths of splits will not be stored. If
<code class="docutils literal notranslate"><span class="pre">False</span></code>, then edge lengths will be stored.</p></li>
<li><p><strong>ignore_node_ages</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then node ages of splits will not be stored. If
<code class="docutils literal notranslate"><span class="pre">False</span></code>, then node ages will be stored.</p></li>
<li><p><strong>use_tree_weights</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code>, then tree weights will not be used to weight splits.</p></li>
</ul>
</dd>
</dl>
<dl class="py exception">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleEdgeLengthsTreeArrayUpdate">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">IncompatibleEdgeLengthsTreeArrayUpdate</span></span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.IncompatibleEdgeLengthsTreeArrayUpdate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleEdgeLengthsTreeArrayUpdate" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleNodeAgesTreeArrayUpdate">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">IncompatibleNodeAgesTreeArrayUpdate</span></span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.IncompatibleNodeAgesTreeArrayUpdate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleNodeAgesTreeArrayUpdate" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleRootingTreeArrayUpdate">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">IncompatibleRootingTreeArrayUpdate</span></span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.IncompatibleRootingTreeArrayUpdate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleRootingTreeArrayUpdate" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleTreeArrayUpdate">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">IncompatibleTreeArrayUpdate</span></span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.IncompatibleTreeArrayUpdate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleTreeArrayUpdate" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleTreeWeightsTreeArrayUpdate">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">IncompatibleTreeWeightsTreeArrayUpdate</span></span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.IncompatibleTreeWeightsTreeArrayUpdate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleTreeWeightsTreeArrayUpdate" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.__add__">
<span class="sig-name descname"><span class="pre">__add__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.__add__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.__add__" title="Link to this definition">¶</a></dt>
<dd><p>Creates and returns new <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray" title="dendropy.datamodel.treecollectionmodel.TreeArray"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeArray</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> (iterable of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects) – </p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>tlist</strong> (<em>|TreeArray| object</em>) – <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray" title="dendropy.datamodel.treecollectionmodel.TreeArray"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeArray</span></code></a> object containing clones of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects
in <code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.__iadd__">
<span class="sig-name descname"><span class="pre">__iadd__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree_array</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.__iadd__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.__iadd__" title="Link to this definition">¶</a></dt>
<dd><p>Accession of data from <code class="docutils literal notranslate"><span class="pre">tree_array</span></code> to self.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>tree_array</strong> (<a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray" title="dendropy.datamodel.treecollectionmodel.TreeArray"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeArray</span></code></a>) – A <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray" title="dendropy.datamodel.treecollectionmodel.TreeArray"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeArray</span></code></a> instance from which to add data.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.__iter__">
<span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.__iter__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.__iter__" title="Link to this definition">¶</a></dt>
<dd><p>Yields pairs of (split, edge_length) from the store.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.add_tree">
<span class="sig-name descname"><span class="pre">add_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_bipartitions_updated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.add_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.add_tree" title="Link to this definition">¶</a></dt>
<dd><p>Adds the structure represented by a <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance to the
collection.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (<a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>) – A <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance. This must have the same rooting state as
all the other trees accessioned into this collection as well as
that of <code class="docutils literal notranslate"><span class="pre">self.is_rooted_trees</span></code>.</p></li>
<li><p><strong>is_bipartitions_updated</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> [default], then the tree will have its splits encoded or
updated. Otherwise, if <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the tree is assumed to have its
splits already encoded and updated.</p></li>
<li><p><strong>index</strong> (<em>integer</em>) – Insert before index.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>index</strong> (<em>int</em>) – The index of the accession.</p></li>
<li><p><strong>s</strong> (<em>iterable of splits</em>) – A list of split bitmasks from <code class="docutils literal notranslate"><span class="pre">tree</span></code>.</p></li>
<li><p><em>e</em> – A list of edge length values from <code class="docutils literal notranslate"><span class="pre">tree</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.add_trees">
<span class="sig-name descname"><span class="pre">add_trees</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">trees</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_bipartitions_updated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.add_trees"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.add_trees" title="Link to this definition">¶</a></dt>
<dd><p>Adds multiple structures represneted by an iterator over or iterable of
<a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instances to the collection.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trees</strong> (iterator over or iterable of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instances) – An iterator over or iterable of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instances. Thess must
have the same rooting state as all the other trees accessioned into
this collection as well as that of <code class="docutils literal notranslate"><span class="pre">self.is_rooted_trees</span></code>.</p></li>
<li><p><strong>is_bipartitions_updated</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> [default], then the tree will have its splits encoded or
updated. Otherwise, if <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the tree is assumed to have its
splits already encoded and updated.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_bipartitions_updated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.append"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.append" title="Link to this definition">¶</a></dt>
<dd><p>Adds a <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance to the collection before position given
by <code class="docutils literal notranslate"><span class="pre">index</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (<a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>) – A <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance. This must have the same rooting state as
all the other trees accessioned into this collection as well as
that of <code class="docutils literal notranslate"><span class="pre">self.is_rooted_trees</span></code>.</p></li>
<li><p><strong>is_bipartitions_updated</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> [default], then the tree will have its splits encoded or
updated. Otherwise, if <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the tree is assumed to have its
splits already encoded and updated.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.bipartition_encoding_frequencies">
<span class="sig-name descname"><span class="pre">bipartition_encoding_frequencies</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.bipartition_encoding_frequencies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.bipartition_encoding_frequencies" title="Link to this definition">¶</a></dt>
<dd><p>Returns a dictionary with keys being bipartition encodings of trees
(as <code class="docutils literal notranslate"><span class="pre">frozenset</span></code> collections of <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Bipartition" title="dendropy.datamodel.treemodel.Bipartition"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bipartition</span></code></a> objects) and
values the frequency of occurrence of trees represented by that
encoding in the collection.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.calculate_log_product_of_split_supports">
<span class="sig-name descname"><span class="pre">calculate_log_product_of_split_supports</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_external_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.calculate_log_product_of_split_supports"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.calculate_log_product_of_split_supports" title="Link to this definition">¶</a></dt>
<dd><p>Calculates the log product of split support for each of the trees in
the collection.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>include_external_splits</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then non-internal split posteriors will be included in
the score. Defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>: these are skipped. This should only
make a difference when dealing with splits collected from trees of
different leaf sets.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>s</strong> (<em>tuple(list[numeric], integer)</em>) – Returns a tuple, with the first element being the list of scores
and the second being the index of the highest score. The element order
corresponds to the trees accessioned in the collection.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.calculate_sum_of_split_supports">
<span class="sig-name descname"><span class="pre">calculate_sum_of_split_supports</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_external_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.calculate_sum_of_split_supports"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.calculate_sum_of_split_supports" title="Link to this definition">¶</a></dt>
<dd><p>Calculates the <em>sum</em> of split support for all trees in the
collection.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>include_external_splits</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then non-internal split posteriors will be included in
the score. Defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>: these are skipped. This should only
make a difference when dealing with splits collected from trees of
different leaf sets.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>s</strong> (<em>tuple(list[numeric], integer)</em>) – Returns a tuple, with the first element being the list of scores
and the second being the index of the highest score. The element order
corresponds to the trees accessioned in the collection.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.consensus_tree">
<span class="sig-name descname"><span class="pre">consensus_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">min_freq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summarize_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">split_summarization_kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.consensus_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.consensus_tree" title="Link to this definition">¶</a></dt>
<dd><p>Returns a consensus tree from splits in <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>min_freq</strong> (<em>real</em>) – The minimum frequency of a split in this distribution for it to be
added to the tree.</p></li>
<li><p><strong>is_rooted</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Should tree be rooted or not? If <em>all</em> trees counted for splits are
explicitly rooted or unrooted, then this will default to <code class="docutils literal notranslate"><span class="pre">True</span></code> or
<code class="docutils literal notranslate"><span class="pre">False</span></code>, respectively. Otherwise it defaults to <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>**split_summarization_kwargs</strong> (<em>keyword arguments</em>) – These will be passed directly to the underlying
<a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer" title="dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer"><code class="xref any py py-class docutils literal notranslate"><span class="pre">SplitDistributionSummarizer</span></code></a> object. See
<a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer.configure" title="dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer.configure"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SplitDistributionSummarizer.configure</span></code></a> for options.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>t</strong> (<em>consensus tree</em>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.extend">
<span class="sig-name descname"><span class="pre">extend</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree_array</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.extend"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.extend" title="Link to this definition">¶</a></dt>
<dd><p>Accession of data from <code class="docutils literal notranslate"><span class="pre">tree_array</span></code> to self.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>tree_array</strong> (<a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray" title="dendropy.datamodel.treecollectionmodel.TreeArray"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeArray</span></code></a>) – A <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray" title="dendropy.datamodel.treecollectionmodel.TreeArray"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeArray</span></code></a> instance from which to add data.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.get_split_bitmask_and_edge_tuple">
<span class="sig-name descname"><span class="pre">get_split_bitmask_and_edge_tuple</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.get_split_bitmask_and_edge_tuple"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.get_split_bitmask_and_edge_tuple" title="Link to this definition">¶</a></dt>
<dd><p>Returns a pair of tuples, ( (splits…), (lengths…) ), corresponding
to the “tree” at <code class="docutils literal notranslate"><span class="pre">index</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_bipartitions_updated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.insert"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.insert" title="Link to this definition">¶</a></dt>
<dd><p>Adds a <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance to the collection before position given
by <code class="docutils literal notranslate"><span class="pre">index</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>index</strong> (<em>integer</em>) – Insert before index.</p></li>
<li><p><strong>tree</strong> (<a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>) – A <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance. This must have the same rooting state as
all the other trees accessioned into this collection as well as
that of <code class="docutils literal notranslate"><span class="pre">self.is_rooted_trees</span></code>.</p></li>
<li><p><strong>is_bipartitions_updated</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> [default], then the tree will have its splits encoded or
updated. Otherwise, if <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the tree is assumed to have its
splits already encoded and updated.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>index</strong> (<em>int</em>) – The index of the accession.</p></li>
<li><p><strong>s</strong> (<em>iterable of splits</em>) – A list of split bitmasks from <code class="docutils literal notranslate"><span class="pre">tree</span></code>.</p></li>
<li><p><em>e</em> – A list of edge length values <code class="docutils literal notranslate"><span class="pre">tree</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.maximum_product_of_split_support_tree">
<span class="sig-name descname"><span class="pre">maximum_product_of_split_support_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_external_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summarize_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">split_summarization_kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.maximum_product_of_split_support_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.maximum_product_of_split_support_tree" title="Link to this definition">¶</a></dt>
<dd><p>Return the tree with that maximizes the product of split supports, also
known as the “Maximum Clade Credibility Tree” or MCCT.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>include_external_splits</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then non-internal split posteriors will be included in
the score. Defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>: these are skipped. This should only
make a difference when dealing with splits collected from trees of
different leaf sets.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mcct_tree</strong> (<em>Tree</em>) – Tree that maximizes the product of split supports.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.maximum_sum_of_split_support_tree">
<span class="sig-name descname"><span class="pre">maximum_sum_of_split_support_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_external_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summarize_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">split_summarization_kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.maximum_sum_of_split_support_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.maximum_sum_of_split_support_tree" title="Link to this definition">¶</a></dt>
<dd><p>Return the tree with that maximizes the <em>sum</em> of split supports.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>include_external_splits</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then non-internal split posteriors will be included in
the score. Defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>: these are skipped. This should only
make a difference when dealing with splits collected from trees of
different leaf sets.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mst_tree</strong> (<em>Tree</em>) – Tree that maximizes the sum of split supports.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.read">
<span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.read"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.read" title="Link to this definition">¶</a></dt>
<dd><p>Add <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects to existing <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> from data source providing
one or more collections of trees.</p>
<p><strong>Mandatory Source-Specification Keyword Argument (Exactly One Required):</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>file</strong> (<em>file</em>) – File or file-like object of data opened for reading.</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path to file of data.</p></li>
<li><p><strong>url</strong> (<em>str</em>) – URL of data.</p></li>
<li><p><strong>data</strong> (<em>str</em>) – Data given directly.</p></li>
</ul>
</div></blockquote>
<p><strong>Mandatory Schema-Specification Keyword Argument:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>schema</strong> (<em>str</em>) – Identifier of format of data given by the
“<code class="docutils literal notranslate"><span class="pre">file</span></code>”, “<code class="docutils literal notranslate"><span class="pre">path</span></code>”, “<code class="docutils literal notranslate"><span class="pre">data</span></code>”, or “<code class="docutils literal notranslate"><span class="pre">url</span></code>” argument
specified above: “<a class="reference internal" href="../schemas/newick.html"><span class="doc">newick</span></a>”, “<a class="reference internal" href="../schemas/nexus.html"><span class="doc">nexus</span></a>”, or “<a class="reference internal" href="../schemas/nexml.html"><span class="doc">nexml</span></a>”. See
“<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional General Keyword Arguments:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><strong>collection_offset</strong> (<em>int</em>) – 0-based index of tree block or
collection in source to be parsed. If not specified then the
first collection (offset = 0) is assumed.</p></li>
<li><p><strong>tree_offset</strong> (<em>int</em>) – 0-based index of first tree within the
collection specified by <code class="docutils literal notranslate"><span class="pre">collection_offset</span></code> to be parsed (i.e.,
skipping the first <code class="docutils literal notranslate"><span class="pre">tree_offset</span></code> trees). If not
specified, then the first tree (offset = 0) is assumed (i.e., no
trees within the specified collection will be skipped). Use this
to specify, e.g. a burn-in.</p></li>
<li><p><strong>ignore_unrecognized_keyword_arguments</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>,
then unsupported or unrecognized keyword arguments will not
result in an error. Default is <code class="docutils literal notranslate"><span class="pre">False</span></code>: unsupported keyword
arguments will result in an error.</p></li>
</ul>
</div></blockquote>
<p><strong>Optional Schema-Specific Keyword Arguments:</strong></p>
<blockquote>
<div><p>These provide control over how the data is interpreted and
processed, and supported argument names and values depend on
the schema as specified by the value passed as the “<code class="docutils literal notranslate"><span class="pre">schema</span></code>”
argument. See “<a class="reference internal" href="../schemas/index.html"><span class="doc">DendroPy Schemas: Phylogenetic and Evolutionary Biology Data Formats</span></a>” for more details.</p>
</div></blockquote>
<p><strong>Examples:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tree_array</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TreeArray</span><span class="p">()</span>
<span class="n">tree_array</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">file</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;treefile.tre&#39;</span><span class="p">,</span> <span class="s1">&#39;rU&#39;</span><span class="p">),</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">,</span>
        <span class="n">tree_offset</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">tree_array</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">path</span><span class="o">=</span><span class="s1">&#39;sometrees.nexus&#39;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">,</span>
        <span class="n">collection_offset</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">tree_offset</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">tree_array</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="s2">&quot;((A,B),(C,D));((A,C),(B,D));&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>
<span class="n">tree_array</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">url</span><span class="o">=</span><span class="s2">&quot;http://api.opentreeoflife.org/v2/study/pg_1144/tree/tree2324.nex&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.read_from_files">
<span class="sig-name descname"><span class="pre">read_from_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">schema</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.read_from_files"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.read_from_files" title="Link to this definition">¶</a></dt>
<dd><p>Adds multiple structures from one or more external file sources to the
collection.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>files</strong> (<em>iterable</em><em> of </em><em>strings and/or file objects</em>) – A list or some other iterable of file paths or file-like objects
(string elements will be assumed to be paths to files, while all
other types of elements will be assumed to be file-like
objects opened for reading).</p></li>
<li><p><strong>schema</strong> (<em>string</em>) – The data format of the source. E.g., “nexus”, “newick”, “nexml”.</p></li>
<li><p><strong>**kwargs</strong> (<em>keyword arguments</em>) – These will be passed directly to the underlying schema-specific
reader implementation.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.split_bitmask_set_frequencies">
<span class="sig-name descname"><span class="pre">split_bitmask_set_frequencies</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.split_bitmask_set_frequencies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.split_bitmask_set_frequencies" title="Link to this definition">¶</a></dt>
<dd><p>Returns a dictionary with keys being sets of split bitmasks and values
being the frequency of occurrence of trees represented by those split
bitmask sets in the collection.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.TreeArray.topologies">
<span class="sig-name descname"><span class="pre">topologies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sort_descending</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frequency_attr_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'frequency'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frequency_annotation_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'frequency'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#TreeArray.topologies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.TreeArray.topologies" title="Link to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> instance containing the reconstructed tree
topologies (i.e. <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instances with no edge weights) in the
collection, with the frequency added as an attributed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sort_descending</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then topologies will be sorted in <em>descending</em> frequency
order (i.e., topologies with the highest frequencies will be listed
first). If <code class="docutils literal notranslate"><span class="pre">False</span></code>, then they will be sorted in <em>ascending</em>
frequency. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), then they will not be sorted.</p></li>
<li><p><strong>frequency_attr_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – Name of attribute to add to each <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> representing
the frequency of that topology in the collection. If <code class="docutils literal notranslate"><span class="pre">None</span></code>
then the attribute will not be added.</p></li>
<li><p><strong>frequency_annotation_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – Name of annotation to add to the annotations of each <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>,
representing the frequency of that topology in the collection. The
value of this annotation will be dynamically-bound to the attribute
specified by <code class="docutils literal notranslate"><span class="pre">frequency_attr_name</span></code> unless that is <code class="docutils literal notranslate"><span class="pre">None</span></code>. If
<code class="docutils literal notranslate"><span class="pre">frequency_annotation_name</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code> then the annotation will not
be added.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="the-splitdistribution-class">
<h2>The <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution" title="dendropy.datamodel.treecollectionmodel.SplitDistribution"><code class="xref py py-class docutils literal notranslate"><span class="pre">SplitDistribution</span></code></a> Class<a class="headerlink" href="#the-splitdistribution-class" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistribution">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dendropy.datamodel.treecollectionmodel.</span></span><span class="sig-name descname"><span class="pre">SplitDistribution</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon_namespace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_edge_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_node_ages</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_tree_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ultrametricity_precision</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_force_max_age</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_label_age_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistribution"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution" title="Link to this definition">¶</a></dt>
<dd><p>Collects information regarding splits over multiple trees.</p>
<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistribution.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">split_bitmask</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistribution.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.__getitem__" title="Link to this definition">¶</a></dt>
<dd><p>Returns freqency of split_bitmask.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistribution.calc_freqs">
<span class="sig-name descname"><span class="pre">calc_freqs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistribution.calc_freqs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.calc_freqs" title="Link to this definition">¶</a></dt>
<dd><p>Forces recalculation of frequencies.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistribution.collapse_edges_with_less_than_minimum_support">
<span class="sig-name descname"><span class="pre">collapse_edges_with_less_than_minimum_support</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_freq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistribution.collapse_edges_with_less_than_minimum_support"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.collapse_edges_with_less_than_minimum_support" title="Link to this definition">¶</a></dt>
<dd><p>Collapse edges on tree that have support less than indicated by
<code class="docutils literal notranslate"><span class="pre">min_freq</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistribution.consensus_tree">
<span class="sig-name descname"><span class="pre">consensus_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">min_freq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_rooted</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summarize_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">split_summarization_kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistribution.consensus_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.consensus_tree" title="Link to this definition">¶</a></dt>
<dd><p>Returns a consensus tree from splits in <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>min_freq</strong> (<em>real</em>) – The minimum frequency of a split in this distribution for it to be
added to the tree.</p></li>
<li><p><strong>is_rooted</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Should tree be rooted or not? If <em>all</em> trees counted for splits are
explicitly rooted or unrooted, then this will default to <code class="docutils literal notranslate"><span class="pre">True</span></code> or
<code class="docutils literal notranslate"><span class="pre">False</span></code>, respectively. Otherwise it defaults to <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>**split_summarization_kwargs</strong> (<em>keyword arguments</em>) – These will be passed directly to the underlying
<a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer" title="dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer"><code class="xref any py py-class docutils literal notranslate"><span class="pre">SplitDistributionSummarizer</span></code></a> object. See
<a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer.configure" title="dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer.configure"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SplitDistributionSummarizer.configure</span></code></a> for options.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>t</strong> (<em>consensus tree</em>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistribution.count_splits_on_tree">
<span class="sig-name descname"><span class="pre">count_splits_on_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_bipartitions_updated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_edge_length_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistribution.count_splits_on_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.count_splits_on_tree" title="Link to this definition">¶</a></dt>
<dd><p>Counts splits in this tree and add to totals. <code class="docutils literal notranslate"><span class="pre">tree</span></code> must be decorated
with splits, and no attempt is made to normalize taxa.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (a <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> object.) – The tree on which to count the splits.</p></li>
<li><p><strong>is_bipartitions_updated</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> [default], then the tree will have its splits encoded or
updated. Otherwise, if <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the tree is assumed to have its
splits already encoded and updated.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>s</strong> (<em>iterable of splits</em>) – A list of split bitmasks from <code class="docutils literal notranslate"><span class="pre">tree</span></code>.</p></li>
<li><p><em>e</em> – A list of edge length values from <code class="docutils literal notranslate"><span class="pre">tree</span></code>.</p></li>
<li><p><em>a</em> – A list of node age values from <code class="docutils literal notranslate"><span class="pre">tree</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistribution.log_product_of_split_support_on_tree">
<span class="sig-name descname"><span class="pre">log_product_of_split_support_on_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_bipartitions_updated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_external_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistribution.log_product_of_split_support_on_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.log_product_of_split_support_on_tree" title="Link to this definition">¶</a></dt>
<dd><p>Calculates the (log) product of the support of the splits of the
tree, where the support is given by the proportional frequency of the
split in the current split distribution.</p>
<p>The tree that has the highest product of split support out of a sample
of trees corresponds to the “maximum credibility tree” for that sample.
This can also be referred to as the “maximum clade credibility tree”,
though this latter term is sometimes use for the tree that has the
highest <em>sum</em> of split support (see
<a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.sum_of_split_support_on_tree" title="dendropy.datamodel.treecollectionmodel.SplitDistribution.sum_of_split_support_on_tree"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SplitDistribution.sum_of_split_support_on_tree</span></code></a>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (<a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>) – The tree for which the score should be calculated.</p></li>
<li><p><strong>is_bipartitions_updated</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the splits are assumed to have already been encoded
and will not be updated on the trees.</p></li>
<li><p><strong>include_external_splits</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then non-internal split posteriors will be included in
the score. Defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>: these are skipped. This should only
make a difference when dealing with splits collected from trees of
different leaf sets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>s</strong> (<em>numeric</em>) – The log product of the support of the splits of the tree.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistribution.normalize_bitmask">
<span class="sig-name descname"><span class="pre">normalize_bitmask</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bitmask</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistribution.normalize_bitmask"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.normalize_bitmask" title="Link to this definition">¶</a></dt>
<dd><p>“Normalizes” split, by ensuring that the least-significant bit is
always 1 (used on unrooted trees to establish split identity
independent of rotation).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>bitmask</strong> (<em>integer</em>) – Split bitmask hash to be normalized.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>h</strong> (<em>integer</em>) – Normalized split bitmask.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistribution.split_support_iter">
<span class="sig-name descname"><span class="pre">split_support_iter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_bipartitions_updated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_external_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">traversal_strategy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'preorder'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_support_attr_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_support_attr_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistribution.split_support_iter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.split_support_iter" title="Link to this definition">¶</a></dt>
<dd><p>Returns iterator over support values for the splits of a given tree,
where the support value is given by the proportional frequency of the
split in the current split distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (<a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>) – The <a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> which will be scored.</p></li>
<li><p><strong>is_bipartitions_updated</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> [default], then the tree will have its splits encoded or
updated. Otherwise, if <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the tree is assumed to have its
splits already encoded and updated.</p></li>
<li><p><strong>include_external_splits</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then non-internal split posteriors will be included.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, then these are skipped. This should only make a
difference when dealing with splits collected from trees of
different leaf sets.</p></li>
<li><p><strong>traversal_strategy</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – One of: “preorder” or “postorder”. Specfies order in which splits
are visited.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>s</strong> (<em>list of floats</em>) – List of values for splits in the tree corresponding to the
proportional frequency that the split is found in the current
distribution.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistribution.splits_considered">
<span class="sig-name descname"><span class="pre">splits_considered</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistribution.splits_considered"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.splits_considered" title="Link to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Returns 4 values:</dt><dd><p>total number of splits counted
total <em>weighted</em> number of unique splits counted
total number of non-trivial splits counted
total <em>weighted</em> number of unique non-trivial splits counted</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistribution.sum_of_split_support_on_tree">
<span class="sig-name descname"><span class="pre">sum_of_split_support_on_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_bipartitions_updated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_external_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistribution.sum_of_split_support_on_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.sum_of_split_support_on_tree" title="Link to this definition">¶</a></dt>
<dd><p>Calculates the sum of the support of the splits of the tree, where the
support is given by the proportional frequency of the split in the
current distribtion.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (<a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>) – The tree for which the score should be calculated.</p></li>
<li><p><strong>is_bipartitions_updated</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the splits are assumed to have already been encoded
and will not be updated on the trees.</p></li>
<li><p><strong>include_external_splits</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then non-internal split posteriors will be included in
the score. Defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>: these are skipped. This should only
make a difference when dealing with splits collected from trees of
different leaf sets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>s</strong> (<em>numeric</em>) – The sum of the support of the splits of the tree.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistribution.summarize_splits_on_tree">
<span class="sig-name descname"><span class="pre">summarize_splits_on_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_bipartitions_updated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">split_summarization_kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistribution.summarize_splits_on_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.summarize_splits_on_tree" title="Link to this definition">¶</a></dt>
<dd><p>Summarizes support of splits/edges/node on tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (<a class="reference internal" href="treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance) – Tree to be decorated with support values.</p></li>
<li><p><strong>is_bipartitions_updated</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then bipartitions will not be recalculated.</p></li>
<li><p><strong>**split_summarization_kwargs</strong> (<em>keyword arguments</em>) – These will be passed directly to the underlying
<a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer" title="dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer"><code class="xref any py py-class docutils literal notranslate"><span class="pre">SplitDistributionSummarizer</span></code></a> object. See
<a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer.configure" title="dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer.configure"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SplitDistributionSummarizer.configure</span></code></a> for options.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="the-splitdistributionsummarizer-class">
<h2>The <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer" title="dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">SplitDistributionSummarizer</span></code></a> Class<a class="headerlink" href="#the-splitdistributionsummarizer-class" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dendropy.datamodel.treecollectionmodel.</span></span><span class="sig-name descname"><span class="pre">SplitDistributionSummarizer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistributionSummarizer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer" title="Link to this definition">¶</a></dt>
<dd><p>See <a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer.configure" title="dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer.configure"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SplitDistributionSummarizer.configure</span></code></a> for configuration
options.</p>
<dl class="py method">
<dt class="sig sig-object py" id="dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer.configure">
<span class="sig-name descname"><span class="pre">configure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/datamodel/treecollectionmodel.html#SplitDistributionSummarizer.configure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer.configure" title="Link to this definition">¶</a></dt>
<dd><p>Configure rendition/mark-up.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>set_edge_lengths</strong> (<em>string</em>) – <p>For each edge, set the length based on:</p>
<blockquote>
<div><ul>
<li><p>”support”: use support values split corresponding to edge</p></li>
<li><p>”mean-length”: mean of edge lengths for split</p></li>
<li><p>”median-length”: median of edge lengths for split</p></li>
<li><p>”mean-age”: such that split age is equal to mean of ages</p></li>
<li><p>”median-age”: such that split age is equal to mean of ages</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">None</span></code>: do not set edge lengths</p></li>
</ul>
</div></blockquote>
</p></li>
<li><p><strong>add_support_as_node_attribute</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Adds each node’s support value as an attribute of the node,
“<code class="docutils literal notranslate"><span class="pre">support</span></code>”.</p></li>
<li><p><strong>add_support_as_node_annotation</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Adds support as a metadata annotation, “<code class="docutils literal notranslate"><span class="pre">support</span></code>”. If
<code class="docutils literal notranslate"><span class="pre">add_support_as_node_attribute</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the value will be
dynamically-bound to the value of the node’s “<code class="docutils literal notranslate"><span class="pre">support</span></code>” attribute.</p></li>
<li><p><strong>set_support_as_node_label</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Sets the <code class="docutils literal notranslate"><span class="pre">label</span></code> attribute of each node to the support value.</p></li>
<li><p><strong>add_node_age_summaries_as_node_attributes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – <p>Summarizes the distribution of the ages of each node in the
following attributes:</p>
<blockquote>
<div><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">age_mean</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">age_median</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">age_sd</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">age_hpd95</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">age_range</span></code></p></li>
</ul>
</div></blockquote>
</p></li>
<li><p><strong>add_node_age_summaries_as_node_annotations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – <p>Summarizes the distribution of the ages of each node in the
following metadata annotations:</p>
<blockquote>
<div><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">age_mean</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">age_median</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">age_sd</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">age_hpd95</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">age_range</span></code></p></li>
</ul>
</div></blockquote>
<p>If <code class="docutils literal notranslate"><span class="pre">add_node_age_summaries_as_node_attributes</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the
values will be dynamically-bound to the corresponding node
attributes.</p>
</p></li>
<li><p><strong>add_edge_length_summaries_as_edge_attributes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – <p>Summarizes the distribution of the lengths of each edge in the
following attribtutes:</p>
<blockquote>
<div><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">length_mean</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">length_median</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">length_sd</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">length_hpd95</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">length_range</span></code></p></li>
</ul>
</div></blockquote>
</p></li>
<li><p><strong>add_edge_length_summaries_as_edge_annotations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – <p>Summarizes the distribution of the lengths of each edge in the
following metadata annotations:</p>
<blockquote>
<div><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">length_mean</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">length_median</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">length_sd</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">length_hpd95</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">length_range</span></code></p></li>
</ul>
</div></blockquote>
<p>If <code class="docutils literal notranslate"><span class="pre">add_edge_length_summaries_as_edge_attributes</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the
values will be dynamically-bound to the corresponding edge
attributes.</p>
</p></li>
<li><p><strong>support_label_decimals</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – Number of decimal places to express when rendering the support
value as a string for the node label.</p></li>
<li><p><strong>support_as_percentages</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Whether or not to express the support value as percentages (default
is probability or proportion).</p></li>
<li><p><strong>minimum_edge_length</strong> (<em>numeric</em>) – All edge lengths calculated to have a value less than this will be
set to this.</p></li>
<li><p><strong>error_on_negative_edge_lengths</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, an inferred edge length that is less than 0 will result
in a ValueError.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div style="text-align: center; padding-top: 20px; padding-bottom: 5px; width: 100%;">
    <a href="../index.html"><img src="../_static/dendropy_logo.png" /></a>
</div><div style="clear:both; width: 100%; height:1px;"></div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h3><a href="../index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.datamodel.treecollectionmodel</span></code>: Collections of Trees</a><ul>
<li><a class="reference internal" href="#the-treelist-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code> Class</a><ul>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList"><code class="docutils literal notranslate"><span class="pre">TreeList</span></code></a><ul>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.DEFAULT_TREE_TYPE"><code class="docutils literal notranslate"><span class="pre">TreeList.DEFAULT_TREE_TYPE</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.__add__"><code class="docutils literal notranslate"><span class="pre">TreeList.__add__</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.__getitem__"><code class="docutils literal notranslate"><span class="pre">TreeList.__getitem__</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.__iadd__"><code class="docutils literal notranslate"><span class="pre">TreeList.__iadd__</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.append"><code class="docutils literal notranslate"><span class="pre">TreeList.append</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.as_string"><code class="docutils literal notranslate"><span class="pre">TreeList.as_string</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.as_tree_array"><code class="docutils literal notranslate"><span class="pre">TreeList.as_tree_array</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.clone"><code class="docutils literal notranslate"><span class="pre">TreeList.clone</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.consensus"><code class="docutils literal notranslate"><span class="pre">TreeList.consensus</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.copy_annotations_from"><code class="docutils literal notranslate"><span class="pre">TreeList.copy_annotations_from</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.deep_copy_annotations_from"><code class="docutils literal notranslate"><span class="pre">TreeList.deep_copy_annotations_from</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.extend"><code class="docutils literal notranslate"><span class="pre">TreeList.extend</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.frequency_of_bipartition"><code class="docutils literal notranslate"><span class="pre">TreeList.frequency_of_bipartition</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.frequency_of_split"><code class="docutils literal notranslate"><span class="pre">TreeList.frequency_of_split</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.get"><code class="docutils literal notranslate"><span class="pre">TreeList.get</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.get_from_path"><code class="docutils literal notranslate"><span class="pre">TreeList.get_from_path</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.get_from_stream"><code class="docutils literal notranslate"><span class="pre">TreeList.get_from_stream</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.get_from_string"><code class="docutils literal notranslate"><span class="pre">TreeList.get_from_string</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.get_from_url"><code class="docutils literal notranslate"><span class="pre">TreeList.get_from_url</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.insert"><code class="docutils literal notranslate"><span class="pre">TreeList.insert</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.maximum_product_of_split_support_tree"><code class="docutils literal notranslate"><span class="pre">TreeList.maximum_product_of_split_support_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.maximum_sum_of_split_support_tree"><code class="docutils literal notranslate"><span class="pre">TreeList.maximum_sum_of_split_support_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.migrate_taxon_namespace"><code class="docutils literal notranslate"><span class="pre">TreeList.migrate_taxon_namespace</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.poll_taxa"><code class="docutils literal notranslate"><span class="pre">TreeList.poll_taxa</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.purge_taxon_namespace"><code class="docutils literal notranslate"><span class="pre">TreeList.purge_taxon_namespace</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.read"><code class="docutils literal notranslate"><span class="pre">TreeList.read</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.read_from_path"><code class="docutils literal notranslate"><span class="pre">TreeList.read_from_path</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.read_from_stream"><code class="docutils literal notranslate"><span class="pre">TreeList.read_from_stream</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.read_from_string"><code class="docutils literal notranslate"><span class="pre">TreeList.read_from_string</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.read_from_url"><code class="docutils literal notranslate"><span class="pre">TreeList.read_from_url</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.reconstruct_taxon_namespace"><code class="docutils literal notranslate"><span class="pre">TreeList.reconstruct_taxon_namespace</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.reindex_subcomponent_taxa"><code class="docutils literal notranslate"><span class="pre">TreeList.reindex_subcomponent_taxa</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.reindex_taxa"><code class="docutils literal notranslate"><span class="pre">TreeList.reindex_taxa</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.split_distribution"><code class="docutils literal notranslate"><span class="pre">TreeList.split_distribution</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.taxon_namespace_scoped_copy"><code class="docutils literal notranslate"><span class="pre">TreeList.taxon_namespace_scoped_copy</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.tree_factory"><code class="docutils literal notranslate"><span class="pre">TreeList.tree_factory</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.update_taxon_namespace"><code class="docutils literal notranslate"><span class="pre">TreeList.update_taxon_namespace</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.write"><code class="docutils literal notranslate"><span class="pre">TreeList.write</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.write_to_path"><code class="docutils literal notranslate"><span class="pre">TreeList.write_to_path</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.write_to_stream"><code class="docutils literal notranslate"><span class="pre">TreeList.write_to_stream</span></code></a></li>
<li><a class="reference internal" href="#id0"><code class="docutils literal notranslate"><span class="pre">TreeList.read</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeList.put"><code class="docutils literal notranslate"><span class="pre">TreeList.put</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#the-treearray-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">TreeArray</span></code> Class</a><ul>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray"><code class="docutils literal notranslate"><span class="pre">TreeArray</span></code></a><ul>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleEdgeLengthsTreeArrayUpdate"><code class="docutils literal notranslate"><span class="pre">TreeArray.IncompatibleEdgeLengthsTreeArrayUpdate</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleNodeAgesTreeArrayUpdate"><code class="docutils literal notranslate"><span class="pre">TreeArray.IncompatibleNodeAgesTreeArrayUpdate</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleRootingTreeArrayUpdate"><code class="docutils literal notranslate"><span class="pre">TreeArray.IncompatibleRootingTreeArrayUpdate</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleTreeArrayUpdate"><code class="docutils literal notranslate"><span class="pre">TreeArray.IncompatibleTreeArrayUpdate</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.IncompatibleTreeWeightsTreeArrayUpdate"><code class="docutils literal notranslate"><span class="pre">TreeArray.IncompatibleTreeWeightsTreeArrayUpdate</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.__add__"><code class="docutils literal notranslate"><span class="pre">TreeArray.__add__</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.__iadd__"><code class="docutils literal notranslate"><span class="pre">TreeArray.__iadd__</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.__iter__"><code class="docutils literal notranslate"><span class="pre">TreeArray.__iter__</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.add_tree"><code class="docutils literal notranslate"><span class="pre">TreeArray.add_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.add_trees"><code class="docutils literal notranslate"><span class="pre">TreeArray.add_trees</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.append"><code class="docutils literal notranslate"><span class="pre">TreeArray.append</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.bipartition_encoding_frequencies"><code class="docutils literal notranslate"><span class="pre">TreeArray.bipartition_encoding_frequencies</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.calculate_log_product_of_split_supports"><code class="docutils literal notranslate"><span class="pre">TreeArray.calculate_log_product_of_split_supports</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.calculate_sum_of_split_supports"><code class="docutils literal notranslate"><span class="pre">TreeArray.calculate_sum_of_split_supports</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.consensus_tree"><code class="docutils literal notranslate"><span class="pre">TreeArray.consensus_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.extend"><code class="docutils literal notranslate"><span class="pre">TreeArray.extend</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.get_split_bitmask_and_edge_tuple"><code class="docutils literal notranslate"><span class="pre">TreeArray.get_split_bitmask_and_edge_tuple</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.insert"><code class="docutils literal notranslate"><span class="pre">TreeArray.insert</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.maximum_product_of_split_support_tree"><code class="docutils literal notranslate"><span class="pre">TreeArray.maximum_product_of_split_support_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.maximum_sum_of_split_support_tree"><code class="docutils literal notranslate"><span class="pre">TreeArray.maximum_sum_of_split_support_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.read"><code class="docutils literal notranslate"><span class="pre">TreeArray.read</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.read_from_files"><code class="docutils literal notranslate"><span class="pre">TreeArray.read_from_files</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.split_bitmask_set_frequencies"><code class="docutils literal notranslate"><span class="pre">TreeArray.split_bitmask_set_frequencies</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.TreeArray.topologies"><code class="docutils literal notranslate"><span class="pre">TreeArray.topologies</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#the-splitdistribution-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">SplitDistribution</span></code> Class</a><ul>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution"><code class="docutils literal notranslate"><span class="pre">SplitDistribution</span></code></a><ul>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.__getitem__"><code class="docutils literal notranslate"><span class="pre">SplitDistribution.__getitem__</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.calc_freqs"><code class="docutils literal notranslate"><span class="pre">SplitDistribution.calc_freqs</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.collapse_edges_with_less_than_minimum_support"><code class="docutils literal notranslate"><span class="pre">SplitDistribution.collapse_edges_with_less_than_minimum_support</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.consensus_tree"><code class="docutils literal notranslate"><span class="pre">SplitDistribution.consensus_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.count_splits_on_tree"><code class="docutils literal notranslate"><span class="pre">SplitDistribution.count_splits_on_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.log_product_of_split_support_on_tree"><code class="docutils literal notranslate"><span class="pre">SplitDistribution.log_product_of_split_support_on_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.normalize_bitmask"><code class="docutils literal notranslate"><span class="pre">SplitDistribution.normalize_bitmask</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.split_support_iter"><code class="docutils literal notranslate"><span class="pre">SplitDistribution.split_support_iter</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.splits_considered"><code class="docutils literal notranslate"><span class="pre">SplitDistribution.splits_considered</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.sum_of_split_support_on_tree"><code class="docutils literal notranslate"><span class="pre">SplitDistribution.sum_of_split_support_on_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistribution.summarize_splits_on_tree"><code class="docutils literal notranslate"><span class="pre">SplitDistribution.summarize_splits_on_tree</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#the-splitdistributionsummarizer-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">SplitDistributionSummarizer</span></code> Class</a><ul>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer"><code class="docutils literal notranslate"><span class="pre">SplitDistributionSummarizer</span></code></a><ul>
<li><a class="reference internal" href="#dendropy.datamodel.treecollectionmodel.SplitDistributionSummarizer.configure"><code class="docutils literal notranslate"><span class="pre">SplitDistributionSummarizer.configure</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

  </div><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h4>Previous topic</h4>
    <p class="topless"><a href="treemodel.html"
                          title="previous chapter"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.datamodel.treemodel</span></code>: Trees</a></p>
  </div>
  <div>
    <h4>Next topic</h4>
    <p class="topless"><a href="charstatemodel.html"
                          title="next chapter"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.datamodel.charstatemodel</span></code>: Character State Identities and Alphabets</a></p>
  </div><div style="clear:both; width: 100%; height:1px;"></div>
    <!-- Documentation -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Documentation</h3>
        <ul>
            <li><a href="../primer/downloading.html">Downloading and Installing DendroPy</a></li>
            <li><a href="../primer/index.html">The DendroPy Primer</a></li>
            <li><a href="index.html">Library API Reference</a></li>
            <li>
                <a href="../schemas/index.html">Schemas</a>
                <ul>
                    <li><a href="../schemas/fasta.html">FASTA</a></li>
                    <li><a href="../schemas/newick.html">Newick</a></li>
                    <li><a href="../schemas/nexml.html">NeXML</a></li>
                    <li><a href="../schemas/nexus.html">Nexus</a></li>
                    <li><a href="../schemas/phylip.html">PHYLIP</a></li>
                </ul>
            </li>
            <li>
                <a href="../programs/index.html">Programs</a>
                <ul>
                    <li><a href="../programs/sumtrees.html">SumTrees</a></li>
                </ul>
            </li>
            <li><a href="../glossary.html">Glossary and Terminological Reference</a></li>
            <li><a href="../developer.html">Developer Guide</a></li>
            <li><a href="../planning.html">Ongoing Development</a></li>
            <li><a href="../changes.html">Change History</a></li>
        </ul>
    </div>

    <!-- Downloads -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Obtaining</h3>
        <ul>
            <li><a target="_blank" href="http://pypi.python.org/pypi/DendroPy">Install from the Python Package Index</a></li>
            <li><a target="_blank" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.5.tar.gz">Download the Source Code Archive</a></li>
            <li><a target="_blank" href="http://github.com/jeetsukumaran/DendroPy">Clone the Source Code Repository</a></li>
        </ul>
    </div>

    <!-- Discussions -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Discussion</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot; DendroPy Users&quot; </a>group to follow and participate in discussion, troubleshooting, help, information, suggestions, etc. on the usage and development of the DendroPy phylogenetic computing library.</p>
            <form action="http://groups.google.com/group/dendropy-users/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot;dendropy-users&quot;</a> group, or click <a href="http://groups.google.com/group/dendropy-users?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Announcements -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Announcements</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; DendroPy Announcements&quot; </a>group to receive announcements of new releases, updates, changes and other news of interest to DendroPy users and developers.</p>
            <form action="http://groups.google.com/group/dendropy-announce/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; dendropy-announce&quot; </a>group, or click <a href="http://groups.google.com/group/dendropy-announce?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Development -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative; padding-bottom: 15px; margin-bottom:5px;">
        <h3><span style="text-align: left"><a href="https://github.com/jeetsukumaran/DendroPy/">Development</a></span><a href="https://github.com/jeetsukumaran/DendroPy/"><span style="position: absolute; right: 0; top: 10px "><img src="../_static/Octocat.png" height="30px" alt="GitHub" /></span></a></h3>
        <div style="margin-top: 15px;">
            <!-- <a href="https://github.com/jeetsukumaran/DendroPy/issues">Issues</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/subscription">Watch</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/fork">Fork</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/stargazers">Star</a> &bull; <a href="https://github.com/jeetsukumaran/">Follow</a> -->
            <ul>
                <li>                <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/issues"><code>Issues</code></a></span> <span style="font-style: italic; font-size:80%;"> - Report bugs or request features</span></li>
                <li>     <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/subscription"><code>&nbsp;Watch</code></a></span> <span style="font-style: italic; font-size:80%;"> - Follow development activity</span></li>
                <li>        <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/fork"><code>&nbsp;&nbsp;Fork</code></a></span> <span style="font-style: italic; font-size:80%;"> - Contribute and collaborate</span></li>
                <li>  <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/stargazers"><code>&nbsp;&nbsp;Star</code></a></span> <span style="font-style: italic; font-size:80%;"> - Throw some glitter, add some glamour</span></li>
        </div>
    </div>


        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="charstatemodel.html" title="dendropy.datamodel.charstatemodel: Character State Identities and Alphabets"
             >next</a> |</li>
        <li class="right" >
          <a href="treemodel.html" title="dendropy.datamodel.treemodel: Trees"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.5 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >DendroPy Library Reference</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href=""><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.datamodel.treecollectionmodel</span></code>: Collections of Trees</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2009-2025, Jeet Sukumaran and Mark T. Holder.
    </div>
  </body>
</html>