<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>dendropy.simulate.treesim: Unified Namespace Aggregating Functions and Classes for Tree Simulations &#8212; DendroPy 5.0.5 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/rtd.css?v=cedc6a2f" />
    
    <script src="../_static/documentation_options.js?v=967d581c"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="dendropy.simulate.popgensim: Population Genetics Simulations" href="popgensim.html" />
    <link rel="prev" title="dendropy.calculate.statistics: General Statistics" href="statistics.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="popgensim.html" title="dendropy.simulate.popgensim: Population Genetics Simulations"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="statistics.html" title="dendropy.calculate.statistics: General Statistics"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.5 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">DendroPy Library Reference</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href=""><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.simulate.treesim</span></code>: Unified Namespace Aggregating Functions and Classes for Tree Simulations</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="module-dendropy.simulate.treesim">
<span id="dendropy-simulate-treesim-unified-namespace-aggregating-functions-and-classes-for-tree-simulations"></span><h1><a class="reference internal" href="#module-dendropy.simulate.treesim" title="dendropy.simulate.treesim"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.simulate.treesim</span></code></a>: Unified Namespace Aggregating Functions and Classes for Tree Simulations<a class="headerlink" href="#module-dendropy.simulate.treesim" title="Link to this heading">¶</a></h1>
<p>This module provides a convenient interface that aggregates, wraps, and/or
implements functions and classes that simulate trees under various
models and processes. This module just exposes these function and classes under
the <code class="docutils literal notranslate"><span class="pre">dendropy.simulate.treesim</span></code> namespace. The actual functions and classes
are defined under the the appropriate model namespace in the <code class="docutils literal notranslate"><span class="pre">dendropy.model</span></code>
sub-package.</p>
<dl class="py function">
<dt class="sig sig-object py" id="dendropy.simulate.treesim.birth_death_tree">
<span class="sig-prename descclassname"><span class="pre">dendropy.simulate.treesim.</span></span><span class="sig-name descname"><span class="pre">birth_death_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">birth_rate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">death_rate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">birth_rate_sd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">death_rate_sd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/model/birthdeath.html#birth_death_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.simulate.treesim.birth_death_tree" title="Link to this definition">¶</a></dt>
<dd><p>Returns a birth-death tree with birth rate specified by <code class="docutils literal notranslate"><span class="pre">birth_rate</span></code>, and
death rate specified by <code class="docutils literal notranslate"><span class="pre">death_rate</span></code>, with edge lengths in continuous (real)
units.</p>
<p>Tree growth is controlled by one or more of the following arguments, of which
at least one must be specified:</p>
<blockquote>
<div><ul class="simple">
<li><p>If <code class="docutils literal notranslate"><span class="pre">num_extant_tips</span></code> is given as a keyword argument, tree is grown until the
number of EXTANT tips equals this number.</p></li>
<li><p>If <code class="docutils literal notranslate"><span class="pre">num_extinct_tips</span></code> is given as a keyword argument, tree is grown until the
number of EXTINCT tips equals this number.</p></li>
<li><p>If <code class="docutils literal notranslate"><span class="pre">num_total_tips</span></code> is given as a keyword argument, tree is grown until the
number of EXTANT plus EXTINCT tips equals this number.</p></li>
<li><p>If ‘max_time’ is given as a keyword argument, tree is grown for
a maximum of <code class="docutils literal notranslate"><span class="pre">max_time</span></code>.</p></li>
<li><p>If <code class="docutils literal notranslate"><span class="pre">gsa_ntax</span></code> is given then the tree will be simulated up to this number of
EXTANT tips (or 0 tips), then a tree will be randomly selected from the
intervals which corresond to times at which the tree had exactly <code class="docutils literal notranslate"><span class="pre">num_extant_tips</span></code>
leaves. This allows for simulations according to the “General
Sampling Approach” of Hartmann et al. (2010). If this option is
specified, then <code class="docutils literal notranslate"><span class="pre">num_extant_tips</span></code> MUST be specified and
<code class="docutils literal notranslate"><span class="pre">num_extinct_tips</span></code> and <code class="docutils literal notranslate"><span class="pre">num_total_tips</span></code> CANNOT be specified.</p></li>
</ul>
</div></blockquote>
<p>If more than one of the above is given, then tree growth will terminate when
<em>any</em> one of the termination conditions are met.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>birth_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – The birth rate.</p></li>
<li><p><strong>death_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – The death rate.</p></li>
<li><p><strong>birth_rate_sd</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – The standard deviation of the normally-distributed mutation added to
the birth rate as it is inherited by daughter nodes; if 0, birth rate
does not evolve on the tree.</p></li>
<li><p><strong>death_rate_sd</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – The standard deviation of the normally-distributed mutation added to
the death rate as it is inherited by daughter nodes; if 0, death rate
does not evolve on the tree.</p></li>
</ul>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>num_extant_tips</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – If specified, branching process is terminated when number of EXTANT
tips equals this number.</p></li>
<li><p><strong>num_extinct_tips</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – If specified, branching process is terminated when number of EXTINCT
tips equals this number.</p></li>
<li><p><strong>num_total_tips</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – If specified, branching process is terminated when number of EXTINCT
plus EXTANT tips equals this number.</p></li>
<li><p><strong>max_time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – If specified, branching process is terminated when time reaches or
exceeds this value.</p></li>
<li><p><strong>gsa_ntax</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The General Sampling Approach threshold for number of taxa. See above
for details.</p></li>
<li><p><strong>tree</strong> (<em>Tree instance</em>) – If given, then this tree will be used; otherwise a new one will be created.</p></li>
<li><p><strong>taxon_namespace</strong> (<em>TaxonNamespace instance</em>) – If given, then this will be assigned to the new tree, and, in addition,
taxa assigned to tips will be sourced from or otherwise created with
reference to this.</p></li>
<li><p><strong>is_assign_extant_taxa</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em> [</em><em>default: True</em><em>]</em>) – If False, then taxa will not be assigned to extant tips. If True
(default), then taxa will be assigned to extant tips. Taxa will be
assigned from the specified <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> or
<code class="docutils literal notranslate"><span class="pre">tree.taxon_namespace</span></code>. If the number of taxa required exceeds the
number of taxa existing in the taxon namespace, new <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects
will be created as needed and added to the taxon namespace.</p></li>
<li><p><strong>is_assign_extinct_taxa</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em> [</em><em>default: True</em><em>]</em>) – If False, then taxa will not be assigned to extant tips. If True
(default), then taxa will be assigned to extant tips. Taxa will be
assigned from the specified <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> or
<code class="docutils literal notranslate"><span class="pre">tree.taxon_namespace</span></code>. If the number of taxa required exceeds the
number of taxa existing in the taxon namespace, new <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects
will be created as needed and added to the taxon namespace. Note that
this option only makes sense if extinct tips are retained (specified via
‘is_retain_extinct_tips’ option), and will otherwise be ignored.</p></li>
<li><p><strong>is_add_extinct_attr</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em> [</em><em>default: True</em><em>]</em>) – If True (default), add an boolean attribute indicating whether or not a
node is an extinct tip or not. False will skip this. Name of attribute
is set by ‘extinct_attr_name’ argument, defaulting to ‘is_extinct’.
Note that this option only makes sense if extinct tips are retained
(specified via ‘is_retain_extinct_tips’ option), and will otherwise be
ignored.</p></li>
<li><p><strong>extinct_attr_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em> [</em><em>default: 'is_extinct'</em><em>]</em>) – Name of attribute to add to nodes indicating whether or not tip is extinct.
Note that this option only makes sense if extinct tips are retained
(specified via ‘is_retain_extinct_tips’ option), and will otherwise be
ignored.</p></li>
<li><p><strong>is_retain_extinct_tips</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em> [</em><em>default: False</em><em>]</em>) – If True, extinct tips will be retained on tree. Defaults to False:
extinct lineages removed from tree.</p></li>
<li><p><strong>repeat_until_success</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em> [</em><em>default: True</em><em>]</em>) – Under some conditions, it is possible for all lineages on a tree to go
extinct. In this case, if this argument is given as <code class="docutils literal notranslate"><span class="pre">True</span></code> (the
default), then a new branching process is initiated. If <code class="docutils literal notranslate"><span class="pre">False</span></code>
(default), then a TreeSimTotalExtinctionException is raised.</p></li>
<li><p><strong>rng</strong> (<a class="reference external" href="https://docs.python.org/3/library/random.html#random.Random" title="(in Python v3.13)"><em>random.Random</em></a><em>(</em><em>) or </em><em>equivalent instance</em>) – A Random() object or equivalent can be passed using the <code class="docutils literal notranslate"><span class="pre">rng</span></code> keyword;
otherwise GLOBAL_RNG is used.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>Hartmann, Wong, and Stadler “Sampling Trees from Evolutionary Models” Systematic Biology. 2010. 59(4). 465-476</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dendropy.simulate.treesim.constrained_kingman_tree">
<span class="sig-prename descclassname"><span class="pre">dendropy.simulate.treesim.</span></span><span class="sig-name descname"><span class="pre">constrained_kingman_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pop_tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_tree_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_node_label_fn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_sampling_strategy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'random_uniform'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_genes_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'num_genes'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_size_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pop_size'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">decorate_original_tree</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/model/coalescent.html#constrained_kingman_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.simulate.treesim.constrained_kingman_tree" title="Link to this definition">¶</a></dt>
<dd><p>Given a population tree, <code class="docutils literal notranslate"><span class="pre">pop_tree</span></code> this will return a <em>pair of
trees</em>: a gene tree simulated on this population tree based on
Kingman’s n-coalescent, and population tree with the additional
attribute ‘gene_nodes’ on each node, which is a list of
uncoalesced nodes from the gene tree associated with the given
node from the population tree.</p>
<p><code class="docutils literal notranslate"><span class="pre">pop_tree</span></code>: a Tree object.</p>
<dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">gene_sampling_strategy</span></code>: string</dt><dd><ul class="simple">
<li><p>“node_attribute”: Will expect each leaf of <code class="docutils literal notranslate"><span class="pre">pop_tree</span></code> to
have an attribute, <code class="docutils literal notranslate"><span class="pre">num_genes</span></code>, that specifies the number
of genes to be sampled from that population.</p></li>
<li><p>“fixed_per_population”: Will assign <code class="docutils literal notranslate"><span class="pre">num_genes</span></code> to each population.</p></li>
<li><p>“random_uniform”: Will assign genes to leaves with
uniform probability until <code class="docutils literal notranslate"><span class="pre">num_genes</span></code> genes have been
assigned.</p></li>
</ul>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">pop_size_attr</span></code>: string</dt><dd><p>The attribute name of the edges of <code class="docutils literal notranslate"><span class="pre">pop_tree</span></code> that
specify the population size. By default it is <code class="docutils literal notranslate"><span class="pre">pop_size</span></code>. The should
specify the effective <em>haploid</em> population size; i.e., number of gene
in the population: 2 * N in a diploid population of N individuals,
or N in a haploid population of N individuals.</p>
</dd>
</dl>
<p>If <code class="docutils literal notranslate"><span class="pre">pop_size</span></code> is 1 or 0 or None, then the edge lengths of <code class="docutils literal notranslate"><span class="pre">pop_tree</span></code> is
taken to be in haploid population units; i.e. where 1 unit equals 2N
generations for a diploid population of size N, or N generations for a
haploid population of size N. Otherwise the edge lengths of <code class="docutils literal notranslate"><span class="pre">pop_tree</span></code> is
taken to be in generations.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">gene_tree_list</span></code> is given, then the gene tree is added to the
tree block, and the tree block’s taxa block will be used to manage
the gene tree’s <code class="docutils literal notranslate"><span class="pre">taxa</span></code>.</p>
<p><code class="docutils literal notranslate"><span class="pre">gene_node_label_fn</span></code> is a function that takes two arguments (a string
and an integer, respectively, where the string is the containing species
taxon label and the integer is the gene index) and returns a label for
the corresponding the gene node.</p>
<p>if <code class="docutils literal notranslate"><span class="pre">decorate_original_tree</span></code> is True, then the list of uncoalesced nodes at
each node of the population tree is added to the original (input) population
tree instead of a copy.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">num_genes</span></code> is None, then it will be set to 1 under the
“node_attribute” strategy (serving as a fallback default for nodes that do
not spcify <code class="docutils literal notranslate"><span class="pre">num_genes_attr</span></code>) or the leaf count of <code class="docutils literal notranslate"><span class="pre">pop_tree</span></code> under the
<code class="docutils literal notranslate"><span class="pre">random_uniform</span></code> strategy.</p>
<p>Note that this function does very much the same thing as
<code class="docutils literal notranslate"><span class="pre">contained_coalescent_tree()</span></code>,
but provides a very different API.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dendropy.simulate.treesim.contained_coalescent_tree">
<span class="sig-prename descclassname"><span class="pre">dendropy.simulate.treesim.</span></span><span class="sig-name descname"><span class="pre">contained_coalescent_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">containing_tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_to_containing_taxon_map</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_pop_size_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pop_size'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_pop_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/model/coalescent.html#contained_coalescent_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.simulate.treesim.contained_coalescent_tree" title="Link to this definition">¶</a></dt>
<dd><p>Returns a gene tree simulated under the coalescent contained within a
population or species tree.</p>
<blockquote>
<div><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">containing_tree</span></code></dt><dd><p>The population or species tree. If <code class="docutils literal notranslate"><span class="pre">edge_pop_size_map</span></code> is not None,
and population sizes given are non-trivial (i.e., &gt;1), then edge
lengths on this tree are in units of generations. Otherwise edge
lengths are in population units; i.e. 2N generations for diploid
populations of size N, or N generations for diploid populations of
size N.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">gene_to_containing_taxon_map</span></code></dt><dd><p>A TaxonNamespaceMapping object mapping Taxon objects in the
<code class="docutils literal notranslate"><span class="pre">containing_tree</span></code> TaxonNamespace to corresponding Taxon objects in the
resulting gene tree.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">edge_pop_size_attr</span></code></dt><dd><p>Name of attribute of edges that specify population size. By default
this is “pop_size”. If this attribute does not exist,
<code class="docutils literal notranslate"><span class="pre">default_pop_size</span></code> will be used.  The value for this attribute
should be the haploid population size or the number of genes;
i.e.  2N for a diploid population of N individuals, or N for a
haploid population of N individuals. This value determines how
branch length units are interpreted in the input tree,
<code class="docutils literal notranslate"><span class="pre">containing_tree</span></code>.  If a biologically-meaningful value, then branch
lengths on the <code class="docutils literal notranslate"><span class="pre">containing_tree</span></code> are properly read as generations.
If not (e.g. 1 or 0), then they are in population units, i.e. where
1 unit of time equals 2N generations for a diploid population of
size N, or N generations for a haploid population of size N.
Otherwise time is in generations. If this argument is None, then
population sizes default to <code class="docutils literal notranslate"><span class="pre">default_pop_size</span></code>.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">default_pop_size</span></code></dt><dd><p>Population size to use if <code class="docutils literal notranslate"><span class="pre">edge_pop_size_attr</span></code> is None or
if an edge does not have the attribute. Defaults to 1.</p>
</dd>
</dl>
</div></blockquote>
<p>The returned gene tree will have the following extra attributes:</p>
<blockquote>
<div><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">pop_node_genes</span></code></dt><dd><p>A dictionary with nodes of <code class="docutils literal notranslate"><span class="pre">containing_tree</span></code> as keys and a list of gene
tree nodes that are uncoalesced as values.</p>
</dd>
</dl>
</div></blockquote>
<p>Note that this function does very much the same thing as
<code class="docutils literal notranslate"><span class="pre">constrained_kingman_tree()</span></code>, but provides a very different API.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dendropy.simulate.treesim.discrete_birth_death_tree">
<span class="sig-prename descclassname"><span class="pre">dendropy.simulate.treesim.</span></span><span class="sig-name descname"><span class="pre">discrete_birth_death_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">birth_rate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">death_rate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">birth_rate_sd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">death_rate_sd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/model/birthdeath.html#discrete_birth_death_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.simulate.treesim.discrete_birth_death_tree" title="Link to this definition">¶</a></dt>
<dd><p>Returns a birth-death tree with birth rate specified by <code class="docutils literal notranslate"><span class="pre">birth_rate</span></code>, and
death rate specified by <code class="docutils literal notranslate"><span class="pre">death_rate</span></code>, with edge lengths in discrete (integer)
units.</p>
<p><code class="docutils literal notranslate"><span class="pre">birth_rate_sd</span></code> is the standard deviation of the normally-distributed mutation
added to the birth rate as it is inherited by daughter nodes; if 0, birth
rate does not evolve on the tree.</p>
<p><code class="docutils literal notranslate"><span class="pre">death_rate_sd</span></code> is the standard deviation of the normally-distributed mutation
added to the death rate as it is inherited by daughter nodes; if 0, death
rate does not evolve on the tree.</p>
<p>Tree growth is controlled by one or more of the following arguments, of which
at least one must be specified:</p>
<blockquote>
<div><ul class="simple">
<li><p>If <code class="docutils literal notranslate"><span class="pre">ntax</span></code> is given as a keyword argument, tree is grown until the number of
tips == ntax.</p></li>
<li><p>If <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> is given as a keyword argument, tree is grown until the
number of tips == len(taxon_namespace), and the taxa are assigned randomly to the
tips.</p></li>
<li><p>If ‘max_time’ is given as a keyword argument, tree is grown for <code class="docutils literal notranslate"><span class="pre">max_time</span></code>
number of generations.</p></li>
</ul>
</div></blockquote>
<p>If more than one of the above is given, then tree growth will terminate when
<em>any</em> of the termination conditions (i.e., number of tips == <code class="docutils literal notranslate"><span class="pre">ntax</span></code>, or number
of tips == len(taxon_namespace) or number of generations = <code class="docutils literal notranslate"><span class="pre">max_time</span></code>) are met.</p>
<p>Also accepts a Tree object (with valid branch lengths) as an argument passed
using the keyword <code class="docutils literal notranslate"><span class="pre">tree</span></code>: if given, then this tree will be used; otherwise
a new one will be created.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">assign_taxa</span></code> is False, then taxa will <em>not</em> be assigned to the tips;
otherwise (default), taxa will be assigned. If <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> is given
(<code class="docutils literal notranslate"><span class="pre">tree.taxon_namespace</span></code>, if <code class="docutils literal notranslate"><span class="pre">tree</span></code> is given), and the final number of tips on the
tree after the termination condition is reached is less then the number of
taxa in <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> (as will be the case, for example, when
<code class="docutils literal notranslate"><span class="pre">ntax</span></code> &lt; len(<code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code>)), then a random subset of taxa in <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> will
be assigned to the tips of tree. If the number of tips is more than the number
of taxa in the <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code>, new Taxon objects will be created and added
to the <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> if the keyword argument <code class="docutils literal notranslate"><span class="pre">create_required_taxa</span></code> is not given as
False.</p>
<p>Under some conditions, it is possible for all lineages on a tree to go extinct.
In this case, if the keyword argument <code class="docutils literal notranslate"><span class="pre">repeat_until_success</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>, then a new
branching process is initiated.
If <code class="docutils literal notranslate"><span class="pre">False</span></code> (default), then a TreeSimTotalExtinctionException is raised.</p>
<p>A Random() object or equivalent can be passed using the <code class="docutils literal notranslate"><span class="pre">rng</span></code> keyword;
otherwise GLOBAL_RNG is used.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dendropy.simulate.treesim.mean_kingman_tree">
<span class="sig-prename descclassname"><span class="pre">dendropy.simulate.treesim.</span></span><span class="sig-name descname"><span class="pre">mean_kingman_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon_namespace</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/model/coalescent.html#mean_kingman_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.simulate.treesim.mean_kingman_tree" title="Link to this definition">¶</a></dt>
<dd><p>Returns a tree with coalescent intervals given by the expected times under
Kingman’s neutral coalescent.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dendropy.simulate.treesim.pure_kingman_tree">
<span class="sig-prename descclassname"><span class="pre">dendropy.simulate.treesim.</span></span><span class="sig-name descname"><span class="pre">pure_kingman_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon_namespace</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/model/coalescent.html#pure_kingman_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.simulate.treesim.pure_kingman_tree" title="Link to this definition">¶</a></dt>
<dd><p>Generates a tree under the unconstrained Kingman’s coalescent process.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>taxon_namespace</strong> (<a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> instance) – A pre-populated <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> where the contained <a class="reference internal" href="taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> instances
represent the genes or individuals sampled from the population.</p></li>
<li><p><strong>pop_size</strong> (<em>numeric</em>) – The size of the population from the which the coalescent process is
sampled.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>t</strong> (<em>|Tree|</em>) – A tree sampled from the Kingman’s neutral coalescent.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dendropy.simulate.treesim.rand_trees">
<span class="sig-prename descclassname"><span class="pre">dendropy.simulate.treesim.</span></span><span class="sig-name descname"><span class="pre">rand_trees</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rng</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_fn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_kwargs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_replicates</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/simulate/treesim.html#rand_trees"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.simulate.treesim.rand_trees" title="Link to this definition">¶</a></dt>
<dd><p>The model parameters may be specified as:
-   A single dict or map, in which case it will be repeated</p>
<blockquote>
<div><p>through the replicates,</p>
</div></blockquote>
<ul class="simple">
<li><p>A function, in which case it will be called for each replicate
with two positional arguments: the 0-based replicate index and
the random numberg generator object to use; the function should
return a dict or mapping of keyword-value pairs for the model
simulation call.</p></li>
<li><p>An iterable of dicts or maps, for <em>each</em> of which
<code class="xref any docutils literal notranslate"><span class="pre">n_replicates</span></code> simulations will be generated, in order.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dendropy.simulate.treesim.star_tree">
<span class="sig-prename descclassname"><span class="pre">dendropy.simulate.treesim.</span></span><span class="sig-name descname"><span class="pre">star_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon_namespace</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dendropy/model/treeshape.html#star_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dendropy.simulate.treesim.star_tree" title="Link to this definition">¶</a></dt>
<dd><p>Builds and returns a star tree from the given taxa block.</p>
</dd></dl>

</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div style="text-align: center; padding-top: 20px; padding-bottom: 5px; width: 100%;">
    <a href="../index.html"><img src="../_static/dendropy_logo.png" /></a>
</div><div style="clear:both; width: 100%; height:1px;"></div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h3><a href="../index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.simulate.treesim</span></code>: Unified Namespace Aggregating Functions and Classes for Tree Simulations</a><ul>
<li><a class="reference internal" href="#dendropy.simulate.treesim.birth_death_tree"><code class="docutils literal notranslate"><span class="pre">birth_death_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.simulate.treesim.constrained_kingman_tree"><code class="docutils literal notranslate"><span class="pre">constrained_kingman_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.simulate.treesim.contained_coalescent_tree"><code class="docutils literal notranslate"><span class="pre">contained_coalescent_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.simulate.treesim.discrete_birth_death_tree"><code class="docutils literal notranslate"><span class="pre">discrete_birth_death_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.simulate.treesim.mean_kingman_tree"><code class="docutils literal notranslate"><span class="pre">mean_kingman_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.simulate.treesim.pure_kingman_tree"><code class="docutils literal notranslate"><span class="pre">pure_kingman_tree</span></code></a></li>
<li><a class="reference internal" href="#dendropy.simulate.treesim.rand_trees"><code class="docutils literal notranslate"><span class="pre">rand_trees</span></code></a></li>
<li><a class="reference internal" href="#dendropy.simulate.treesim.star_tree"><code class="docutils literal notranslate"><span class="pre">star_tree</span></code></a></li>
</ul>
</li>
</ul>

  </div><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h4>Previous topic</h4>
    <p class="topless"><a href="statistics.html"
                          title="previous chapter"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.calculate.statistics</span></code>: General Statistics</a></p>
  </div>
  <div>
    <h4>Next topic</h4>
    <p class="topless"><a href="popgensim.html"
                          title="next chapter"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.simulate.popgensim</span></code>: Population Genetics Simulations</a></p>
  </div><div style="clear:both; width: 100%; height:1px;"></div>
    <!-- Documentation -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Documentation</h3>
        <ul>
            <li><a href="../primer/downloading.html">Downloading and Installing DendroPy</a></li>
            <li><a href="../primer/index.html">The DendroPy Primer</a></li>
            <li><a href="index.html">Library API Reference</a></li>
            <li>
                <a href="../schemas/index.html">Schemas</a>
                <ul>
                    <li><a href="../schemas/fasta.html">FASTA</a></li>
                    <li><a href="../schemas/newick.html">Newick</a></li>
                    <li><a href="../schemas/nexml.html">NeXML</a></li>
                    <li><a href="../schemas/nexus.html">Nexus</a></li>
                    <li><a href="../schemas/phylip.html">PHYLIP</a></li>
                </ul>
            </li>
            <li>
                <a href="../programs/index.html">Programs</a>
                <ul>
                    <li><a href="../programs/sumtrees.html">SumTrees</a></li>
                </ul>
            </li>
            <li><a href="../glossary.html">Glossary and Terminological Reference</a></li>
            <li><a href="../developer.html">Developer Guide</a></li>
            <li><a href="../planning.html">Ongoing Development</a></li>
            <li><a href="../changes.html">Change History</a></li>
        </ul>
    </div>

    <!-- Downloads -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Obtaining</h3>
        <ul>
            <li><a target="_blank" href="http://pypi.python.org/pypi/DendroPy">Install from the Python Package Index</a></li>
            <li><a target="_blank" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.5.tar.gz">Download the Source Code Archive</a></li>
            <li><a target="_blank" href="http://github.com/jeetsukumaran/DendroPy">Clone the Source Code Repository</a></li>
        </ul>
    </div>

    <!-- Discussions -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Discussion</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot; DendroPy Users&quot; </a>group to follow and participate in discussion, troubleshooting, help, information, suggestions, etc. on the usage and development of the DendroPy phylogenetic computing library.</p>
            <form action="http://groups.google.com/group/dendropy-users/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot;dendropy-users&quot;</a> group, or click <a href="http://groups.google.com/group/dendropy-users?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Announcements -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Announcements</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; DendroPy Announcements&quot; </a>group to receive announcements of new releases, updates, changes and other news of interest to DendroPy users and developers.</p>
            <form action="http://groups.google.com/group/dendropy-announce/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; dendropy-announce&quot; </a>group, or click <a href="http://groups.google.com/group/dendropy-announce?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Development -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative; padding-bottom: 15px; margin-bottom:5px;">
        <h3><span style="text-align: left"><a href="https://github.com/jeetsukumaran/DendroPy/">Development</a></span><a href="https://github.com/jeetsukumaran/DendroPy/"><span style="position: absolute; right: 0; top: 10px "><img src="../_static/Octocat.png" height="30px" alt="GitHub" /></span></a></h3>
        <div style="margin-top: 15px;">
            <!-- <a href="https://github.com/jeetsukumaran/DendroPy/issues">Issues</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/subscription">Watch</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/fork">Fork</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/stargazers">Star</a> &bull; <a href="https://github.com/jeetsukumaran/">Follow</a> -->
            <ul>
                <li>                <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/issues"><code>Issues</code></a></span> <span style="font-style: italic; font-size:80%;"> - Report bugs or request features</span></li>
                <li>     <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/subscription"><code>&nbsp;Watch</code></a></span> <span style="font-style: italic; font-size:80%;"> - Follow development activity</span></li>
                <li>        <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/fork"><code>&nbsp;&nbsp;Fork</code></a></span> <span style="font-style: italic; font-size:80%;"> - Contribute and collaborate</span></li>
                <li>  <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/stargazers"><code>&nbsp;&nbsp;Star</code></a></span> <span style="font-style: italic; font-size:80%;"> - Throw some glitter, add some glamour</span></li>
        </div>
    </div>


        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="popgensim.html" title="dendropy.simulate.popgensim: Population Genetics Simulations"
             >next</a> |</li>
        <li class="right" >
          <a href="statistics.html" title="dendropy.calculate.statistics: General Statistics"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.5 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >DendroPy Library Reference</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href=""><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.simulate.treesim</span></code>: Unified Namespace Aggregating Functions and Classes for Tree Simulations</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2009-2025, Jeet Sukumaran and Mark T. Holder.
    </div>
  </body>
</html>