<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Data Sets &#8212; DendroPy 5.0.5 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/rtd.css?v=cedc6a2f" />
    
    <script src="../_static/documentation_options.js?v=967d581c"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Working with Metadata Annotations" href="working_with_metadata_annotations.html" />
    <link rel="prev" title="Phylogenetic Distance-Based Operations" href="phylogenetic_distances.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="working_with_metadata_annotations.html" title="Working with Metadata Annotations"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="phylogenetic_distances.html" title="Phylogenetic Distance-Based Operations"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.5 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">The DendroPy Primer</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Data Sets</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="data-sets">
<h1>Data Sets<a class="headerlink" href="#data-sets" title="Link to this heading">¶</a></h1>
<p>The <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> class provides for objects that allow you to manage multiple types of phylogenetic data.</p>
<p>It has three primary attributes:</p>
<blockquote>
<div><dl class="simple">
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">taxon_namespaces</span></code></dt><dd><p>A list of all <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> objects in the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>, in the order that they were added or read, include <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> objects added implicitly through being associated with added <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> or <a class="reference internal" href="../library/charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a> objects.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_lists</span></code></dt><dd><p>A list of all <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> objects in the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>, in the order that they were added or read.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">char_matrices</span></code></dt><dd><p>A list of all <a class="reference internal" href="../library/charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a> objects in the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>, in the order that they were added or read.</p>
</dd>
</dl>
</div></blockquote>
<section id="dataset-creation-and-reading">
<h2><a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> Creation and Reading<a class="headerlink" href="#dataset-creation-and-reading" title="Link to this heading">¶</a></h2>
<section id="reading-and-writing-dataset-objects">
<h3>Reading and Writing <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> Objects<a class="headerlink" href="#reading-and-writing-dataset-objects" title="Link to this heading">¶</a></h3>
<p>You can use the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet.get" title="dendropy.datamodel.datasetmodel.DataSet.get"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get</span></code></a> factory class method for simultaneously instantiating and populating <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object, taking a data source as the first argument and a <a class="reference internal" href="../schemas/index.html#specifying-the-data-source-format"><span class="std std-ref">schema specification string</span></a> (”<code class="docutils literal notranslate"><span class="pre">nexus</span></code>”, “<code class="docutils literal notranslate"><span class="pre">newick</span></code>”, “<code class="docutils literal notranslate"><span class="pre">nexml</span></code>”, “<code class="docutils literal notranslate"><span class="pre">fasta</span></code>”, “<code class="docutils literal notranslate"><span class="pre">phylip</span></code>”, etc.) as the second:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
<span class="go">    path=&#39;pythonidae.nex&#39;,</span>
<span class="go">    schema=&#39;nexus&#39;)</span>
</pre></div>
</div>
<p>The <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet.read" title="dendropy.datamodel.datasetmodel.DataSet.read"><code class="xref py py-meth docutils literal notranslate"><span class="pre">read</span></code></a> instance method for reading additional data into existing objects are also supported, taking the same arguments (i.e., a data source, a <a class="reference internal" href="../schemas/index.html#specifying-the-data-source-format"><span class="std std-ref">schema specification string</span></a>, as well as optional :keyword arguments to customize the parse behavior):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>

<span class="c1"># Create the DataSet to store data</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="p">()</span>

<span class="c1"># Set it up to manage all data under a single taxon namespace.</span>
<span class="c1"># HIGHLY RECOMMENDED!</span>
<span class="n">taxon_namespace</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">()</span>
<span class="n">ds</span><span class="o">.</span><span class="n">attach_taxon_namespace</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="p">)</span>

<span class="c1"># Read from multiple sources</span>

<span class="c1"># Add a collection of trees</span>
<span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
    <span class="n">path</span><span class="o">=</span><span class="s1">&#39;pythonidae.mle.nex&#39;</span><span class="p">,</span>
    <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;nexus&#39;</span><span class="p">,)</span>

<span class="c1"># Add a collection of characters from a Nexus source</span>
<span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
    <span class="n">path</span><span class="o">=</span><span class="s1">&#39;pythonidae.chars.nexus&#39;</span><span class="p">,</span>
    <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;nexus&#39;</span><span class="p">,)</span>

<span class="c1"># Add a collection of characters from a FASTA source</span>
<span class="c1"># Note that with this format, we have to explicitly provide the type of data</span>
<span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
    <span class="n">path</span><span class="o">=</span><span class="s1">&#39;pythonidae_cytb.fasta&#39;</span><span class="p">,</span>
    <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">,</span>
    <span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>

<span class="c1"># Add a collection of characters from a PHYLIP source</span>
<span class="c1"># Note that with this format, we have to explicitly provide the type of data</span>
<span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
    <span class="n">path</span><span class="o">=</span><span class="s1">&#39;pythonidae.chars.phylip&#39;</span><span class="p">,</span>
    <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;phylip&#39;</span><span class="p">,</span>
    <span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>

<span class="c1"># Add a collection of continuous characters from a NeXML source</span>
<span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
    <span class="n">path</span><span class="o">=</span><span class="s1">&#39;pythonidae_continuous.chars.nexml&#39;</span><span class="p">,</span>
    <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;nexml&#39;</span><span class="p">,)</span>



</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note how the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet.attach_taxon_namespace" title="dendropy.datamodel.datasetmodel.DataSet.attach_taxon_namespace"><code class="xref py py-meth docutils literal notranslate"><span class="pre">attach_taxon_namespace</span></code></a> method is called before invoking any “<a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet.read" title="dendropy.datamodel.datasetmodel.DataSet.read"><code class="xref py py-meth docutils literal notranslate"><span class="pre">read</span></code></a>” statements, to ensure that all the taxon references in the data sources get mapped to the same <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> instance.
It is <strong>HIGHLY</strong> recommended that you do this, i.e., manage all data with the same <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> instance under the same taxonomic namespace, unless you have a special reason to include multiple independent taxon “domains” in the same data set.</p>
</div>
<p>The “<a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet.write" title="dendropy.datamodel.datasetmodel.DataSet.write"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write</span></code></a>” method allows you to write the data of a <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> to a file-like object or a file path
The following example aggregates the post-burn in MCMC samples from a series of NEXUS-formatted tree files into a single <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>, then, adds the <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> as well as the original character data into a single <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object, which is then written out as NEXUS-formatted file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">taxa</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">()</span>
<span class="n">trees</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TreeList</span><span class="p">(</span><span class="n">taxon_namespace</span><span class="o">=</span><span class="n">taxa</span><span class="p">)</span>
<span class="n">trees</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s1">&#39;pythonidae.mb.run1.t&#39;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;nexus&#39;</span><span class="p">,</span> <span class="n">tree_offset</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">trees</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s1">&#39;pythonidae.mb.run2.t&#39;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;nexus&#39;</span><span class="p">,</span> <span class="n">tree_offset</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">trees</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s1">&#39;pythonidae.mb.run3.t&#39;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;nexus&#39;</span><span class="p">,</span> <span class="n">tree_offset</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">trees</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s1">&#39;pythonidae.mb.run4.t&#39;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;nexus&#39;</span><span class="p">,</span> <span class="n">tree_offset</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="p">([</span><span class="n">trees</span><span class="p">])</span>
<span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s1">&#39;pythonidae_cytb.fasta&#39;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">,</span>
        <span class="n">data_type</span><span class="o">=</span><span class="s1">&#39;dna&#39;</span><span class="p">,</span>
        <span class="p">)</span>
<span class="n">ds</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s1">&#39;pythonidae_combined.nex&#39;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;nexus&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>If you do not want to actually write to a file, but instead simply need a string representing the data in a particular format, you can call the instance method <code class="xref py py-meth docutils literal notranslate"><span class="pre">as_string</span></code>, passing a <a class="reference internal" href="../schemas/index.html#specifying-the-data-writing-format"><span class="std std-ref">schema specification string</span></a> as the first argument:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="p">()</span>
<span class="n">ds</span><span class="o">.</span><span class="n">read_from_path</span><span class="p">(</span><span class="s1">&#39;pythonidae.cytb.fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;dnafasta&#39;</span><span class="p">)</span>
<span class="n">s</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">as_string</span><span class="p">(</span><span class="s1">&#39;nexus&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>or:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dna1</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="s2">&quot;pythonidae.nex&quot;</span><span class="p">),</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="n">s</span> <span class="o">=</span> <span class="n">dna1</span><span class="o">.</span><span class="n">as_string</span><span class="p">(</span><span class="n">schema</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
</pre></div>
</div>
<p>In addition, fine-grained control over the reading and writing of data is available through various keyword arguments.
More information on reading operations is available in the <a class="reference internal" href="reading_and_writing.html"><span class="doc">Reading and Writing Phylogenetic Data</span></a> section.</p>
</section>
<section id="creating-a-new-dataset-from-existing-treelist-and-charactermatrix-objects">
<h3>Creating a New <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> from Existing <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> and <a class="reference internal" href="../library/charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a> Objects<a class="headerlink" href="#creating-a-new-dataset-from-existing-treelist-and-charactermatrix-objects" title="Link to this heading">¶</a></h3>
<p>You can add independentally created or parsed data objects to a <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> by passing them as unnamed arguments to the constructor:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">treelist1</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">path</span><span class="o">=</span><span class="s1">&#39;pythonidae.mle.nex&#39;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;nexus&#39;</span><span class="p">)</span>
<span class="n">cytb</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DnaCharacterMatrix</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
    <span class="n">path</span><span class="o">=</span><span class="s1">&#39;pythonidae_cytb.fasta&#39;</span><span class="p">,</span>
    <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="p">([</span><span class="n">cytb</span><span class="p">,</span> <span class="n">treelist1</span><span class="p">])</span>
<span class="n">ds</span><span class="o">.</span><span class="n">unify_taxon_namespaces</span><span class="p">()</span>

</pre></div>
</div>
<p>Note how we call the instance method <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet.unify_taxon_namespaces" title="dendropy.datamodel.datasetmodel.DataSet.unify_taxon_namespaces"><code class="xref py py-meth docutils literal notranslate"><span class="pre">unify_taxon_namespaces</span></code></a> after the creation of the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object.
This method will remove all existing <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> objects from the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>, create and add a new one, and then map all taxon references in all contained <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> and <a class="reference internal" href="../library/charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a> objects to this new, unified <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>.</p>
</section>
<section id="adding-data-to-an-exisiting-dataset">
<h3>Adding Data to an Exisiting <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a><a class="headerlink" href="#adding-data-to-an-exisiting-dataset" title="Link to this heading">¶</a></h3>
<p>You can add independentally created or parsed data objects to a <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> using the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet.add" title="dendropy.datamodel.datasetmodel.DataSet.add"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add</span></code></a> method:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">..</span> <span class="n">literalinclude</span><span class="p">::</span> <span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">ds4</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
<p>Here, again, we call the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet.unify_taxon_namespaces" title="dendropy.datamodel.datasetmodel.DataSet.unify_taxon_namespaces"><code class="xref py py-meth docutils literal notranslate"><span class="pre">unify_taxon_namespaces</span></code></a> to map all taxon references to the same, common, unified <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>.</p>
</section>
</section>
<section id="taxon-management-with-data-sets">
<h2>Taxon Management with Data Sets<a class="headerlink" href="#taxon-management-with-data-sets" title="Link to this heading">¶</a></h2>
<p>The <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object, representing a meta-collection of phylogenetic data, differs in one important way from all the other phylogenetic data objects discussed so far with respect to taxon management, in that it is not associated with any particular <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> object.
Rather, it maintains a list (in the property <code class="xref py py-attr docutils literal notranslate"><span class="pre">taxon_namespaces</span></code>) of <em>all</em> the <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> objects referenced by its contained <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> objects (in the property <code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_lists</span></code>) and <a class="reference internal" href="../library/charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a> objects (in the property <code class="xref py py-attr docutils literal notranslate"><span class="pre">char_matrices</span></code>).</p>
<p>With respect to taxon management, <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> objects operate in one of two modes: “detached taxon set” mode and “attached taxon set” mode.</p>
<section id="detached-multiple-taxon-set-mode">
<h3>Detached (Multiple) Taxon Set Mode<a class="headerlink" href="#detached-multiple-taxon-set-mode" title="Link to this heading">¶</a></h3>
<p>In the “detached taxon set” mode, which is the default, <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object tracks all <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> references of their other data members in the property <code class="xref py py-attr docutils literal notranslate"><span class="pre">taxon_namespaces</span></code>, but no effort is made at taxon management as such.
Thus, every time a data source is read with a “detached taxon set” mode <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object, by default, a new  <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> object will be created and associated with the <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>, <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>, or <a class="reference internal" href="../library/charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a> objects created from each data source, resulting in multiple <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> independent references.
As such, “detached taxon set” mode <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> objects are suitable for handling data with multiple distinct sets of taxa.</p>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;primates.nex&quot;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;snakes.nex&quot;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The dataset, <code class="docutils literal notranslate"><span class="pre">ds</span></code>, will now contain two distinct sets of <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> objects, one for the taxa defined in “primates.nex”, and the other for the taxa defined for “snakes.nex”.
In this case, this behavior is correct, as the two files do indeed refer to different sets of taxa.</p>
<p>However, consider the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;pythonidae_cytb.fasta&quot;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;pythonidae_aa.nex&quot;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;pythonidae_morphological.nex&quot;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;pythonidae.mle.tre&quot;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Here, even though all the data files refer to the same set of taxa, the resulting  <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object will actually have 4 distinct  <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> objects, one for each of the independent reads, and a taxon with a particular label in the first file (e.g., “Python regius” of “pythonidae_cytb.fasta”) will map to a completely distinct <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> object than a taxon with the same label in the second file (e.g., “Python regius” of “pythonidae_aa.nex”).
This is incorrect behavior, and to achieve the correct behavior with a multiple taxon set mode <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object, we need to explicitly pass a <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> object to each of the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet.read_from_path" title="dendropy.datamodel.datasetmodel.DataSet.read_from_path"><code class="xref py py-meth docutils literal notranslate"><span class="pre">read_from_path</span></code></a> statements:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;pythonidae_cytb.fasta&quot;</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">schema</span><span class="o">=</span><span class="s2">&quot;pythonidae_aa.nex&quot;</span><span class="p">,</span> <span class="s2">&quot;nexus&quot;</span><span class="p">,</span> <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">ds</span><span class="o">.</span><span class="n">taxon_namespaces</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">schema</span><span class="o">=</span><span class="s2">&quot;pythonidae_morphological.nex&quot;</span><span class="p">,</span> <span class="s2">&quot;nexus&quot;</span><span class="p">,</span> <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">ds</span><span class="o">.</span><span class="n">taxon_namespaces</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">schema</span><span class="o">=</span><span class="s2">&quot;pythonidae.mle.tre&quot;</span><span class="p">,</span> <span class="s2">&quot;nexus&quot;</span><span class="p">,</span> <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">ds</span><span class="o">.</span><span class="n">taxon_namespaces</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">write_to_path</span><span class="p">(</span><span class="s2">&quot;pythonidae_combined.nex&quot;</span><span class="p">,</span> <span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>In the previous example, the first <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet.read" title="dendropy.datamodel.datasetmodel.DataSet.read"><code class="xref py py-meth docutils literal notranslate"><span class="pre">read</span></code></a> statement results in a new <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> object, which is added to the <code class="xref py py-attr docutils literal notranslate"><span class="pre">taxon_namespaces</span></code> property of the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object <code class="docutils literal notranslate"><span class="pre">ds</span></code>.
This <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> object gets passed via the <code class="docutils literal notranslate"><span class="pre">taxon_namespace</span></code> keyword to subsequent <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet.read_from_path" title="dendropy.datamodel.datasetmodel.DataSet.read_from_path"><code class="xref py py-meth docutils literal notranslate"><span class="pre">read_from_path</span></code></a> statements, and thus as each of the data sources are processed, the taxon references get mapped to <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects in the same, single, <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> object.</p>
<p>While this approach works to ensure correct taxon mapping across multiple data object reads and instantiation, in this context, it is probably more convenient to use the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> in “attached taxon set” mode.
In fact, it is highly recommended that <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> instances <em>always</em> use the “attached taxon set” mode, as, conceptually there are very few cases where a collection of data should span multiple independent taxon namespaces.</p>
</section>
<section id="attached-single-taxon-set-mode">
<h3>Attached (Single) Taxon Set Mode<a class="headerlink" href="#attached-single-taxon-set-mode" title="Link to this heading">¶</a></h3>
<p>In the “attached taxon set” mode, <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> objects ensure that the taxon references of all data objects that are added to them are mapped to the same <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> object (at least one for each independent read or creation operation).
The “attached taxon set” mode is activated by calling the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet.attach_taxon_namespace" title="dendropy.datamodel.datasetmodel.DataSet.attach_taxon_namespace"><code class="xref py py-meth docutils literal notranslate"><span class="pre">attach_taxon_namespace</span></code></a> method on a <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> and passing in the <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> to use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">taxa</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="s2">&quot;global&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">attach_taxon_namespace</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read_from_path</span><span class="p">(</span><span class="s2">&quot;pythonidae_cytb.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;dnafasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read_from_path</span><span class="p">(</span><span class="s2">&quot;pythonidae_aa.nex&quot;</span><span class="p">,</span> <span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read_from_path</span><span class="p">(</span><span class="s2">&quot;pythonidae_morphological.nex&quot;</span><span class="p">,</span> <span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read_from_path</span><span class="p">(</span><span class="s2">&quot;pythonidae.mle.tre&quot;</span><span class="p">,</span> <span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="switching-between-attached-and-detached-taxon-set-modes">
<h3>Switching Between Attached and Detached Taxon Set Modes<a class="headerlink" href="#switching-between-attached-and-detached-taxon-set-modes" title="Link to this heading">¶</a></h3>
<p>As noted above, you can use the <code class="xref py py-meth docutils literal notranslate"><span class="pre">attached_taxon_namespace</span></code> method to switch a <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object to attached taxon set mode.
To restore it to multiple taxon set mode, you would use the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet.detach_taxon_namespace" title="dendropy.datamodel.datasetmodel.DataSet.detach_taxon_namespace"><code class="xref py py-meth docutils literal notranslate"><span class="pre">detach_taxon_namespace</span></code></a> method:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">taxa</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="s2">&quot;global&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">attach_taxon_namespace</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read_from_path</span><span class="p">(</span><span class="s2">&quot;pythonidae_cytb.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;dnafasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read_from_path</span><span class="p">(</span><span class="s2">&quot;pythonidae_aa.nex&quot;</span><span class="p">,</span> <span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read_from_path</span><span class="p">(</span><span class="s2">&quot;pythonidae_morphological.nex&quot;</span><span class="p">,</span> <span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read_from_path</span><span class="p">(</span><span class="s2">&quot;pythonidae.mle.tre&quot;</span><span class="p">,</span> <span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">detach_taxon_namespace</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">read_from_path</span><span class="p">(</span><span class="s2">&quot;primates.nex&quot;</span><span class="p">,</span> <span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Here, the same <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> object is used to manage taxon references for data parsed from the first four files, while the data from the fifth and final file gets its own, distinct, <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> object and associated <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> object references.</p>
</section>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div style="text-align: center; padding-top: 20px; padding-bottom: 5px; width: 100%;">
    <a href="../index.html"><img src="../_static/dendropy_logo.png" /></a>
</div><div style="clear:both; width: 100%; height:1px;"></div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h3><a href="../index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Data Sets</a><ul>
<li><a class="reference internal" href="#dataset-creation-and-reading"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code> Creation and Reading</a><ul>
<li><a class="reference internal" href="#reading-and-writing-dataset-objects">Reading and Writing <code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code> Objects</a></li>
<li><a class="reference internal" href="#creating-a-new-dataset-from-existing-treelist-and-charactermatrix-objects">Creating a New <code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code> from Existing <code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code> and <code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code> Objects</a></li>
<li><a class="reference internal" href="#adding-data-to-an-exisiting-dataset">Adding Data to an Exisiting <code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#taxon-management-with-data-sets">Taxon Management with Data Sets</a><ul>
<li><a class="reference internal" href="#detached-multiple-taxon-set-mode">Detached (Multiple) Taxon Set Mode</a></li>
<li><a class="reference internal" href="#attached-single-taxon-set-mode">Attached (Single) Taxon Set Mode</a></li>
<li><a class="reference internal" href="#switching-between-attached-and-detached-taxon-set-modes">Switching Between Attached and Detached Taxon Set Modes</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  </div><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h4>Previous topic</h4>
    <p class="topless"><a href="phylogenetic_distances.html"
                          title="previous chapter">Phylogenetic Distance-Based Operations</a></p>
  </div>
  <div>
    <h4>Next topic</h4>
    <p class="topless"><a href="working_with_metadata_annotations.html"
                          title="next chapter">Working with Metadata Annotations</a></p>
  </div><div style="clear:both; width: 100%; height:1px;"></div>
    <!-- Documentation -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Documentation</h3>
        <ul>
            <li><a href="downloading.html">Downloading and Installing DendroPy</a></li>
            <li><a href="index.html">The DendroPy Primer</a></li>
            <li><a href="../library/index.html">Library API Reference</a></li>
            <li>
                <a href="../schemas/index.html">Schemas</a>
                <ul>
                    <li><a href="../schemas/fasta.html">FASTA</a></li>
                    <li><a href="../schemas/newick.html">Newick</a></li>
                    <li><a href="../schemas/nexml.html">NeXML</a></li>
                    <li><a href="../schemas/nexus.html">Nexus</a></li>
                    <li><a href="../schemas/phylip.html">PHYLIP</a></li>
                </ul>
            </li>
            <li>
                <a href="../programs/index.html">Programs</a>
                <ul>
                    <li><a href="../programs/sumtrees.html">SumTrees</a></li>
                </ul>
            </li>
            <li><a href="../glossary.html">Glossary and Terminological Reference</a></li>
            <li><a href="../developer.html">Developer Guide</a></li>
            <li><a href="../planning.html">Ongoing Development</a></li>
            <li><a href="../changes.html">Change History</a></li>
        </ul>
    </div>

    <!-- Downloads -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Obtaining</h3>
        <ul>
            <li><a target="_blank" href="http://pypi.python.org/pypi/DendroPy">Install from the Python Package Index</a></li>
            <li><a target="_blank" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.5.tar.gz">Download the Source Code Archive</a></li>
            <li><a target="_blank" href="http://github.com/jeetsukumaran/DendroPy">Clone the Source Code Repository</a></li>
        </ul>
    </div>

    <!-- Discussions -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Discussion</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot; DendroPy Users&quot; </a>group to follow and participate in discussion, troubleshooting, help, information, suggestions, etc. on the usage and development of the DendroPy phylogenetic computing library.</p>
            <form action="http://groups.google.com/group/dendropy-users/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot;dendropy-users&quot;</a> group, or click <a href="http://groups.google.com/group/dendropy-users?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Announcements -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Announcements</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; DendroPy Announcements&quot; </a>group to receive announcements of new releases, updates, changes and other news of interest to DendroPy users and developers.</p>
            <form action="http://groups.google.com/group/dendropy-announce/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; dendropy-announce&quot; </a>group, or click <a href="http://groups.google.com/group/dendropy-announce?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Development -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative; padding-bottom: 15px; margin-bottom:5px;">
        <h3><span style="text-align: left"><a href="https://github.com/jeetsukumaran/DendroPy/">Development</a></span><a href="https://github.com/jeetsukumaran/DendroPy/"><span style="position: absolute; right: 0; top: 10px "><img src="../_static/Octocat.png" height="30px" alt="GitHub" /></span></a></h3>
        <div style="margin-top: 15px;">
            <!-- <a href="https://github.com/jeetsukumaran/DendroPy/issues">Issues</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/subscription">Watch</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/fork">Fork</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/stargazers">Star</a> &bull; <a href="https://github.com/jeetsukumaran/">Follow</a> -->
            <ul>
                <li>                <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/issues"><code>Issues</code></a></span> <span style="font-style: italic; font-size:80%;"> - Report bugs or request features</span></li>
                <li>     <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/subscription"><code>&nbsp;Watch</code></a></span> <span style="font-style: italic; font-size:80%;"> - Follow development activity</span></li>
                <li>        <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/fork"><code>&nbsp;&nbsp;Fork</code></a></span> <span style="font-style: italic; font-size:80%;"> - Contribute and collaborate</span></li>
                <li>  <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/stargazers"><code>&nbsp;&nbsp;Star</code></a></span> <span style="font-style: italic; font-size:80%;"> - Throw some glitter, add some glamour</span></li>
        </div>
    </div>


        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="working_with_metadata_annotations.html" title="Working with Metadata Annotations"
             >next</a> |</li>
        <li class="right" >
          <a href="phylogenetic_distances.html" title="Phylogenetic Distance-Based Operations"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.5 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >The DendroPy Primer</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Data Sets</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2009-2025, Jeet Sukumaran and Mark T. Holder.
    </div>
  </body>
</html>