<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Population Genetic Summary Statistics &#8212; DendroPy 5.0.2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/rtd.css?v=cedc6a2f" />
    
    <script src="../_static/documentation_options.js?v=5df4c8d1"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Phylogenetic Distance-Based Operations" href="phylogenetic_distances.html" />
    <link rel="prev" title="Phylogenetic Character Analyses" href="phylogenetic_character_analyses.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="phylogenetic_distances.html" title="Phylogenetic Distance-Based Operations"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="phylogenetic_character_analyses.html" title="Phylogenetic Character Analyses"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">The DendroPy Primer</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Population Genetic Summary Statistics</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="population-genetic-summary-statistics">
<h1>Population Genetic Summary Statistics<a class="headerlink" href="#population-genetic-summary-statistics" title="Link to this heading">¶</a></h1>
<p>The <a class="reference internal" href="../library/popgenstat.html#module-dendropy.calculate.popgenstat" title="dendropy.calculate.popgenstat"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dendropy.calculate.popgenstat</span></code></a> module provides functions that calculate some common population genetic summary statistics.</p>
<p>For example, given a <a class="reference internal" href="../library/charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.DnaCharacterMatrix" title="dendropy.datamodel.charmatrixmodel.DnaCharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">DnaCharacterMatrix</span></code></a> as an argument, the <a class="reference internal" href="../library/popgenstat.html#dendropy.calculate.popgenstat.num_segregating_sites" title="dendropy.calculate.popgenstat.num_segregating_sites"><code class="xref py py-func docutils literal notranslate"><span class="pre">num_segregating_sites</span></code></a> function returns the raw number of segregating sites, <a class="reference internal" href="../library/popgenstat.html#dendropy.calculate.popgenstat.average_number_of_pairwise_differences" title="dendropy.calculate.popgenstat.average_number_of_pairwise_differences"><code class="xref py py-func docutils literal notranslate"><span class="pre">average_number_of_pairwise_differences</span></code></a> returns the average number of pairwise differences, and <a class="reference internal" href="../library/popgenstat.html#dendropy.calculate.popgenstat.nucleotide_diversity" title="dendropy.calculate.popgenstat.nucleotide_diversity"><code class="xref py py-func docutils literal notranslate"><span class="pre">nucleotide_diversity</span></code></a> returns the nucleotide diversity.</p>
<p>More complex statistics are provided by the <code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationPairSummaryStatistics</span></code> class.
Objects of this class are instantatiated with two lists of <code class="xref py py-class docutils literal notranslate"><span class="pre">DnaCharacterDataSequence</span></code> objects as arguments, each representing a sample of DNA sequences drawn from two distinct but related populations.
Once instantiated, the following attributes of the <code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationPairSummaryStatistics</span></code> object are available:</p>
<blockquote>
<div><dl class="simple">
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">average_number_of_pairwise_differences</span></code></dt><dd><p>The average number of pairwise differences between every sequence across both populations.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">average_number_of_pairwise_differences_between</span></code></dt><dd><p>The average number of pairwise differences between every sequence between both populations.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">average_number_of_pairwise_differences_within</span></code></dt><dd><p>The average number of pairwise differences between every sequence within each population.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">average_number_of_pairwise_differences_net</span></code></dt><dd><p>The net number of pairwise differences.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">num_segregating_sites</span></code></dt><dd><p>The number of segregating sites.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">wattersons_theta</span></code></dt><dd><p>Watterson’s theta.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">wakeleys_psi</span></code></dt><dd><p>Wakeley’s psi.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">tajimas_d</span></code></dt><dd><p>Tajima’s D.</p>
</dd>
</dl>
</div></blockquote>
<p>The following example calculates the suite of population genetic summary statistics for sequences drawn from two populations of sticklebacks.
The original data consists of 23 sequences, with individuals from Eastern Pacific populations identified by their taxon labels beginning with “<code class="docutils literal notranslate"><span class="pre">EPAC</span></code>” and individuals from Western Pacific populations identified by their taxon labels beginning with “<code class="docutils literal notranslate"><span class="pre">WPAC</span></code>”.
The taxon labels thus are used as the basis for sorting the sequences into the required lists of <code class="xref py py-class docutils literal notranslate"><span class="pre">DnaCharacterDataSequence</span></code> objects, <code class="docutils literal notranslate"><span class="pre">p1</span></code> and <code class="docutils literal notranslate"><span class="pre">p2</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="linenos"> 1</span><span class="ch">#! /usr/bin/env python</span>
<span class="linenos"> 2</span><span class="c1"># -*- coding: utf-8 -*-</span>
<span class="linenos"> 3</span>
<span class="linenos"> 4</span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="linenos"> 5</span><span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">popgenstat</span>
<span class="linenos"> 6</span>
<span class="linenos"> 7</span><span class="n">seqs</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DnaCharacterMatrix</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
<span class="linenos"> 8</span>        <span class="n">path</span><span class="o">=</span><span class="s2">&quot;orti1994.nex&quot;</span><span class="p">,</span>
<span class="linenos"> 9</span>        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="linenos">10</span><span class="n">p1</span> <span class="o">=</span> <span class="p">[]</span>
<span class="linenos">11</span><span class="n">p2</span> <span class="o">=</span> <span class="p">[]</span>
<span class="linenos">12</span><span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">t</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">seqs</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">):</span>
<span class="linenos">13</span>    <span class="k">if</span> <span class="n">t</span><span class="o">.</span><span class="n">label</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;EPAC&#39;</span><span class="p">):</span>
<span class="linenos">14</span>        <span class="n">p1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqs</span><span class="p">[</span><span class="n">t</span><span class="p">])</span>
<span class="linenos">15</span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos">16</span>        <span class="n">p2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqs</span><span class="p">[</span><span class="n">t</span><span class="p">])</span>
<span class="linenos">17</span><span class="n">pp</span> <span class="o">=</span> <span class="n">popgenstat</span><span class="o">.</span><span class="n">PopulationPairSummaryStatistics</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">)</span>
<span class="linenos">18</span>
<span class="linenos">19</span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Average number of pairwise differences (total): </span><span class="si">%s</span><span class="s1">&#39;</span> \
<span class="linenos">20</span>    <span class="o">%</span> <span class="n">pp</span><span class="o">.</span><span class="n">average_number_of_pairwise_differences</span><span class="p">)</span>
<span class="linenos">21</span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Average number of pairwise differences (between populations): </span><span class="si">%s</span><span class="s1">&#39;</span> \
<span class="linenos">22</span>    <span class="o">%</span> <span class="n">pp</span><span class="o">.</span><span class="n">average_number_of_pairwise_differences_between</span><span class="p">)</span>
<span class="linenos">23</span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Average number of pairwise differences (within populations): </span><span class="si">%s</span><span class="s1">&#39;</span> \
<span class="linenos">24</span>    <span class="o">%</span> <span class="n">pp</span><span class="o">.</span><span class="n">average_number_of_pairwise_differences_within</span><span class="p">)</span>
<span class="linenos">25</span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Average number of pairwise differences (net): </span><span class="si">%s</span><span class="s1">&#39;</span> \
<span class="linenos">26</span>    <span class="o">%</span> <span class="n">pp</span><span class="o">.</span><span class="n">average_number_of_pairwise_differences_net</span><span class="p">)</span>
<span class="linenos">27</span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Number of segregating sites: </span><span class="si">%s</span><span class="s1">&#39;</span> \
<span class="linenos">28</span>    <span class="o">%</span> <span class="n">pp</span><span class="o">.</span><span class="n">num_segregating_sites</span><span class="p">)</span>
<span class="linenos">29</span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Watterson&#39;s theta: </span><span class="si">%s</span><span class="s2">&quot;</span> \
<span class="linenos">30</span>    <span class="o">%</span> <span class="n">pp</span><span class="o">.</span><span class="n">wattersons_theta</span><span class="p">)</span>
<span class="linenos">31</span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Wakeley&#39;s Psi: </span><span class="si">%s</span><span class="s2">&quot;</span> \
<span class="linenos">32</span>    <span class="o">%</span> <span class="n">pp</span><span class="o">.</span><span class="n">wakeleys_psi</span><span class="p">)</span>
<span class="linenos">33</span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Tajima&#39;s D: </span><span class="si">%s</span><span class="s2">&quot;</span> \
<span class="linenos">34</span>    <span class="o">%</span> <span class="n">pp</span><span class="o">.</span><span class="n">tajimas_d</span><span class="p">)</span>
</pre></div>
</div>
<p>Lines 6-12 build up the two lists of <code class="xref py py-class docutils literal notranslate"><span class="pre">DnaCharacterDataSequence</span></code> objects by sorting the original sequences into their source populations based on the taxon label (with operational taxonomic units with labels beginning with “<code class="docutils literal notranslate"><span class="pre">EPAC</span></code>” coming from the Eastern Pacific, and assigned to the list <code class="docutils literal notranslate"><span class="pre">p1</span></code>, while those that begin with “<code class="docutils literal notranslate"><span class="pre">WPAC</span></code>” coming from the Western Pacific, and assigned to the list <code class="docutils literal notranslate"><span class="pre">p2</span></code>).
These lists are then passed as the instantiation arguments to the <code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationPairSummaryStatistics</span></code> constructor in line 14.
The calculations are performed immediately, and the results reported in the following lines.</p>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div style="text-align: center; padding-top: 20px; padding-bottom: 5px; width: 100%;">
    <a href="../index.html"><img src="../_static/dendropy_logo.png" /></a>
</div><div style="clear:both; width: 100%; height:1px;"></div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script><div style="clear:both; width: 100%; height:1px;"></div><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h4>Previous topic</h4>
    <p class="topless"><a href="phylogenetic_character_analyses.html"
                          title="previous chapter">Phylogenetic Character Analyses</a></p>
  </div>
  <div>
    <h4>Next topic</h4>
    <p class="topless"><a href="phylogenetic_distances.html"
                          title="next chapter">Phylogenetic Distance-Based Operations</a></p>
  </div><div style="clear:both; width: 100%; height:1px;"></div>
    <!-- Documentation -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Documentation</h3>
        <ul>
            <li><a href="downloading.html">Downloading and Installing DendroPy</a></li>
            <li><a href="index.html">The DendroPy Primer</a></li>
            <li><a href="../library/index.html">Library API Reference</a></li>
            <li>
                <a href="../schemas/index.html">Schemas</a>
                <ul>
                    <li><a href="../schemas/fasta.html">FASTA</a></li>
                    <li><a href="../schemas/newick.html">Newick</a></li>
                    <li><a href="../schemas/nexml.html">NeXML</a></li>
                    <li><a href="../schemas/nexus.html">Nexus</a></li>
                    <li><a href="../schemas/phylip.html">PHYLIP</a></li>
                </ul>
            </li>
            <li>
                <a href="../programs/index.html">Programs</a>
                <ul>
                    <li><a href="../programs/sumtrees.html">SumTrees</a></li>
                </ul>
            </li>
            <li><a href="../glossary.html">Glossary and Terminological Reference</a></li>
            <li><a href="../developer.html">Developer Guide</a></li>
            <li><a href="../planning.html">Ongoing Development</a></li>
            <li><a href="../changes.html">Change History</a></li>
        </ul>
    </div>

    <!-- Downloads -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Obtaining</h3>
        <ul>
            <li><a target="_blank" href="http://pypi.python.org/pypi/DendroPy">Install from the Python Package Index</a></li>
            <li><a target="_blank" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.2.tar.gz">Download the Source Code Archive</a></li>
            <li><a target="_blank" href="http://github.com/jeetsukumaran/DendroPy">Clone the Source Code Repository</a></li>
        </ul>
    </div>

    <!-- Discussions -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Discussion</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot; DendroPy Users&quot; </a>group to follow and participate in discussion, troubleshooting, help, information, suggestions, etc. on the usage and development of the DendroPy phylogenetic computing library.</p>
            <form action="http://groups.google.com/group/dendropy-users/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot;dendropy-users&quot;</a> group, or click <a href="http://groups.google.com/group/dendropy-users?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Announcements -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Announcements</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; DendroPy Announcements&quot; </a>group to receive announcements of new releases, updates, changes and other news of interest to DendroPy users and developers.</p>
            <form action="http://groups.google.com/group/dendropy-announce/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; dendropy-announce&quot; </a>group, or click <a href="http://groups.google.com/group/dendropy-announce?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Development -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative; padding-bottom: 15px; margin-bottom:5px;">
        <h3><span style="text-align: left"><a href="https://github.com/jeetsukumaran/DendroPy/">Development</a></span><a href="https://github.com/jeetsukumaran/DendroPy/"><span style="position: absolute; right: 0; top: 10px "><img src="../_static/Octocat.png" height="30px" alt="GitHub" /></span></a></h3>
        <div style="margin-top: 15px;">
            <!-- <a href="https://github.com/jeetsukumaran/DendroPy/issues">Issues</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/subscription">Watch</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/fork">Fork</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/stargazers">Star</a> &bull; <a href="https://github.com/jeetsukumaran/">Follow</a> -->
            <ul>
                <li>                <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/issues"><code>Issues</code></a></span> <span style="font-style: italic; font-size:80%;"> - Report bugs or request features</span></li>
                <li>     <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/subscription"><code>&nbsp;Watch</code></a></span> <span style="font-style: italic; font-size:80%;"> - Follow development activity</span></li>
                <li>        <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/fork"><code>&nbsp;&nbsp;Fork</code></a></span> <span style="font-style: italic; font-size:80%;"> - Contribute and collaborate</span></li>
                <li>  <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/stargazers"><code>&nbsp;&nbsp;Star</code></a></span> <span style="font-style: italic; font-size:80%;"> - Throw some glitter, add some glamour</span></li>
        </div>
    </div>


        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="phylogenetic_distances.html" title="Phylogenetic Distance-Based Operations"
             >next</a> |</li>
        <li class="right" >
          <a href="phylogenetic_character_analyses.html" title="Phylogenetic Character Analyses"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >The DendroPy Primer</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Population Genetic Summary Statistics</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2009-2024, Jeet Sukumaran and Mark T. Holder.
    </div>
  </body>
</html>