<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Tree Statistics, Metrics, Summarizations, and Other Calculations &#8212; DendroPy 5.0.5 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/rtd.css?v=cedc6a2f" />
    
    <script src="../_static/documentation_options.js?v=967d581c"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Tree Manipulation and Restructuring" href="treemanips.html" />
    <link rel="prev" title="Collections of Trees" href="treecollections.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="treemanips.html" title="Tree Manipulation and Restructuring"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="treecollections.html" title="Collections of Trees"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.5 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">The DendroPy Primer</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Tree Statistics, Metrics, Summarizations, and Other Calculations</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="tree-statistics-metrics-summarizations-and-other-calculations">
<h1>Tree Statistics, Metrics, Summarizations, and Other Calculations<a class="headerlink" href="#tree-statistics-metrics-summarizations-and-other-calculations" title="Link to this heading">¶</a></h1>
<p>Some general tree metrics that are calculated without reference to any particular model or data and general report some tree metadata (e.g., tree length, node ages, etc.) are available as instance methods.
More specialized tree statistics, however, are available through functions in various other modules:</p>
<ul class="simple">
<li><p>The <a class="reference internal" href="../library/treemeasure.html#module-dendropy.calculate.treemeasure" title="dendropy.calculate.treemeasure"><code class="xref py py-mod docutils literal notranslate"><span class="pre">treemeasure</span></code></a> module provides for calculation of statistics that are typically calculated on a single tree.</p></li>
<li><p>The <a class="reference internal" href="../library/treecompare.html#module-dendropy.calculate.treecompare" title="dendropy.calculate.treecompare"><code class="xref py py-mod docutils literal notranslate"><span class="pre">treecompare</span></code></a> module provides for calculation of statistics that are typically calculated between trees</p></li>
<li><p>The <a class="reference internal" href="../library/treescore.html#module-dendropy.calculate.treescore" title="dendropy.calculate.treescore"><code class="xref py py-mod docutils literal notranslate"><span class="pre">treescore</span></code></a> module provides for statistics that typically score a tree under a model and with reference to some sort of data.</p></li>
<li><p>The <a class="reference internal" href="../library/coalescent.html#module-dendropy.model.coalescent" title="dendropy.model.coalescent"><code class="xref py py-mod docutils literal notranslate"><span class="pre">coalescent</span></code></a> module provides for calcuations on trees under the coalescent model.</p></li>
</ul>
<p>In addition, see the <a class="reference internal" href="phylogenetic_distances.html"><span class="doc">PhylogeneticDistanceMatrix class</span></a> for statistics, operations, and inferences based on phylogenetic (taxon-to-taxon) distances, including rapid calculation of MRCA’s, patristics distances, neighbor-joining (NJ) and Unweighted Pair Group with Mathematical Average (UPGMA) trees, phylogenetic community statistics (such as the Mean Pairwise Distance, MPD, or Mean Nearest Taxon Distance, MNTD), and more.</p>
<section id="native-tree-statistic-and-metric-methods">
<h2>Native Tree Statistic and Metric Methods<a class="headerlink" href="#native-tree-statistic-and-metric-methods" title="Link to this heading">¶</a></h2>
<p>Basic meta-information about tree structure are available as native <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> methods.</p>
<section id="tree-length">
<h3>Tree Length<a class="headerlink" href="#tree-length" title="Link to this heading">¶</a></h3>
<p>The <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree.length" title="dendropy.datamodel.treemodel.Tree.length"><code class="xref py py-meth docutils literal notranslate"><span class="pre">length</span></code></a> method returns the sum of edge lengths of a <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> object, with edges that do not have any length assigned being treated as edges with length 0.
The following example shows how to identify the “critical” value for an <a class="reference external" href="http://hymenoptera.tamu.edu/courses/ento606/Suggested%20Readings/Slowinksi_Crother_1998.pdf">Archie-Faith-Cranston or PTP test</a> from a sample of <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects, i.e. a tree length equal to or greater than 95% of the trees in the sample:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="kn">import</span> <span class="nn">dendropy</span>

<span class="n">trees</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
    <span class="n">path</span><span class="o">=</span><span class="s2">&quot;pythonidae.random.bd0301.tre&quot;</span><span class="p">,</span>
    <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="n">tree_lengths</span> <span class="o">=</span> <span class="p">[</span><span class="n">tree</span><span class="o">.</span><span class="n">length</span><span class="p">()</span> <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">trees</span><span class="p">]</span>
<span class="n">tree_lengths</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="n">crit_index_95</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="mf">0.95</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">tree_lengths</span><span class="p">))</span>
<span class="n">crit_index_99</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="mf">0.99</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">tree_lengths</span><span class="p">))</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;95</span><span class="si">%%</span><span class="s2"> critical value: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">tree_lengths</span><span class="p">[</span><span class="n">crit_index_95</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;99</span><span class="si">%%</span><span class="s2"> critical value: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">tree_lengths</span><span class="p">[</span><span class="n">crit_index_99</span><span class="p">])</span>
</pre></div>
</div>
</section>
<section id="node-ages">
<h3>Node Ages<a class="headerlink" href="#node-ages" title="Link to this heading">¶</a></h3>
<p>The <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree.calc_node_ages" title="dendropy.datamodel.treemodel.Tree.calc_node_ages"><code class="xref py py-meth docutils literal notranslate"><span class="pre">calc_node_ages</span></code></a> method calculates the age of a node (i.e., the sum of edge lengths from the node to a tip) and assigns it to the <code class="xref py py-attr docutils literal notranslate"><span class="pre">age</span></code> attribute. The following example iterates through the post-burnin of an MCMC sample of ultrametric trees, calculating the age of the MRCA of two taxa, and reports the mean age of the node.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="kn">import</span> <span class="nn">dendropy</span>
<span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">treemeasure</span>

<span class="n">trees</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">path</span><span class="o">=</span><span class="s2">&quot;pythonidae.beast-mcmc.trees&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">,</span>
        <span class="n">tree_offset</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">maculosa_childreni_ages</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">tree</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">trees</span><span class="p">):</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">calc_node_ages</span><span class="p">()</span>
    <span class="n">taxon_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Antaresia maculosa&quot;</span><span class="p">,</span><span class="s2">&quot;Antaresia childreni&quot;</span><span class="p">]</span>
    <span class="n">mrca</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">mrca</span><span class="p">(</span><span class="n">taxon_labels</span><span class="o">=</span><span class="n">taxon_labels</span><span class="p">)</span>
    <span class="n">maculosa_childreni_ages</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mrca</span><span class="o">.</span><span class="n">age</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Mean age of MRCA of &#39;Antaresia maculosa&#39; and &#39;Antaresia childreni&#39;: </span><span class="si">%s</span><span class="s2">&quot;</span> \
    <span class="o">%</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">maculosa_childreni_ages</span><span class="p">))</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">maculosa_childreni_ages</span><span class="p">)))</span>


</pre></div>
</div>
</section>
<section id="number-of-lineages-at-a-particular-time-and-lineage-through-time-plots">
<h3>Number of Lineages at a Particular Time and Lineage Through Time Plots<a class="headerlink" href="#number-of-lineages-at-a-particular-time-and-lineage-through-time-plots" title="Link to this heading">¶</a></h3>
<p>The <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree.num_lineages_at" title="dendropy.datamodel.treemodel.Tree.num_lineages_at"><code class="xref py py-meth docutils literal notranslate"><span class="pre">num_lineages_at</span></code></a> method of the <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> class returns the number of lineages at a particular time given in terms of distance from the root.
The following example extracts the number of lineages at fixed intervals along the length of the tree to use in an Lineage Through Time (LTT) plot:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>

<span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
            <span class="n">path</span><span class="o">=</span><span class="s2">&quot;hiv1.nexus&quot;</span><span class="p">,</span>
            <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>

<span class="c1"># Returns distance of node furthest from root, i.e., maximum time available on</span>
<span class="c1"># tree</span>
<span class="n">total_time</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">max_distance_from_root</span><span class="p">()</span>

<span class="c1"># Divide time span into 10 steps</span>
<span class="n">step</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">total_time</span><span class="p">)</span> <span class="o">/</span> <span class="mi">10</span>

<span class="c1"># To store tuples of (time, number of lineages)</span>
<span class="n">ltt</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># Start at first time step</span>
<span class="n">current_time</span> <span class="o">=</span> <span class="n">step</span>
<span class="k">while</span> <span class="n">current_time</span> <span class="o">&lt;=</span> <span class="n">total_time</span><span class="p">:</span>
    <span class="c1"># Get number of lineages at current time</span>
    <span class="n">num_lineages</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">num_lineages_at</span><span class="p">(</span><span class="n">current_time</span><span class="p">)</span>
    <span class="c1"># Store it</span>
    <span class="n">ltt</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="p">(</span><span class="n">current_time</span><span class="p">,</span> <span class="n">num_lineages</span><span class="p">)</span> <span class="p">)</span>
    <span class="c1"># Move to next time step</span>
    <span class="n">current_time</span> <span class="o">+=</span> <span class="n">step</span>

<span class="c1"># Get the final number of lineages</span>
<span class="c1"># Note: may not be the same as the number of tips if the tree has extinct</span>
<span class="c1"># tips/taxa; though, if this were the case, we would not be dealing with an</span>
<span class="c1"># ultrametric tree.</span>
<span class="k">if</span> <span class="n">current_time</span> <span class="o">&lt;</span> <span class="n">total_time</span><span class="p">:</span>
    <span class="n">ltt</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="p">(</span><span class="n">total_time</span><span class="p">,</span> <span class="n">tree</span><span class="o">.</span><span class="n">num_lineages_at</span><span class="p">(</span><span class="n">total_time</span><span class="p">))</span> <span class="p">)</span>

<span class="c1"># Print results</span>
<span class="k">for</span> <span class="n">t</span><span class="p">,</span> <span class="n">num_lineages</span> <span class="ow">in</span> <span class="n">ltt</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{:12.8f}</span><span class="se">\t</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">num_lineages</span><span class="p">))</span>
</pre></div>
</div>
</section>
</section>
<section id="unary-tree-statistics-and-metrics">
<h2>Unary Tree Statistics and Metrics<a class="headerlink" href="#unary-tree-statistics-and-metrics" title="Link to this heading">¶</a></h2>
<p>Numerous specialized statistics and indexes of tree shape and structure (B1, Colless’ imbalance, Pybus-Harvey-Gamma, etc.) are available through the <a class="reference internal" href="../library/treemeasure.html#module-dendropy.calculate.treemeasure" title="dendropy.calculate.treemeasure"><code class="xref py py-mod docutils literal notranslate"><span class="pre">treemeasure</span></code></a> module:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">dendropy</span>
<span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">treemeasure</span>
<span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">statistics</span>

<span class="c1"># Since we do not want to waste memory by keeping the actual trees around</span>
<span class="c1"># after we are done calculating the statistics, we use the tree yielder</span>
<span class="c1"># instead of:</span>
<span class="c1">#       dendropy.TreeList.get(</span>
<span class="c1">#           path=&quot;pythonidae.beast-mcmc.trees&quot;,</span>
<span class="c1">#           schema=&quot;nexus&quot;,</span>
<span class="c1">#           tree_offset=200)</span>

<span class="n">tree_stats</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
<span class="k">for</span> <span class="n">tree_idx</span><span class="p">,</span> <span class="n">tree</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">yield_from_files</span><span class="p">(</span>
            <span class="n">files</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;pythonidae.beast-mcmc.trees&quot;</span><span class="p">],</span>
            <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)):</span>
    <span class="k">if</span> <span class="n">tree_idx</span> <span class="o">&lt;</span> <span class="mi">200</span><span class="p">:</span>
        <span class="k">continue</span> <span class="c1"># burnin</span>
    <span class="n">tree_stats</span><span class="p">[</span><span class="s2">&quot;B1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">treemeasure</span><span class="o">.</span><span class="n">B1</span><span class="p">(</span><span class="n">tree</span><span class="p">))</span>
    <span class="n">tree_stats</span><span class="p">[</span><span class="s2">&quot;colless&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">treemeasure</span><span class="o">.</span><span class="n">colless_tree_imbalance</span><span class="p">(</span><span class="n">tree</span><span class="p">))</span>
    <span class="n">tree_stats</span><span class="p">[</span><span class="s2">&quot;PBH&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">treemeasure</span><span class="o">.</span><span class="n">pybus_harvey_gamma</span><span class="p">(</span><span class="n">tree</span><span class="p">))</span>
    <span class="n">tree_stats</span><span class="p">[</span><span class="s2">&quot;sackin&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">treemeasure</span><span class="o">.</span><span class="n">sackin_index</span><span class="p">(</span><span class="n">tree</span><span class="p">))</span>
    <span class="n">tree_stats</span><span class="p">[</span><span class="s2">&quot;treeness&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">treemeasure</span><span class="o">.</span><span class="n">treeness</span><span class="p">(</span><span class="n">tree</span><span class="p">))</span>

<span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">tree_stats</span><span class="p">:</span>
    <span class="n">values</span> <span class="o">=</span> <span class="n">tree_stats</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
    <span class="n">mean</span><span class="p">,</span> <span class="n">var</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">mean_and_sample_variance</span><span class="p">(</span><span class="n">values</span><span class="p">)</span>
    <span class="n">hpd</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">empirical_hpd</span><span class="p">(</span><span class="n">values</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{:15}</span><span class="s2">: mean = </span><span class="si">{}</span><span class="s2">, variance = </span><span class="si">{}</span><span class="s2">, hpd = (</span><span class="si">{}</span><span class="s2">, </span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">mean</span><span class="p">,</span> <span class="n">var</span><span class="p">,</span> <span class="n">hpd</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">hpd</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
</pre></div>
</div>
<section id="pybus-harvey-gamma">
<h3>Pybus-Harvey Gamma<a class="headerlink" href="#pybus-harvey-gamma" title="Link to this heading">¶</a></h3>
<p>The Pybus-Harvey Gamma statistic is given by the <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree.pybus_harvey_gamma" title="dendropy.datamodel.treemodel.Tree.pybus_harvey_gamma"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pybus_harvey_gamma</span></code></a> instance method. The following example iterates through the post-burn-in of an MCMC sample of trees, reporting the mean Pybus-Harvey Gamma statistic:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="kn">import</span> <span class="nn">dendropy</span>
<span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">treemeasure</span>

<span class="n">trees</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TreeList</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">path</span><span class="o">=</span><span class="s2">&quot;pythonidae.beast-mcmc.trees&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">,</span>
        <span class="n">tree_offset</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">pbhg</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">tree</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">trees</span><span class="p">):</span>
    <span class="n">pbhg</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">treemeasure</span><span class="o">.</span><span class="n">pybus_harvey_gamma</span><span class="p">(</span><span class="n">tree</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Mean Pybus-Harvey-Gamma: </span><span class="si">%s</span><span class="s2">&quot;</span> \
    <span class="o">%</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">pbhg</span><span class="p">))</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">pbhg</span><span class="p">)))</span>


</pre></div>
</div>
</section>
<section id="patristic-distances">
<h3>Patristic Distances<a class="headerlink" href="#patristic-distances" title="Link to this heading">¶</a></h3>
<p>The <a class="reference internal" href="../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix" title="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix</span></code></a> is the most efficient way to calculate the patristic distances between taxa or leaves on a tree, when doing multiple such calculations.
The easiest way to get an object of this class for a particular tree is to call <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree.phylogenetic_distance_matrix" title="dendropy.datamodel.treemodel.Tree.phylogenetic_distance_matrix"><code class="xref py py-meth docutils literal notranslate"><span class="pre">phylogenetic_distance_matrix</span></code></a>.
The object is callable, taking two <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects as arguments and returning the sum of edge lengths between the two. The following example reports the pairwise distances between all taxa on the input tree:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>

<span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
    <span class="n">path</span><span class="o">=</span><span class="s2">&quot;pythonidae.mle.nex&quot;</span><span class="p">,</span>
    <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="n">pdc</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">phylogenetic_distance_matrix</span><span class="p">()</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">t1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]):</span>
    <span class="k">for</span> <span class="n">t2</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">:]:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Distance between &#39;</span><span class="si">%s</span><span class="s2">&#39; and &#39;</span><span class="si">%s</span><span class="s2">&#39;: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">t1</span><span class="o">.</span><span class="n">label</span><span class="p">,</span> <span class="n">t2</span><span class="o">.</span><span class="n">label</span><span class="p">,</span> <span class="n">pdc</span><span class="p">(</span><span class="n">t1</span><span class="p">,</span> <span class="n">t2</span><span class="p">)))</span>
</pre></div>
</div>
<p>Note that the <a class="reference internal" href="../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix" title="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix</span></code></a> object does not automatically update if the original <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> changes: it is essentially a snapshot of <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> at the point in which it is instantiated.
If the original <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> changes, you should create a new instance of the corresponding <a class="reference internal" href="../library/phylogeneticdistance.html#dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix" title="dendropy.calculate.phylogeneticdistance.PhylogeneticDistanceMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">PhylogeneticDistanceMatrix</span></code></a> object.</p>
</section>
</section>
<section id="comparing-and-summarizing-trees">
<h2>Comparing and Summarizing Trees<a class="headerlink" href="#comparing-and-summarizing-trees" title="Link to this heading">¶</a></h2>
<section id="distances-between-trees">
<h3>Distances Between Trees<a class="headerlink" href="#distances-between-trees" title="Link to this heading">¶</a></h3>
<section id="unweighted-robinson-foulds-distance">
<h4>Unweighted Robinson-Foulds Distance<a class="headerlink" href="#unweighted-robinson-foulds-distance" title="Link to this heading">¶</a></h4>
<p>The <em>unweighted</em> Robinson-Foulds distance (often referred to as just the Robinson-Foulds distance) is given by the <a class="reference internal" href="../library/treecompare.html#dendropy.calculate.treecompare.symmetric_difference" title="dendropy.calculate.treecompare.symmetric_difference"><code class="xref py py-func docutils literal notranslate"><span class="pre">dendropy.calculate.treecompare.symmetric_difference</span></code></a> function:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">treecompare</span>

<span class="n">s1</span> <span class="o">=</span> <span class="s2">&quot;(a,(b,(c,d)));&quot;</span>
<span class="n">s2</span> <span class="o">=</span> <span class="s2">&quot;(a,(d,(b,c)));&quot;</span>

<span class="c1"># establish common taxon namespace</span>
<span class="n">tns</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">()</span>

<span class="c1"># ensure all trees loaded use common namespace</span>
<span class="n">tree1</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">s1</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;newick&#39;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">tns</span><span class="p">)</span>
<span class="n">tree2</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">s2</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;newick&#39;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">tns</span><span class="p">)</span>

<span class="c1">## Unweighted Robinson-Foulds distance</span>
<span class="nb">print</span><span class="p">(</span><span class="n">treecompare</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">tree1</span><span class="p">,</span> <span class="n">tree2</span><span class="p">))</span>
</pre></div>
</div>
<p>Note that the two trees <em>must</em> share the same <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> reference, otherwise an error will be raised:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;&gt;</span> <span class="kn">import</span> <span class="nn">dendropy</span>
<span class="o">&gt;&gt;</span> <span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">treecompare</span>
<span class="o">&gt;&gt;</span> <span class="n">s1</span> <span class="o">=</span> <span class="s2">&quot;(a,(b,(c,d)));&quot;</span>
<span class="o">&gt;&gt;</span> <span class="n">s2</span> <span class="o">=</span> <span class="s2">&quot;(a,(d,(b,c)));&quot;</span>
<span class="o">&gt;&gt;</span> <span class="n">tree1</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">s1</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;newick&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;</span> <span class="n">tree2</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">s2</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;newick&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">treecompare</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">tree1</span><span class="p">,</span> <span class="n">tree2</span><span class="p">))</span>
<span class="n">Traceback</span> <span class="p">(</span><span class="n">most</span> <span class="n">recent</span> <span class="n">call</span> <span class="n">last</span><span class="p">):</span>
    <span class="n">File</span> <span class="s2">&quot;&lt;stdin&gt;&quot;</span><span class="p">,</span> <span class="n">line</span> <span class="mi">1</span><span class="p">,</span> <span class="ow">in</span> <span class="o">&lt;</span><span class="n">module</span><span class="o">&gt;</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">treecompare</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">tree1</span><span class="p">,</span> <span class="n">tree2</span><span class="p">))</span>
    <span class="n">File</span> <span class="s2">&quot;/Users/jeet/Documents/Projects/Phyloinformatics/DendroPy/dendropy/dendropy/calculate/treecompare.py&quot;</span><span class="p">,</span> <span class="n">line</span> <span class="mi">85</span><span class="p">,</span> <span class="ow">in</span> <span class="n">symmetric_difference</span>
        <span class="n">is_bipartitions_updated</span><span class="o">=</span><span class="n">is_bipartitions_updated</span><span class="p">)</span>
    <span class="n">File</span> <span class="s2">&quot;/Users/jeet/Documents/Projects/Phyloinformatics/DendroPy/dendropy/dendropy/calculate/treecompare.py&quot;</span><span class="p">,</span> <span class="n">line</span> <span class="mi">221</span><span class="p">,</span> <span class="ow">in</span> <span class="n">false_positives_and_negatives</span>
        <span class="k">raise</span> <span class="n">error</span><span class="o">.</span><span class="n">TaxonNamespaceIdentityError</span><span class="p">(</span><span class="n">reference_tree</span><span class="p">,</span> <span class="n">comparison_tree</span><span class="p">)</span>
    <span class="n">dendropy</span><span class="o">.</span><span class="n">utility</span><span class="o">.</span><span class="n">error</span><span class="o">.</span><span class="n">TaxonNamespaceIdentityError</span><span class="p">:</span> <span class="n">Non</span><span class="o">-</span><span class="n">identical</span> <span class="n">taxon</span> <span class="n">namespace</span> <span class="n">references</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">TaxonNamespace</span> <span class="nb">object</span> <span class="n">at</span> <span class="mh">0x10052d310</span><span class="o">&gt;</span> <span class="ow">is</span> <span class="ow">not</span> <span class="o">&lt;</span><span class="n">TaxonNamespace</span> <span class="nb">object</span> <span class="n">at</span> <span class="mh">0x101572210</span><span class="o">&gt;</span>
</pre></div>
</div>
<p>Note, too, that results very much depend on the rooting states of the tree:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">treecompare</span>

<span class="n">s1</span> <span class="o">=</span> <span class="s2">&quot;(a,(b,(c,d)));&quot;</span>
<span class="n">s2</span> <span class="o">=</span> <span class="s2">&quot;((a,b),(c,d));&quot;</span>

<span class="n">tns</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">()</span>

<span class="n">unrooted_tree1</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">s1</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;newick&#39;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">tns</span><span class="p">)</span>
<span class="n">unrooted_tree2</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">s2</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;newick&#39;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">tns</span><span class="p">)</span>

<span class="n">rooted_tree1</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">s1</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;newick&#39;</span><span class="p">,</span>
        <span class="n">rooting</span><span class="o">=</span><span class="s2">&quot;force-rooted&quot;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">tns</span><span class="p">)</span>
<span class="n">rooted_tree2</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">s2</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;newick&#39;</span><span class="p">,</span>
        <span class="n">rooting</span><span class="o">=</span><span class="s2">&quot;force-rooted&quot;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">tns</span><span class="p">)</span>

<span class="c1">## Unweighted Robinson-Foulds distance (rooted) = 2</span>
<span class="nb">print</span><span class="p">(</span><span class="n">treecompare</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">rooted_tree1</span><span class="p">,</span> <span class="n">rooted_tree2</span><span class="p">))</span>
<span class="c1">## Unweighted Robinson-Foulds distance (unrooted) = 0</span>
<span class="nb">print</span><span class="p">(</span><span class="n">treecompare</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">unrooted_tree1</span><span class="p">,</span> <span class="n">unrooted_tree2</span><span class="p">))</span>
<span class="c1">## Unweighted Robinson-Foulds distance (rooted1 to unrooted2) = 3</span>
<span class="nb">print</span><span class="p">(</span><span class="n">treecompare</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">rooted_tree1</span><span class="p">,</span> <span class="n">unrooted_tree2</span><span class="p">))</span>
<span class="c1">## Unweighted Robinson-Foulds distance (unrooted1 to rooted2) = 5</span>
<span class="nb">print</span><span class="p">(</span><span class="n">treecompare</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">unrooted_tree1</span><span class="p">,</span> <span class="n">rooted_tree2</span><span class="p">))</span>

</pre></div>
</div>
</section>
<section id="weighted-robinson-foulds-distance">
<h4>Weighted Robinson-Foulds Distance<a class="headerlink" href="#weighted-robinson-foulds-distance" title="Link to this heading">¶</a></h4>
<p>The <em>weighted</em> Robinson-Foulds distance takes edge lengths into account, and is given by the <a class="reference internal" href="../library/treecompare.html#dendropy.calculate.treecompare.weighted_robinson_foulds_distance" title="dendropy.calculate.treecompare.weighted_robinson_foulds_distance"><code class="xref py py-func docutils literal notranslate"><span class="pre">dendropy.calculate.treecompare.weighted_robinson_foulds_distance</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">treecompare</span>

<span class="n">s1</span> <span class="o">=</span> <span class="s2">&quot;((t5:0.161175,t6:0.161175):0.392293,((t4:0.104381,(t2:0.075411,t1:0.075411):0.028969):0.065840,t3:0.170221):0.383247);&quot;</span>
<span class="n">s2</span> <span class="o">=</span> <span class="s2">&quot;((t5:2.161175,t6:0.161175):0.392293,((t4:0.104381,(t2:0.075411,t1:0.075411):1):0.065840,t3:0.170221):0.383247);&quot;</span>

<span class="n">tns</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">()</span>

<span class="n">tree1</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">s1</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;newick&#39;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">tns</span><span class="p">)</span>
<span class="n">tree2</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">s2</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;newick&#39;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">tns</span><span class="p">)</span>

<span class="c1">## Weighted Robinson-Foulds distance = 2.971031</span>
<span class="nb">print</span><span class="p">(</span><span class="n">treecompare</span><span class="o">.</span><span class="n">weighted_robinson_foulds_distance</span><span class="p">(</span><span class="n">tree1</span><span class="p">,</span> <span class="n">tree2</span><span class="p">))</span>

<span class="c1">## Compare to unweighted Robinson-Foulds distance: 0</span>
<span class="nb">print</span><span class="p">(</span><span class="n">treecompare</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">tree1</span><span class="p">,</span> <span class="n">tree2</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="euclidean-distance">
<h4>Euclidean Distance<a class="headerlink" href="#euclidean-distance" title="Link to this heading">¶</a></h4>
<p>The Euclidean distance, like the weighted Robinson-Foulds distance takes edge lengths into account, but squares the edge lengths instead of taking the absolute distance, and is given by the <a class="reference internal" href="../library/treecompare.html#dendropy.calculate.treecompare.euclidean_distance" title="dendropy.calculate.treecompare.euclidean_distance"><code class="xref py py-func docutils literal notranslate"><span class="pre">dendropy.calculate.treecompare.euclidean_distance</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="kn">from</span> <span class="nn">dendropy.calculate</span> <span class="kn">import</span> <span class="n">treecompare</span>

<span class="n">s1</span> <span class="o">=</span> <span class="s2">&quot;((t5:0.161175,t6:0.161175):0.392293,((t4:0.104381,(t2:0.075411,t1:0.075411):0.028969):0.065840,t3:0.170221):0.383247);&quot;</span>
<span class="n">s2</span> <span class="o">=</span> <span class="s2">&quot;((t5:2.161175,t6:0.161175):0.392293,((t4:0.104381,(t2:0.075411,t1:0.075411):1):0.065840,t3:0.170221):0.383247);&quot;</span>

<span class="n">tns</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">()</span>

<span class="n">tree1</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">s1</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;newick&#39;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">tns</span><span class="p">)</span>
<span class="n">tree2</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">s2</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;newick&#39;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">tns</span><span class="p">)</span>

<span class="c1">## Euclidean distance = 2.22326363775</span>
<span class="nb">print</span><span class="p">(</span><span class="n">treecompare</span><span class="o">.</span><span class="n">euclidean_distance</span><span class="p">(</span><span class="n">tree1</span><span class="p">,</span> <span class="n">tree2</span><span class="p">))</span>
</pre></div>
</div>
</section>
</section>
<section id="majority-rule-consensus-tree-from-a-collection-of-trees">
<h3>Majority-Rule Consensus Tree from a Collection of Trees<a class="headerlink" href="#majority-rule-consensus-tree-from-a-collection-of-trees" title="Link to this heading">¶</a></h3>
<p>To get the majority-rule consensus tree of a <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> object, you can call the <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList.consensus" title="dendropy.datamodel.treecollectionmodel.TreeList.consensus"><code class="xref py py-meth docutils literal notranslate"><span class="pre">consensus</span></code></a> instance method.
You can specify the frequency threshold for the consensus tree by the <code class="docutils literal notranslate"><span class="pre">min_freq</span></code> argument, which default to 0.5 (i.e., a 50% majority rule tree).
The following example aggregates the post-burn-in trees from four MCMC samples into a single <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> object, and prints the 95% majority-rule consensus as a Newick string:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="kn">import</span> <span class="nn">dendropy</span>

<span class="n">trees</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TreeList</span><span class="p">()</span>
<span class="k">for</span> <span class="n">tree_file</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;pythonidae.mb.run1.t&#39;</span><span class="p">,</span>
        <span class="s1">&#39;pythonidae.mb.run2.t&#39;</span><span class="p">,</span>
        <span class="s1">&#39;pythonidae.mb.run3.t&#39;</span><span class="p">,</span>
        <span class="s1">&#39;pythonidae.mb.run4.t&#39;</span><span class="p">]:</span>
    <span class="n">trees</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
            <span class="n">path</span><span class="o">=</span><span class="n">tree_file</span><span class="p">,</span>
            <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;nexus&#39;</span><span class="p">,</span>
            <span class="n">tree_offset</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">con_tree</span> <span class="o">=</span> <span class="n">trees</span><span class="o">.</span><span class="n">consensus</span><span class="p">(</span><span class="n">min_freq</span><span class="o">=</span><span class="mf">0.95</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">con_tree</span><span class="o">.</span><span class="n">as_string</span><span class="p">(</span><span class="n">schema</span><span class="o">=</span><span class="s1">&#39;newick&#39;</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="frequency-of-a-split-in-a-collection-of-trees">
<h3>Frequency of a Split in a Collection of Trees<a class="headerlink" href="#frequency-of-a-split-in-a-collection-of-trees" title="Link to this heading">¶</a></h3>
<p>The <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList.frequency_of_split" title="dendropy.datamodel.treecollectionmodel.TreeList.frequency_of_split"><code class="xref py py-meth docutils literal notranslate"><span class="pre">frequency_of_split</span></code></a> method of a <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> object returns the frequency of occurrence of a single split across all the <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects in the <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>.
The split can be specified by passing a split bitmask directly using the <code class="docutils literal notranslate"><span class="pre">split_bitmask</span></code> keyword argument, as a list of <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects using the <code class="docutils literal notranslate"><span class="pre">taxa</span></code> keyword argument, or as a list of taxon labels using the <code class="docutils literal notranslate"><span class="pre">labels</span></code> keyword argument.
The following example shows how to calculate the frequency of a split defined by two taxa, “Morelia amethistina” and “Morelia tracyae”, from the post-burn-in trees aggregated across four MCMC samples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="kn">import</span> <span class="nn">dendropy</span>

<span class="n">trees</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TreeList</span><span class="p">()</span>
<span class="k">for</span> <span class="n">tree_file</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;pythonidae.mb.run1.t&#39;</span><span class="p">,</span>
        <span class="s1">&#39;pythonidae.mb.run2.t&#39;</span><span class="p">,</span>
        <span class="s1">&#39;pythonidae.mb.run3.t&#39;</span><span class="p">,</span>
        <span class="s1">&#39;pythonidae.mb.run4.t&#39;</span><span class="p">]:</span>
    <span class="n">trees</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
            <span class="n">path</span><span class="o">=</span><span class="n">tree_file</span><span class="p">,</span>
            <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;nexus&#39;</span><span class="p">,</span>
            <span class="n">tree_offset</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">split_leaves</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Morelia amethistina&#39;</span><span class="p">,</span> <span class="s1">&#39;Morelia tracyae&#39;</span><span class="p">]</span>
<span class="n">f</span> <span class="o">=</span> <span class="n">trees</span><span class="o">.</span><span class="n">frequency_of_bipartition</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="n">split_leaves</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Frequency of split </span><span class="si">%s</span><span class="s1">: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">split_leaves</span><span class="p">,</span> <span class="n">f</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="the-maximum-clade-credibility-tree-the-tree-that-maximizes-the-product-of-split-support">
<h3>The Maximum Clade Credibility Tree: The Tree that Maximizes the Product of Split Support<a class="headerlink" href="#the-maximum-clade-credibility-tree-the-tree-that-maximizes-the-product-of-split-support" title="Link to this heading">¶</a></h3>
<p>The Maximum Clade Credibility Tree (MCCT) is one that maximize the <em>product</em> of split support, and is returned for a collection of trees managed in a <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> instance by the <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList.maximum_product_of_split_support_tree" title="dendropy.datamodel.treecollectionmodel.TreeList.maximum_product_of_split_support_tree"><code class="xref py py-meth docutils literal notranslate"><span class="pre">maximum_product_of_split_support_tree</span></code></a> method:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="kn">import</span> <span class="nn">dendropy</span>

<span class="n">trees</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TreeList</span><span class="p">()</span>
<span class="k">for</span> <span class="n">tree_file</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;pythonidae.mb.run1.t&#39;</span><span class="p">,</span>
        <span class="s1">&#39;pythonidae.mb.run2.t&#39;</span><span class="p">,</span>
        <span class="s1">&#39;pythonidae.mb.run3.t&#39;</span><span class="p">,</span>
        <span class="s1">&#39;pythonidae.mb.run4.t&#39;</span><span class="p">]:</span>
    <span class="n">trees</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
            <span class="n">path</span><span class="o">=</span><span class="n">tree_file</span><span class="p">,</span>
            <span class="n">schema</span><span class="o">=</span><span class="s1">&#39;nexus&#39;</span><span class="p">,</span>
            <span class="n">tree_offset</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

<span class="n">mcct</span> <span class="o">=</span> <span class="n">trees</span><span class="o">.</span><span class="n">maximum_product_of_split_support_tree</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Tree </span><span class="si">{}</span><span class="s2"> maximizes the product of split support (log product = </span><span class="si">{}</span><span class="s2">): </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">trees</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">mcct</span><span class="p">),</span> <span class="n">mcct</span><span class="o">.</span><span class="n">log_product_of_split_support</span><span class="p">,</span> <span class="n">mcct</span><span class="p">))</span>

<span class="n">msst</span> <span class="o">=</span> <span class="n">trees</span><span class="o">.</span><span class="n">maximum_sum_of_split_support_tree</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Tree </span><span class="si">{}</span><span class="s2"> maximizes the sum of split support (sum = </span><span class="si">{}</span><span class="s2">): </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">trees</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">msst</span><span class="p">),</span> <span class="n">msst</span><span class="o">.</span><span class="n">sum_of_split_support</span><span class="p">,</span> <span class="n">msst</span><span class="p">))</span>
</pre></div>
</div>
<p>Unfortunately, terminology in usage and literature regarding this type of summary is <em>very</em> confusing, and sometimes the term “MCCT” is used to refer to the tree that maximizes the <em>sum</em> of split support and “MCT” to the tree that maximizes the product of split support.
If the tree that maximizes the <em>sum</em> of split support is the criteria required, then the <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList.maximum_sum_of_split_support_tree" title="dendropy.datamodel.treecollectionmodel.TreeList.maximum_sum_of_split_support_tree"><code class="xref py py-meth docutils literal notranslate"><span class="pre">maximum_sum_of_split_support_tree</span></code></a> method of the <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a> object should be used.</p>
</section>
</section>
<section id="scoring-trees-under-the-coalescent">
<h2>Scoring Trees Under the Coalescent<a class="headerlink" href="#scoring-trees-under-the-coalescent" title="Link to this heading">¶</a></h2>
<section id="probability-under-the-coalescent-model">
<h3>Probability Under the Coalescent Model<a class="headerlink" href="#probability-under-the-coalescent-model" title="Link to this heading">¶</a></h3>
<p>The <a class="reference internal" href="../library/coalescent.html#module-dendropy.model.coalescent" title="dendropy.model.coalescent"><code class="xref py py-mod docutils literal notranslate"><span class="pre">coalescent</span></code></a> module provides a range of methods for simulations and calculations under Kingman’s coalescent framework and related models. For example:</p>
<blockquote>
<div><dl class="simple">
<dt><a class="reference internal" href="../library/coalescent.html#dendropy.model.coalescent.log_probability_of_coalescent_tree" title="dendropy.model.coalescent.log_probability_of_coalescent_tree"><code class="xref py py-func docutils literal notranslate"><span class="pre">log_probability_of_coalescent_tree</span></code></a></dt><dd><p>Given a <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> object as the first argument, and the haploid population size as the second, returns the log probability of the <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> under the neutral coalescent.</p>
</dd>
</dl>
</div></blockquote>
</section>
<section id="numbers-of-deep-coalescences">
<h3>Numbers of Deep Coalescences<a class="headerlink" href="#numbers-of-deep-coalescences" title="Link to this heading">¶</a></h3>
<blockquote>
<div><dl class="simple">
<dt><a class="reference internal" href="../library/reconcile.html#dendropy.model.reconcile.reconciliation_discordance" title="dendropy.model.reconcile.reconciliation_discordance"><code class="xref py py-func docutils literal notranslate"><span class="pre">reconciliation_discordance</span></code></a></dt><dd><p>Given two <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects <em>sharing the same leaf-set</em>, this returns the number of deep coalescences resulting from fitting the first tree (e.g., a gene tree) to the second (e.g., a species tree). This is based on the algorithm described <a class="reference external" href="bioinformatics.oxfordjournals.org/cgi/reprint/14/9/819.pdf">Goodman, et al.</a> (Goodman, et al., 1979. Fitting the gene lineage into its species lineage,a parsimony strategy illustrated by cladograms constructed from globin sequences. Syst. Zool. 19: 99-113).</p>
</dd>
<dt><a class="reference internal" href="../library/reconcile.html#dendropy.model.reconcile.monophyletic_partition_discordance" title="dendropy.model.reconcile.monophyletic_partition_discordance"><code class="xref py py-func docutils literal notranslate"><span class="pre">monophyletic_partition_discordance</span></code></a></dt><dd><p>Given a <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> object as the first argument, and a list of lists of
<a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects representing the expected monophyletic partitioning of the <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> of the <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> as the second argument, this returns the number of deep coalescences found in the relationships implied by the <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> object, conditional on the taxon groupings given by the second argument. This statistic corresponds to the Slatkin and Maddison (1989) <strong>s</strong> statistic, as described <a class="reference external" href="http://mesquiteproject.org/Mesquite_Folder/docs/mesquite/popGen/popGen.html#s">here</a>.</p>
</dd>
</dl>
</div></blockquote>
</section>
<section id="number-of-deep-coalescences-when-embedding-one-tree-in-another-e-g-gene-species-trees">
<h3>Number of Deep Coalescences when Embedding One Tree in Another (e.g. Gene/Species Trees)<a class="headerlink" href="#number-of-deep-coalescences-when-embedding-one-tree-in-another-e-g-gene-species-trees" title="Link to this heading">¶</a></h3>
<p>Imagine we wanted to generate the distribution of the number of deep coalescences under two scenarios: one in which a population underwent sequential or step-wise vicariance, and another when there was simultaneous fragmentation.
In this case, the containing tree and the embedded trees have different leaf sets, and there is a many-to-one mapping of embedded tree taxa to containing tree taxa.</p>
<p>The <a class="reference internal" href="../library/reconcile.html#dendropy.model.reconcile.ContainingTree" title="dendropy.model.reconcile.ContainingTree"><code class="xref py py-class docutils literal notranslate"><span class="pre">ContainingTree</span></code></a> class is designed to allow for counting deep coalescences in cases like this.
It requires a <code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespaceMapping</span></code> object, which provides an association between the embedded taxa and the containing taxa.
The easiest way to get a <code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespaceMapping</span></code> object is to call the special factory function <code class="xref py py-meth docutils literal notranslate"><span class="pre">create_contained_taxon_mapping</span></code>.
This will create a new <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> to manage the gene taxa, and create the associations between the gene taxa and the containing tree taxa for you.
It takes two arguments: the <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> of the containing tree, and the number of genes you want sampled from each species.
If the gene-species associations are more complex, e.g., different numbers of genes per species, we can pass in a list of values as the second argument to <code class="xref py py-meth docutils literal notranslate"><span class="pre">create_contained_taxon_mapping</span></code>.
This approach should be used with caution if we cannot be certain of the order of taxa (as is the case with data read in Newick formats). In these case, and in more complex cases, we might need to directly instantiate the <code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespaceMapping</span></code> object. The API to describe the associations when constructing this object is very similar to that of the <code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespacePartition</span></code> object: you can use a function, attribute or dictionary.</p>
<p>The <a class="reference internal" href="../library/reconcile.html#dendropy.model.reconcile.ContainingTree" title="dendropy.model.reconcile.ContainingTree"><code class="xref py py-class docutils literal notranslate"><span class="pre">ContainingTree</span></code></a> class has its own native contained coalescent simulator, <a class="reference internal" href="../library/reconcile.html#dendropy.model.reconcile.ContainingTree.embed_contained_kingman" title="dendropy.model.reconcile.ContainingTree.embed_contained_kingman"><code class="xref py py-meth docutils literal notranslate"><span class="pre">embed_contained_kingman</span></code></a>, which simulates <em>and</em> embeds a contained coalescent tree at the same time.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="kn">import</span> <span class="nn">dendropy</span>
<span class="kn">from</span> <span class="nn">dendropy.simulate</span> <span class="kn">import</span> <span class="n">treesim</span>
<span class="kn">from</span> <span class="nn">dendropy.model</span> <span class="kn">import</span> <span class="n">reconcile</span>

<span class="c1"># simulation parameters and output</span>
<span class="n">num_reps</span> <span class="o">=</span> <span class="mi">10</span>

<span class="c1"># population tree descriptions</span>
<span class="n">stepwise_tree_str</span> <span class="o">=</span> <span class="s2">&quot;[&amp;R](A:120000,(B:80000,(C:40000,D:40000):40000):40000):100000;&quot;</span>
<span class="n">frag_tree_str</span> <span class="o">=</span> <span class="s2">&quot;[&amp;R](A:120000,B:120000,C:120000,D:120000):100000;&quot;</span>

<span class="c1"># taxa and trees</span>
<span class="n">containing_taxa</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">()</span>
<span class="n">stepwise_tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">stepwise_tree_str</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">containing_taxa</span><span class="p">)</span>
<span class="n">frag_tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">frag_tree_str</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">containing_taxa</span><span class="p">)</span>

<span class="c1"># taxon set association</span>
<span class="n">genes_to_species</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespaceMapping</span><span class="o">.</span><span class="n">create_contained_taxon_mapping</span><span class="p">(</span>
        <span class="n">containing_taxon_namespace</span><span class="o">=</span><span class="n">containing_taxa</span><span class="p">,</span>
        <span class="n">num_contained</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>

<span class="c1"># convert to containing tree</span>
<span class="n">stepwise_tree</span> <span class="o">=</span> <span class="n">reconcile</span><span class="o">.</span><span class="n">ContainingTree</span><span class="p">(</span><span class="n">stepwise_tree</span><span class="p">,</span>
            <span class="n">contained_taxon_namespace</span><span class="o">=</span><span class="n">genes_to_species</span><span class="o">.</span><span class="n">domain_taxon_namespace</span><span class="p">,</span>
            <span class="n">contained_to_containing_taxon_map</span><span class="o">=</span><span class="n">genes_to_species</span><span class="p">)</span>
<span class="n">frag_tree</span> <span class="o">=</span> <span class="n">reconcile</span><span class="o">.</span><span class="n">ContainingTree</span><span class="p">(</span><span class="n">frag_tree</span><span class="p">,</span>
            <span class="n">contained_taxon_namespace</span><span class="o">=</span><span class="n">genes_to_species</span><span class="o">.</span><span class="n">domain_taxon_namespace</span><span class="p">,</span>
            <span class="n">contained_to_containing_taxon_map</span><span class="o">=</span><span class="n">genes_to_species</span><span class="p">)</span>

<span class="c1"># for each rep</span>
<span class="k">for</span> <span class="n">rep</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_reps</span><span class="p">):</span>
    <span class="n">stepwise_tree</span><span class="o">.</span><span class="n">embed_contained_kingman</span><span class="p">(</span><span class="n">default_pop_size</span><span class="o">=</span><span class="mi">40000</span><span class="p">)</span>
    <span class="n">frag_tree</span><span class="o">.</span><span class="n">embed_contained_kingman</span><span class="p">(</span><span class="n">default_pop_size</span><span class="o">=</span><span class="mi">40000</span><span class="p">)</span>

<span class="c1"># write results</span>

<span class="c1"># returns dictionary with contained trees as keys</span>
<span class="c1"># and number of deep coalescences as values</span>
<span class="n">stepwise_deep_coals</span> <span class="o">=</span> <span class="n">stepwise_tree</span><span class="o">.</span><span class="n">deep_coalescences</span><span class="p">()</span>
<span class="n">stepwise_out</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;stepwise.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">stepwise_deep_coals</span><span class="p">:</span>
    <span class="n">stepwise_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%d</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">stepwise_deep_coals</span><span class="p">[</span><span class="n">tree</span><span class="p">])</span>

<span class="c1"># returns dictionary with contained trees as keys</span>
<span class="c1"># and number of deep coalescences as values</span>
<span class="n">frag_deep_coals</span> <span class="o">=</span> <span class="n">frag_tree</span><span class="o">.</span><span class="n">deep_coalescences</span><span class="p">()</span>
<span class="n">frag_out</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;frag.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">frag_deep_coals</span><span class="p">:</span>
    <span class="n">frag_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%d</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">frag_deep_coals</span><span class="p">[</span><span class="n">tree</span><span class="p">])</span>








</pre></div>
</div>
<p>If you have used some other method to simulate your trees, you can use <a class="reference internal" href="../library/reconcile.html#dendropy.model.reconcile.ContainingTree.embed_tree" title="dendropy.model.reconcile.ContainingTree.embed_tree"><code class="xref py py-meth docutils literal notranslate"><span class="pre">embed_tree</span></code></a> to embed the trees and count then number of deep coalescences.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="kn">import</span> <span class="nn">dendropy</span>
<span class="kn">from</span> <span class="nn">dendropy.simulate</span> <span class="kn">import</span> <span class="n">treesim</span>
<span class="kn">from</span> <span class="nn">dendropy.model</span> <span class="kn">import</span> <span class="n">reconcile</span>

<span class="c1"># simulation parameters and output</span>
<span class="n">num_reps</span> <span class="o">=</span> <span class="mi">10</span>

<span class="c1"># population tree descriptions</span>
<span class="n">stepwise_tree_str</span> <span class="o">=</span> <span class="s2">&quot;[&amp;R](A:120000,(B:80000,(C:40000,D:40000):40000):40000):100000;&quot;</span>
<span class="n">frag_tree_str</span> <span class="o">=</span> <span class="s2">&quot;[&amp;R](A:120000,B:120000,C:120000,D:120000):100000;&quot;</span>

<span class="c1"># taxa and trees</span>
<span class="n">containing_taxa</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespace</span><span class="p">()</span>
<span class="n">stepwise_tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">stepwise_tree_str</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">containing_taxa</span><span class="p">)</span>
<span class="n">frag_tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">frag_tree_str</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">,</span>
        <span class="n">taxon_namespace</span><span class="o">=</span><span class="n">containing_taxa</span><span class="p">)</span>

<span class="c1"># taxon set association</span>
<span class="n">genes_to_species</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">TaxonNamespaceMapping</span><span class="o">.</span><span class="n">create_contained_taxon_mapping</span><span class="p">(</span>
        <span class="n">containing_taxon_namespace</span><span class="o">=</span><span class="n">containing_taxa</span><span class="p">,</span>
        <span class="n">num_contained</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>

<span class="c1"># convert to containing tree</span>
<span class="n">stepwise_tree</span> <span class="o">=</span> <span class="n">reconcile</span><span class="o">.</span><span class="n">ContainingTree</span><span class="p">(</span><span class="n">stepwise_tree</span><span class="p">,</span>
            <span class="n">contained_taxon_namespace</span><span class="o">=</span><span class="n">genes_to_species</span><span class="o">.</span><span class="n">domain_taxon_namespace</span><span class="p">,</span>
            <span class="n">contained_to_containing_taxon_map</span><span class="o">=</span><span class="n">genes_to_species</span><span class="p">)</span>
<span class="n">frag_tree</span> <span class="o">=</span> <span class="n">reconcile</span><span class="o">.</span><span class="n">ContainingTree</span><span class="p">(</span><span class="n">frag_tree</span><span class="p">,</span>
            <span class="n">contained_taxon_namespace</span><span class="o">=</span><span class="n">genes_to_species</span><span class="o">.</span><span class="n">domain_taxon_namespace</span><span class="p">,</span>
            <span class="n">contained_to_containing_taxon_map</span><span class="o">=</span><span class="n">genes_to_species</span><span class="p">)</span>

<span class="c1"># for each rep</span>
<span class="k">for</span> <span class="n">rep</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_reps</span><span class="p">):</span>
    <span class="n">gene_tree1</span> <span class="o">=</span> <span class="n">treesim</span><span class="o">.</span><span class="n">contained_coalescent_tree</span><span class="p">(</span><span class="n">containing_tree</span><span class="o">=</span><span class="n">stepwise_tree</span><span class="p">,</span>
        <span class="n">gene_to_containing_taxon_map</span><span class="o">=</span><span class="n">genes_to_species</span><span class="p">,</span>
        <span class="n">default_pop_size</span><span class="o">=</span><span class="mi">40000</span><span class="p">)</span>
    <span class="n">stepwise_tree</span><span class="o">.</span><span class="n">embed_tree</span><span class="p">(</span><span class="n">gene_tree1</span><span class="p">)</span>
    <span class="n">gene_tree2</span> <span class="o">=</span> <span class="n">treesim</span><span class="o">.</span><span class="n">contained_coalescent_tree</span><span class="p">(</span><span class="n">containing_tree</span><span class="o">=</span><span class="n">frag_tree</span><span class="p">,</span>
        <span class="n">gene_to_containing_taxon_map</span><span class="o">=</span><span class="n">genes_to_species</span><span class="p">,</span>
        <span class="n">default_pop_size</span><span class="o">=</span><span class="mi">40000</span><span class="p">)</span>
    <span class="n">frag_tree</span><span class="o">.</span><span class="n">embed_tree</span><span class="p">(</span><span class="n">gene_tree2</span><span class="p">)</span>

<span class="c1"># write results</span>

<span class="c1"># returns dictionary with contained trees as keys</span>
<span class="c1"># and number of deep coalescences as values</span>
<span class="n">stepwise_deep_coals</span> <span class="o">=</span> <span class="n">stepwise_tree</span><span class="o">.</span><span class="n">deep_coalescences</span><span class="p">()</span>
<span class="n">stepwise_out</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;stepwise.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">stepwise_deep_coals</span><span class="p">:</span>
    <span class="n">stepwise_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%d</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">stepwise_deep_coals</span><span class="p">[</span><span class="n">tree</span><span class="p">])</span>

<span class="c1"># returns dictionary with contained trees as keys</span>
<span class="c1"># and number of deep coalescences as values</span>
<span class="n">frag_deep_coals</span> <span class="o">=</span> <span class="n">frag_tree</span><span class="o">.</span><span class="n">deep_coalescences</span><span class="p">()</span>
<span class="n">frag_out</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;frag.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">frag_deep_coals</span><span class="p">:</span>
    <span class="n">frag_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%d</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">frag_deep_coals</span><span class="p">[</span><span class="n">tree</span><span class="p">])</span>








</pre></div>
</div>
<p>For more details on simulating contained coalescent trees and counting numbers of deep coalescences on them, see “<a class="reference internal" href="treesims.html#simulating-contained-coalescent-trees"><span class="std std-ref">Multispecies Coalescent (“Contained Coalescent” or “Censored Coalescent”) Trees</span></a>” or “<a class="reference internal" href="treesims.html#simulating-and-counting-deep-coalescences"><span class="std std-ref">Simulating the Distribution of Number Deep Coalescences Under Different Phylogeographic History Scenarios</span></a>”.</p>
</section>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div style="text-align: center; padding-top: 20px; padding-bottom: 5px; width: 100%;">
    <a href="../index.html"><img src="../_static/dendropy_logo.png" /></a>
</div><div style="clear:both; width: 100%; height:1px;"></div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h3><a href="../index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Tree Statistics, Metrics, Summarizations, and Other Calculations</a><ul>
<li><a class="reference internal" href="#native-tree-statistic-and-metric-methods">Native Tree Statistic and Metric Methods</a><ul>
<li><a class="reference internal" href="#tree-length">Tree Length</a></li>
<li><a class="reference internal" href="#node-ages">Node Ages</a></li>
<li><a class="reference internal" href="#number-of-lineages-at-a-particular-time-and-lineage-through-time-plots">Number of Lineages at a Particular Time and Lineage Through Time Plots</a></li>
</ul>
</li>
<li><a class="reference internal" href="#unary-tree-statistics-and-metrics">Unary Tree Statistics and Metrics</a><ul>
<li><a class="reference internal" href="#pybus-harvey-gamma">Pybus-Harvey Gamma</a></li>
<li><a class="reference internal" href="#patristic-distances">Patristic Distances</a></li>
</ul>
</li>
<li><a class="reference internal" href="#comparing-and-summarizing-trees">Comparing and Summarizing Trees</a><ul>
<li><a class="reference internal" href="#distances-between-trees">Distances Between Trees</a><ul>
<li><a class="reference internal" href="#unweighted-robinson-foulds-distance">Unweighted Robinson-Foulds Distance</a></li>
<li><a class="reference internal" href="#weighted-robinson-foulds-distance">Weighted Robinson-Foulds Distance</a></li>
<li><a class="reference internal" href="#euclidean-distance">Euclidean Distance</a></li>
</ul>
</li>
<li><a class="reference internal" href="#majority-rule-consensus-tree-from-a-collection-of-trees">Majority-Rule Consensus Tree from a Collection of Trees</a></li>
<li><a class="reference internal" href="#frequency-of-a-split-in-a-collection-of-trees">Frequency of a Split in a Collection of Trees</a></li>
<li><a class="reference internal" href="#the-maximum-clade-credibility-tree-the-tree-that-maximizes-the-product-of-split-support">The Maximum Clade Credibility Tree: The Tree that Maximizes the Product of Split Support</a></li>
</ul>
</li>
<li><a class="reference internal" href="#scoring-trees-under-the-coalescent">Scoring Trees Under the Coalescent</a><ul>
<li><a class="reference internal" href="#probability-under-the-coalescent-model">Probability Under the Coalescent Model</a></li>
<li><a class="reference internal" href="#numbers-of-deep-coalescences">Numbers of Deep Coalescences</a></li>
<li><a class="reference internal" href="#number-of-deep-coalescences-when-embedding-one-tree-in-another-e-g-gene-species-trees">Number of Deep Coalescences when Embedding One Tree in Another (e.g. Gene/Species Trees)</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  </div><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h4>Previous topic</h4>
    <p class="topless"><a href="treecollections.html"
                          title="previous chapter">Collections of Trees</a></p>
  </div>
  <div>
    <h4>Next topic</h4>
    <p class="topless"><a href="treemanips.html"
                          title="next chapter">Tree Manipulation and Restructuring</a></p>
  </div><div style="clear:both; width: 100%; height:1px;"></div>
    <!-- Documentation -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Documentation</h3>
        <ul>
            <li><a href="downloading.html">Downloading and Installing DendroPy</a></li>
            <li><a href="index.html">The DendroPy Primer</a></li>
            <li><a href="../library/index.html">Library API Reference</a></li>
            <li>
                <a href="../schemas/index.html">Schemas</a>
                <ul>
                    <li><a href="../schemas/fasta.html">FASTA</a></li>
                    <li><a href="../schemas/newick.html">Newick</a></li>
                    <li><a href="../schemas/nexml.html">NeXML</a></li>
                    <li><a href="../schemas/nexus.html">Nexus</a></li>
                    <li><a href="../schemas/phylip.html">PHYLIP</a></li>
                </ul>
            </li>
            <li>
                <a href="../programs/index.html">Programs</a>
                <ul>
                    <li><a href="../programs/sumtrees.html">SumTrees</a></li>
                </ul>
            </li>
            <li><a href="../glossary.html">Glossary and Terminological Reference</a></li>
            <li><a href="../developer.html">Developer Guide</a></li>
            <li><a href="../planning.html">Ongoing Development</a></li>
            <li><a href="../changes.html">Change History</a></li>
        </ul>
    </div>

    <!-- Downloads -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Obtaining</h3>
        <ul>
            <li><a target="_blank" href="http://pypi.python.org/pypi/DendroPy">Install from the Python Package Index</a></li>
            <li><a target="_blank" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.5.tar.gz">Download the Source Code Archive</a></li>
            <li><a target="_blank" href="http://github.com/jeetsukumaran/DendroPy">Clone the Source Code Repository</a></li>
        </ul>
    </div>

    <!-- Discussions -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Discussion</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot; DendroPy Users&quot; </a>group to follow and participate in discussion, troubleshooting, help, information, suggestions, etc. on the usage and development of the DendroPy phylogenetic computing library.</p>
            <form action="http://groups.google.com/group/dendropy-users/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot;dendropy-users&quot;</a> group, or click <a href="http://groups.google.com/group/dendropy-users?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Announcements -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Announcements</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; DendroPy Announcements&quot; </a>group to receive announcements of new releases, updates, changes and other news of interest to DendroPy users and developers.</p>
            <form action="http://groups.google.com/group/dendropy-announce/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; dendropy-announce&quot; </a>group, or click <a href="http://groups.google.com/group/dendropy-announce?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Development -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative; padding-bottom: 15px; margin-bottom:5px;">
        <h3><span style="text-align: left"><a href="https://github.com/jeetsukumaran/DendroPy/">Development</a></span><a href="https://github.com/jeetsukumaran/DendroPy/"><span style="position: absolute; right: 0; top: 10px "><img src="../_static/Octocat.png" height="30px" alt="GitHub" /></span></a></h3>
        <div style="margin-top: 15px;">
            <!-- <a href="https://github.com/jeetsukumaran/DendroPy/issues">Issues</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/subscription">Watch</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/fork">Fork</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/stargazers">Star</a> &bull; <a href="https://github.com/jeetsukumaran/">Follow</a> -->
            <ul>
                <li>                <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/issues"><code>Issues</code></a></span> <span style="font-style: italic; font-size:80%;"> - Report bugs or request features</span></li>
                <li>     <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/subscription"><code>&nbsp;Watch</code></a></span> <span style="font-style: italic; font-size:80%;"> - Follow development activity</span></li>
                <li>        <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/fork"><code>&nbsp;&nbsp;Fork</code></a></span> <span style="font-style: italic; font-size:80%;"> - Contribute and collaborate</span></li>
                <li>  <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/stargazers"><code>&nbsp;&nbsp;Star</code></a></span> <span style="font-style: italic; font-size:80%;"> - Throw some glitter, add some glamour</span></li>
        </div>
    </div>


        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="treemanips.html" title="Tree Manipulation and Restructuring"
             >next</a> |</li>
        <li class="right" >
          <a href="treecollections.html" title="Collections of Trees"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.5 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >The DendroPy Primer</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Tree Statistics, Metrics, Summarizations, and Other Calculations</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2009-2025, Jeet Sukumaran and Mark T. Holder.
    </div>
  </body>
</html>