<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Working with Metadata Annotations &#8212; DendroPy 5.0.4 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/rtd.css?v=cedc6a2f" />
    
    <script src="../_static/documentation_options.js?v=3205dc05"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Working with GenBank Molecular Sequence Databases" href="genbank.html" />
    <link rel="prev" title="Data Sets" href="datasets.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="genbank.html" title="Working with GenBank Molecular Sequence Databases"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="datasets.html" title="Data Sets"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.4 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">The DendroPy Primer</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Working with Metadata Annotations</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="working-with-metadata-annotations">
<h1>Working with Metadata Annotations<a class="headerlink" href="#working-with-metadata-annotations" title="Link to this heading">¶</a></h1>
<p>DendroPy provides a rich infrastructure for decorating most types of phylogenetic objects (e.g., the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>, <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>, <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>, <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>, and various <a class="reference internal" href="../library/charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a> classes) with metadata information.
These phylogenetic objects have an attribute, <code class="xref py py-attr docutils literal notranslate"><span class="pre">annotations</span></code>, that is an instance of the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a> class, which is an iterable (derived from <code class="xref py py-class docutils literal notranslate"><span class="pre">dendropy.utility.containers.OrderedSet</span></code>) that serves to manage a collection of <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects.
Each <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> object tracks a single annotation element.
These annotations will be rendered as <code class="docutils literal notranslate"><span class="pre">meta</span></code> elements when writing to NeXML format or ampersand-prepended comemnt strings when writing to NEXUS/NEWICK format.
Note that full and robust expression of metadata annotations, including stable and consistent round-tripping of information, can only be achieved while in the NeXML format.</p>
<section id="overview-of-the-infrastructure-for-metadata-annotation-in-dendropy">
<h2>Overview of the Infrastructure for Metadata Annotation in DendroPy<a class="headerlink" href="#overview-of-the-infrastructure-for-metadata-annotation-in-dendropy" title="Link to this heading">¶</a></h2>
<p>Each item of metadata is maintained in an object of the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> class.
This class has the following attributes:</p>
<blockquote>
<div><dl class="simple">
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">name</span></code></dt><dd><p>The name of the metadata item or annotation.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">value</span></code></dt><dd><p>The value or content of the metadata item or annotation.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">datatype_hint</span></code></dt><dd><p>Custom data type indication for NeXML output (e.g. “xsd:string”).</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">name_prefix</span></code></dt><dd><p>Prefix that represents an abbreviation of the namespace associated with
this metadata item.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">namespace</span></code></dt><dd><p>The namespace (e.g. “<a class="reference external" href="http://www.w3.org/XML/1998/namespace">http://www.w3.org/XML/1998/namespace</a>”) of this
metadata item (NeXML output).</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">annotate_as_reference</span></code></dt><dd><p>If <code class="docutils literal notranslate"><span class="pre">True</span></code>, indicates that this annotation should not be interpreted semantically as a literal value, but rather as a source to be dereferenced.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">is_hidden</span></code></dt><dd><p>If <code class="docutils literal notranslate"><span class="pre">True</span></code>, indicates that this annotation should not be printed or written out.</p>
</dd>
<dt><code class="xref py py-attr docutils literal notranslate"><span class="pre">prefixed_name</span></code></dt><dd><p>Returns the name of this annotation with its namespace prefix (e.g. “dc:subject”).</p>
</dd>
</dl>
</div></blockquote>
<p>These <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects are typically collected and managed in a “annotations manager” container class, <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a>.
This is a specialization of <code class="xref py py-class docutils literal notranslate"><span class="pre">dendropy.utility.containers.OrderedSet</span></code> whose elements are instances of <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a>.
The full set of annotations associated with each object of <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a>, <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a>, <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> <a class="reference internal" href="../library/treecollectionmodel.html#dendropy.datamodel.treecollectionmodel.TreeList" title="dendropy.datamodel.treecollectionmodel.TreeList"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeList</span></code></a>, <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>, various <a class="reference internal" href="../library/charmatrixmodel.html#dendropy.datamodel.charmatrixmodel.CharacterMatrix" title="dendropy.datamodel.charmatrixmodel.CharacterMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">CharacterMatrix</span></code></a> and other phylogenetic data class types is available through the <code class="xref py py-attr docutils literal notranslate"><span class="pre">annotations</span></code> attribute of those objects, which is an instance of <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a>.
The <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a> includes the following additional methods to support the creation, access, and management of the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> object elements contained within it:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.add_new" title="dendropy.datamodel.basemodel.AnnotationSet.add_new"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_new</span></code></a></p></li>
<li><p><a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.add_bound_attribute" title="dendropy.datamodel.basemodel.AnnotationSet.add_bound_attribute"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_bound_attribute</span></code></a></p></li>
<li><p><a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.add_citation" title="dendropy.datamodel.basemodel.AnnotationSet.add_citation"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_citation</span></code></a></p></li>
<li><p><a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.findall" title="dendropy.datamodel.basemodel.AnnotationSet.findall"><code class="xref py py-meth docutils literal notranslate"><span class="pre">findall</span></code></a></p></li>
<li><p><a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.find" title="dendropy.datamodel.basemodel.AnnotationSet.find"><code class="xref py py-meth docutils literal notranslate"><span class="pre">find</span></code></a></p></li>
<li><p><a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.drop" title="dendropy.datamodel.basemodel.AnnotationSet.drop"><code class="xref py py-meth docutils literal notranslate"><span class="pre">drop</span></code></a></p></li>
<li><p><a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.values_as_dict" title="dendropy.datamodel.basemodel.AnnotationSet.values_as_dict"><code class="xref py py-meth docutils literal notranslate"><span class="pre">values_as_dict</span></code></a></p></li>
</ul>
</div></blockquote>
<p>The following code snippet reads in a data file in NeXML format, and dumps out the annotations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>

<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span> <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="nb">print</span> <span class="s2">&quot;-- (dataset) ---</span><span class="se">\n</span><span class="s2">&quot;</span>
<span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
    <span class="nb">print</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> = &#39;</span><span class="si">%s</span><span class="s2">&#39;&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">a</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
<span class="k">for</span> <span class="n">tree_list</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">tree_lists</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">tree_list</span><span class="p">:</span>
        <span class="nb">print</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">-- (tree &#39;</span><span class="si">%s</span><span class="s2">&#39;) --</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">tree</span><span class="o">.</span><span class="n">label</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
            <span class="nb">print</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> = &#39;</span><span class="si">%s</span><span class="s2">&#39;&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">a</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
</pre></div>
</div>
<p>Running the above results in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">--</span> <span class="p">(</span><span class="n">dataset</span><span class="p">)</span> <span class="o">---</span>

<span class="n">bibliographicCitation</span> <span class="o">=</span> <span class="s1">&#39;Wiklund H., Altamira I.V., Glover A., Smith C., Baco A., &amp; Dahlgren T.G. 2012. Systematics and biodiversity of Ophryotrocha (Annelida, Dorvilleidae) with descriptions of six new species from deep-sea whale-fall and wood-fall habitats in the north-east Pacific. Systematics and Biodiversity, .&#39;</span>
<span class="n">subject</span> <span class="o">=</span> <span class="s1">&#39;whale-fall&#39;</span>
<span class="n">changeNote</span> <span class="o">=</span> <span class="s1">&#39;Generated on Wed Jun 06 11:02:45 EDT 2012&#39;</span>
<span class="n">subject</span> <span class="o">=</span> <span class="s1">&#39;wood-fall&#39;</span>
<span class="n">title</span> <span class="o">=</span> <span class="s1">&#39;Systematics and biodiversity of Ophryotrocha (Annelida, Dorvilleidae) with descriptions of six new species from deep-sea whale-fall and wood-fall habitats in the north-east Pacific&#39;</span>
<span class="n">publicationName</span> <span class="o">=</span> <span class="s1">&#39;Systematics and Biodiversity&#39;</span>
<span class="n">creator</span> <span class="o">=</span> <span class="s1">&#39;Wiklund H., Altamira I.V., Glover A., Smith C., Baco A., &amp; Dahlgren T.G.&#39;</span>
<span class="n">publisher</span> <span class="o">=</span> <span class="s1">&#39;Systematics and Biodiversity&#39;</span>
<span class="n">contributor</span> <span class="o">=</span> <span class="s1">&#39;Wiklund H.&#39;</span>
<span class="n">volume</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="n">contributor</span> <span class="o">=</span> <span class="s1">&#39;Altamira I.V.&#39;</span>
<span class="n">number</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="n">contributor</span> <span class="o">=</span> <span class="s1">&#39;Glover A.&#39;</span>
<span class="n">historyNote</span> <span class="o">=</span> <span class="s1">&#39;Mapped from TreeBASE schema using org.cipres.treebase.domain.nexus.nexml.NexmlDocumentWriter@645f9132 $Rev: 1060 $&#39;</span>
<span class="n">contributor</span> <span class="o">=</span> <span class="s1">&#39;Smith C.&#39;</span>
<span class="n">modificationDate</span> <span class="o">=</span> <span class="s1">&#39;2012-06-04&#39;</span>
<span class="n">contributor</span> <span class="o">=</span> <span class="s1">&#39;Baco A.&#39;</span>
<span class="n">contributor</span> <span class="o">=</span> <span class="s1">&#39;Dahlgren T.G.&#39;</span>
<span class="n">identifier</span><span class="o">.</span><span class="n">study</span><span class="o">.</span><span class="n">tb1</span> <span class="o">=</span> <span class="s1">&#39;None&#39;</span>
<span class="n">publicationDate</span> <span class="o">=</span> <span class="s1">&#39;2012&#39;</span>
<span class="n">section</span> <span class="o">=</span> <span class="s1">&#39;Study&#39;</span>
<span class="n">doi</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="n">title</span><span class="o">.</span><span class="n">study</span> <span class="o">=</span> <span class="s1">&#39;Systematics and biodiversity of Ophryotrocha (Annelida, Dorvilleidae) with descriptions of six new species from deep-sea whale-fall and wood-fall habitats in the north-east Pacific&#39;</span>
<span class="n">subject</span> <span class="o">=</span> <span class="s1">&#39;New species&#39;</span>
<span class="n">subject</span> <span class="o">=</span> <span class="s1">&#39;Ophryotrocha&#39;</span>
<span class="n">creationDate</span> <span class="o">=</span> <span class="s1">&#39;2012-05-09&#39;</span>
<span class="n">subject</span> <span class="o">=</span> <span class="s1">&#39;polychaeta&#39;</span>
<span class="n">date</span> <span class="o">=</span> <span class="s1">&#39;2012-06-04&#39;</span>
<span class="n">subject</span> <span class="o">=</span> <span class="s1">&#39;molecular phylogeny&#39;</span>
<span class="n">identifier</span><span class="o">.</span><span class="n">study</span> <span class="o">=</span> <span class="s1">&#39;12713&#39;</span>

<span class="o">--</span> <span class="p">(</span><span class="n">tree</span> <span class="s1">&#39;con 50 majrule&#39;</span><span class="p">)</span> <span class="o">--</span>

<span class="n">ntax</span><span class="o">.</span><span class="n">tree</span> <span class="o">=</span> <span class="s1">&#39;41&#39;</span>
<span class="n">kind</span><span class="o">.</span><span class="n">tree</span> <span class="o">=</span> <span class="s1">&#39;Species Tree&#39;</span>
<span class="n">quality</span><span class="o">.</span><span class="n">tree</span> <span class="o">=</span> <span class="s1">&#39;Unrated&#39;</span>
<span class="n">isDefinedBy</span> <span class="o">=</span> <span class="s1">&#39;http://purl.org/phylo/treebase/phylows/study/TB2:S12713&#39;</span>
<span class="nb">type</span><span class="o">.</span><span class="n">tree</span> <span class="o">=</span> <span class="s1">&#39;Consensus&#39;</span>
</pre></div>
</div>
<p>The following sections discuss these methods and attributes in detail, describing how the create, read, write, search, and manipulate annotations.</p>
</section>
<section id="metadata-annotation-creation">
<h2>Metadata Annotation Creation<a class="headerlink" href="#metadata-annotation-creation" title="Link to this heading">¶</a></h2>
<section id="reading-data-from-an-external-source">
<h3>Reading Data from an External Source<a class="headerlink" href="#reading-data-from-an-external-source" title="Link to this heading">¶</a></h3>
<p>When reading data in NeXML format, metadata annotations given in the source are automatically created and associated with the corresponding data objects.</p>
<p>The metadata annotations associated with the phylogenetic data objects are collected in the attribute <code class="docutils literal notranslate"><span class="pre">annotations</span></code> of the objects, which is an object of type <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a>.
Each annotation item is represented as an
object of type <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a>.</p>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>
<span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;pythonidae.annotated.nexml&quot;</span><span class="p">,</span>
<span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
    <span class="nb">print</span> <span class="s2">&quot;Data Set &#39;</span><span class="si">%s</span><span class="s2">&#39;: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">label</span><span class="p">,</span> <span class="n">a</span><span class="p">)</span>
<span class="k">for</span> <span class="n">taxon_namespace</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">taxon_namespaces</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">taxon_namespace</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
        <span class="nb">print</span> <span class="s2">&quot;Taxon Set &#39;</span><span class="si">%s</span><span class="s2">&#39;: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">taxon_namespace</span><span class="o">.</span><span class="n">label</span><span class="p">,</span> <span class="n">a</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">taxon_namespace</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">taxon</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
            <span class="nb">print</span> <span class="s2">&quot;Taxon &#39;</span><span class="si">%s</span><span class="s2">&#39;: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">taxon</span><span class="o">.</span><span class="n">label</span><span class="p">,</span> <span class="n">a</span><span class="p">)</span>
<span class="k">for</span> <span class="n">tree_list</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">tree_lists</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">tree_list</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
        <span class="nb">print</span> <span class="s2">&quot;Tree List &#39;</span><span class="si">%s</span><span class="s2">&#39;: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">tree_list</span><span class="o">.</span><span class="n">label</span><span class="p">,</span> <span class="n">a</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">tree_list</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
            <span class="nb">print</span> <span class="s2">&quot;Tree &#39;</span><span class="si">%s</span><span class="s2">&#39;: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">label</span><span class="p">,</span> <span class="n">a</span><span class="p">)</span>
</pre></div>
</div>
<p>produces:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Data</span> <span class="n">Set</span> <span class="s1">&#39;None&#39;</span><span class="p">:</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;composite dataset of Pythonid sequences and trees&quot;</span>
<span class="n">Data</span> <span class="n">Set</span> <span class="s1">&#39;None&#39;</span><span class="p">:</span> <span class="n">subject</span><span class="o">=</span><span class="s2">&quot;Pythonidae&quot;</span>
<span class="n">Taxon</span> <span class="n">Set</span> <span class="s1">&#39;None&#39;</span><span class="p">:</span> <span class="n">subject</span><span class="o">=</span><span class="s2">&quot;Pythonidae&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Python regius&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/51751&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Python sebae&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/51752&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Python molurus&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/51750&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Python curtus&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/143436&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Morelia bredli&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/461327&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Morelia spilota&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/51896&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Morelia tracyae&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/129332&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Morelia clastolepis&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/129329&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Morelia kinghorni&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/129330&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Morelia nauta&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/129331&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Morelia amethistina&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/51895&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Morelia oenpelliensis&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/461329&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Antaresia maculosa&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/51891&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Antaresia perthensis&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/461324&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Antaresia stimsoni&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/461325&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Antaresia childreni&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/51888&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Morelia carinata&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/461328&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Morelia viridisN&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/129333&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Morelia viridisS&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/129333&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Apodora papuana&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/129310&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Liasis olivaceus&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/283338&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Liasis fuscus&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/129327&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Liasis mackloti&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/51889&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Antaresia melanocephalus&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/51883&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Antaresia ramsayi&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/461326&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Liasis albertisii&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/129326&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Bothrochilus boa&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/461341&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Morelia boeleni&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/129328&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Python timoriensis&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/51753&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Python reticulatus&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/37580&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Xenopeltis unicolor&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/196253&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Candoia aspera&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/51853&quot;</span>
<span class="n">Taxon</span> <span class="s1">&#39;Loxocemus bicolor&#39;</span><span class="p">:</span> <span class="n">closeMatch</span><span class="o">=</span><span class="s2">&quot;http://purl.uniprot.org/taxonomy/39078&quot;</span>
<span class="n">Tree</span> <span class="s1">&#39;0&#39;</span><span class="p">:</span> <span class="n">treeEstimator</span><span class="o">=</span><span class="s2">&quot;RAxML&quot;</span>
<span class="n">Tree</span> <span class="s1">&#39;0&#39;</span><span class="p">:</span> <span class="n">substitutionModel</span><span class="o">=</span><span class="s2">&quot;GTR+G+I&quot;</span>
</pre></div>
</div>
<p>Metadata annotations in NEXUS and NEWICK must be given in the form of “hot comments” either in BEAST/FigTree syntax:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="o">&amp;</span><span class="n">subject</span><span class="o">=</span><span class="s1">&#39;Pythonidae&#39;</span><span class="p">]</span>

<span class="p">[</span><span class="o">&amp;</span><span class="n">length_hpd95</span><span class="o">=</span><span class="p">{</span><span class="mf">0.01917252</span><span class="p">,</span><span class="mf">0.06241567</span><span class="p">},</span><span class="n">length_quant_5_95</span><span class="o">=</span><span class="p">{</span><span class="mf">0.02461821</span><span class="p">,</span><span class="mf">0.06197141</span><span class="p">},</span><span class="n">length_range</span><span class="o">=</span><span class="p">{</span><span class="mf">0.01570374</span><span class="p">,</span><span class="mf">0.07787249</span><span class="p">},</span><span class="n">length_mean</span><span class="o">=</span><span class="mf">0.0418470252488</span><span class="p">,</span><span class="n">length_median</span><span class="o">=</span><span class="mf">0.04091105</span><span class="p">,</span><span class="n">length_sd</span><span class="o">=</span><span class="mf">0.0113086027131</span><span class="p">]</span>
</pre></div>
</div>
<p>or NHX-like syntax:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">subject</span><span class="o">=</span><span class="s1">&#39;Pythonidae&#39;</span><span class="p">]</span>

<span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">length_hpd95</span><span class="o">=</span><span class="p">{</span><span class="mf">0.01917252</span><span class="p">,</span><span class="mf">0.06241567</span><span class="p">},</span><span class="n">length_quant_5_95</span><span class="o">=</span><span class="p">{</span><span class="mf">0.02461821</span><span class="p">,</span><span class="mf">0.06197141</span><span class="p">},</span><span class="n">length_range</span><span class="o">=</span><span class="p">{</span><span class="mf">0.01570374</span><span class="p">,</span><span class="mf">0.07787249</span><span class="p">},</span><span class="n">length_mean</span><span class="o">=</span><span class="mf">0.0418470252488</span><span class="p">,</span><span class="n">length_median</span><span class="o">=</span><span class="mf">0.04091105</span><span class="p">,</span><span class="n">length_sd</span><span class="o">=</span><span class="mf">0.0113086027131</span><span class="p">]</span>
</pre></div>
</div>
<p>However, by default these annotations are not parsed into DendroPy data model
unless the keyword argument <code class="docutils literal notranslate"><span class="pre">extract_comment_metadata=True</span></code> is passed in to the call:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;data.nex&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="s2">&quot;nexus&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">extract_comment_metadata</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>In general, support for metadata in NEXUS and NEWICK formats is very basic and lossy, and is limited to a small range of phylogenetic data types (taxa, trees, nodes, edges).
These issues and limits are fundamental to the NEXUS and NEWICK formats, and thus if metadata is important to you and your work, you should be working with NeXML format.
The NeXML format provides for rich, flexible and robust metadata annotation for the broad range of phylogenetic data, and DendroPy provides full support for metadata reading and writing in NeXML.</p>
</section>
<section id="direct-composition-with-literal-values">
<h3>Direct Composition with Literal Values<a class="headerlink" href="#direct-composition-with-literal-values" title="Link to this heading">¶</a></h3>
<p>The <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.add_new" title="dendropy.datamodel.basemodel.AnnotationSet.add_new"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_new</span></code></a> method of the <code class="xref py py-attr docutils literal notranslate"><span class="pre">annotations</span></code> attribute allows for direct adding of metadata. This method has two mandatory arguments, “<code class="docutils literal notranslate"><span class="pre">name</span></code>” and “<code class="docutils literal notranslate"><span class="pre">value</span></code>”:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s1">&#39;pythonidae.mle.tree&#39;</span><span class="p">,</span> <span class="s1">&#39;nexus&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s1">&#39;examples/pythonidae.mle.nex&#39;</span><span class="p">,</span> <span class="s1">&#39;nexus&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_new</span><span class="p">(</span>
<span class="gp">... </span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;subject&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">value</span><span class="o">=</span><span class="s2">&quot;Python phylogenetics&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
<p>When printing the tree in NeXML, the metadata will be rendered as a “<code class="docutils literal notranslate"><span class="pre">&lt;meta&gt;</span></code>” tag child element of the associated “<code class="docutils literal notranslate"><span class="pre">&lt;tree&gt;</span></code>” element:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;</span><span class="n">nex</span><span class="p">:</span><span class="n">nexml</span>
    <span class="n">version</span><span class="o">=</span><span class="s2">&quot;0.9&quot;</span>
    <span class="n">xsi</span><span class="p">:</span><span class="n">schemaLocation</span><span class="o">=</span><span class="s2">&quot;http://www.nexml.org/2009&quot;</span>
    <span class="n">xmlns</span><span class="p">:</span><span class="n">dendropy</span><span class="o">=</span><span class="s2">&quot;http://pypi.org/project/DendroPy//&quot;</span>
    <span class="n">xmlns</span><span class="o">=</span><span class="s2">&quot;http://www.nexml.org/2009&quot;</span>
    <span class="n">xmlns</span><span class="p">:</span><span class="n">xsi</span><span class="o">=</span><span class="s2">&quot;http://www.w3.org/2001/XMLSchema-instance&quot;</span>
    <span class="n">xmlns</span><span class="p">:</span><span class="n">xml</span><span class="o">=</span><span class="s2">&quot;http://www.w3.org/XML/1998/namespace&quot;</span>
    <span class="n">xmlns</span><span class="p">:</span><span class="n">nex</span><span class="o">=</span><span class="s2">&quot;http://www.nexml.org/2009&quot;</span>
<span class="o">&gt;</span>
<span class="o">.</span>
<span class="o">.</span>
<span class="o">.</span>
    <span class="o">&lt;</span><span class="n">trees</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;x4320340992&quot;</span> <span class="n">otus</span><span class="o">=</span><span class="s2">&quot;x4320340552&quot;</span><span class="o">&gt;</span>
        <span class="o">&lt;</span><span class="n">tree</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;x4320381904&quot;</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;0&quot;</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:FloatTree&quot;</span><span class="o">&gt;</span>
            <span class="o">&lt;</span><span class="n">meta</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:LiteralMeta&quot;</span> <span class="nb">property</span><span class="o">=</span><span class="s2">&quot;dendropy:subject&quot;</span> <span class="n">content</span><span class="o">=</span><span class="s2">&quot;Python phylogenetics&quot;</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;meta4320379536&quot;</span> <span class="o">/&gt;</span>
<span class="o">.</span>
<span class="o">.</span>
<span class="o">.</span>
</pre></div>
</div>
<p>As can be seen, by default, the metadata property is mapped to the “<code class="docutils literal notranslate"><span class="pre">dendropy</span></code>” namespace (i.e., ‘<code class="docutils literal notranslate"><span class="pre">xmlns:dendropy=&quot;http://pypi.org/project/DendroPy//&quot;</span></code>’).
This can be customized by using the “<code class="docutils literal notranslate"><span class="pre">name_prefix</span></code>” and “<code class="docutils literal notranslate"><span class="pre">namespace</span></code>” arguments to the call to <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.add_new" title="dendropy.datamodel.basemodel.AnnotationSet.add_new"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_new</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_new</span><span class="p">(</span>
<span class="gp">... </span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;subject&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">value</span><span class="o">=</span><span class="s2">&quot;Python phylogenetics&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">name_prefix</span><span class="o">=</span><span class="s2">&quot;dc&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">namespace</span><span class="o">=</span><span class="s2">&quot;http://purl.org/dc/elements/1.1/&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
<p>This will result in the following NeXML fragment:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;</span><span class="n">nex</span><span class="p">:</span><span class="n">nexml</span>
    <span class="n">version</span><span class="o">=</span><span class="s2">&quot;0.9&quot;</span>
    <span class="n">xsi</span><span class="p">:</span><span class="n">schemaLocation</span><span class="o">=</span><span class="s2">&quot;http://www.nexml.org/2009&quot;</span>
    <span class="n">xmlns</span><span class="p">:</span><span class="n">dc</span><span class="o">=</span><span class="s2">&quot;http://purl.org/dc/elements/1.1/&quot;</span>
    <span class="n">xmlns</span><span class="o">=</span><span class="s2">&quot;http://www.nexml.org/2009&quot;</span>
    <span class="n">xmlns</span><span class="p">:</span><span class="n">xsi</span><span class="o">=</span><span class="s2">&quot;http://www.w3.org/2001/XMLSchema-instance&quot;</span>
    <span class="n">xmlns</span><span class="p">:</span><span class="n">xml</span><span class="o">=</span><span class="s2">&quot;http://www.w3.org/XML/1998/namespace&quot;</span>
    <span class="n">xmlns</span><span class="p">:</span><span class="n">nex</span><span class="o">=</span><span class="s2">&quot;http://www.nexml.org/2009&quot;</span>
    <span class="n">xmlns</span><span class="p">:</span><span class="n">dendropy</span><span class="o">=</span><span class="s2">&quot;http://pypi.org/project/DendroPy//&quot;</span>
<span class="o">&gt;</span>
<span class="o">.</span>
<span class="o">.</span>
<span class="o">.</span>
    <span class="o">&lt;</span><span class="n">trees</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;x4320340904&quot;</span> <span class="n">otus</span><span class="o">=</span><span class="s2">&quot;x4320340464&quot;</span><span class="o">&gt;</span>
        <span class="o">&lt;</span><span class="n">tree</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;x4320377872&quot;</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;0&quot;</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:FloatTree&quot;</span><span class="o">&gt;</span>
            <span class="o">&lt;</span><span class="n">meta</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:LiteralMeta&quot;</span> <span class="nb">property</span><span class="o">=</span><span class="s2">&quot;dc:subject&quot;</span> <span class="n">content</span><span class="o">=</span><span class="s2">&quot;Python phylogenetics&quot;</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;meta4320375440&quot;</span> <span class="o">/&gt;</span>
<span class="o">.</span>
<span class="o">.</span>
<span class="o">.</span>
</pre></div>
</div>
<p>Note that the “<code class="docutils literal notranslate"><span class="pre">name_prefix</span></code>” or “<code class="docutils literal notranslate"><span class="pre">namespace</span></code>” must be specified simultaneously; that is, if one is specified, then the other must be specified as well.
For convenience, you can specify the name of the annotation with the name prefix prepended by specifying “<code class="docutils literal notranslate"><span class="pre">name_is_prefixed=True</span></code>”, though the namespace must still be provided separately:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_new</span><span class="p">(</span>
<span class="gp">... </span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;dc:subject&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">value</span><span class="o">=</span><span class="s2">&quot;Python phylogenetics&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">name_is_prefixed</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span><span class="n">namespace</span><span class="o">=</span><span class="s2">&quot;http://purl.org/dc/elements/1.1/&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
<p>For NeXML output, you can also specify a datatype:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_new</span><span class="p">(</span>
<span class="gp">... </span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;subject&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">value</span><span class="o">=</span><span class="s2">&quot;Python phylogenetics&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">datatype_hint</span><span class="o">=</span><span class="s2">&quot;xsd:string&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_new</span><span class="p">(</span>
<span class="gp">... </span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;answer&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">value</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span>
<span class="gp">... </span><span class="n">datatype_hint</span><span class="o">=</span><span class="s2">&quot;xsd:integer&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
<p>When writing to NeXML, this will result in the following fragment:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;</span><span class="n">trees</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;x4320340992&quot;</span> <span class="n">otus</span><span class="o">=</span><span class="s2">&quot;x4320340552&quot;</span><span class="o">&gt;</span>
    <span class="o">&lt;</span><span class="n">tree</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;x4320381968&quot;</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;0&quot;</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:FloatTree&quot;</span><span class="o">&gt;</span>
        <span class="o">&lt;</span><span class="n">meta</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:LiteralMeta&quot;</span> <span class="nb">property</span><span class="o">=</span><span class="s2">&quot;dendropy:answer&quot;</span> <span class="n">content</span><span class="o">=</span><span class="s2">&quot;42&quot;</span> <span class="n">datatype</span><span class="o">=</span><span class="s2">&quot;xsd:integer&quot;</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;meta4320379536&quot;</span> <span class="o">/&gt;</span>
        <span class="o">&lt;</span><span class="n">meta</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:LiteralMeta&quot;</span> <span class="nb">property</span><span class="o">=</span><span class="s2">&quot;dendropy:subject&quot;</span> <span class="n">content</span><span class="o">=</span><span class="s2">&quot;Python phylogenetics&quot;</span> <span class="n">datatype</span><span class="o">=</span><span class="s2">&quot;xsd:string&quot;</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;meta4320379472&quot;</span> <span class="o">/&gt;</span>
</pre></div>
</div>
<p>You can also specify that the data should be interpreted as a source to be dereferenced in NeXML by passing in <code class="docutils literal notranslate"><span class="pre">annotate_as_reference=True</span></code>.
Note that this does not actually populate the contents of the annotation from the source (unlike the dynamic attribute value binding discussed below), but just indicates the the contents of the annotation should be <em>interpreted</em> differently by semantic readers.
Thus, the following annotation:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_new</span><span class="p">(</span>
<span class="gp">... </span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;subject&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">value</span><span class="o">=</span><span class="s2">&quot;http://en.wikipedia.org/wiki/Pythonidae&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">name_prefix</span><span class="o">=</span><span class="s2">&quot;dc&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">namespace</span><span class="o">=</span><span class="s2">&quot;http://purl.org/dc/elements/1.1/&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">annotate_as_reference</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
<p>will be rendered in NeXML as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;</span><span class="n">meta</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:ResourceMeta&quot;</span> <span class="n">rel</span><span class="o">=</span><span class="s2">&quot;dc:subject&quot;</span> <span class="n">href</span><span class="o">=</span><span class="s2">&quot;http://en.wikipedia.org/wiki/Pythonidae&quot;</span> <span class="o">/&gt;</span>
</pre></div>
</div>
<p>Sometimes, you may want to annotate an object with metadata, but do not want it to be printed or written out.
Passing the <code class="docutils literal notranslate"><span class="pre">is_hidden=True</span></code> argument will result in the annotation being suppressed in all output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_new</span><span class="p">(</span>
<span class="gp">... </span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;subject&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">value</span><span class="o">=</span><span class="s2">&quot;Python phylogenetics&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">name_prefix</span><span class="o">=</span><span class="s2">&quot;dc&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">namespace</span><span class="o">=</span><span class="s2">&quot;http://purl.org/dc/elements/1.1/&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">is_hidden</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">is_hidden</span></code> attribute of the an <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> object can also be set directly:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">subject_annotations</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;citation&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">subject_annotations</span><span class="p">:</span>
<span class="gp">... </span>   <span class="n">a</span><span class="o">.</span><span class="n">is_hidden</span> <span class="o">=</span> <span class="kc">True</span>
</pre></div>
</div>
</section>
<section id="dynamically-binding-annotation-values-to-object-attribute-values">
<h3>Dynamically Binding Annotation Values to Object Attribute Values<a class="headerlink" href="#dynamically-binding-annotation-values-to-object-attribute-values" title="Link to this heading">¶</a></h3>
<p>In some cases, instead of “hard-wiring” in metadata for an object, you may want to write out metadata that takes its value from the value of an attribute of the object.
The <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.add_bound_attribute" title="dendropy.datamodel.basemodel.AnnotationSet.add_bound_attribute"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_bound_attribute</span></code></a> method allows you to do this.
This method takes, as a minimum, a <em>string</em> specifying the <em>name</em> of an existing attribute to which the value of the annotation will be dynamically bound.</p>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="kn">import</span> <span class="nn">dendropy</span>
<span class="kn">import</span> <span class="nn">random</span>

<span class="n">categories</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;A&quot;</span> <span class="p">:</span> <span class="s2">&quot;N/A&quot;</span><span class="p">,</span>
    <span class="s2">&quot;B&quot;</span> <span class="p">:</span> <span class="s2">&quot;N/A&quot;</span><span class="p">,</span>
    <span class="s2">&quot;C&quot;</span> <span class="p">:</span> <span class="s2">&quot;N/A&quot;</span><span class="p">,</span>
    <span class="s2">&quot;D&quot;</span> <span class="p">:</span> <span class="s2">&quot;N/A&quot;</span><span class="p">,</span>
    <span class="s2">&quot;E&quot;</span> <span class="p">:</span> <span class="s2">&quot;N/A&quot;</span>
<span class="p">}</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">Tree</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="s2">&quot;(A,(B,(C,(D,E))));&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">taxon_namespace</span><span class="p">:</span>
    <span class="n">taxon</span><span class="o">.</span><span class="n">category</span> <span class="o">=</span> <span class="n">categories</span><span class="p">[</span><span class="n">taxon</span><span class="o">.</span><span class="n">label</span><span class="p">]</span>
    <span class="n">taxon</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_bound_attribute</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">postorder_node_iter</span><span class="p">():</span>
    <span class="n">node</span><span class="o">.</span><span class="n">pop_size</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">node</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_bound_attribute</span><span class="p">(</span><span class="s2">&quot;pop_size&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">postorder_node_iter</span><span class="p">():</span>
    <span class="n">node</span><span class="o">.</span><span class="n">pop_size</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">10000</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">taxon</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">pop_size</span> <span class="o">&gt;=</span> <span class="mi">8000</span><span class="p">:</span>
            <span class="n">node</span><span class="o">.</span><span class="n">taxon</span><span class="o">.</span><span class="n">category</span> <span class="o">=</span> <span class="s2">&quot;large&quot;</span>
        <span class="k">elif</span> <span class="n">node</span><span class="o">.</span><span class="n">pop_size</span> <span class="o">&gt;=</span> <span class="mi">6000</span><span class="p">:</span>
            <span class="n">node</span><span class="o">.</span><span class="n">taxon</span><span class="o">.</span><span class="n">category</span> <span class="o">=</span> <span class="s2">&quot;medium&quot;</span>
        <span class="k">elif</span> <span class="n">node</span><span class="o">.</span><span class="n">pop_size</span> <span class="o">&gt;=</span> <span class="mi">4000</span><span class="p">:</span>
            <span class="n">node</span><span class="o">.</span><span class="n">taxon</span><span class="o">.</span><span class="n">category</span> <span class="o">=</span> <span class="s2">&quot;small&quot;</span>
        <span class="k">elif</span> <span class="n">node</span><span class="o">.</span><span class="n">pop_size</span> <span class="o">&gt;=</span> <span class="mi">2000</span><span class="p">:</span>
            <span class="n">node</span><span class="o">.</span><span class="n">taxon</span><span class="o">.</span><span class="n">category</span> <span class="o">=</span> <span class="s2">&quot;tiny&quot;</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">as_string</span><span class="p">(</span><span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexml&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p>results in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&lt;?xml version=&quot;1.0&quot; encoding=&quot;ISO-8859-1&quot;?&gt;
&lt;nex:nexml
    version=&quot;0.9&quot;
    xsi:schemaLocation=&quot;http://www.nexml.org/2009&quot;
    xmlns:dendropy=&quot;http://pypi.org/project/DendroPy//&quot;
    xmlns=&quot;http://www.nexml.org/2009&quot;
    xmlns:xsi=&quot;http://www.w3.org/2001/XMLSchema-instance&quot;
    xmlns:xml=&quot;http://www.w3.org/XML/1998/namespace&quot;
    xmlns:nex=&quot;http://www.nexml.org/2009&quot;
&gt;
    &lt;otus id=&quot;x4320344648&quot;&gt;
        &lt;otu id=&quot;x4320380112&quot; label=&quot;A&quot;&gt;
            &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:category&quot; content=&quot;tiny&quot; id=&quot;meta4320379472&quot; /&gt;
        &lt;/otu&gt;
        &lt;otu id=&quot;x4320380432&quot; label=&quot;B&quot;&gt;
            &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:category&quot; content=&quot;medium&quot; id=&quot;meta4320379536&quot; /&gt;
        &lt;/otu&gt;
        &lt;otu id=&quot;x4320380752&quot; label=&quot;C&quot;&gt;
            &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:category&quot; content=&quot;N/A&quot; id=&quot;meta4320379792&quot; /&gt;
        &lt;/otu&gt;
        &lt;otu id=&quot;x4320381072&quot; label=&quot;D&quot;&gt;
            &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:category&quot; content=&quot;tiny&quot; id=&quot;meta4320381328&quot; /&gt;
        &lt;/otu&gt;
        &lt;otu id=&quot;x4320381264&quot; label=&quot;E&quot;&gt;
            &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:category&quot; content=&quot;tiny&quot; id=&quot;meta4320381392&quot; /&gt;
        &lt;/otu&gt;
    &lt;/otus&gt;
    &lt;trees id=&quot;x4320344560&quot; otus=&quot;x4320344648&quot;&gt;
        &lt;tree id=&quot;x4320379600&quot; xsi:type=&quot;nex:FloatTree&quot;&gt;
            &lt;node id=&quot;x4320379856&quot;&gt;
                &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:pop_size&quot; content=&quot;5491&quot; id=&quot;meta4320379280&quot; /&gt;
            &lt;/node&gt;
            &lt;node id=&quot;x4320379984&quot; otu=&quot;x4320380112&quot;&gt;
                &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:pop_size&quot; content=&quot;2721&quot; id=&quot;meta4320379408&quot; /&gt;
            &lt;/node&gt;
            &lt;node id=&quot;x4320380176&quot;&gt;
                &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:pop_size&quot; content=&quot;4627&quot; id=&quot;meta4320379344&quot; /&gt;
            &lt;/node&gt;
            &lt;node id=&quot;x4320380304&quot; otu=&quot;x4320380432&quot;&gt;
                &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:pop_size&quot; content=&quot;7202&quot; id=&quot;meta4320381456&quot; /&gt;
            &lt;/node&gt;
            &lt;node id=&quot;x4320380496&quot;&gt;
                &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:pop_size&quot; content=&quot;5337&quot; id=&quot;meta4320379664&quot; /&gt;
            &lt;/node&gt;
            &lt;node id=&quot;x4320380624&quot; otu=&quot;x4320380752&quot;&gt;
                &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:pop_size&quot; content=&quot;1478&quot; id=&quot;meta4320381520&quot; /&gt;
            &lt;/node&gt;
            &lt;node id=&quot;x4320380816&quot;&gt;
                &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:pop_size&quot; content=&quot;1539&quot; id=&quot;meta4320379728&quot; /&gt;
            &lt;/node&gt;
            &lt;node id=&quot;x4320380944&quot; otu=&quot;x4320381072&quot;&gt;
                &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:pop_size&quot; content=&quot;3457&quot; id=&quot;meta4320381584&quot; /&gt;
            &lt;/node&gt;
            &lt;node id=&quot;x4320381136&quot; otu=&quot;x4320381264&quot;&gt;
                &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;dendropy:pop_size&quot; content=&quot;3895&quot; id=&quot;meta4320381648&quot; /&gt;
            &lt;/node&gt;
            &lt;rootedge id=&quot;x4320379920&quot; target=&quot;x4320379856&quot; /&gt;
            &lt;edge id=&quot;x4320380048&quot; source=&quot;x4320379856&quot; target=&quot;x4320379984&quot; /&gt;
            &lt;edge id=&quot;x4320380240&quot; source=&quot;x4320379856&quot; target=&quot;x4320380176&quot; /&gt;
            &lt;edge id=&quot;x4320380368&quot; source=&quot;x4320380176&quot; target=&quot;x4320380304&quot; /&gt;
            &lt;edge id=&quot;x4320380560&quot; source=&quot;x4320380176&quot; target=&quot;x4320380496&quot; /&gt;
            &lt;edge id=&quot;x4320380688&quot; source=&quot;x4320380496&quot; target=&quot;x4320380624&quot; /&gt;
            &lt;edge id=&quot;x4320380880&quot; source=&quot;x4320380496&quot; target=&quot;x4320380816&quot; /&gt;
            &lt;edge id=&quot;x4320381008&quot; source=&quot;x4320380816&quot; target=&quot;x4320380944&quot; /&gt;
            &lt;edge id=&quot;x4320381200&quot; source=&quot;x4320380816&quot; target=&quot;x4320381136&quot; /&gt;
        &lt;/tree&gt;
    &lt;/trees&gt;
&lt;/nex:nexml&gt;
</pre></div>
</div>
<p>By default, the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.add_bound_attribute" title="dendropy.datamodel.basemodel.AnnotationSet.add_bound_attribute"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_bound_attribute</span></code></a> method uses the name of the attribute as the name of the annotation.
The “<code class="docutils literal notranslate"><span class="pre">annotation_name</span></code>” argument allows you explictly set the name of the annotation.
In addition, the method call also supports the other customization arguments of the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.add_new" title="dendropy.datamodel.basemodel.AnnotationSet.add_new"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_new</span></code></a> method: “<code class="docutils literal notranslate"><span class="pre">datatype_hint</span></code>”, “<code class="docutils literal notranslate"><span class="pre">name_prefix</span></code>”, “<code class="docutils literal notranslate"><span class="pre">namespace</span></code>”, “<code class="docutils literal notranslate"><span class="pre">name_is_prefixed</span></code>”, “<code class="docutils literal notranslate"><span class="pre">annotate_as_reference</span></code>”, “<code class="docutils literal notranslate"><span class="pre">is_hidden</span></code>”, etc.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">source_uri</span> <span class="o">=</span> <span class="kc">None</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_bound_attribute</span><span class="p">(</span>
<span class="gp">... </span><span class="s2">&quot;source_uri&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">annotation_name</span><span class="o">=</span><span class="s2">&quot;dc:subject&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">namespace</span><span class="o">=</span><span class="s2">&quot;http://purl.org/dc/elements/1.1/&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">annotate_as_reference</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="adding-citation-metadata">
<h3>Adding Citation Metadata<a class="headerlink" href="#adding-citation-metadata" title="Link to this heading">¶</a></h3>
<p>You can add citation annotations using the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.add_citation" title="dendropy.datamodel.basemodel.AnnotationSet.add_citation"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_citation</span></code></a> method.
This method takes at least one argument, <code class="docutils literal notranslate"><span class="pre">citation</span></code>.
This can be a string representing the citation as a BibTex record or a dictionary with BibTex fields as keys and field content as values.</p>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">dendropy</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
    <span class="s2">&quot;This example is known to be broken! &quot;</span>
    <span class="s2">&quot;It will be fixed or removed in the future. &quot;</span>
    <span class="s2">&quot;See https://github.com/jeetsukumaran/DendroPy/issues/160 for details. &quot;</span>
    <span class="s2">&quot;Patch contributions are welcome.&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">citation</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span><span class="se">\</span>
<span class="s2">@article{HeathHH2012,</span>
<span class="s2">	Author = {Tracy A. Heath and Mark T. Holder and John P. Huelsenbeck},</span>
<span class="s2">	Doi = {10.1093/molbev/msr255},</span>
<span class="s2">	Journal = {Molecular Biology and Evolution},</span>
<span class="s2">	Number = </span><span class="si">{3}</span><span class="s2">,</span>
<span class="s2">	Pages = {939-955},</span>
<span class="s2">	Title = {A </span><span class="si">{Dirichlet}</span><span class="s2"> Process Prior for Estimating Lineage-Specific Substitution Rates.},</span>
<span class="s2">	Url = {http://mbe.oxfordjournals.org/content/early/2011/11/04/molbev.msr255.abstract},</span>
<span class="s2">	Volume = </span><span class="si">{29}</span><span class="s2">,</span>
<span class="s2">	Year = </span><span class="si">{2012}</span>
<span class="s2">	}</span>
<span class="s2">&quot;&quot;&quot;</span>


<span class="n">dataset</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="s2">&quot;(A,(B,(C,(D,E))));&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_citation</span><span class="p">(</span><span class="n">citation</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">as_string</span><span class="p">(</span><span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexml&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p>will result in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&lt;?xml version=&quot;1.0&quot; encoding=&quot;ISO-8859-1&quot;?&gt;
&lt;nex:nexml
    version=&quot;0.9&quot;
    xsi:schemaLocation=&quot;http://www.nexml.org/2009&quot;
    xmlns:bibtex=&quot;http://www.edutella.org/bibtex#&quot;
    xmlns=&quot;http://www.nexml.org/2009&quot;
    xmlns:xsi=&quot;http://www.w3.org/2001/XMLSchema-instance&quot;
    xmlns:xml=&quot;http://www.w3.org/XML/1998/namespace&quot;
    xmlns:nex=&quot;http://www.nexml.org/2009&quot;
    xmlns:dendropy=&quot;http://pypi.org/project/DendroPy//&quot;
&gt;
    &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;bibtex:journal&quot; content=&quot;Molecular Biology and Evolution&quot; datatype=&quot;xsd:string&quot; id=&quot;meta4320453648&quot; /&gt;
    &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;bibtex:bibtype&quot; content=&quot;article&quot; datatype=&quot;xsd:string&quot; id=&quot;meta4320453200&quot; /&gt;
    &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;bibtex:number&quot; content=&quot;3&quot; datatype=&quot;xsd:string&quot; id=&quot;meta4320453776&quot; /&gt;
    &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;bibtex:citekey&quot; content=&quot;HeathHH2012&quot; datatype=&quot;xsd:string&quot; id=&quot;meta4320453328&quot; /&gt;
    &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;bibtex:pages&quot; content=&quot;939-955&quot; datatype=&quot;xsd:string&quot; id=&quot;meta4320453968&quot; /&gt;
    &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;bibtex:volume&quot; content=&quot;29&quot; datatype=&quot;xsd:string&quot; id=&quot;meta4320453840&quot; /&gt;
    &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;bibtex:year&quot; content=&quot;2012&quot; datatype=&quot;xsd:string&quot; id=&quot;meta4320453904&quot; /&gt;
    &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;bibtex:doi&quot; content=&quot;10.1093/molbev/msr255&quot; datatype=&quot;xsd:string&quot; id=&quot;meta4320453456&quot; /&gt;
    &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;bibtex:title&quot; content=&quot;A {Dirichlet} Process Prior for Estimating Lineage-Specific Substitution Rates.&quot; datatype=&quot;xsd:string&quot; id=&quot;meta4320453520&quot; /&gt;
    &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;bibtex:url&quot; content=&quot;http://mbe.oxfordjournals.org/content/early/2011/11/04/molbev.msr255.abstract&quot; datatype=&quot;xsd:string&quot; id=&quot;meta4320453584&quot; /&gt;
    &lt;meta xsi:type=&quot;nex:LiteralMeta&quot; property=&quot;bibtex:author&quot; content=&quot;Tracy A. Heath and Mark T. Holder and John P. Huelsenbeck&quot; datatype=&quot;xsd:string&quot; id=&quot;meta4320453712&quot; /&gt;
    .
    .
    .
</pre></div>
</div>
<p>The following results in the same output as above, but the citation is given as a dictionary with BibTex fields as keys and content as values:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">dendropy</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
    <span class="s2">&quot;This example is known to be broken! &quot;</span>
    <span class="s2">&quot;It will be fixed or removed in the future. &quot;</span>
    <span class="s2">&quot;See https://github.com/jeetsukumaran/DendroPy/issues/160 for details. &quot;</span>
    <span class="s2">&quot;Patch contributions are welcome.&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">citation</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;BibType&quot;</span><span class="p">:</span> <span class="s2">&quot;article&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Author&quot;</span><span class="p">:</span> <span class="s2">&quot;Tracy A. Heath and Mark T. Holder and John P. Huelsenbeck&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Doi&quot;</span><span class="p">:</span> <span class="s2">&quot;10.1093/molbev/msr255&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Journal&quot;</span><span class="p">:</span> <span class="s2">&quot;Molecular Biology and Evolution&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;3&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Pages&quot;</span><span class="p">:</span> <span class="s2">&quot;939-955&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Title&quot;</span><span class="p">:</span> <span class="s2">&quot;A </span><span class="si">{Dirichlet}</span><span class="s2"> Process Prior for Estimating Lineage-Specific Substitution Rates.&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Url&quot;</span><span class="p">:</span> <span class="s2">&quot;http://mbe.oxfordjournals.org/content/early/2011/11/04/molbev.msr255.abstract&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Volume&quot;</span><span class="p">:</span> <span class="s2">&quot;29&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Year&quot;</span><span class="p">:</span> <span class="s2">&quot;2012&quot;</span><span class="p">,</span>
<span class="p">}</span>


<span class="n">dataset</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="s2">&quot;(A,(B,(C,(D,E))));&quot;</span><span class="p">,</span>
        <span class="n">schema</span><span class="o">=</span><span class="s2">&quot;newick&quot;</span><span class="p">)</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_citation</span><span class="p">(</span><span class="n">citation</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">as_string</span><span class="p">(</span><span class="n">schema</span><span class="o">=</span><span class="s2">&quot;nexml&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p>By default, the citation gets annotated as a series of separate BibTex elements.
You can specify alternate formats by using the “<code class="docutils literal notranslate"><span class="pre">store_as</span></code>” argument.
This argument can take one of the following values:</p>
<blockquote>
<div><ul>
<li><dl class="simple">
<dt>“<code class="docutils literal notranslate"><span class="pre">bibtex</span></code>”</dt><dd><p>Each BibTex field gets recorded as a separate annotation, with name
given by the field name, content by the field value.
This is the default, and the results in NeXML are shown above.</p>
</dd>
</dl>
</li>
<li><dl>
<dt>“<code class="docutils literal notranslate"><span class="pre">dublin</span></code>”</dt><dd><p>A subset of the BibTex fields gets recorded as a set of Dublin Core (Publishing Requirements for Industry Standard Metadata) annotations, one per field:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;</span><span class="n">meta</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:LiteralMeta&quot;</span> <span class="nb">property</span><span class="o">=</span><span class="s2">&quot;dc:date&quot;</span> <span class="n">content</span><span class="o">=</span><span class="s2">&quot;2012&quot;</span> <span class="n">datatype</span><span class="o">=</span><span class="s2">&quot;xsd:string&quot;</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;meta4320461584&quot;</span> <span class="o">/&gt;</span>
<span class="o">&lt;</span><span class="n">meta</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:LiteralMeta&quot;</span> <span class="nb">property</span><span class="o">=</span><span class="s2">&quot;dc:publisher&quot;</span> <span class="n">content</span><span class="o">=</span><span class="s2">&quot;Molecular Biology and Evolution&quot;</span> <span class="n">datatype</span><span class="o">=</span><span class="s2">&quot;xsd:string&quot;</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;meta4320461648&quot;</span> <span class="o">/&gt;</span>
<span class="o">&lt;</span><span class="n">meta</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:LiteralMeta&quot;</span> <span class="nb">property</span><span class="o">=</span><span class="s2">&quot;dc:title&quot;</span> <span class="n">content</span><span class="o">=</span><span class="s2">&quot;A </span><span class="si">{Dirichlet}</span><span class="s2"> Process Prior for Estimating Lineage-Specific Substitution Rates.&quot;</span> <span class="n">datatype</span><span class="o">=</span><span class="s2">&quot;xsd:string&quot;</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;meta4320461776&quot;</span> <span class="o">/&gt;</span>
<span class="o">&lt;</span><span class="n">meta</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:LiteralMeta&quot;</span> <span class="nb">property</span><span class="o">=</span><span class="s2">&quot;dc:creator&quot;</span> <span class="n">content</span><span class="o">=</span><span class="s2">&quot;Tracy A. Heath and Mark T. Holder and John P. Huelsenbeck&quot;</span> <span class="n">datatype</span><span class="o">=</span><span class="s2">&quot;xsd:string&quot;</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;meta4320461712&quot;</span> <span class="o">/&gt;</span>
</pre></div>
</div>
</dd>
</dl>
</li>
<li><dl>
<dt>“<code class="docutils literal notranslate"><span class="pre">prism</span></code>”</dt><dd><p>A subset of the BibTex fields gets recorded as a set of PRISM (Publishing Requirements for Industry Standard Metadata) annotations, one per field:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;</span><span class="n">meta</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:LiteralMeta&quot;</span> <span class="nb">property</span><span class="o">=</span><span class="s2">&quot;prism:volume&quot;</span> <span class="n">content</span><span class="o">=</span><span class="s2">&quot;29&quot;</span> <span class="n">datatype</span><span class="o">=</span><span class="s2">&quot;xsd:string&quot;</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;meta4320461584&quot;</span> <span class="o">/&gt;</span>
<span class="o">&lt;</span><span class="n">meta</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:LiteralMeta&quot;</span> <span class="nb">property</span><span class="o">=</span><span class="s2">&quot;prism:pageRange&quot;</span> <span class="n">content</span><span class="o">=</span><span class="s2">&quot;939-955&quot;</span> <span class="n">datatype</span><span class="o">=</span><span class="s2">&quot;xsd:string&quot;</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;meta4320461648&quot;</span> <span class="o">/&gt;</span>
<span class="o">&lt;</span><span class="n">meta</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:LiteralMeta&quot;</span> <span class="nb">property</span><span class="o">=</span><span class="s2">&quot;prism:publicationDate&quot;</span> <span class="n">content</span><span class="o">=</span><span class="s2">&quot;2012&quot;</span> <span class="n">datatype</span><span class="o">=</span><span class="s2">&quot;xsd:string&quot;</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;meta4320461776&quot;</span> <span class="o">/&gt;</span>
<span class="o">&lt;</span><span class="n">meta</span> <span class="n">xsi</span><span class="p">:</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;nex:LiteralMeta&quot;</span> <span class="nb">property</span><span class="o">=</span><span class="s2">&quot;prism:publicationName&quot;</span> <span class="n">content</span><span class="o">=</span><span class="s2">&quot;Molecular Biology and Evolution&quot;</span> <span class="n">datatype</span><span class="o">=</span><span class="s2">&quot;xsd:string&quot;</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;meta4320461712&quot;</span> <span class="o">/&gt;</span>
</pre></div>
</div>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
<p>In addition, the method call also supports some of the other customization arguments of the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.add_new" title="dendropy.datamodel.basemodel.AnnotationSet.add_new"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_new</span></code></a> method:  “<code class="docutils literal notranslate"><span class="pre">name_prefix</span></code>”, “<code class="docutils literal notranslate"><span class="pre">namespace</span></code>”, “<code class="docutils literal notranslate"><span class="pre">name_is_prefixed</span></code>”, “<code class="docutils literal notranslate"><span class="pre">is_hidden</span></code>”.</p>
</section>
<section id="copying-metadata-annotations-from-one-phylogenetic-data-object-to-another">
<h3>Copying Metadata Annotations from One Phylogenetic Data Object to Another<a class="headerlink" href="#copying-metadata-annotations-from-one-phylogenetic-data-object-to-another" title="Link to this heading">¶</a></h3>
<p>As the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a> is derived from <code class="xref py py-class docutils literal notranslate"><span class="pre">dendropy.utility.containers.OrderedSet</span></code>, it has the <code class="xref py py-meth docutils literal notranslate"><span class="pre">dendropy.utility.containers.OrderedSet.add</span></code> and <code class="xref py py-meth docutils literal notranslate"><span class="pre">dendropy.utility.containers.OrderedSet.update</span></code> methods available for direct addition of <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects.
The following example shows how to add metadata annotations associated with a <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object to all its <a class="reference internal" href="../library/treemodel.html#dendropy.datamodel.treemodel.Tree" title="dendropy.datamodel.treemodel.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> objects:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="n">ds_annotes</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">name_prefix</span><span class="o">=</span><span class="s2">&quot;dc&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">values_as_dict</span><span class="p">()</span>
<span class="k">for</span> <span class="n">tree_list</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">tree_lists</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">tree_list</span><span class="p">:</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">ds_annotes</span><span class="p">)</span>
</pre></div>
</div>
<p>Or, alternatively:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="n">ds_annotes</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">name_prefix</span><span class="o">=</span><span class="s2">&quot;dc&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">values_as_dict</span><span class="p">()</span>
<span class="k">for</span> <span class="n">tree_list</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">tree_lists</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">tree_list</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">ds_annotes</span><span class="p">:</span>
            <span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="metadata-annotation-access-and-manipulation">
<h2>Metadata Annotation Access and Manipulation<a class="headerlink" href="#metadata-annotation-access-and-manipulation" title="Link to this heading">¶</a></h2>
<section id="iterating-over-collections-of-annotations">
<h3>Iterating Over Collections of Annotations<a class="headerlink" href="#iterating-over-collections-of-annotations" title="Link to this heading">¶</a></h3>
<p>The collection of <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects representing metadata annotations associated with particular phylgoenetic data objects can be accessed through the <code class="xref py py-attr docutils literal notranslate"><span class="pre">annotations</span></code> attribute of each particular object.</p>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;pythonidae.annotated.nexml&quot;</span><span class="p">,</span>
<span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
    <span class="nb">print</span> <span class="s2">&quot;The dataset has metadata annotation &#39;</span><span class="si">%s</span><span class="s2">&#39; with content &#39;</span><span class="si">%s</span><span class="s2">&#39;&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">a</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">tree_lists</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
    <span class="nb">print</span> <span class="s2">&quot;Tree &#39;</span><span class="si">%s</span><span class="s2">&#39; has metadata annotation &#39;</span><span class="si">%s</span><span class="s2">&#39; with content &#39;</span><span class="si">%s</span><span class="s2">&#39;&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">label</span><span class="p">,</span> <span class="n">a</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">a</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
</pre></div>
</div>
<p>will result in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">The</span> <span class="n">dataset</span> <span class="n">has</span> <span class="n">metadata</span> <span class="n">annotation</span> <span class="s1">&#39;description&#39;</span> <span class="k">with</span> <span class="n">content</span> <span class="s1">&#39;composite dataset of Pythonid sequences and trees&#39;</span>
<span class="n">The</span> <span class="n">dataset</span> <span class="n">has</span> <span class="n">metadata</span> <span class="n">annotation</span> <span class="s1">&#39;subject&#39;</span> <span class="k">with</span> <span class="n">content</span> <span class="s1">&#39;Pythonidae&#39;</span>
<span class="n">Tree</span> <span class="s1">&#39;0&#39;</span> <span class="n">has</span> <span class="n">metadata</span> <span class="n">annotation</span> <span class="s1">&#39;treeEstimator&#39;</span> <span class="k">with</span> <span class="n">content</span> <span class="s1">&#39;RAxML&#39;</span>
<span class="n">Tree</span> <span class="s1">&#39;0&#39;</span> <span class="n">has</span> <span class="n">metadata</span> <span class="n">annotation</span> <span class="s1">&#39;substitutionModel&#39;</span> <span class="k">with</span> <span class="n">content</span> <span class="s1">&#39;GTR+G+I&#39;</span>
</pre></div>
</div>
</section>
<section id="retrieving-annotations-by-search-criteria">
<h3>Retrieving Annotations By Search Criteria<a class="headerlink" href="#retrieving-annotations-by-search-criteria" title="Link to this heading">¶</a></h3>
<p>Instead of interating through every element in the <code class="xref py py-attr docutils literal notranslate"><span class="pre">annotations</span></code> attribute of data objects, you can use the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.findall" title="dendropy.datamodel.basemodel.AnnotationSet.findall"><code class="xref py py-meth docutils literal notranslate"><span class="pre">findall</span></code></a> method of the the <code class="xref py py-attr docutils literal notranslate"><span class="pre">annotations</span></code> object to return a <em>collection</em> of <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects that match the search or filter criteria specified in keyword arguments to the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.findall" title="dendropy.datamodel.basemodel.AnnotationSet.findall"><code class="xref py py-meth docutils literal notranslate"><span class="pre">findall</span></code></a> call.
These keyword arguments should specify attributes of <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> and the corresponding value to be matched.
Multiple keyword-value pairs can be specified, and only <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects that match <em>all</em> the criteria will be returned.</p>
<p>For example, the following returns a collection of annotations that have a name of “contributor”:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;contributor&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
    <span class="nb">print</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">=&#39;</span><span class="si">%s</span><span class="s2">&#39;&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">a</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
</pre></div>
</div>
<p>and will result in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">contributor</span><span class="o">=</span><span class="s1">&#39;Dahlgren T.G.&#39;</span>
<span class="n">contributor</span><span class="o">=</span><span class="s1">&#39;Baco A.&#39;</span>
<span class="n">contributor</span><span class="o">=</span><span class="s1">&#39;Smith C.&#39;</span>
<span class="n">contributor</span><span class="o">=</span><span class="s1">&#39;Glover A.&#39;</span>
<span class="n">contributor</span><span class="o">=</span><span class="s1">&#39;Altamira I.V.&#39;</span>
<span class="n">contributor</span><span class="o">=</span><span class="s1">&#39;Wiklund H.&#39;</span>
</pre></div>
</div>
<p>While the following returns a collection of annotations that are in the Dublin Core namespace:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">namespace</span><span class="o">=</span><span class="s2">&quot;http://purl.org/dc/elements/1.1/&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
    <span class="nb">print</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">=&#39;</span><span class="si">%s</span><span class="s2">&#39;&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">a</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
</pre></div>
</div>
<p>and results in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">subject</span><span class="o">=</span><span class="s1">&#39;wood-fall&#39;</span>
<span class="n">contributor</span><span class="o">=</span><span class="s1">&#39;Wiklund H.&#39;</span>
<span class="n">publisher</span><span class="o">=</span><span class="s1">&#39;Systematics and Biodiversity&#39;</span>
<span class="n">subject</span><span class="o">=</span><span class="s1">&#39;whale-fall&#39;</span>
<span class="n">contributor</span><span class="o">=</span><span class="s1">&#39;Dahlgren T.G.&#39;</span>
<span class="n">contributor</span><span class="o">=</span><span class="s1">&#39;Smith C.&#39;</span>
<span class="n">date</span><span class="o">=</span><span class="s1">&#39;2012-06-04&#39;</span>
<span class="n">subject</span><span class="o">=</span><span class="s1">&#39;polychaeta&#39;</span>
<span class="n">contributor</span><span class="o">=</span><span class="s1">&#39;Glover A.&#39;</span>
<span class="n">subject</span><span class="o">=</span><span class="s1">&#39;Ophryotrocha&#39;</span>
<span class="n">title</span><span class="o">=</span><span class="s1">&#39;Systematics and biodiversity of Ophryotrocha (Annelida, Dorvilleidae) with descriptions of six new species from deep-sea whale-fall and wood-fall habitats in the north-east Pacific&#39;</span>
<span class="n">subject</span><span class="o">=</span><span class="s1">&#39;New species&#39;</span>
<span class="n">subject</span><span class="o">=</span><span class="s1">&#39;molecular phylogeny&#39;</span>
<span class="n">contributor</span><span class="o">=</span><span class="s1">&#39;Altamira I.V.&#39;</span>
<span class="n">creator</span><span class="o">=</span><span class="s1">&#39;Wiklund H., Altamira I.V., Glover A., Smith C., Baco A., &amp; Dahlgren T.G.&#39;</span>
<span class="n">contributor</span><span class="o">=</span><span class="s1">&#39;Baco A.&#39;</span>
</pre></div>
</div>
<p>The following, in turn, searches for and suppresses printing of annotations that have a name prefix of “dc” <em>and</em> have empty values:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">name_prefix</span><span class="o">=</span><span class="s2">&quot;dc&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
    <span class="n">a</span><span class="o">.</span><span class="n">is_hidden</span> <span class="o">=</span> <span class="kc">True</span>
</pre></div>
</div>
<p>Modifying the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects in a returned collection modifies the metadata of the parent data object. For example, the following sets all the field values to upper case characters:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;contributor&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
    <span class="n">a</span><span class="o">.</span><span class="n">value</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">value</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;contributor&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
    <span class="nb">print</span> <span class="n">a</span><span class="o">.</span><span class="n">value</span>
</pre></div>
</div>
<p>and results in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">DAHLGREN</span> <span class="n">T</span><span class="o">.</span><span class="n">G</span><span class="o">.</span>
<span class="n">BACO</span> <span class="n">A</span><span class="o">.</span>
<span class="n">SMITH</span> <span class="n">C</span><span class="o">.</span>
<span class="n">GLOVER</span> <span class="n">A</span><span class="o">.</span>
<span class="n">ALTAMIRA</span> <span class="n">I</span><span class="o">.</span><span class="n">V</span><span class="o">.</span>
<span class="n">WIKLUND</span> <span class="n">H</span><span class="o">.</span>
</pre></div>
</div>
<p>The collection returned by the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.findall" title="dendropy.datamodel.basemodel.AnnotationSet.findall"><code class="xref py py-meth docutils literal notranslate"><span class="pre">findall</span></code></a> method is an object of type <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a>.
However, while modifying <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects in this collection will result in the metadata of the parent object being modified (as in the previous example), adding new annotations to this returned collection will <em>not</em>  add them to the collection of metadata annotations of the parent object.
Thus, the following example shows that the size of the annotations collection associated with the dataset is unchanged by adding new annotations to the results of a <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.findall" title="dendropy.datamodel.basemodel.AnnotationSet.findall"><code class="xref py py-meth docutils literal notranslate"><span class="pre">findall</span></code></a> call:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="nb">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">namespace</span><span class="o">=</span><span class="s2">&quot;http://purl.org/dc/elements/1.1/&quot;</span><span class="p">)</span>
<span class="n">results</span><span class="o">.</span><span class="n">add_new</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;color&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;blue&quot;</span><span class="p">)</span>
<span class="n">results</span><span class="o">.</span><span class="n">add_new</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;height&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;100&quot;</span><span class="p">)</span>
<span class="n">results</span><span class="o">.</span><span class="n">add_new</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;length&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;200&quot;</span><span class="p">)</span>
<span class="n">results</span><span class="o">.</span><span class="n">add_new</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;width&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;50&quot;</span><span class="p">)</span>
<span class="nb">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
</pre></div>
</div>
<p>The above produces:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">30</span>
<span class="mi">30</span>
</pre></div>
</div>
<p>As can be seen, no new annotations are added to the data set metadata.</p>
<p>If no matching <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects are found then the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a> that is returned is empty.</p>
<p>If <em>no</em> keyword arguments are passed to <code class="xref py py-meth docutils literal notranslate"><span class="pre">findall</span></code>, then <em>all</em> annotations are returned:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">findall</span><span class="p">()</span>
<span class="nb">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">results</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
</pre></div>
</div>
<p>The above produces:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kc">True</span>
</pre></div>
</div>
</section>
<section id="retrieving-a-single-annotation-by-search-criteria">
<h3>Retrieving a Single Annotation By Search Criteria<a class="headerlink" href="#retrieving-a-single-annotation-by-search-criteria" title="Link to this heading">¶</a></h3>
<p>The <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.find" title="dendropy.datamodel.basemodel.AnnotationSet.find"><code class="xref py py-meth docutils literal notranslate"><span class="pre">find</span></code></a> method of the the <code class="xref py py-attr docutils literal notranslate"><span class="pre">annotations</span></code> object return a the <em>first</em> <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> object that matches the search or filter criteria specified in keyword arguments to the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.findall" title="dendropy.datamodel.basemodel.AnnotationSet.findall"><code class="xref py py-meth docutils literal notranslate"><span class="pre">findall</span></code></a> call.
These keyword arguments should specify attributes of <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> and the corresponding value to be matched.
Multiple keyword-value pairs can be specified, and only the first <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> object that matches <em>all</em> the criteria will be returned.</p>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="nb">print</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;contributor&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>and will result in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">contributor</span><span class="o">=</span><span class="s1">&#39;Dahlgren T.G.&#39;</span>
</pre></div>
</div>
<p>While the following returns the first annotation in the Dublin Core namespace:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="nb">print</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">namespace</span><span class="o">=</span><span class="s2">&quot;http://purl.org/dc/elements/1.1/&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>and results in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">subject</span><span class="o">=</span><span class="s1">&#39;wood-fall&#39;</span>
</pre></div>
</div>
<p>If no matching <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects are found then a default of <code class="docutils literal notranslate"><span class="pre">None</span></code> is returned:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;author&quot;</span><span class="p">)</span>
<span class="go">None</span>
</pre></div>
</div>
<p>Unlike <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.findall" title="dendropy.datamodel.basemodel.AnnotationSet.findall"><code class="xref py py-meth docutils literal notranslate"><span class="pre">findall</span></code></a>, it is invalid to call <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.find" title="dendropy.datamodel.basemodel.AnnotationSet.find"><code class="xref py py-meth docutils literal notranslate"><span class="pre">find</span></code></a> with no search criteria keyword arguments, and an <code class="docutils literal notranslate"><span class="pre">TypeError</span></code> exception will be raised.</p>
</section>
<section id="retrieving-the-value-of-a-single-annotation">
<h3>Retrieving the Value of a Single Annotation<a class="headerlink" href="#retrieving-the-value-of-a-single-annotation" title="Link to this heading">¶</a></h3>
<p>For convenience, the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.get_value" title="dendropy.datamodel.basemodel.AnnotationSet.get_value"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_value</span></code></a>, method is provided.
This will search the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a> for the <em>first</em> <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> that has its name field equal to the first argument passed to the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.get_value" title="dendropy.datamodel.basemodel.AnnotationSet.get_value"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_value</span></code></a> method, and return its value.
If no match is found, the second argument is returned (or <code class="docutils literal notranslate"><span class="pre">None</span></code>, if no second argument is specified).
Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">get_value</span><span class="p">(</span><span class="s2">&quot;subject&quot;</span><span class="p">)</span>
<span class="go">molecular phylogeny</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">get_value</span><span class="p">(</span><span class="s2">&quot;creator&quot;</span><span class="p">)</span>
<span class="go">Yoder A.D., &amp; Yang Z.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">get_value</span><span class="p">(</span><span class="s2">&quot;generator&quot;</span><span class="p">)</span>
<span class="go">None</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">tree</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">get_value</span><span class="p">(</span><span class="s2">&quot;generator&quot;</span><span class="p">,</span> <span class="s2">&quot;unspecified&quot;</span><span class="p">)</span>
<span class="go">unspecified</span>
</pre></div>
</div>
</section>
<section id="transforming-annotations-to-a-dictionary">
<h3>Transforming Annotations to a Dictionary<a class="headerlink" href="#transforming-annotations-to-a-dictionary" title="Link to this heading">¶</a></h3>
<p>In some applications, it might be more convenient to work with dictionaries rather than <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a> objects.
The <code class="xref py py-meth docutils literal notranslate"><span class="pre">values_as_dict</span></code> methods creates a dictionary populated with key-value pairs from the collection.
By default, the keys are the <code class="docutils literal notranslate"><span class="pre">name</span></code> attribute of the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> object and the values are the <code class="docutils literal notranslate"><span class="pre">value</span></code> attribute.
Thus, the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">values_as_dict</span><span class="p">()</span>
<span class="nb">print</span> <span class="n">a</span>
</pre></div>
</div>
<p>results in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="s1">&#39;volume&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;doi&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;date&#39;</span><span class="p">:</span> <span class="s1">&#39;2012-06-04&#39;</span><span class="p">,</span>
<span class="s1">&#39;bibliographicCitation&#39;</span><span class="p">:</span> <span class="s1">&#39;Wiklund H., Altamira I.V., Glover A., Smith C., Baco A., &amp; Dahlgren T.G. 2012. Systematics and biodiversity of Ophryotrocha (Annelida, Dorvilleidae) with descriptions of six new species from deep-sea whale-fall and wood-fall habitats in the north-east Pacific. Systematics and Biodiversity, .&#39;</span><span class="p">,</span>
<span class="s1">&#39;changeNote&#39;</span><span class="p">:</span> <span class="s1">&#39;Generated on Wed Jun 06 11:02:45 EDT 2012&#39;</span><span class="p">,</span>
<span class="s1">&#39;creator&#39;</span><span class="p">:</span> <span class="s1">&#39;Wiklund H., Altamira I.V., Glover A., Smith C., Baco A., &amp; Dahlgren T.G.&#39;</span><span class="p">,</span>
<span class="s1">&#39;section&#39;</span><span class="p">:</span> <span class="s1">&#39;Study&#39;</span><span class="p">,</span>
<span class="s1">&#39;title&#39;</span><span class="p">:</span> <span class="s1">&#39;Systematics and biodiversity of Ophryotrocha (Annelida, Dorvilleidae) with descriptions of six new species from deep-sea whale-fall and wood-fall habitats in the north-east Pacific&#39;</span><span class="p">,</span>
<span class="s1">&#39;publisher&#39;</span><span class="p">:</span> <span class="s1">&#39;Systematics and Biodiversity&#39;</span><span class="p">,</span>
<span class="s1">&#39;identifier.study.tb1&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
<span class="s1">&#39;number&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;identifier.study&#39;</span><span class="p">:</span> <span class="s1">&#39;12713&#39;</span><span class="p">,</span>
<span class="s1">&#39;modificationDate&#39;</span><span class="p">:</span> <span class="s1">&#39;2012-06-04&#39;</span><span class="p">,</span>
<span class="s1">&#39;historyNote&#39;</span><span class="p">:</span> <span class="s1">&#39;Mapped from TreeBASE schema using org.cipres.treebase.domain.nexus.nexml.NexmlDocumentWriter@645f9132 $Rev: 1060 $&#39;</span><span class="p">,</span>
<span class="s1">&#39;publicationDate&#39;</span><span class="p">:</span> <span class="s1">&#39;2012&#39;</span><span class="p">,</span>
<span class="s1">&#39;contributor&#39;</span><span class="p">:</span> <span class="s1">&#39;Wiklund H.&#39;</span><span class="p">,</span>
<span class="s1">&#39;publicationName&#39;</span><span class="p">:</span> <span class="s1">&#39;Systematics and Biodiversity&#39;</span><span class="p">,</span>
<span class="s1">&#39;creationDate&#39;</span><span class="p">:</span> <span class="s1">&#39;2012-05-09&#39;</span><span class="p">,</span>
<span class="s1">&#39;title.study&#39;</span><span class="p">:</span> <span class="s1">&#39;Systematics and biodiversity of Ophryotrocha (Annelida, Dorvilleidae) with descriptions of six new species from deep-sea whale-fall and wood-fall habitats in the north-east Pacific&#39;</span><span class="p">,</span>
<span class="s1">&#39;subject&#39;</span><span class="p">:</span> <span class="s1">&#39;molecular phylogeny&#39;</span><span class="p">}</span>
</pre></div>
</div>
<p>Note that no attempt is made to prevent or account for key collision: <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> with the same name value will overwrite each other in the dictionary.
Custom control of the dictionary key/value generation can be specified via keyword arguments:</p>
<blockquote>
<div><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">key_attr</span></code></dt><dd><p>String specifying an Annotation object attribute name to be used
as keys for the dictionary.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">key_func</span></code></dt><dd><p>Function that takes an Annotation object as an argument and returns
the value to be used as a key for the dictionary.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">value_attr</span></code></dt><dd><p>String specifying an Annotation object attribute name to be used
as values for the dictionary.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">value_func</span></code></dt><dd><p>Function that takes an Annotation object as an argument and returns
the value to be used as a value for the dictionary.</p>
</dd>
</dl>
</div></blockquote>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">values_as_dict</span><span class="p">(</span><span class="n">key_attr</span><span class="o">=</span><span class="s2">&quot;prefixed_name&quot;</span><span class="p">)</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">values_as_dict</span><span class="p">(</span><span class="n">key_attr</span><span class="o">=</span><span class="s2">&quot;prefixed_name&quot;</span><span class="p">,</span> <span class="n">value_attr</span><span class="o">=</span><span class="s2">&quot;namespace&quot;</span><span class="p">)</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">values_as_dict</span><span class="p">(</span><span class="n">key_func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">a</span><span class="p">:</span> <span class="n">a</span><span class="o">.</span><span class="n">namespace</span> <span class="o">+</span> <span class="n">a</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">values_as_dict</span><span class="p">(</span><span class="n">key_func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">a</span><span class="p">:</span> <span class="n">a</span><span class="o">.</span><span class="n">namespace</span> <span class="o">+</span> <span class="n">a</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
        <span class="n">value_attr</span><span class="o">=</span><span class="s2">&quot;value&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>As the collection returned by the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.findall" title="dendropy.datamodel.basemodel.AnnotationSet.findall"><code class="xref py py-meth docutils literal notranslate"><span class="pre">findall</span></code></a> method is an object of type <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a>, this can also be transformed to a dictionary.
For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">name_prefix</span><span class="o">=</span><span class="s2">&quot;dc&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">values_as_dict</span><span class="p">()</span>
<span class="nb">print</span> <span class="n">a</span>
</pre></div>
</div>
<p>will result in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="s1">&#39;publisher&#39;</span><span class="p">:</span> <span class="s1">&#39;Systematics and Biodiversity&#39;</span><span class="p">,</span>
<span class="s1">&#39;creator&#39;</span><span class="p">:</span> <span class="s1">&#39;Wiklund H., Altamira I.V., Glover A., Smith C., Baco A., &amp; Dahlgren T.G.&#39;</span><span class="p">,</span>
<span class="s1">&#39;title&#39;</span><span class="p">:</span> <span class="s1">&#39;Systematics and biodiversity of Ophryotrocha (Annelida, Dorvilleidae) with descriptions of six new species from deep-sea whale-fall and wood-fall habitats in the north-east Pacific&#39;</span><span class="p">,</span>
<span class="s1">&#39;date&#39;</span><span class="p">:</span> <span class="s1">&#39;2012-06-04&#39;</span><span class="p">,</span>
<span class="s1">&#39;contributor&#39;</span><span class="p">:</span> <span class="s1">&#39;Baco A.&#39;</span><span class="p">,</span>
<span class="s1">&#39;subject&#39;</span><span class="p">:</span> <span class="s1">&#39;molecular phylogeny&#39;</span><span class="p">}</span>
</pre></div>
</div>
<p>Note how only one entry for “contributor” is present: the others were overwritten/replaced.</p>
<p>Adding to, deleting, or modifying either the keys or the values of the dictionary returned by <code class="xref py py-meth docutils literal notranslate"><span class="pre">values_as_dict</span></code> in <em>no way</em> changes any of the original metadata: it is serves as snapshot copy of literal values of the metadata.</p>
</section>
<section id="deleting-or-removing-metadata-annotations">
<h3>Deleting or Removing Metadata Annotations<a class="headerlink" href="#deleting-or-removing-metadata-annotations" title="Link to this heading">¶</a></h3>
<p>The <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.drop" title="dendropy.datamodel.basemodel.AnnotationSet.drop"><code class="xref py py-meth docutils literal notranslate"><span class="pre">drop</span></code></a> method of <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a> objects takes search criteria similar to <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.findall" title="dendropy.datamodel.basemodel.AnnotationSet.findall"><code class="xref py py-meth docutils literal notranslate"><span class="pre">findall</span></code></a>, but instead of returning the matched  <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects, it <em>removes</em> them from the parent collection.
For example, the following removes all metadata annotations with the name prefix “dc” from the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object <code class="docutils literal notranslate"><span class="pre">ds</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="nb">print</span> <span class="s2">&quot;Original: </span><span class="si">%d</span><span class="s2"> items&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
<span class="n">removed</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">name_prefix</span><span class="o">=</span><span class="s2">&quot;dc&quot;</span><span class="p">)</span>
<span class="nb">print</span> <span class="s2">&quot;Removed: </span><span class="si">%d</span><span class="s2"> items&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">removed</span><span class="p">)</span>
<span class="nb">print</span> <span class="s2">&quot;Current: </span><span class="si">%d</span><span class="s2"> items&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
</pre></div>
</div>
<p>and results in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Original</span><span class="p">:</span> <span class="mi">30</span> <span class="n">items</span>
<span class="n">Removed</span><span class="p">:</span> <span class="mi">16</span> <span class="n">items</span>
<span class="n">Current</span><span class="p">:</span> <span class="mi">14</span> <span class="n">items</span>
</pre></div>
</div>
<p>As can be seen, the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.drop" title="dendropy.datamodel.basemodel.AnnotationSet.drop"><code class="xref py py-meth docutils literal notranslate"><span class="pre">drop</span></code></a> method returns the individual <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> removed as a new <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet" title="dendropy.datamodel.basemodel.AnnotationSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationSet</span></code></a> collection.
This is useful if you still want to use the removed <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.Annotation" title="dendropy.datamodel.basemodel.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a> objects elsewhere.</p>
<p>As with the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.findall" title="dendropy.datamodel.basemodel.AnnotationSet.findall"><code class="xref py py-meth docutils literal notranslate"><span class="pre">findall</span></code></a> method, multiple keyword criteria can be specified:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">name_prefix</span><span class="o">=</span><span class="s2">&quot;dc&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;contributor&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>In addition, again similar in behavior to the <a class="reference internal" href="../library/basemodel.html#dendropy.datamodel.basemodel.AnnotationSet.findall" title="dendropy.datamodel.basemodel.AnnotationSet.findall"><code class="xref py py-meth docutils literal notranslate"><span class="pre">findall</span></code></a> method, <em>no</em> keyword arguments result in <em>all</em> the annotations being removed.
Thus, the following results in all metadata annotations being deleted from the <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> object <code class="docutils literal notranslate"><span class="pre">ds</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dendropy</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">dendropy</span><span class="o">.</span><span class="n">DataSet</span><span class="o">.</span><span class="n">get_from_path</span><span class="p">(</span><span class="s2">&quot;sample1.xml&quot;</span><span class="p">,</span>
    <span class="s2">&quot;nexml&quot;</span><span class="p">)</span>
<span class="nb">print</span> <span class="s2">&quot;Original: </span><span class="si">%d</span><span class="s2"> items&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
<span class="n">removed</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">drop</span><span class="p">()</span>
<span class="nb">print</span> <span class="s2">&quot;Removed: </span><span class="si">%d</span><span class="s2"> items&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">removed</span><span class="p">)</span>
<span class="nb">print</span> <span class="s2">&quot;Current: </span><span class="si">%d</span><span class="s2"> items&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
</pre></div>
</div>
<p>and results in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Original</span><span class="p">:</span> <span class="mi">30</span> <span class="n">items</span>
<span class="n">Removed</span><span class="p">:</span> <span class="mi">30</span> <span class="n">items</span>
<span class="n">Current</span><span class="p">:</span> <span class="mi">0</span> <span class="n">items</span>
</pre></div>
</div>
</section>
</section>
<section id="writing-or-saving-metadata">
<h2>Writing or Saving Metadata<a class="headerlink" href="#writing-or-saving-metadata" title="Link to this heading">¶</a></h2>
<p>When writing to NeXML format, all metadata annotations are preserved and can be fully round-tripped.
Currently, this is the only data format that allows for robust treatment of metadata.</p>
<p>Due to the fundamental limitations of the NEXUS/Newick format, metadata handling in this format is limited and rather idiosyncratic.
Currently, metadata will be written out as name-value pairs (separated by “=”) in ampersand-prepended comments associated with the particular phylogenetic data object.
This syntax corresponds to the BEAST or FigTree style of metadata annotation.
However, this association might not be preserved.
For example, metadata annotations associated with edges and nodes of trees will be written out fully in NEXUS and NEWICK formats, but when read in again will all be associated with nodes.
The keyword argument <code class="docutils literal notranslate"><span class="pre">annotations_as_nhx=True</span></code> passed to the call to write the data in NEXUS/NEWICK format will result in a double ampersand prefix to the comment, thus (partially) conforming to NHX specifications.
Metadata associated with <a class="reference internal" href="../library/datasetmodel.html#dendropy.datamodel.datasetmodel.DataSet" title="dendropy.datamodel.datasetmodel.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSet</span></code></a> objects will be written in out in the same BEAST/FigTree/NHX syntax at the top of the file, while metadata associated with <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.TaxonNamespace" title="dendropy.datamodel.taxonmodel.TaxonNamespace"><code class="xref py py-class docutils literal notranslate"><span class="pre">TaxonNamespace</span></code></a> and <a class="reference internal" href="../library/taxonmodel.html#dendropy.datamodel.taxonmodel.Taxon" title="dendropy.datamodel.taxonmodel.Taxon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Taxon</span></code></a> objects will be written out immediately after the start of the Taxa Block and taxon labels respectively.
This is very fragile: for example, a metadata annotation <em>before</em> a taxon label will be associated with the <em>previous</em> taxon when being read in again.
As noted above, if metadata annotations are important for yourself, your workflow, or your task, then the NeXML format should be used rather than NEXUS or NEWICK.</p>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div style="text-align: center; padding-top: 20px; padding-bottom: 5px; width: 100%;">
    <a href="../index.html"><img src="../_static/dendropy_logo.png" /></a>
</div><div style="clear:both; width: 100%; height:1px;"></div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h3><a href="../index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Working with Metadata Annotations</a><ul>
<li><a class="reference internal" href="#overview-of-the-infrastructure-for-metadata-annotation-in-dendropy">Overview of the Infrastructure for Metadata Annotation in DendroPy</a></li>
<li><a class="reference internal" href="#metadata-annotation-creation">Metadata Annotation Creation</a><ul>
<li><a class="reference internal" href="#reading-data-from-an-external-source">Reading Data from an External Source</a></li>
<li><a class="reference internal" href="#direct-composition-with-literal-values">Direct Composition with Literal Values</a></li>
<li><a class="reference internal" href="#dynamically-binding-annotation-values-to-object-attribute-values">Dynamically Binding Annotation Values to Object Attribute Values</a></li>
<li><a class="reference internal" href="#adding-citation-metadata">Adding Citation Metadata</a></li>
<li><a class="reference internal" href="#copying-metadata-annotations-from-one-phylogenetic-data-object-to-another">Copying Metadata Annotations from One Phylogenetic Data Object to Another</a></li>
</ul>
</li>
<li><a class="reference internal" href="#metadata-annotation-access-and-manipulation">Metadata Annotation Access and Manipulation</a><ul>
<li><a class="reference internal" href="#iterating-over-collections-of-annotations">Iterating Over Collections of Annotations</a></li>
<li><a class="reference internal" href="#retrieving-annotations-by-search-criteria">Retrieving Annotations By Search Criteria</a></li>
<li><a class="reference internal" href="#retrieving-a-single-annotation-by-search-criteria">Retrieving a Single Annotation By Search Criteria</a></li>
<li><a class="reference internal" href="#retrieving-the-value-of-a-single-annotation">Retrieving the Value of a Single Annotation</a></li>
<li><a class="reference internal" href="#transforming-annotations-to-a-dictionary">Transforming Annotations to a Dictionary</a></li>
<li><a class="reference internal" href="#deleting-or-removing-metadata-annotations">Deleting or Removing Metadata Annotations</a></li>
</ul>
</li>
<li><a class="reference internal" href="#writing-or-saving-metadata">Writing or Saving Metadata</a></li>
</ul>
</li>
</ul>

  </div><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h4>Previous topic</h4>
    <p class="topless"><a href="datasets.html"
                          title="previous chapter">Data Sets</a></p>
  </div>
  <div>
    <h4>Next topic</h4>
    <p class="topless"><a href="genbank.html"
                          title="next chapter">Working with GenBank Molecular Sequence Databases</a></p>
  </div><div style="clear:both; width: 100%; height:1px;"></div>
    <!-- Documentation -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Documentation</h3>
        <ul>
            <li><a href="downloading.html">Downloading and Installing DendroPy</a></li>
            <li><a href="index.html">The DendroPy Primer</a></li>
            <li><a href="../library/index.html">Library API Reference</a></li>
            <li>
                <a href="../schemas/index.html">Schemas</a>
                <ul>
                    <li><a href="../schemas/fasta.html">FASTA</a></li>
                    <li><a href="../schemas/newick.html">Newick</a></li>
                    <li><a href="../schemas/nexml.html">NeXML</a></li>
                    <li><a href="../schemas/nexus.html">Nexus</a></li>
                    <li><a href="../schemas/phylip.html">PHYLIP</a></li>
                </ul>
            </li>
            <li>
                <a href="../programs/index.html">Programs</a>
                <ul>
                    <li><a href="../programs/sumtrees.html">SumTrees</a></li>
                </ul>
            </li>
            <li><a href="../glossary.html">Glossary and Terminological Reference</a></li>
            <li><a href="../developer.html">Developer Guide</a></li>
            <li><a href="../planning.html">Ongoing Development</a></li>
            <li><a href="../changes.html">Change History</a></li>
        </ul>
    </div>

    <!-- Downloads -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Obtaining</h3>
        <ul>
            <li><a target="_blank" href="http://pypi.python.org/pypi/DendroPy">Install from the Python Package Index</a></li>
            <li><a target="_blank" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.4.tar.gz">Download the Source Code Archive</a></li>
            <li><a target="_blank" href="http://github.com/jeetsukumaran/DendroPy">Clone the Source Code Repository</a></li>
        </ul>
    </div>

    <!-- Discussions -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Discussion</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot; DendroPy Users&quot; </a>group to follow and participate in discussion, troubleshooting, help, information, suggestions, etc. on the usage and development of the DendroPy phylogenetic computing library.</p>
            <form action="http://groups.google.com/group/dendropy-users/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot;dendropy-users&quot;</a> group, or click <a href="http://groups.google.com/group/dendropy-users?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Announcements -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Announcements</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; DendroPy Announcements&quot; </a>group to receive announcements of new releases, updates, changes and other news of interest to DendroPy users and developers.</p>
            <form action="http://groups.google.com/group/dendropy-announce/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; dendropy-announce&quot; </a>group, or click <a href="http://groups.google.com/group/dendropy-announce?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Development -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative; padding-bottom: 15px; margin-bottom:5px;">
        <h3><span style="text-align: left"><a href="https://github.com/jeetsukumaran/DendroPy/">Development</a></span><a href="https://github.com/jeetsukumaran/DendroPy/"><span style="position: absolute; right: 0; top: 10px "><img src="../_static/Octocat.png" height="30px" alt="GitHub" /></span></a></h3>
        <div style="margin-top: 15px;">
            <!-- <a href="https://github.com/jeetsukumaran/DendroPy/issues">Issues</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/subscription">Watch</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/fork">Fork</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/stargazers">Star</a> &bull; <a href="https://github.com/jeetsukumaran/">Follow</a> -->
            <ul>
                <li>                <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/issues"><code>Issues</code></a></span> <span style="font-style: italic; font-size:80%;"> - Report bugs or request features</span></li>
                <li>     <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/subscription"><code>&nbsp;Watch</code></a></span> <span style="font-style: italic; font-size:80%;"> - Follow development activity</span></li>
                <li>        <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/fork"><code>&nbsp;&nbsp;Fork</code></a></span> <span style="font-style: italic; font-size:80%;"> - Contribute and collaborate</span></li>
                <li>  <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/stargazers"><code>&nbsp;&nbsp;Star</code></a></span> <span style="font-style: italic; font-size:80%;"> - Throw some glitter, add some glamour</span></li>
        </div>
    </div>


        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="genbank.html" title="Working with GenBank Molecular Sequence Databases"
             >next</a> |</li>
        <li class="right" >
          <a href="datasets.html" title="Data Sets"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.4 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >The DendroPy Primer</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Working with Metadata Annotations</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2009-2025, Jeet Sukumaran and Mark T. Holder.
    </div>
  </body>
</html>