<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>SumTrees: Phylogenetic Tree Summarization and Annotation &#8212; DendroPy 5.0.2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/rtd.css?v=cedc6a2f" />
    
    <script src="../_static/documentation_options.js?v=5df4c8d1"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Glossary and Terminological Reference" href="../glossary.html" />
    <link rel="prev" title="Programs and Applications" href="index.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../glossary.html" title="Glossary and Terminological Reference"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Programs and Applications"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Programs and Applications</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">SumTrees: Phylogenetic Tree Summarization and Annotation</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="sumtrees-phylogenetic-tree-summarization-and-annotation">
<h1>SumTrees: Phylogenetic Tree Summarization and Annotation<a class="headerlink" href="#sumtrees-phylogenetic-tree-summarization-and-annotation" title="Link to this heading">¶</a></h1>
<section id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Link to this heading">¶</a></h2>
<p><a class="reference external" href="sumtrees.html">SumTrees</a> is a program by <a class="reference external" href="http://www.jeetworks.org/about">Jeet Sukumaran</a> and <a class="reference external" href="http://people.ku.edu/~mtholder">Mark T. Holder</a> to summarize non-parameteric bootstrap or Bayesian posterior probability support for splits or clades on phylogenetic trees.</p>
<p>The basis of the support assessment is typically given by a set of non-parametric bootstrap replicate tree samples produced by programs such as GARLI or RAxML, or by a set of MCMC tree samples produced by programs such as Mr. Bayes or BEAST.
The proportion of trees out of the samples in which a particular split is found is taken to be the degree of support for that split as indicated by the samples.
The samples that are the basis of the support can be distributed across multiple files, and a burn-in option allows for an initial number of trees in each file to be excluded from the analysis if they are not considered to be drawn from the true support distribution.</p>
<p>The support for the splits will be mapped onto one or more target trees either in terms of node labels or branch lengths.
The target trees can be supplied by yourself, or, if no target trees are given, then a
a <em>summary</em> tree can be constructed.
This summary tree can be a Maximum Clade Credibility Topology (i.e., MCCT, a topology that maximizes the product of the the clade posterior probabilities), the majority-rule clade consensus tree, or some other type.
If a majority-rule consensus tree is selected, you have the option of specifying the minimum posterior probability or proportional frequency threshold for a clade to be included on the consensus tree.</p>
<p>By default SumTrees will provide summaries of edge lengths (i.e., mean, median, standard deviation, range, 95% HPD, 5% and 95% quantiles, etc.) as special node comments. These can be visualized in <a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a> by, for example, checking “Node Labels”, then selecting one of “length_mean”, “length_median”, “length_sd”, “length_hpd95”, etc.
If the trees are ultrametric and the “<code class="docutils literal notranslate"><span class="pre">--summarize-node-ages</span></code>” flag is used, or edge lengths are set so that node ages on the output trees correspond to be mean or median of the node ages of the input trees (”<code class="docutils literal notranslate"><span class="pre">--edges=mean-age</span></code>” or “<code class="docutils literal notranslate"><span class="pre">--edges=median-age</span></code>”), then node ages will be summarized as well. In all cases, the flag “<code class="docutils literal notranslate"><span class="pre">--suppress-annotations</span></code>” will suppress calculation and output of these summaries.</p>
<p>If you are processing multiple source files and you have multiple cores available on your machine, you can specify the “<code class="docutils literal notranslate"><span class="pre">-M</span></code>” flag to use all the cores or, e.g., “<code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">4</span></code>” to use 4 cores.
Using multiple cores will, of course, speed up processing of your files.</p>
</section>
<section id="where-to-find-the-package">
<h2>Where to Find the Package<a class="headerlink" href="#where-to-find-the-package" title="Link to this heading">¶</a></h2>
<p>SumTrees is distributed and installed as part of the <a class="reference external" href="https://jeetsukumaran.github.io/DendroPy/">DendroPy</a> phylogenetic computing library, which has its homepage here:</p>
<blockquote>
<div><p><a href="#id3"><span class="problematic" id="id4">|dendropy_homepage_url|</span></a></p>
</div></blockquote>
<p>with the main download page here:</p>
<blockquote>
<div><p><a class="reference external" href="http://pypi.python.org/pypi/DendroPy">http://pypi.python.org/pypi/DendroPy</a></p>
</div></blockquote>
</section>
<section id="how-to-install-the-package">
<h2>How to Install the Package<a class="headerlink" href="#how-to-install-the-package" title="Link to this heading">¶</a></h2>
<p>DendroPy is fully easy installable and can be installed using <a class="reference external" href="http://pypi.python.org/pypi/pip">pip</a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python -m pip install dendropy
</pre></div>
</div>
<p>If you do not have <a class="reference external" href="http://pypi.python.org/pypi/pip">pip</a> installed, you should <em>definitely</em> <a class="reference external" href="https://pip.pypa.io/en/latest/installing.html">install it</a> !
You may wish to use the “<code class="docutils literal notranslate"><span class="pre">--user</span></code>” flag for a local user install:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python3 -m pip install --user dendropy
</pre></div>
</div>
<p>These, and other ways of obtaining and installing DendroPy (e.g., by downloading the <a class="reference external" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.2.tar.gz">DendroPy source code archive</a>, or by cloning the <a class="reference external" href="http://github.com/jeetsukumaran/DendroPy">DendroPy Git repository</a>), are discussed in detail in the “<a class="reference internal" href="../downloading.html"><span class="doc">Downloading and Installing DendroPy</span></a>” section.</p>
</section>
<section id="checking-the-installation">
<h2>Checking the Installation<a class="headerlink" href="#checking-the-installation" title="Link to this heading">¶</a></h2>
<p>If the installation was successful, you should be able to type “<code class="docutils literal notranslate"><span class="pre">sumtrees</span></code>” in the terminal window and see something like the following (with possibly a different date or version number):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/==========================================================================</span>\
<span class="o">|</span>                                 <span class="n">SumTrees</span>                                 <span class="o">|</span>
<span class="o">|</span>                     <span class="n">Phylogenetic</span> <span class="n">Tree</span> <span class="n">Summarization</span>                      <span class="o">|</span>
<span class="o">|</span>                       <span class="n">Version</span> <span class="mf">4.0.0</span> <span class="p">(</span><span class="n">Jan</span> <span class="mi">31</span> <span class="mi">2015</span><span class="p">)</span>                        <span class="o">|</span>
<span class="o">|</span>                   <span class="n">By</span> <span class="n">Jeet</span> <span class="n">Sukumaran</span> <span class="ow">and</span> <span class="n">Mark</span> <span class="n">T</span><span class="o">.</span> <span class="n">Holder</span>                   <span class="o">|</span>
<span class="o">|</span>    <span class="n">Using</span><span class="p">:</span> <span class="n">DendroPy</span> <span class="mf">4.0.0</span><span class="o">.</span><span class="n">dev</span> <span class="p">(</span><span class="n">DendroPy4</span><span class="o">-</span><span class="n">feea2b0</span><span class="p">,</span> <span class="mi">2015</span><span class="o">-</span><span class="mi">06</span><span class="o">-</span><span class="mi">02</span> <span class="mi">20</span><span class="p">:</span><span class="mi">13</span><span class="p">:</span><span class="mi">49</span><span class="p">)</span>    <span class="o">|</span>
<span class="o">+--------------------------------------------------------------------------+</span>
<span class="o">|</span>                                 <span class="n">Citation</span>                                 <span class="o">|</span>
<span class="o">|</span>                                 <span class="o">~~~~~~~~</span>                                 <span class="o">|</span>
<span class="o">|</span> <span class="n">If</span> <span class="nb">any</span> <span class="n">stage</span> <span class="n">of</span> <span class="n">your</span> <span class="n">work</span> <span class="ow">or</span> <span class="n">analyses</span> <span class="n">relies</span> <span class="n">on</span> <span class="n">code</span> <span class="ow">or</span> <span class="n">programs</span> <span class="kn">from</span>    <span class="o">|</span>
<span class="o">|</span> <span class="n">this</span> <span class="n">library</span><span class="p">,</span> <span class="n">either</span> <span class="n">directly</span> <span class="ow">or</span> <span class="n">indirectly</span> <span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span><span class="p">,</span> <span class="n">through</span> <span class="n">usage</span> <span class="n">of</span> <span class="n">your</span> <span class="o">|</span>
<span class="o">|</span> <span class="n">own</span> <span class="ow">or</span> <span class="n">third</span><span class="o">-</span><span class="n">party</span> <span class="n">programs</span><span class="p">,</span> <span class="n">pipelines</span><span class="p">,</span> <span class="ow">or</span> <span class="n">toolkits</span> <span class="n">which</span> <span class="n">use</span><span class="p">,</span> <span class="n">rely</span> <span class="n">on</span><span class="p">,</span>  <span class="o">|</span>
<span class="o">|</span> <span class="n">incorporate</span><span class="p">,</span> <span class="ow">or</span> <span class="n">are</span> <span class="n">otherwise</span> <span class="n">primarily</span> <span class="n">derivative</span> <span class="n">of</span> <span class="n">code</span><span class="o">/</span><span class="n">programs</span> <span class="ow">in</span>   <span class="o">|</span>
<span class="o">|</span> <span class="n">this</span> <span class="n">library</span><span class="p">),</span> <span class="n">please</span> <span class="n">cite</span><span class="p">:</span>                                              <span class="o">|</span>
<span class="o">|</span>                                                                          <span class="o">|</span>
<span class="o">|</span>   <span class="n">Sukumaran</span><span class="p">,</span> <span class="n">J</span> <span class="ow">and</span> <span class="n">MT</span> <span class="n">Holder</span><span class="o">.</span> <span class="mf">2010.</span> <span class="n">DendroPy</span><span class="p">:</span> <span class="n">a</span> <span class="n">Python</span> <span class="n">library</span> <span class="k">for</span>       <span class="o">|</span>
<span class="o">|</span>     <span class="n">phylogenetic</span> <span class="n">computing</span><span class="o">.</span> <span class="n">Bioinformatics</span> <span class="mi">26</span><span class="p">:</span> <span class="mi">1569</span><span class="o">-</span><span class="mf">1571.</span>                <span class="o">|</span>
<span class="o">|</span>                                                                          <span class="o">|</span>
<span class="o">|</span>   <span class="n">Sukumaran</span><span class="p">,</span> <span class="n">J</span> <span class="ow">and</span> <span class="n">MT</span> <span class="n">Holder</span><span class="o">.</span> <span class="n">SumTrees</span><span class="p">:</span> <span class="n">Phylogenetic</span> <span class="n">Tree</span> <span class="n">Summarization</span><span class="o">.</span> <span class="o">|</span>
<span class="o">|</span>     <span class="mf">4.0.0</span> <span class="p">(</span><span class="n">Jan</span> <span class="mi">31</span> <span class="mi">2015</span><span class="p">)</span><span class="o">.</span> <span class="n">Available</span> <span class="n">at</span>                                    <span class="o">|</span>
<span class="o">|</span>     <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">jeetsukumaran</span><span class="o">/</span><span class="n">DendroPy</span><span class="o">.</span>                           <span class="o">|</span>
<span class="o">|</span>                                                                          <span class="o">|</span>
<span class="o">|</span> <span class="n">Note</span> <span class="n">that</span><span class="p">,</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">interests</span> <span class="n">of</span> <span class="n">scientific</span> <span class="n">reproducibility</span><span class="p">,</span> <span class="n">you</span> <span class="n">should</span>    <span class="o">|</span>
<span class="o">|</span> <span class="n">describe</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">text</span> <span class="n">of</span> <span class="n">your</span> <span class="n">publications</span> <span class="ow">not</span> <span class="n">only</span> <span class="n">the</span> <span class="n">specific</span>          <span class="o">|</span>
<span class="o">|</span> <span class="n">version</span> <span class="n">of</span> <span class="n">the</span> <span class="n">SumTrees</span> <span class="n">program</span><span class="p">,</span> <span class="n">but</span> <span class="n">also</span> <span class="n">the</span> <span class="n">DendroPy</span> <span class="n">library</span> <span class="n">used</span> <span class="ow">in</span>   <span class="o">|</span>
<span class="o">|</span> <span class="n">your</span> <span class="n">analysis</span><span class="o">.</span> <span class="n">For</span> <span class="n">your</span> <span class="n">information</span><span class="p">,</span> <span class="n">you</span> <span class="n">are</span> <span class="n">running</span> <span class="n">DendroPy</span>            <span class="o">|</span>
<span class="o">|</span> <span class="mf">4.0.0</span><span class="o">.</span><span class="n">dev</span> <span class="p">(</span><span class="n">DendroPy4</span><span class="o">-</span><span class="n">feea2b0</span><span class="p">,</span> <span class="mi">2015</span><span class="o">-</span><span class="mi">06</span><span class="o">-</span><span class="mi">02</span> <span class="mi">20</span><span class="p">:</span><span class="mi">13</span><span class="p">:</span><span class="mi">49</span><span class="p">)</span><span class="o">.</span>                      <span class="o">|</span>
\<span class="o">==========================================================================/</span>

<span class="n">usage</span><span class="p">:</span> <span class="n">sumtrees</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">FORMAT</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BURNIN</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">force</span><span class="o">-</span><span class="n">rooted</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">force</span><span class="o">-</span><span class="n">unrooted</span><span class="p">]</span>
                <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">weighted</span><span class="o">-</span><span class="n">trees</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">preserve</span><span class="o">-</span><span class="n">underscores</span><span class="p">]</span>
                <span class="p">[</span><span class="o">--</span><span class="n">taxon</span><span class="o">-</span><span class="n">name</span><span class="o">-</span><span class="n">file</span> <span class="n">FILEPATH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">t</span> <span class="n">FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SUMMARY</span><span class="o">-</span><span class="n">TYPE</span><span class="p">]</span>
                <span class="p">[</span><span class="o">-</span><span class="n">f</span> <span class="c1">#.##] [--allow-unknown-target-tree-taxa]</span>
                <span class="p">[</span><span class="o">--</span><span class="n">root</span><span class="o">-</span><span class="n">target</span><span class="o">-</span><span class="n">at</span><span class="o">-</span><span class="n">outgroup</span> <span class="n">TAXON</span><span class="o">-</span><span class="n">LABEL</span><span class="p">]</span>
                <span class="p">[</span><span class="o">--</span><span class="n">root</span><span class="o">-</span><span class="n">target</span><span class="o">-</span><span class="n">at</span><span class="o">-</span><span class="n">midpoint</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">set</span><span class="o">-</span><span class="n">outgroup</span> <span class="n">TAXON</span><span class="o">-</span><span class="n">LABEL</span><span class="p">]</span>
                <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">STRATEGY</span><span class="p">]</span>
                <span class="p">[</span><span class="o">--</span><span class="n">force</span><span class="o">-</span><span class="n">minimum</span><span class="o">-</span><span class="n">edge</span><span class="o">-</span><span class="n">length</span> <span class="n">FORCE_MINIMUM_EDGE_LENGTH</span><span class="p">]</span>
                <span class="p">[</span><span class="o">--</span><span class="n">collapse</span><span class="o">-</span><span class="n">negative</span><span class="o">-</span><span class="n">edges</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">summarize</span><span class="o">-</span><span class="n">node</span><span class="o">-</span><span class="n">ages</span><span class="p">]</span>
                <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="p">{</span><span class="n">support</span><span class="p">,</span><span class="n">keep</span><span class="p">,</span><span class="n">clear</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">suppress</span><span class="o">-</span><span class="n">annotations</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">p</span><span class="p">]</span>
                <span class="p">[</span><span class="o">-</span><span class="n">d</span> <span class="c1">#] [-o FILEPATH] [-F {nexus,newick,phylip,nexml}]</span>
                <span class="p">[</span><span class="o">-</span><span class="n">x</span> <span class="n">PREFIX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">taxa</span><span class="o">-</span><span class="n">block</span><span class="p">]</span>
                <span class="p">[</span><span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">analysis</span><span class="o">-</span><span class="n">metainformation</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span> <span class="n">ADDITIONAL_COMMENTS</span><span class="p">]</span>
                <span class="p">[</span><span class="o">-</span><span class="n">r</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">M</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">m</span> <span class="n">NUM</span><span class="o">-</span><span class="n">PROCESSES</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span> <span class="n">LOG</span><span class="o">-</span><span class="n">FREQUENCY</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">q</span><span class="p">]</span>
                <span class="p">[</span><span class="o">--</span><span class="n">ignore</span><span class="o">-</span><span class="n">missing</span><span class="o">-</span><span class="n">support</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">citation</span><span class="p">]</span>
                <span class="p">[</span><span class="o">--</span><span class="n">usage</span><span class="o">-</span><span class="n">examples</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">describe</span><span class="p">]</span>
                <span class="p">[</span><span class="n">TREE</span><span class="o">-</span><span class="n">FILEPATH</span> <span class="p">[</span><span class="n">TREE</span><span class="o">-</span><span class="n">FILEPATH</span> <span class="o">...</span><span class="p">]]</span>

<span class="n">Type</span> <span class="s1">&#39;sumtrees --help&#39;</span> <span class="k">for</span> <span class="n">details</span> <span class="n">on</span> <span class="n">usage</span><span class="o">.</span>
<span class="n">Type</span> <span class="s1">&#39;sumtrees --usage-examples&#39;</span> <span class="k">for</span> <span class="n">examples</span> <span class="n">of</span> <span class="n">usage</span><span class="o">.</span>
</pre></div>
</div>
<p>You can now delete the original downloaded archive and unpacked directory if you want.</p>
</section>
<section id="how-to-use-the-program">
<h2>How to Use the Program<a class="headerlink" href="#how-to-use-the-program" title="Link to this heading">¶</a></h2>
<p>SumTrees is typically invoked by providing it a list of one or more tree files to be summarized:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees [OPTIONS] &lt;TREEFILE&gt; [&lt;TREEFILE&gt; [&lt;TREEFILE&gt; ...]]]
</pre></div>
</div>
<p>Common options include specification of a target topology onto which to map the support (”<code class="docutils literal notranslate"><span class="pre">-t</span></code>” or “<code class="docutils literal notranslate"><span class="pre">--target</span></code>”), a summary tree to use (e.g., “<code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">mcct</span></code>” or “<code class="docutils literal notranslate"><span class="pre">--summary-target=mcct</span></code>” an output file (”<code class="docutils literal notranslate"><span class="pre">-o</span></code>” or “<code class="docutils literal notranslate"><span class="pre">--output</span></code>”), and a burn-in value (”<code class="docutils literal notranslate"><span class="pre">-b</span></code>” or “<code class="docutils literal notranslate"><span class="pre">--burnin</span></code>”).</p>
<p>Full help on program usage and options is given by using the “<code class="docutils literal notranslate"><span class="pre">--help</span></code>” option:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --help
</pre></div>
</div>
</section>
<section id="quick-recipes">
<h2>Quick Recipes<a class="headerlink" href="#quick-recipes" title="Link to this heading">¶</a></h2>
<section id="summarization-of-posterior-probabilities-of-clades-with-a-consensus-tree">
<h3>Summarization of Posterior Probabilities of Clades with a Consensus Tree<a class="headerlink" href="#summarization-of-posterior-probabilities-of-clades-with-a-consensus-tree" title="Link to this heading">¶</a></h3>
<p>Summarize a set of tree files using a 95% rule consensus tree, with support for clades indicated as proportions (posterior probabilities) using branch labels, and branch lengths being the mean across all trees, dropping the first 200 trees in each file as a burn-in, and saving the result to “<code class="docutils literal notranslate"><span class="pre">result.tre</span></code>”:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --min-clade-freq=0.95 --burnin=200 --output-tree-filepath=result.tre treefile1.tre treefile2.tre treefile3.tre
$ sumtrees -f0.95 -b200 -o result.tre treefile1.tre treefile2.tre treefile3.tre
</pre></div>
</div>
</section>
<section id="summarization-of-posterior-probabilities-of-clades-with-a-maximum-clade-credibility-tree-mcct">
<h3>Summarization of Posterior Probabilities of Clades with a Maximum Clade Credibility Tree (MCCT)<a class="headerlink" href="#summarization-of-posterior-probabilities-of-clades-with-a-maximum-clade-credibility-tree-mcct" title="Link to this heading">¶</a></h3>
<p>Summarize a set of tree files using a tree in the input set that maximizes the product of clade support, with support for clades indicated as proportions (posterior probabilities) using branch labels, and branch lengths the mean across all trees, dropping the first 200 trees in each file as a burn-in, and saving the result to “<code class="docutils literal notranslate"><span class="pre">result.tre</span></code>”:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --summary-target=mcct --burnin=200 --support-as-labels --output-tree-filepath=result.tre treefile1.tre treefile2.tre treefile3.tre
$ sumtrees -s mcct -b200 -l -o result.tre treefile1.tre treefile2.tre treefile3.tre
</pre></div>
</div>
</section>
<section id="non-parametric-bootstrap-support-of-a-model-tree">
<h3>Non-parametric Bootstrap Support of a Model Tree<a class="headerlink" href="#non-parametric-bootstrap-support-of-a-model-tree" title="Link to this heading">¶</a></h3>
<p>Calculate support for nodes on a specific tree, “<code class="docutils literal notranslate"><span class="pre">best.tre</span></code>” as given by a set of tree files, with support reported as percentages rounded to integers, and saving the result to “<code class="docutils literal notranslate"><span class="pre">result.tre</span></code>”:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --decimals=0 --percentages --output-tree-filepath=result.tre --target=best.tre treefile1.tre treefile2.tre treefile3.tre
$ sumtrees -d0 -p -o result.tre -t best.tre treefile1.tre treefile2.tre treefile3.tre
</pre></div>
</div>
</section>
<section id="set-node-ages-of-target-or-summary-tree-s-to-mean-median-node-age-of-input-trees">
<h3>Set Node Ages of Target or Summary Tree(s) to Mean/Median Node Age of Input Trees<a class="headerlink" href="#set-node-ages-of-target-or-summary-tree-s-to-mean-median-node-age-of-input-trees" title="Link to this heading">¶</a></h3>
<p>Summarize a set of ultrametric tree files using a 95% majority-rule consensus tree, with support for clades indicated as proportions (posterior probabilities) using branch labels, and branch lengths adjusted so the ages of internal nodes are the mean across all trees, dropping the first 200 trees in each file as a burn-in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --min-clade-freq=0.95 --burnin=200 --edges=mean-age --output-tree-filepath=result.tre treefile1.tre treefile2.tre treefile3.tre
$ sumtrees -f0.95 -b200 -o result.tre -l -e mean-age treefile1.tre treefile2.tre treefile3.tre
</pre></div>
</div>
<p>To use the median age instead:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --min-clade-freq=0.95 --burnin=200 --edges=median-age --output-tree-filepath=result.tre treefile1.tre treefile2.tre treefile3.tre
$ sumtrees -f0.95 -b200 -o result.tre -e median-age treefile1.tre treefile2.tre treefile3.tre
</pre></div>
</div>
</section>
<section id="running-in-parallel-mode">
<h3>Running in Parallel Mode<a class="headerlink" href="#running-in-parallel-mode" title="Link to this heading">¶</a></h3>
<p>Running in parallel mode will analyze each input source in its own independent process, with multiple processes running in parallel.
Multiprocessing analysis is invoked by adding the “<code class="docutils literal notranslate"><span class="pre">-m</span></code>” or “<code class="docutils literal notranslate"><span class="pre">--multiprocessing</span></code>”  flag to the SumTrees command, and passing in the maximum number of processes to run in parallel.
For example, if your machine has two cores, and you want to run the previous analyses using both of them, you would specify that SumTrees run in parallel mode with two processes by adding “<code class="docutils literal notranslate"><span class="pre">-m2</span></code>” or “<code class="docutils literal notranslate"><span class="pre">--multiprocessing=2</span></code>” to the SumTrees command invocation:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --multiprocessing=2 --decimals=0 --percentages --output-tree-filepath=result.tre --target=best.tre treefile1.tre treefile2.tre treefile3.tre
$ sumtrees -m2 -d0 -p -o result.tre -t best.tre treefile1.tre treefile2.tre treefile3.tre
$ sumtrees --multiprocessing=2 --min-clade-freq=0.95 --burnin=200 --support-as-labels --output-tree-filepath=result.tre treefile1.tre treefile2.tre treefile3.tre
$ sumtrees -m2 -f0.95 -b200 -l -o result.tre treefile1.tre treefile2.tre treefile3.tre
</pre></div>
</div>
<p>You can specify as many processes as you want, up to the total number of tree support files passed as input sources.
If you want to use <em>all</em> the available cores on your machine, you can use the “<code class="docutils literal notranslate"><span class="pre">-M</span></code>” or “<code class="docutils literal notranslate"><span class="pre">--maximum-multiprocessing</span></code>” flag:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --maximum-multiprocessing --decimals=0 --percentages --output-tree-filepath=result.tre --target=best.tre treefile1.tre treefile2.tre treefile3.tre
$ sumtrees -M -d0 -p -o result.tre -t best.tre treefile1.tre treefile2.tre treefile3.tre
</pre></div>
</div>
<p>If you specify fewer processes than input sources, then the files will be cycled through the processes.</p>
</section>
</section>
<section id="primers-and-examples">
<h2>Primers and Examples<a class="headerlink" href="#primers-and-examples" title="Link to this heading">¶</a></h2>
<p>At its most basic, you will need to supply SumTrees with the path to one or more tree files in Newick or NEXUS format that you want to summarize:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees phylo.tre
</pre></div>
</div>
<p>As no target tree was provided and no summary tree type was specified, SumTrees will, by default construct ad 50% majority-rule clade consensus tree of all the trees found in the file “<code class="docutils literal notranslate"><span class="pre">phylo.tre</span></code>” as  the summarization target.
The internal node labels of the resulting consensus tree will, by default, indicate the proportion of trees in “<code class="docutils literal notranslate"><span class="pre">phylo.tre</span></code>” in which that clade was found, while the branch lengths of the resulting consensus tree being set to the mean of the branch lengths of that clade across all the trees in “<code class="docutils literal notranslate"><span class="pre">phylo.tre</span></code>”.</p>
<p>If you have split searches across multiple runs (across, for example, multiple computers, so as to speed up the search time), such that you have multiple tree files (”<code class="docutils literal notranslate"><span class="pre">phylo.run1.tre</span></code>”, “<code class="docutils literal notranslate"><span class="pre">phylo.run2.tre</span></code>”, “<code class="docutils literal notranslate"><span class="pre">phylo.run3.tre</span></code>”, …), you can instruct SumTrees to consider all these files together when summarizing the support by simply listing them one after another separated by spaces:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees phylo.run1.tre phylo.run2.tre phylo.run3.tre
</pre></div>
</div>
<p>As before, the above command will construct a 50% majority-rule consensus tree with clade supported indicated by internal node labels and branch lengths being the mean across all trees, but this time it will use all the trees found across all the files listed: “<code class="docutils literal notranslate"><span class="pre">phylo.run1.tre</span></code>”, “<code class="docutils literal notranslate"><span class="pre">phylo.run2.tre</span></code>”, and “<code class="docutils literal notranslate"><span class="pre">phylo.run3.tre</span></code>”.</p>
<p>You will notice that the final resulting tree is displayed to the terminal and not saved anywhere.
It will probably be more useful if we can save it to a file for visualization for further analysis.
This can be done in one of two ways, either by redirecting the screen output to a file, using the standard (at least on UNIX and UNIX-like systems) redirection operator, <code class="docutils literal notranslate"><span class="pre">&gt;</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees phylo.tre &gt; phylo.consensus.sumtrees
</pre></div>
</div>
<p>or by using the or “<code class="docutils literal notranslate"><span class="pre">--output-tree-filepath</span></code>” option:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --output-tree-filepath=phylo.consensus.sumtrees phylo.tre
</pre></div>
</div>
<p>If the files are in different directories, or you are not in the same directory as the files, you should use the full directory path specification:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --output-tree-filepath=/Users/myself/MyProjects/phylo1/final/phylo.consensus.sumtrees /Users/myself/MyProjects/phylo1/phylo.tre
</pre></div>
</div>
<p>More extended options specify things like: where to save the output (by default it goes to the screen), the topology or tree to which to map the support (user-supplied or consensus tree), the output format (NEXUS by default, but it can also be Newick), whether support is indicated in terms of proportions or percentages etc.
All of these options are specified on the command line when invoking the program, with multiple options separated by spaces.
Many of the options have two alternate forms, a long form (a word or phrase preceded by two dashes, e.g., “<code class="docutils literal notranslate"><span class="pre">--option</span></code>”) and a short form (a single letter preceded by a single dash, “<code class="docutils literal notranslate"><span class="pre">-o</span></code>”).
The long form of the options needs an equals sign before setting the paramater (e.g., “<code class="docutils literal notranslate"><span class="pre">--option=1</span></code>”), while the short one does not (e.g., “<code class="docutils literal notranslate"><span class="pre">-o1</span></code>” or “<code class="docutils literal notranslate"><span class="pre">-o</span> <span class="pre">1</span></code>”).
Most of the options have default values that will be used if not explicitly set when the program is invoked.
The order that the options are given does <em>not</em> matter, i.e., “<code class="docutils literal notranslate"><span class="pre">sumtrees</span> <span class="pre">--option1=something</span> <span class="pre">--option2=something</span></code>” is the same as “<code class="docutils literal notranslate"><span class="pre">sumtrees</span> <span class="pre">--option2=something</span> <span class="pre">--option1=something</span></code>”.
As mentioned above, full details on these options, their long and short forms, as well as their default values will be given by invoking the program with the “<code class="docutils literal notranslate"><span class="pre">--help</span></code>” or “<code class="docutils literal notranslate"><span class="pre">-h</span></code>” option: “<code class="docutils literal notranslate"><span class="pre">sumtrees</span> <span class="pre">--help</span></code>”.</p>
</section>
<section id="specifying-and-customization-of-the-summarization-target">
<h2>Specifying and Customization of the Summarization Target<a class="headerlink" href="#specifying-and-customization-of-the-summarization-target" title="Link to this heading">¶</a></h2>
<p>SumTrees maps support values calculated from the input set of trees onto a <em>target</em> topology.
This target topology can be a <em>summary</em> topology constructed from the input set of trees based on a strategy specified by the user (using the the “<code class="docutils literal notranslate"><span class="pre">--summary-target</span></code>” or “<code class="docutils literal notranslate"><span class="pre">-s</span></code>” flag to specify, for example, a majority-rule consensus tree or a maximum credbility tree) or a topology provided by the user (using the “<code class="docutils literal notranslate"><span class="pre">--target-tree-filepath</span></code>” or “<code class="docutils literal notranslate"><span class="pre">-t</span></code>” option to provide, e.g., a maximum-likehood estimate or some other topology sourced by other means).</p>
<section id="specifying-a-summarization-topology-type">
<h3>Specifying a Summarization Topology Type<a class="headerlink" href="#specifying-a-summarization-topology-type" title="Link to this heading">¶</a></h3>
<p>You can specify the type of summary topology onto which the support values are mapped using the “<code class="docutils literal notranslate"><span class="pre">--summary-target</span></code>” or “<code class="docutils literal notranslate"><span class="pre">-s</span></code>” option.
This option takes one of three values as an argument:</p>
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>“consensus”</p></td>
<td><p>The majority-rule consensus tree (default)</p></td>
</tr>
<tr class="row-odd"><td><p>“mcct”</p></td>
<td><p>The Maximum Credibility Tree: the topology from the input set that maximizes the <em>product</em> of the support of the clades</p></td>
</tr>
<tr class="row-even"><td><p>“msct”</p></td>
<td><p>The Maximum Sum of Credibilities Tree: the topology from the input set that maximizes the <em>sum</em> of support of the clades.</p></td>
</tr>
</tbody>
</table>
<section id="majority-rule-consensus-topology">
<h4>Majority-Rule Consensus Topology<a class="headerlink" href="#majority-rule-consensus-topology" title="Link to this heading">¶</a></h4>
<p>For example, say you have completed a 1000-replicate non-parametric analysis of your dataset using a program such as GARLI or RAxML.
You want to construct a 70% majority-rule consensus tree of the replicates, with support indicated as percentages on the node labels.
If the bootstrap replicates are in the file “<code class="docutils literal notranslate"><span class="pre">phylo-boots.tre</span></code>”, you would then enter the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --summary-target=consensus --min-clade-freq=0.7 --percentages --decimals=0 phylo-boots.tre
</pre></div>
</div>
<p>Or, using the short option syntax:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees -s consensus -f0.7 -p -d0 phylo-boots.tre
</pre></div>
</div>
<p>Here, the “<code class="docutils literal notranslate"><span class="pre">--min-clade-freq=0.7</span></code>” or “<code class="docutils literal notranslate"><span class="pre">-f0.7</span></code>” option lowers the minimum threshold for clade inclusion to 70%.
If you want a 95% majority-rule consensus tree instead, you would use “<code class="docutils literal notranslate"><span class="pre">--min-clade-freq=0.95</span></code>” or “<code class="docutils literal notranslate"><span class="pre">-f0.95</span></code>”.
The default threshold if none is specified is 0.5 (50%).
The “<code class="docutils literal notranslate"><span class="pre">--percentages</span></code>” or “<code class="docutils literal notranslate"><span class="pre">-p</span></code>” option instructs SumTrees to report the support in terms of percentages, while the “<code class="docutils literal notranslate"><span class="pre">--decimals=0</span></code>” or “<code class="docutils literal notranslate"><span class="pre">-d</span> <span class="pre">0</span></code>” option instructs SumTrees not to bother reporting any decimals.
Note that even if you instruct SumTrees to report the support in terms of percentages, the minimum clade inclusion threshold is still given in terms of proportions.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>As noted, if no target topology is specified (either using the “<code class="docutils literal notranslate"><span class="pre">--summary-target</span></code>”/”<code class="docutils literal notranslate"><span class="pre">-s</span></code>” or “<code class="docutils literal notranslate"><span class="pre">--target-tree-filepath</span></code>”/”<code class="docutils literal notranslate"><span class="pre">-t</span></code>” options), then SumTrees by default will construct and use a majority-rule consensus topology as a target, and hence the explicit specification of this as a target is not needed. Thus, the following will produce exactly the same results as above:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --min-clade-freq=0.7 --percentages --decimals=0 phylo-boots.tre
$ sumtrees -f0.7 -p -d0 phylo-boots.tre
</pre></div>
</div>
</div>
<p>Again, if we want to actually save the results to the file, we should use the “<code class="docutils literal notranslate"><span class="pre">--output-tree-filepath</span></code>” option:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --output-tree-filepath=phylo-mle-support.sumtrees --min-clade-freq=0.7 --percentages --decimals=0 phylo-boots.tre
$ sumtrees -o phylo-mle-support.sumtrees -f0.7 --p --d0 phylo-boots.tre
</pre></div>
</div>
</section>
<section id="maximum-clade-credibility-topology">
<h4>Maximum Clade Credibility Topology<a class="headerlink" href="#maximum-clade-credibility-topology" title="Link to this heading">¶</a></h4>
<p>The Maximum Clade Credibility Topology, or MCCT, is the topology that maximizes the <em>product</em> of the split support. You can use this as the target topology by specifying “<code class="docutils literal notranslate"><span class="pre">--summary-target=mcct</span></code>” or “<code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">mcct</span></code>”:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --summary-target=mcct phylo-boots.tre
$ sumtrees -s mcct phylo-boots.tre
</pre></div>
</div>
<p>As might be expected, in can be combined with other options. For example, to discard the first 200 trees from each of the input sources and write the results to a file, “<code class="docutils literal notranslate"><span class="pre">results.tre</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --summary-target=mcct --burnin=200 --output-tree-filepath=results.tre treefile1.tre treefile2.tre treefile3.tre
$ sumtrees -s mcct -b 200 -o results.tre treefile1.tre treefile2.tre treefile3.tre
</pre></div>
</div>
</section>
</section>
<section id="specifying-a-custom-topology-or-set-of-topologies">
<h3>Specifying a Custom Topology or Set of Topologies<a class="headerlink" href="#specifying-a-custom-topology-or-set-of-topologies" title="Link to this heading">¶</a></h3>
<p>Say you also have a maximum likelihood estimate of the phylogeny, and want to annotate the nodes of the maximum likelihood tree with the proportion of trees out of the bootstrap replicates in which the node is found.
Then, assuming your maximum likelihood tree is in the file, “<code class="docutils literal notranslate"><span class="pre">phylo-mle.tre</span></code>”, and the bootstrap tree file is “<code class="docutils literal notranslate"><span class="pre">phylo-boots.tre</span></code>”, you would use the “<code class="docutils literal notranslate"><span class="pre">--target-tree-filepath</span></code>” options, as in the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --target-tree-filepath=phylo-mle.tre phylo-boots.tre
</pre></div>
</div>
<p>Here, “<code class="docutils literal notranslate"><span class="pre">--target-tree-filepath</span></code>” specifies the target topology onto which the support will be mapped, while the remaining (unprefixed) argument specifies the tree file that is the source of the support.
An equivalent form of the same command, using the short option syntax is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees -t phylo-mle.tre phylo-boots.tre
</pre></div>
</div>
<p>If you want the support expressed in percentages instead of proportions, and the final tree saved to a file, you would enter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --output phylo-mle-support.sumtrees --target-tree-filepath phylo-mle.tre --proportions --decimals=0 phylo-boots.tre
$ sumtrees -o phylo-mle-support.sumtrees -t phylo-mle.tre -p -d0 phylo-boots.tre
</pre></div>
</div>
</section>
<section id="collapsing-clades-edges-splits-below-a-minimum-support-frequency-threshold">
<h3>Collapsing Clades/Edges/Splits Below a Minimum Support Frequency Threshold<a class="headerlink" href="#collapsing-clades-edges-splits-below-a-minimum-support-frequency-threshold" title="Link to this heading">¶</a></h3>
<p>The “<code class="docutils literal notranslate"><span class="pre">-f</span></code>” or “<code class="docutils literal notranslate"><span class="pre">--min-clade-freq</span></code>” flag sets the threshold below which clades will be collapsed.
This applies to all summary targets: user-specified target topologies as well consensus trees, MCCT trees, etc.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --min-clade-freq=0.75 --percentages --decimals=0 phylo-boots.tre
$ sumtrees -f0.75 -p -d0 phylo-boots.tre
$ sumtrees --min-clade-freq=0.95 --summary-target=mcct phylo-boots.tre
$ sumtrees -f0.95 -s mcct phylo-boots.tre
$ sumtrees --min-clade-freq=0.75 --decimals=0 --percentages --target=best.tre treefile1.tre treefile2.tre treefile3.tre
$ sumtrees -f0.75 -d0 -p -t best.tre treefile1.tre treefile2.tre treefile3.tre
</pre></div>
</div>
<p>Note that the minimum support frequency threshold is specified as proportion scaled from 0 to 1, even if the percentages are requested as the support units.</p>
</section>
</section>
<section id="summarizing-edge-lengths">
<h2>Summarizing Edge Lengths<a class="headerlink" href="#summarizing-edge-lengths" title="Link to this heading">¶</a></h2>
<p>If a target topology has been specified using the “<code class="docutils literal notranslate"><span class="pre">--target-tree-filepath</span></code>” or the “<code class="docutils literal notranslate"><span class="pre">-t</span></code>” option, then by default SumTrees retains the edge lengths of the target topologies.
Otherwise, if the input trees are ultrametric and the “<code class="docutils literal notranslate"><span class="pre">--summarize-node-ages</span></code>” option is given, then by default SumTrees will adjust the edge lengths of the target topology so that the ages of the internal nodes are the mean of the ages of the corresponding nodes in the input set of trees.
Otherwise, if no target trees are specified and the “<code class="docutils literal notranslate"><span class="pre">--summarize-node-ages</span></code>” is not given, the edge lengths of the target topology will be set to the mean lengths of the corresponding edges of the input set.
In all cases, these defaults can be overridden by using the “<code class="docutils literal notranslate"><span class="pre">--set-edges</span></code>” or “<code class="docutils literal notranslate"><span class="pre">-e</span></code>” option, which takes one of the following values:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">mean-length</span></code>: sets the edge lengths of the target/consensus tree(s) to the mean of the lengths of the corresponding edges of the input trees.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">median-length</span></code>: sets the edge lengths of the target/consensus tree(s) to the median of the lengths of the corresponding edges of the input trees.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">median-age</span></code>: adjusts the edge lengths of the target/consensus tree(s) such that the node ages correspond to the median age of corresponding nodes of the input trees [requires rooted ultrametric trees].</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mean-age</span></code>: adjusts the edge lengths of the target/consensus tree(s) such that the node ages correspond to the mean age of corresponding nodes of the input trees [requires rooted ultrametric trees].</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">support</span></code>: edge lengths will be set to the support value for the split represented by the edge.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">keep</span></code>: do not change the existing edge lengths of the target topology.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">clear</span></code>: all edge lengths will be removed</p></li>
</ul>
</div></blockquote>
<p>So, for example, to construct a consensus tree of a post-burnin set of ultrametric trees, with the node ages set to the <em>mean</em> instead of the median node age:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --set-edges=mean-age --summarize-node-ages --burnin=200 beast1.trees beast2.trees beast3.trees
$ sumtrees --e mean-age --summarize-node-ages --b 200 beast1.trees beast2.trees beast3.trees
</pre></div>
</div>
<p>Or to set the edges of a user-specifed tree to the median edge length of the input trees:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --set-edges=median-length --target=mle.tre boots1.tre boots2.tre
$ sumtrees --e median-length -t mle.tre boots1.tre boots2.tre
</pre></div>
</div>
</section>
<section id="summarizing-rooted-and-ultrametric-trees">
<h2>Summarizing Rooted and Ultrametric Trees<a class="headerlink" href="#summarizing-rooted-and-ultrametric-trees" title="Link to this heading">¶</a></h2>
<section id="forcing-trees-to-be-treated-as-rooted">
<h3>Forcing Trees to be Treated as Rooted<a class="headerlink" href="#forcing-trees-to-be-treated-as-rooted" title="Link to this heading">¶</a></h3>
<p>SumTrees treats all trees as unrooted unless specified otherwise. You can force SumTrees to treat all trees as rooted by passing it the “<code class="docutils literal notranslate"><span class="pre">--force-rooted</span></code>” flag:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --force-rooted phylo.trees
</pre></div>
</div>
</section>
<section id="summarizing-node-ages">
<h3>Summarizing Node Ages<a class="headerlink" href="#summarizing-node-ages" title="Link to this heading">¶</a></h3>
<p>If the trees are rooted <strong>and</strong> ultrametric, the “<code class="docutils literal notranslate"><span class="pre">--summarize-node-ages</span></code>” flag will result in SumTrees summarizing node age information as well:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --summarize-node-ages phylo.trees
</pre></div>
</div>
<p>This will calculate and report statistics such as the mean, standard deviation, ranges, high posterior density intervals, etc. of the node ages.</p>
</section>
<section id="setting-the-node-ages-of-the-summary-trees">
<h3>Setting the Node Ages of the Summary Trees<a class="headerlink" href="#setting-the-node-ages-of-the-summary-trees" title="Link to this heading">¶</a></h3>
<p>If the tree is rooted and ultrametric, instead of summarizing edge lengths to be the mean of the edge lengths across the input set, you probably want to set the edge lengths of the summary tree so the node ages correspond to the mean or median of the node ages of the input set:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --set-edges=mean-age --summarize-node-ages beast1.trees beast2.trees beast3.trees
$ sumtrees --e mean-age --summarize-node-ages beast1.trees beast2.trees beast3.trees
</pre></div>
</div>
</section>
<section id="rooting-the-target-topology">
<h3>Rooting the Target Topology<a class="headerlink" href="#rooting-the-target-topology" title="Link to this heading">¶</a></h3>
<p>The following options allow for re-rooting of the target topology or topologies:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">--</span><span class="n">root</span><span class="o">-</span><span class="n">target</span><span class="o">-</span><span class="n">at</span><span class="o">-</span><span class="n">outgroup</span> <span class="n">TAXON</span><span class="o">-</span><span class="n">LABEL</span>
                      <span class="n">Root</span> <span class="n">target</span> <span class="n">tree</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="n">using</span> <span class="n">specified</span> <span class="n">taxon</span> <span class="k">as</span> <span class="n">outgroup</span><span class="o">.</span>
<span class="o">--</span><span class="n">root</span><span class="o">-</span><span class="n">target</span><span class="o">-</span><span class="n">at</span><span class="o">-</span><span class="n">midpoint</span>
                      <span class="n">Root</span> <span class="n">target</span> <span class="n">tree</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="n">at</span> <span class="n">midpoint</span><span class="o">.</span>
<span class="o">--</span><span class="nb">set</span><span class="o">-</span><span class="n">outgroup</span> <span class="n">TAXON</span><span class="o">-</span><span class="n">LABEL</span>
                      <span class="n">Rotate</span> <span class="n">the</span> <span class="n">target</span> <span class="n">trees</span> <span class="n">such</span> <span class="n">the</span> <span class="n">specified</span> <span class="n">taxon</span> <span class="ow">is</span> <span class="ow">in</span>
                      <span class="n">the</span> <span class="n">outgroup</span> <span class="n">position</span><span class="p">,</span> <span class="n">but</span> <span class="n">do</span> <span class="ow">not</span> <span class="n">explicitly</span> <span class="n">change</span>
                      <span class="n">the</span> <span class="n">target</span> <span class="n">tree</span> <span class="n">rooting</span><span class="o">.</span>
</pre></div>
</div>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --root-target-at-outgroup Python_regius --target=mle.tre boots1.tre boots2.tre
$ sumtrees --root-target-at-midpoint -s mcct trees1.tre trees2.tre
$ sumtrees --set-outgroup Python_regius -s mcct trees1.tre trees2.tre
</pre></div>
</div>
</section>
</section>
<section id="assigning-the-ages-of-non-contemporaneous-tips-tip-dating">
<h2>Assigning the Ages of Non-Contemporaneous Tips (“Tip-Dating”)<a class="headerlink" href="#assigning-the-ages-of-non-contemporaneous-tips-tip-dating" title="Link to this heading">¶</a></h2>
<p>In some studies, such as those including fossil taxa or serially-sampled
lineages (e.g., viral or bacterial studies), the tips of the tree corresponding
to taxa are non-contemporaneous.
In these studies, the tips corresponding to taxa or samples in the past are usually assigned dates.
Commonly referred to by some folks by the unfortunate and somewhat ambiguous
phrase “tip-dating”, here we refer to this process of assigning ages of
non-contemporaneous tips as somewhat less sloppy and more explicit, “assigning
ages of non-contemporaneous tips”.</p>
<p>To do this using SumTrees, you would use the option <code class="docutils literal notranslate"><span class="pre">--tip-ages</span></code> to specify the path to a file containing the tip age data.
The format of this file is specified by the <code class="docutils literal notranslate"><span class="pre">--tip-ages-format</span></code> option, which can be “tsv” (tab-separated values), “csv” (comma-separated values), or “json” (JSON).
The default format is “tsv”, which specifies a tab delimited file.
For both the tab-separated and comma-separated formats, the data should consist of two columns, with the first column specifying the taxon label and the second column specifying the age of the corresponding tip.
The JSON-format file should consist of a single dictionary with keys being taxon labels and values being the age of the corresponding tips.</p>
<p>Taxon labels must match <em>EXACTLY</em> with the taxon labels in the input tree sources.
This needs to account for things like the conversion of unprotected underscores to spaces as mandated by the Newick/NEXUS standard unless overridden by the <code class="docutils literal notranslate"><span class="pre">--preserve-underscores</span></code> option.
Any taxon not listed in the tip ages file will get assigned an age of 0.0 by default.</p>
<p>So, for example, given a tree file ‘<code class="docutils literal notranslate"><span class="pre">data/x1.tre</span></code>’ with trees like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="o">&amp;</span><span class="n">R</span><span class="p">]</span> <span class="p">((</span><span class="n">d</span><span class="p">:</span><span class="mi">2</span><span class="p">,(</span><span class="n">a</span><span class="p">:</span><span class="mi">3</span><span class="p">,(</span><span class="n">b</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="n">c</span><span class="p">:</span><span class="mi">2</span><span class="p">):</span><span class="mi">1</span><span class="p">):</span><span class="mi">3</span><span class="p">):</span><span class="mi">2</span><span class="p">,(</span><span class="n">g</span><span class="p">:</span><span class="mi">4</span><span class="p">,(</span><span class="n">e</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="n">f</span><span class="p">:</span><span class="mi">1</span><span class="p">):</span><span class="mi">2</span><span class="p">):</span><span class="mi">4</span><span class="p">);</span>
<span class="p">[</span><span class="o">&amp;</span><span class="n">R</span><span class="p">]</span> <span class="p">((</span><span class="n">d</span><span class="p">:</span><span class="mi">2</span><span class="p">,(</span><span class="n">a</span><span class="p">:</span><span class="mi">3</span><span class="p">,(</span><span class="n">b</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="n">c</span><span class="p">:</span><span class="mi">2</span><span class="p">):</span><span class="mi">1</span><span class="p">):</span><span class="mi">3</span><span class="p">):</span><span class="mi">2</span><span class="p">,(</span><span class="n">g</span><span class="p">:</span><span class="mi">4</span><span class="p">,(</span><span class="n">e</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="n">f</span><span class="p">:</span><span class="mi">1</span><span class="p">):</span><span class="mi">2</span><span class="p">):</span><span class="mi">4</span><span class="p">);</span>
<span class="o">.</span>
<span class="o">.</span>
<span class="o">.</span>
</pre></div>
</div>
<p>and a tip age data file, ‘<code class="docutils literal notranslate"><span class="pre">data/x1.ages.tsv</span></code>’, consisting of tab-separated values like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">d</span><span class="o">&lt;</span><span class="n">TAB</span><span class="o">&gt;</span><span class="mi">4</span>
<span class="n">b</span><span class="o">&lt;</span><span class="n">TAB</span><span class="o">&gt;</span><span class="mi">1</span>
<span class="n">e</span><span class="o">&lt;</span><span class="n">TAB</span><span class="o">&gt;</span><span class="mi">1</span>
<span class="n">f</span><span class="o">&lt;</span><span class="n">TAB</span><span class="o">&gt;</span><span class="mi">1</span>
</pre></div>
</div>
<p>then the following is the invocation to summarize the trees and their node ages:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees \
 --tip-ages data/x1.ages.tsv
 --summarize-node-ages \
 data/x1-1.tre
 data/x1-2.tre
</pre></div>
</div>
<p>If the tip age data file is given in a comma-separated file, ‘<code class="docutils literal notranslate"><span class="pre">data/x1.ages.csv</span></code>’, like the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">d</span><span class="p">,</span><span class="mi">4</span>
<span class="n">b</span><span class="p">,</span><span class="mi">1</span>
<span class="n">e</span><span class="p">,</span><span class="mi">1</span>
<span class="n">f</span><span class="p">,</span><span class="mi">1</span>
</pre></div>
</div>
<p>then the invocation is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees \
 --tip-ages data/x1.ages.csv
 --tip-ages-format=csv \
 --summarize-node-ages \
 data/x1-1.tre
 data/x1-2.tre
</pre></div>
</div>
<p>Note that if specifying the tip ages you have to explicitly specify <code class="docutils literal notranslate"><span class="pre">--summarize-node-ages</span></code> or some other option that results in node ages being analyzed (e.g., <code class="docutils literal notranslate"><span class="pre">--set-edges=mean-age</span></code>).</p>
</section>
<section id="parallelizing-sumtrees">
<h2>Parallelizing SumTrees<a class="headerlink" href="#parallelizing-sumtrees" title="Link to this heading">¶</a></h2>
<section id="basics">
<h3>Basics<a class="headerlink" href="#basics" title="Link to this heading">¶</a></h3>
<p>If you have multiple input (support) files, you can greatly increase the performance of SumTrees by running it in parallel mode.
In parallel mode, each input source will be handled in a separate process, resulting in a speed-up linearly proportional to the total number of processes running in parallel.
At its most basic, running in parallel mode involves nothing more than adding the “<code class="docutils literal notranslate"><span class="pre">-m</span></code>” or “<code class="docutils literal notranslate"><span class="pre">--multiprocessing</span></code> option to the SumTrees invocation, and passing in the number of parallel processes to run.
So, for example, if you have four tree files that you want to summarize, and you want to run these using two processes in parallel:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --multiprocessing=2 -o result.tre phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
$ sumtrees -m 2 -o result.tre phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
</pre></div>
</div>
<p>Or, to run in four processes simultaneously:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --multiprocessing=4 phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
$ sumtrees -m 4 phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
</pre></div>
</div>
<p>If you want to use all the available cores on the current machine, use the “<code class="docutils literal notranslate"><span class="pre">-M</span></code>” or “<code class="docutils literal notranslate"><span class="pre">--maximum-multiprocessing</span></code>” flag:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --maximum-multiprocessing phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
$ sumtrees -M phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
</pre></div>
</div>
<p>Other options as described above, can, of course be added as needed:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --multiprocessing=4 --burnin=200 phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
$ sumtrees -m 4 -b 200 phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre

$ sumtrees --maximum-multiprocessing --burnin=200 --min-clade-freq=0.75 phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
$ sumtrees -M -b 200 -f0.75 phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre

$ sumtrees --maximum-multiprocessing --burnin=200 --min-clade-freq=0.75 --output-tree-filepath=con.tre phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
$ sumtrees -M -b 200 -f0.75 -o con.tre phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
</pre></div>
</div>
</section>
<section id="parallelization-strategy-deciding-on-the-number-of-processes">
<h3>Parallelization Strategy: Deciding on the Number of Processes<a class="headerlink" href="#parallelization-strategy-deciding-on-the-number-of-processes" title="Link to this heading">¶</a></h3>
<p>In parallel mode, SumTrees parallelizes by input files, which means that the maximum number of processes that can be run is limited to the number of input files.
If you have four support tree files as input sources, as in the above example (”<code class="docutils literal notranslate"><span class="pre">phylo.run1.tre</span></code>”, “<code class="docutils literal notranslate"><span class="pre">phylo.run2.tre</span></code>”,”<code class="docutils literal notranslate"><span class="pre">phylo.run3.tre</span></code>”, and “<code class="docutils literal notranslate"><span class="pre">phylo.run4.tre</span></code>”), then you can run from 2 up to 4 processes in parallel.
If you have 40 tree source or support files, then you can run from 2 up to 40 processes in parallel.</p>
<p>Note that there is a difference between the number of <em>processes</em> that SumTrees runs, and the number of <em>processors</em> or cores available on your machine or in your given hardware context.
When running SumTrees in parallel mode, you can specify <em>any</em> number of parallel <em>processes</em>, up to the number of support files, even if the number of processes greatly exceeds the number <em>processors</em> available.
For example, you might have an octo-core machine available, which means that you have 8 processors available.
If your analysis has 40 independent support files, you can invoke SumTrees with 40 parallel processes:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --multiprocessing=40 -o result.tre t1.tre ... t40.tre
</pre></div>
</div>
<p>SumTrees will indeed launch 40 parallel processes to process the files, and it will seem like 40 processes are being executed in parallel.
In reality, however, the operating system is actually cycling the processes through the available processors in rapid succession, such that, on a nanosecond time-scale, only 8 processes are actually executing simultaneously: one on each of the available cores.
While your run should complete without problems when oversubscribing the hardware in this way, there is going to be some degree of performance hit due to the overhead involved in managing the cycling of the processes through the processors.
On some operating systems and hardware contexts, depending on the magnitude of oversubscription, this can be considerable.
Thus it probably is a good idea to match the number of processes to the number of processors available.</p>
<p>If you do not mind using all the available cores on your machine, you can use the “<code class="docutils literal notranslate"><span class="pre">--maximum-multiprocessing</span></code>” or “<code class="docutils literal notranslate"><span class="pre">-M</span></code>” flag to request this, instead of using a specific number.</p>
<p>Another issue to consider is an even distribution of workload.
Assuming that each of your input support files have the same number of trees, then it makes sense to specify a number of processes that is a factor of the number of input files.
So, for example, if you have 8 input files to be summarized, you will get the best performance out of SumTrees by specifying 2, 4, or 8 processes, with the actual number given by the maximum number of processors available or that you want to dedicate to this task.</p>
</section>
<section id="running-parallel-mode-sumtrees-in-a-parallel-environment-on-a-high-performance-computing-hpc-cluster">
<h3>Running Parallel-Mode SumTrees in a Parallel Environment on a High-Performance Computing (HPC) Cluster<a class="headerlink" href="#running-parallel-mode-sumtrees-in-a-parallel-environment-on-a-high-performance-computing-hpc-cluster" title="Link to this heading">¶</a></h3>
<p>Unfortunately, the diversity and idiosyncracies in various HPC configurations and set-ups are so great that it is very difficult to provide general recipes on how to run parallel-mode SumTrees in a parallel environment on an HPC cluster.
However, in all cases, all you really need to do is to set up an appropriate parallel execution environment on the cluster, requesting a specific number of processors from the cluster scheduler software, and then tell SumTrees to run the same number of processes.</p>
<p>For example, if your cluster uses the <a class="reference external" href="http://gridengine.sunsource.net/">Sun Grid Engine</a> as its scheduler, then you might use a job script such as the following:</p>
<blockquote>
<div><div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># /bin/sh</span>
<span class="c1">#$ -cwd</span>
<span class="c1">#$ -V</span>
<span class="c1">#$ -pe mpi 4</span>
<span class="n">sumtrees</span> <span class="o">-</span><span class="n">m4</span> <span class="o">-</span><span class="n">o</span> <span class="n">result</span><span class="o">.</span><span class="n">con</span><span class="o">.</span><span class="n">tre</span> <span class="o">-</span><span class="n">burnin</span> <span class="mi">100</span> <span class="n">mcmc1</span><span class="o">.</span><span class="n">tre</span> <span class="n">mcmc2</span><span class="o">.</span><span class="n">tre</span> <span class="n">mcmc3</span><span class="o">.</span><span class="n">tre</span> <span class="n">mcmc4</span><span class="o">.</span><span class="n">tre</span>
</pre></div>
</div>
</div></blockquote>
<p>The “<code class="docutils literal notranslate"><span class="pre">#$</span> <span class="pre">-pe</span> <span class="pre">mpi</span> <span class="pre">4</span></code>” line tells the SGE scheduler to allocate four processors to the job in the “<code class="docutils literal notranslate"><span class="pre">mpi</span></code>” parallel environment, while the “<code class="docutils literal notranslate"><span class="pre">-m4</span></code>” part of the SumTrees command tells SumTrees to run 4 processes in parallel.</p>
<p>If you are using PBS/Torque as a scheduler, the equivalent job script might be:</p>
<blockquote>
<div><div class="highlight-default notranslate"><div class="highlight"><pre><span></span># /bin/sh
#PBS -l nodes=2:ncpus=2
cd $PBS_O_WORKDIR
sumtrees -m4 -o result.con.tre -burnin 100 mcmc1.tre mcmc2.tre mcmc3.tre mcmc4.tre
</pre></div>
</div>
</div></blockquote>
<p>Here, the “<code class="docutils literal notranslate"><span class="pre">#PBS</span> <span class="pre">-l</span> <span class="pre">nodes=2:ncpus=2</span></code>” requests two processors on two nodes, while the “<code class="docutils literal notranslate"><span class="pre">-m4</span></code>” part of the SumTrees command, as before, tells SumTrees to run 4 processes in parallel.</p>
<p>The particular job scripts you use will almost certainly be different, varying with the cluster job/load management software, scheduler and computing resources.
Apart from the parallel environment name, number of processors and/or machine configuration, you might also need to provide a queue name, a maximum run time limit, and a soft or hard memory limit.
None of it should make any difference to how SumTrees is actually invoked: you would still just use the “<code class="docutils literal notranslate"><span class="pre">-m</span></code>” or “<code class="docutils literal notranslate"><span class="pre">--multiprocessing</span></code>” flags to specify that SumTrees runs in parallel mode with a particular number of parallel processes).
You just need to check with your cluster administrators or documentation to make sure your job script or execution context provides sufficient processors to match the number of processes that you want run (as well as other resources, e.g. wall time limit and memory limit, that SumTrees needs to finish its job).</p>
</section>
</section>
<section id="improving-performance">
<h2>Improving Performance<a class="headerlink" href="#improving-performance" title="Link to this heading">¶</a></h2>
<blockquote>
<div><ul>
<li><p>Run SumTrees in parallel mode (see “<a class="reference internal" href="#running-in-parallel-mode">Running in Parallel Mode</a>” or “<a class="reference internal" href="#parallelizing-sumtrees">Parallelizing SumTrees</a>”):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees -m 4 phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
</pre></div>
</div>
</li>
<li><p>Reduce the tree-processing logging frequency.
By default, SumTrees will report back every 500th tree in a tree file, just to let you know where it is and to give you a sense of how long it will take. You can use the “<code class="docutils literal notranslate"><span class="pre">-g</span></code>”  or “<code class="docutils literal notranslate"><span class="pre">--log-frequency</span></code>” flag to control this behavior. If you have very large files, you may be content to have SumTrees report back every 5000th or even every 10000th tree instead:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --log-frequency=10000 phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
$ sumtrees -g10000 phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
</pre></div>
</div>
<p>If you are content to let SumTrees run without reporting its progress <em>within</em> each file (SumTrees will still report back whenever it begins or ends working on a file), then you can switch off tree processing logging altogether by specifying a logging frequency of 0:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sumtrees --log-frequency=0 phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
$ sumtrees -g0 phylo.run1.tre phylo.run2.tre phylo.run3.tre phylo.run4.tre
</pre></div>
</div>
</li>
</ul>
</div></blockquote>
</section>
<section id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Link to this heading">¶</a></h2>
<section id="prerequisites">
<h3>Prerequisites<a class="headerlink" href="#prerequisites" title="Link to this heading">¶</a></h3>
<p><a class="reference external" href="https://jeetsukumaran.github.io/DendroPy/">DendroPy</a> is a <a class="reference external" href="http://www.python.org/">Python</a> library.
It requires and presupposes not only the existence of a <a class="reference external" href="http://www.python.org/">Python</a> installation on your system, but also that this Python installation is available on the system path.</p>
<p>The biggest problem faced by most users is not so much not having <a class="reference external" href="http://www.python.org/">Python</a> installed, but not having the correct version of Python installed. You can check which version of Python you have running by typing:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python -V
</pre></div>
</div>
<p>SumTrees, and the <a class="reference external" href="https://jeetsukumaran.github.io/DendroPy/">DendroPy</a> library that it is part of, works out-of-the-box with any 2.x Python version 2.4 or greater.</p>
<p>SumTrees will not work with versions of <a class="reference external" href="http://www.python.org/">Python</a> prior to 2.4, such as <a class="reference external" href="http://www.python.org/">Python</a> 2.3. It can probably be made to work pretty easily, and if you have strong enough motiviation to use Python 2.3, it might be worth the effort for you.
It is not for me.</p>
<p>SumTrees (and DendroPy) is currently not compatible with Python 3.</p>
</section>
<section id="my-computer-does-not-know-what-a-python-is">
<h3>My Computer Does Not Know What a Python Is<a class="headerlink" href="#my-computer-does-not-know-what-a-python-is" title="Link to this heading">¶</a></h3>
<p>If you get a message like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span><span class="p">:</span> <span class="n">command</span> <span class="ow">not</span> <span class="n">found</span>
</pre></div>
</div>
<p>it is either because <a class="reference external" href="http://www.python.org/">Python</a>  is not installed on your system, or is not found on the system path.</p>
<p>SumTrees is a <a class="reference external" href="http://www.python.org/">Python</a> script, and, as such, you will need to have a <a class="reference external" href="http://www.python.org/">Python</a>  interpreter installed on your system.</p>
<p>Otherwise, you must download and install Python 2.6 from: <a class="reference external" href="http://www.python.org/download/releases/2.6/">http://www.python.org/download/releases/2.6/</a>.
For your convenience, the clicking on the following links should lead you directly to the appropriate pre-compiled download:</p>
<ul class="simple">
<li><p><a class="reference external" href="http://www.python.org/ftp/python/2.6/python-2.6-macosx.dmg">Mac OS X</a></p></li>
<li><p><a class="reference external" href="http://www.python.org/ftp/python/2.6/python-2.6.msi">Microsoft Windows</a></p></li>
</ul>
<p>For other platforms, the usual “<code class="docutils literal notranslate"><span class="pre">./configure</span></code>”, “<code class="docutils literal notranslate"><span class="pre">make</span></code>”, and “<code class="docutils literal notranslate"><span class="pre">sudo</span> <span class="pre">make</span> <span class="pre">install</span></code>” dance should get you up and running with the following:</p>
<ul class="simple">
<li><p><a class="reference external" href="http://www.python.org/ftp/python/2.6/Python-2.6.tgz">Cross-platform Source</a></p></li>
</ul>
<p>Microsoft Windows users should also refer to the <a class="reference external" href="http://www.python.org/doc/faq/windows.html">“Python Windows FAQ”</a>
(<a class="reference external" href="http://www.python.org/doc/faq/windows.html">http://www.python.org/doc/faq/windows.html</a>)
after installing Python, and pay particular attention to the
<a class="reference external" href="http://www.python.org/doc/faq/windows.html#id2">“How do I run a Python program under Windows?”</a> section, as it will
help them greatly in getting Python up and running on the system path.</p>
</section>
</section>
<section id="manual-installation">
<h2>Manual Installation<a class="headerlink" href="#manual-installation" title="Link to this heading">¶</a></h2>
<p>The <a class="reference external" href="https://jeetsukumaran.github.io/DendroPy/">DendroPy</a> library is actually quite straightforward to install manually, especially if you have any familiarity with Python and how Python files are organized.
There are a couple of different things you could do:</p>
<ul class="simple">
<li><p>Add the current location of the “<code class="docutils literal notranslate"><span class="pre">dendropy</span></code>” subdirectory to your Python path environmental variable, “<code class="docutils literal notranslate"><span class="pre">$PYTHONPATH</span></code>”, and place the file “<code class="docutils literal notranslate"><span class="pre">programs/sumtrees</span></code>” on your system path.</p></li>
<li><p>Copy (or symlink) the “<code class="docutils literal notranslate"><span class="pre">dendropy</span></code>” directory to the “<code class="docutils literal notranslate"><span class="pre">site-packages</span></code>” directory of your Python installation, and place the file “<code class="docutils literal notranslate"><span class="pre">programs/sumtrees</span></code>” on your system path.</p></li>
</ul>
</section>
<section id="repository-access">
<h2>Repository Access<a class="headerlink" href="#repository-access" title="Link to this heading">¶</a></h2>
<p>The <a class="reference external" href="https://jeetsukumaran.github.io/DendroPy/">DendroPy</a> public-access <a class="reference external" href="http://git-scm.com/">Git</a> repository can be cloned from:</p>
<blockquote>
<div><p><a class="reference external" href="http://github.com/jeetsukumaran/DendroPy">http://github.com/jeetsukumaran/DendroPy</a></p>
</div></blockquote>
</section>
<section id="bugs-suggestions-comments-etc">
<h2>Bugs, Suggestions, Comments, etc.<a class="headerlink" href="#bugs-suggestions-comments-etc" title="Link to this heading">¶</a></h2>
<p>If you encounter any problems, errors, crashes etc. while using this program, please let me (<a class="reference external" href="http://www.jeetworks.org/about">Jeet Sukumaran</a>) know at <a class="reference external" href="mailto:jeetsukumaran&#37;&#52;&#48;gmail&#46;com">jeetsukumaran<span>&#64;</span>gmail<span>&#46;</span>com</a>. If you include the term “sumtrees” anywhere on the subject line (e.g. “Problem such-and-such with bootscore), it would help greatly with getting through the spam filter. Please include all the datafiles involved, as
well the complete command used (with all the options and parameters) and the complete error message returned (simply cutting-and-pasting the terminal text should work fine).
Please feel free to contact me if you have any other questions, suggestions or comments as well.</p>
</section>
<section id="how-to-cite-this-program">
<h2>How to Cite this Program<a class="headerlink" href="#how-to-cite-this-program" title="Link to this heading">¶</a></h2>
<p>If you use this program in your analysis, please cite it as:</p>
<blockquote>
<div><p>Moreno, M. A., Sukumaran, J., and M. T. Holder. 2024. DendroPy 5: a mature Python library for phylogenetic computing. arXiv preprint arXiv:2405.14120. https://doi.org/10.48550/arXiv.2405.14120</p>
</div></blockquote>
<p>In the text of your paper, if you want to look like you know what you are doing, you should probably also mention explicitly that you specifically used the SumTrees program of the <a class="reference external" href="https://jeetsukumaran.github.io/DendroPy/">DendroPy</a> package, as well as the particular version numbers of SumTrees and <a class="reference external" href="https://jeetsukumaran.github.io/DendroPy/">DendroPy</a> that you used.</p>
</section>
<section id="copyright-license-and-warranty">
<h2>Copyright, License, and Warranty<a class="headerlink" href="#copyright-license-and-warranty" title="Link to this heading">¶</a></h2>
<div class="copyrightblock docutils container">
<p>Copyright 2009-2024, Jeet Sukumaran and Mark T. Holder. All rights reserved.</p>
<p>The DendroPy project is distributed under the BSD 3-clause license except where otherwise explicitly noted.
Refer to our <a class="reference external" href="https://github.com/jeetsukumaran/DendroPy/blob/main/LICENSE.rst">license</a> and <a class="reference external" href="https://github.com/jeetsukumaran/DendroPy/blob/main/NOTICES.rst">notices</a> for details.</p>
</div>
</section>
<section id="acknowledgments">
<h2>Acknowledgments<a class="headerlink" href="#acknowledgments" title="Link to this heading">¶</a></h2>
<p>Portions of <a class="reference external" href="https://jeetsukumaran.github.io/DendroPy/">DendroPy</a> were developed under <a class="reference external" href="http://www.phylo.org">CIPRES</a>, a multi-site collaboration funded by the <a class="reference external" href="http://www.nsf.gov/">NSF</a> Information Technology Research (ITR) program grant entitled “<a class="reference external" href="http://www.phylo.org/about/acknowledgements">BUILDING THE TREE OF LIFE: A National Resource for Phyloinformatics and Computational Phylogenetics</a>” and <a class="reference external" href="http://www.nsf.gov/">NSF</a> ATOL grant 0732920 to Mark Holder.</p>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div style="text-align: center; padding-top: 20px; padding-bottom: 5px; width: 100%;">
    <a href="../index.html"><img src="../_static/dendropy_logo.png" /></a>
</div><div style="clear:both; width: 100%; height:1px;"></div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h3><a href="../index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">SumTrees: Phylogenetic Tree Summarization and Annotation</a><ul>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#where-to-find-the-package">Where to Find the Package</a></li>
<li><a class="reference internal" href="#how-to-install-the-package">How to Install the Package</a></li>
<li><a class="reference internal" href="#checking-the-installation">Checking the Installation</a></li>
<li><a class="reference internal" href="#how-to-use-the-program">How to Use the Program</a></li>
<li><a class="reference internal" href="#quick-recipes">Quick Recipes</a><ul>
<li><a class="reference internal" href="#summarization-of-posterior-probabilities-of-clades-with-a-consensus-tree">Summarization of Posterior Probabilities of Clades with a Consensus Tree</a></li>
<li><a class="reference internal" href="#summarization-of-posterior-probabilities-of-clades-with-a-maximum-clade-credibility-tree-mcct">Summarization of Posterior Probabilities of Clades with a Maximum Clade Credibility Tree (MCCT)</a></li>
<li><a class="reference internal" href="#non-parametric-bootstrap-support-of-a-model-tree">Non-parametric Bootstrap Support of a Model Tree</a></li>
<li><a class="reference internal" href="#set-node-ages-of-target-or-summary-tree-s-to-mean-median-node-age-of-input-trees">Set Node Ages of Target or Summary Tree(s) to Mean/Median Node Age of Input Trees</a></li>
<li><a class="reference internal" href="#running-in-parallel-mode">Running in Parallel Mode</a></li>
</ul>
</li>
<li><a class="reference internal" href="#primers-and-examples">Primers and Examples</a></li>
<li><a class="reference internal" href="#specifying-and-customization-of-the-summarization-target">Specifying and Customization of the Summarization Target</a><ul>
<li><a class="reference internal" href="#specifying-a-summarization-topology-type">Specifying a Summarization Topology Type</a><ul>
<li><a class="reference internal" href="#majority-rule-consensus-topology">Majority-Rule Consensus Topology</a></li>
<li><a class="reference internal" href="#maximum-clade-credibility-topology">Maximum Clade Credibility Topology</a></li>
</ul>
</li>
<li><a class="reference internal" href="#specifying-a-custom-topology-or-set-of-topologies">Specifying a Custom Topology or Set of Topologies</a></li>
<li><a class="reference internal" href="#collapsing-clades-edges-splits-below-a-minimum-support-frequency-threshold">Collapsing Clades/Edges/Splits Below a Minimum Support Frequency Threshold</a></li>
</ul>
</li>
<li><a class="reference internal" href="#summarizing-edge-lengths">Summarizing Edge Lengths</a></li>
<li><a class="reference internal" href="#summarizing-rooted-and-ultrametric-trees">Summarizing Rooted and Ultrametric Trees</a><ul>
<li><a class="reference internal" href="#forcing-trees-to-be-treated-as-rooted">Forcing Trees to be Treated as Rooted</a></li>
<li><a class="reference internal" href="#summarizing-node-ages">Summarizing Node Ages</a></li>
<li><a class="reference internal" href="#setting-the-node-ages-of-the-summary-trees">Setting the Node Ages of the Summary Trees</a></li>
<li><a class="reference internal" href="#rooting-the-target-topology">Rooting the Target Topology</a></li>
</ul>
</li>
<li><a class="reference internal" href="#assigning-the-ages-of-non-contemporaneous-tips-tip-dating">Assigning the Ages of Non-Contemporaneous Tips (“Tip-Dating”)</a></li>
<li><a class="reference internal" href="#parallelizing-sumtrees">Parallelizing SumTrees</a><ul>
<li><a class="reference internal" href="#basics">Basics</a></li>
<li><a class="reference internal" href="#parallelization-strategy-deciding-on-the-number-of-processes">Parallelization Strategy: Deciding on the Number of Processes</a></li>
<li><a class="reference internal" href="#running-parallel-mode-sumtrees-in-a-parallel-environment-on-a-high-performance-computing-hpc-cluster">Running Parallel-Mode SumTrees in a Parallel Environment on a High-Performance Computing (HPC) Cluster</a></li>
</ul>
</li>
<li><a class="reference internal" href="#improving-performance">Improving Performance</a></li>
<li><a class="reference internal" href="#troubleshooting">Troubleshooting</a><ul>
<li><a class="reference internal" href="#prerequisites">Prerequisites</a></li>
<li><a class="reference internal" href="#my-computer-does-not-know-what-a-python-is">My Computer Does Not Know What a Python Is</a></li>
</ul>
</li>
<li><a class="reference internal" href="#manual-installation">Manual Installation</a></li>
<li><a class="reference internal" href="#repository-access">Repository Access</a></li>
<li><a class="reference internal" href="#bugs-suggestions-comments-etc">Bugs, Suggestions, Comments, etc.</a></li>
<li><a class="reference internal" href="#how-to-cite-this-program">How to Cite this Program</a></li>
<li><a class="reference internal" href="#copyright-license-and-warranty">Copyright, License, and Warranty</a></li>
<li><a class="reference internal" href="#acknowledgments">Acknowledgments</a></li>
</ul>
</li>
</ul>

  </div><div style="clear:both; width: 100%; height:1px;"></div>
  <div>
    <h4>Previous topic</h4>
    <p class="topless"><a href="index.html"
                          title="previous chapter">Programs and Applications</a></p>
  </div>
  <div>
    <h4>Next topic</h4>
    <p class="topless"><a href="../glossary.html"
                          title="next chapter">Glossary and Terminological Reference</a></p>
  </div><div style="clear:both; width: 100%; height:1px;"></div>
    <!-- Documentation -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Documentation</h3>
        <ul>
            <li><a href="../primer/downloading.html">Downloading and Installing DendroPy</a></li>
            <li><a href="../primer/index.html">The DendroPy Primer</a></li>
            <li><a href="../library/index.html">Library API Reference</a></li>
            <li>
                <a href="../schemas/index.html">Schemas</a>
                <ul>
                    <li><a href="../schemas/fasta.html">FASTA</a></li>
                    <li><a href="../schemas/newick.html">Newick</a></li>
                    <li><a href="../schemas/nexml.html">NeXML</a></li>
                    <li><a href="../schemas/nexus.html">Nexus</a></li>
                    <li><a href="../schemas/phylip.html">PHYLIP</a></li>
                </ul>
            </li>
            <li>
                <a href="index.html">Programs</a>
                <ul>
                    <li><a href="#">SumTrees</a></li>
                </ul>
            </li>
            <li><a href="../glossary.html">Glossary and Terminological Reference</a></li>
            <li><a href="../developer.html">Developer Guide</a></li>
            <li><a href="../planning.html">Ongoing Development</a></li>
            <li><a href="../changes.html">Change History</a></li>
        </ul>
    </div>

    <!-- Downloads -->
    <div style="border-top: double 1px white; padding-top: 10px;">
        <h3>Obtaining</h3>
        <ul>
            <li><a target="_blank" href="http://pypi.python.org/pypi/DendroPy">Install from the Python Package Index</a></li>
            <li><a target="_blank" href="http://pypi.python.org/packages/source/D/DendroPy/DendroPy-5.0.2.tar.gz">Download the Source Code Archive</a></li>
            <li><a target="_blank" href="http://github.com/jeetsukumaran/DendroPy">Clone the Source Code Repository</a></li>
        </ul>
    </div>

    <!-- Discussions -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Discussion</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot; DendroPy Users&quot; </a>group to follow and participate in discussion, troubleshooting, help, information, suggestions, etc. on the usage and development of the DendroPy phylogenetic computing library.</p>
            <form action="http://groups.google.com/group/dendropy-users/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-users?hl=en">&quot;dendropy-users&quot;</a> group, or click <a href="http://groups.google.com/group/dendropy-users?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Announcements -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative;">
        <h3><span style="text-align: left">Announcements</span><span style="position: absolute; right: 0; top: 10px "><img src="../_static/google-groups-logo1.png" height="20px" alt="Google Groups" /></span></h3>
        <div style="margin-top: 15px;">
            <p style="font-size: 90%; margin-top: 3px; clear: both;">Join the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; DendroPy Announcements&quot; </a>group to receive announcements of new releases, updates, changes and other news of interest to DendroPy users and developers.</p>
            <form action="http://groups.google.com/group/dendropy-announce/boxsubscribe">
                <input type=text name=email>
                <input type=submit name="sub" value="Subscribe">
            </form>
            <p style="font-size: 90%; clear: both; padding-top: 5px; padding-bottom: 10px;">Enter your e-mail address in the box above and click the &quot;subscribe&quot; button to subscribe to the <a href="http://groups.google.com/group/dendropy-announce?hl=en">&quot; dendropy-announce&quot; </a>group, or click <a href="http://groups.google.com/group/dendropy-announce?hl=en">here</a> to visit this group page directly.</p>
        </div>
    </div>

    <!-- Development -->
    <div style="border-top: double 1px white; padding-top: 10px; position: relative; padding-bottom: 15px; margin-bottom:5px;">
        <h3><span style="text-align: left"><a href="https://github.com/jeetsukumaran/DendroPy/">Development</a></span><a href="https://github.com/jeetsukumaran/DendroPy/"><span style="position: absolute; right: 0; top: 10px "><img src="../_static/Octocat.png" height="30px" alt="GitHub" /></span></a></h3>
        <div style="margin-top: 15px;">
            <!-- <a href="https://github.com/jeetsukumaran/DendroPy/issues">Issues</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/subscription">Watch</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/fork">Fork</a> &bull; <a href="https://github.com/jeetsukumaran/DendroPy/stargazers">Star</a> &bull; <a href="https://github.com/jeetsukumaran/">Follow</a> -->
            <ul>
                <li>                <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/issues"><code>Issues</code></a></span> <span style="font-style: italic; font-size:80%;"> - Report bugs or request features</span></li>
                <li>     <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/subscription"><code>&nbsp;Watch</code></a></span> <span style="font-style: italic; font-size:80%;"> - Follow development activity</span></li>
                <li>        <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/fork"><code>&nbsp;&nbsp;Fork</code></a></span> <span style="font-style: italic; font-size:80%;"> - Contribute and collaborate</span></li>
                <li>  <span style="font-weight: bold;"><a href="https://github.com/jeetsukumaran/DendroPy/stargazers"><code>&nbsp;&nbsp;Star</code></a></span> <span style="font-style: italic; font-size:80%;"> - Throw some glitter, add some glamour</span></li>
        </div>
    </div>


        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../glossary.html" title="Glossary and Terminological Reference"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Programs and Applications"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">DendroPy 5.0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Programs and Applications</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">SumTrees: Phylogenetic Tree Summarization and Annotation</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2009-2024, Jeet Sukumaran and Mark T. Holder.
    </div>
  </body>
</html>